## Detecting empty spots

In droplet-based Single-cell RNA-Seq`DropletUtils::emptyDrops()` computes Monte Carlo p-values based on a Dirichlet-multinomial model of sampling molecules into droplets. Same approach can be used for spots in spot-based spatial transcriptomics data.
`emptyDrops()` assumes that libraries with total UMI counts below a certain threshold (`r cfg$EMPTY_DROPLETS_LOWER` by default)
correspond to empty droplets/spots.
These are used to estimate the ambient expression profile against which the remaining libraries are tested.
Under this definition, these low-count libraries cannot be tissue containing spots and are excluded from the hypothesis testing.

**Number of non-empty spots: `r sce_valid_cells_info$dim[2]`**

Non-empty spots should show up with large negative log-probabilities or very large total UMI counts (based on the knee point).

```{r}
is_cell <- empty_droplets$FDR <= cfg$EMPTY_DROPLETS_FDR_THRESHOLD
plot(empty_droplets$Total, -empty_droplets$LogProb, col = ifelse(is_cell, "red", "black"), xlab = "Total UMI count", ylab = "-Log Probability")
```

Spots with empty (spots) FDR > `r cfg$EMPTY_DROPLETS_FDR_THRESHOLD` have been removed. Filtered dataset summary:

```{r}
cat(sce_valid_cells_info$str)
```

`r scdrake::format_used_functions("DropletUtils::emptyDrops()")`
