<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/3/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>A Pipeline For 10x Chromium Single-Cell RNA-seq Data Analysis • scdrake</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/3/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/3/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="preferably.css">
<link id="css-code-light" rel="stylesheet" href="code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="darkswitch.js"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A Pipeline For 10x Chromium Single-Cell RNA-seq Data Analysis">
<meta property="og:description" content="
    This pipeline is implemented in the drake Make-like toolkit and
    provides the following steps: input quality control, filtering of cells/genes,
    normalization, clustering, cluster markers, differential expression analysis
    (constrasts -&gt; cluster vs. cluster), integration of multiple samples.
    All main parts are accompanied by rich graphical outputs and HTML reports
    (RMarkdown).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scdrake</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/scdrake.html">Get started</a>
</li>
<li>
  <a href="articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="divider">
    <li class="dropdown-header">General information</li>
    <li>
      <a href="articles/pipeline_overview.html">Pipeline overview</a>
    </li>
    <li>
      <a href="articles/scdrake_run.html">Running the pipeline, environment variables</a>
    </li>
    <li>
      <a href="articles/cluster_markers.html">Cluster markers</a>
    </li>
    <li>
      <a href="articles/drake_basics.html">drake basics</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Configs</li>
    <li>
      <a href="articles/scdrake_config.html">scdrake config</a>
    </li>
    <li>
      <a href="articles/config_pipeline.html">Pipeline config</a>
    </li>
    <li>
      <a href="articles/config_main.html">Main config</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Single-sample pipeline stages</li>
    <li>
      <a href="articles/stage_input_qc.html">Input and quality control stage</a>
    </li>
    <li>
      <a href="articles/stage_norm_clustering.html">Normalization and clustering stage</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Integration pipeline stages</li>
    <li>
      <a href="articles/stage_integration.html">Integration stage</a>
    </li>
    <li>
      <a href="articles/stage_int_clustering.html">Post-integration clustering stage</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Common stages</li>
    <li>
      <a href="articles/stage_cluster_markers.html">Cluster markers stage</a>
    </li>
    <li>
      <a href="articles/stage_contrasts.html">Contrasts stage</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
          <a href="#" id="css-toggle-btn">
            <span class="fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        <li>
  <a href="https://github.com/bioinfocz/scdrake">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
        
        


      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="scdrake" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#scdrake" class="anchor"></a>scdrake</h1></div>

<p><a href="https://github.com/bioinfocz/scdrake">scdrake</a> is a scalable and reproducible pipeline for downstream processing of 10x Genomics single-cell RNA-seq data. It is built on top of the <a href="https://docs.ropensci.org/drake">drake</a> package, a <a href="https://www.gnu.org/software/make/">Make</a>-like pipeline toolkit for <a href="https://www.r-project.org/">R language</a>.</p>
<p>The main features of the <a href="https://github.com/bioinfocz/scdrake">scdrake</a> pipeline are:</p>
<ul>
<li>Import of <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> output, quality filtering of cells and genes.</li>
<li>Normalization, clustering, and dimensionality reduction.</li>
<li>Integration of multiple datasets.</li>
<li>Computation of cluster markers and differentially expressed genes between clusters (denoted as “contrasts”).</li>
<li>Rich graphical and HTML outputs.</li>
<li>Thanks to <a href="https://docs.ropensci.org/drake">drake</a>, the pipeline is highly scalable and reproducible.
<ul>
<li>Want to change some parameter? No problem! Only parts of the pipeline which changed will rerun, while up-to-date ones will be skipped.</li>
<li>Want to reuse the intermediate results for your own analyses? No problem! The pipeline has smartly defined checkpoints which can be loaded from a <a href="https://docs.ropensci.org/drake">drake</a> cache.</li>
<li>Want to extend the pipeline? No problem! The pipeline definition is just an R object which can be arbitrarily extended.</li>
</ul>
</li>
</ul>
<p><a href="https://github.com/bioinfocz/scdrake">scdrake</a> is aimed at both non-technical and bioinformatic public: both will benefit from reports and visualizations, and the latter also from the possibility to utilize all benefits of <a href="https://docs.ropensci.org/drake">drake</a> for custom analyses.</p>
<div id="installation-instructions" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-instructions" class="anchor"></a>Installation instructions</h2>
<p><a href="https://github.com/bioinfocz/scdrake">scdrake</a> is currently not released on <a href="http://cran.r-project.org/">CRAN</a>, however, you can install it from GitHub using the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"bioinfocz/scdrake"</span><span class="op">)</span></code></pre></div>
<p>To fully utilize the parallelization, you should consider installing <a href="https://zeromq.org/">ZeroMQ</a> and <a href="https://mschubert.github.io/clustermq/">clustermq</a> R package, which is used by <a href="https://docs.ropensci.org/drake">drake</a>. However, it is also possible to use the <a href="https://future.futureverse.org">future</a> package, which doesn’t require any additional system dependencies.</p>
<div id="using-renv" class="section level3">
<h3 class="hasAnchor">
<a href="#using-renv" class="anchor"></a>Using <code>renv</code>
</h3>
<p>Because <a href="https://github.com/bioinfocz/scdrake">scdrake</a> uses a lot of packages, we also encourage users to use the <a href="https://rstudio.github.io/renv/">renv</a> package, which will install all dependencies to a local, private, project library. In case you have already initialized a new <a href="https://github.com/bioinfocz/scdrake">scdrake</a> project (as described in <code><a href="articles/scdrake.html">vignette("scdrake")</a></code>) and your working directory is in the project’s root, you can initialize a new <a href="https://rstudio.github.io/renv/">renv</a> private library with</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/restore.html">restore</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>It will use the <code>renv.lock</code> file (which is bundled with <a href="https://github.com/bioinfocz/scdrake">scdrake</a> package) holding all dependencies and their versions.</p>
<p>To save time and space, you can also symlink the <code>renv/library</code> directory to multiple <a href="https://github.com/bioinfocz/scdrake">scdrake</a> projects.</p>
</div>
</div>
<div id="vignettes-and-other-readings" class="section level2">
<h2 class="hasAnchor">
<a href="#vignettes-and-other-readings" class="anchor"></a>Vignettes and other readings</h2>
<p>See <a href="https://bioinfocz.github.io/scdrake" class="uri">https://bioinfocz.github.io/scdrake</a> for a documentation website.</p>
<ul>
<li>Get started: <code><a href="articles/scdrake.html">vignette("scdrake")</a></code>
</li>
<li>Pipeline overview: <code><a href="articles/pipeline_overview.html">vignette("pipeline_overview")</a></code>
<ul>
<li>Cluster markers: <code><a href="articles/cluster_markers.html">vignette("cluster_markers")</a></code>
</li>
</ul>
</li>
<li>Running the pipeline, environment variables: <code><a href="articles/scdrake_run.html">vignette("scdrake_run")</a></code>
</li>
<li>Config files (basic concept): <code><a href="articles/scdrake_config.html">vignette("scdrake_config")</a></code>
</li>
<li>Targets and config parameters for each stage:
<ul>
<li>Common:
<ul>
<li>Pipeline config: <code><a href="articles/config_pipeline.html">vignette("config_pipeline")</a></code>
</li>
<li>Main config: <code><a href="articles/config_main.html">vignette("config_main")</a></code>
</li>
<li>Cluster markers stage: <code><a href="articles/stage_cluster_markers.html">vignette("stage_cluster_markers")</a></code>
</li>
<li>Contrasts stage: <code><a href="articles/stage_contrasts.html">vignette("stage_contrasts")</a></code>
</li>
</ul>
</li>
<li>Single-sample pipeline:
<ul>
<li>Reading in data, filtering, quality control (<code>01_input_qc</code>): <code><a href="articles/stage_input_qc.html">vignette("stage_input_qc")</a></code>
</li>
<li>Normalization, HVG selection, clustering (<code>02_norm_clustering</code>): <code><a href="articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code>
</li>
</ul>
</li>
<li>Integration pipeline:
<ul>
<li>Reading in data and integration (<code>01_integration</code>): <code><a href="articles/stage_integration.html">vignette("stage_integration")</a></code>
</li>
<li>Clustering (<code>02_int_clustering</code>): <code><a href="articles/stage_int_clustering.html">vignette("stage_int_clustering")</a></code>
</li>
</ul>
</li>
</ul>
</li>
<li>
<a href="https://docs.ropensci.org/drake">drake</a> basics: <code><a href="articles/drake_basics.html">vignette("drake_basics")</a></code>
<ul>
<li>Or the official <a href="https://docs.ropensci.org/drake">drake</a> book: <a href="https://books.ropensci.org/drake/" class="uri">https://books.ropensci.org/drake/</a>
</li>
</ul>
</li>
</ul>
<p>We encourage all users to read <a href="https://books.ropensci.org/drake/">basics</a> of the <a href="https://docs.ropensci.org/drake">drake</a> package. While it is not necessary to know all <a href="https://docs.ropensci.org/drake">drake</a> internals to successfully run the <a href="https://github.com/bioinfocz/scdrake">scdrake</a> pipeline, its knowledge is a plus. You can read the minimum basics in <code><a href="articles/drake_basics.html">vignette("drake_basics")</a></code>.</p>
<p>Also, the prior knowledge of Bioconductor and its classes (especially the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">SingleCellExperiment</a>) is considerable.</p>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>Below is the citation output from using <code><a href="https://rdrr.io/r/utils/citation.html">citation("scdrake")</a></code> in R. Please run this yourself to check for any updates on how to cite <strong>scdrake</strong>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span><span class="op">(</span><span class="st">"scdrake"</span><span class="op">)</span>, bibtex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; To cite package 'scdrake' in publications use:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   Jiri Novotny and Jan Kubovciak (2021). scdrake: A Pipeline For 10x</span>
<span class="co">#&gt;   Chromium Single-Cell RNA-seq Data Analysis.</span>
<span class="co">#&gt;   https://github.com/bioinfocz/scdrake,</span>
<span class="co">#&gt;   https://bioinfocz.github.io/scdrake.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; A BibTeX entry for LaTeX users is</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   @Manual{,</span>
<span class="co">#&gt;     title = {scdrake: A Pipeline For 10x Chromium Single-Cell RNA-seq Data Analysis},</span>
<span class="co">#&gt;     author = {Jiri Novotny and Jan Kubovciak},</span>
<span class="co">#&gt;     year = {2021},</span>
<span class="co">#&gt;     note = {https://github.com/bioinfocz/scdrake, https://bioinfocz.github.io/scdrake},</span>
<span class="co">#&gt;   }</span></code></pre></div>
<p>Please note that the <a href="https://github.com/bioinfocz/scdrake">scdrake</a> was only made possible thanks to many other R and bioinformatics software authors, which are cited either in the vignettes and/or the paper(s) describing this package.</p>
</div>
<div id="help-and-support" class="section level2">
<h2 class="hasAnchor">
<a href="#help-and-support" class="anchor"></a>Help and support</h2>
<p>In case of any problems or suggestions, please, open a new <a href="https://github.com/bioinfocz/scdrake/issues">issue</a>. We will be happy to answer your questions, integrate new ideas, or resolve any problems <span class="emoji" data-emoji="blush">😊</span></p>
</div>
<div id="contribution" class="section level2">
<h2 class="hasAnchor">
<a href="#contribution" class="anchor"></a>Contribution</h2>
<p>If you want to contribute to <a href="https://github.com/bioinfocz/scdrake">scdrake</a>, read the <a href="CONTRIBUTING.html">contribution guide</a>, please. All pull requests are welcome! <span class="emoji" data-emoji="slightly_smiling_face">🙂</span></p>
</div>
<div id="code-of-conduct" class="section level2">
<h2 class="hasAnchor">
<a href="#code-of-conduct" class="anchor"></a>Code of Conduct</h2>
<p>Please note that the <a href="https://github.com/bioinfocz/scdrake">scdrake</a> project is released with a <a href="https://bioinfocz.github.io/scdrake/CODE_OF_CONDUCT.html">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
</div>
<div id="development-tools" class="section level2">
<h2 class="hasAnchor">
<a href="#development-tools" class="anchor"></a>Development tools</h2>
<ul>
<li>Continuous code testing is possible thanks to <a href="https://www.tidyverse.org/blog/2020/04/usethis-1-6-0/">GitHub actions</a> through <a href="https://usethis.r-lib.org">usethis</a>, <a href="https://remotes.r-lib.org">remotes</a>, and <a href="https://r-lib.github.io/rcmdcheck/">rcmdcheck</a>. Customized to use <a href="https://www.bioconductor.org/help/docker/">Bioconductor’s docker containers</a> and <code>{BiocCheck}</code>.</li>
<li>Code coverage assessment is possible thanks to <a href="https://codecov.io/gh">codecov</a> and <a href="https://covr.r-lib.org">covr</a>.</li>
<li>The <a href="http://bioinfocz.github.io/scdrake">documentation website</a> is automatically updated thanks to <a href="https://pkgdown.r-lib.org">pkgdown</a>.</li>
<li>The code is styled automatically thanks to <a href="https://github.com/r-lib/styler">styler</a>.</li>
<li>The documentation is formatted thanks to <a href="https://devtools.r-lib.org/">devtools</a> and <a href="https://roxygen2.r-lib.org/">roxygen2</a>.</li>
</ul>
<p>This package was developed using <a href="https://github.com/lcolladotor/biocthis">biocthis</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/bioinfocz/scdrake">https://​github.com/​bioinfocz/​scdrake</a>
</li>
<li>Report a bug at <br><a href="https://github.com/bioinfocz/scdrake/issues">https://​github.com/​bioinfocz/​scdrake/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="community">
<h2>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Jiri Novotny <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1338-638X" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jan Kubovciak <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://codecov.io/gh/bioinfocz/scdrake?branch=main"><img src="https://codecov.io/gh/gorgitko/bioinfocz/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Jiri Novotny, Jan Kubovciak.</p>
</div>

<div class="pkgdown">
  <p>Made with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1, using <a href="https://preferably.amirmasoudabdol.name/?source=footer">preferably</a> template.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
