<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/3/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>A Pipeline For 10x Chromium Single-Cell RNA-seq Data Analysis ‚Ä¢ scdrake</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><!-- docsearch --><script src="docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="A Pipeline For 10x Chromium Single-Cell RNA-seq Data Analysis">
<meta property="og:description" content="
    This pipeline is implemented in the drake Make-like toolkit and
    provides the following steps: input quality control, filtering of cells/genes,
    normalization, clustering, cluster markers, differential expression analysis
    (constrasts -&gt; cluster vs. cluster), integration of multiple samples.
    All main parts are accompanied by rich graphical outputs and HTML reports
    (RMarkdown).">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/3/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/3/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="preferably.css">
<link id="css-code-light" rel="stylesheet" href="code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="darkswitch.js"></script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scdrake</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/scdrake.html">Get started</a>
</li>
<li>
  <a href="articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="divider">
    </li>
<li class="dropdown-header">General information</li>
    <li>
      <a href="articles/pipeline_overview.html">Pipeline overview</a>
    </li>
    <li>
      <a href="articles/scdrake_run.html">Running the pipeline, environment variables</a>
    </li>
    <li>
      <a href="articles/cluster_markers.html">Cluster markers</a>
    </li>
    <li>
      <a href="articles/drake_basics.html">drake basics</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Configs</li>
    <li>
      <a href="articles/scdrake_config.html">scdrake configs: a general overview</a>
    </li>
    <li>
      <a href="articles/config_pipeline.html">Pipeline config</a>
    </li>
    <li>
      <a href="articles/config_main.html">Main config</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Single-sample pipeline stages</li>
    <li>
      <a href="articles/stage_input_qc.html">Input and quality control stage</a>
    </li>
    <li>
      <a href="articles/stage_norm_clustering.html">Normalization and clustering stage</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Integration pipeline stages</li>
    <li>
      <a href="articles/stage_integration.html">Integration stage</a>
    </li>
    <li>
      <a href="articles/stage_int_clustering.html">Post-integration clustering stage</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Common stages</li>
    <li>
      <a href="articles/stage_cluster_markers.html">Cluster markers stage</a>
    </li>
    <li>
      <a href="articles/stage_contrasts.html">Contrasts stage</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
          <a href="#" id="css-toggle-btn">
            <span class="fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        <li>
  <a href="https://github.com/bioinfocz/scdrake" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
        
        


      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="scdrake">scdrake<a class="anchor" aria-label="anchor" href="#scdrake"></a>
</h1></div>

<p><code>{scdrake}</code> is a scalable and reproducible pipeline for secondary analysis of droplet-based single-cell RNA-seq data. <code>{scdrake}</code> is an R package built on top of the <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> package, a <a href="https://www.gnu.org/software/make" class="external-link">Make</a>-like pipeline toolkit for <a href="https://www.r-project.org" class="external-link">R language</a>.</p>
<p>The main features of the <code>{scdrake}</code> pipeline are:</p>
<ul>
<li>Import of scRNA-seq data: <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">10x Genomics Cell Ranger</a> output, delimited table, or <code>SingleCellExperiment</code> object.</li>
<li>Quality control and filtering of cells and genes, removal of empty droplets.</li>
<li>Higly variable genes detection, cell cycle scoring, normalization, clustering, and dimensionality reduction.</li>
<li>Integration of multiple datasets.</li>
<li>Computation of cluster markers and differentially expressed genes between clusters (denoted as ‚Äúcontrasts‚Äù).</li>
<li>Rich graphical and HTML outputs.
<ul>
<li>You can find links to example outputs <a href="https://bioinfocz.github.io/scdrake/articles/pipeline_overview.html">here</a>.</li>
</ul>
</li>
<li>Thanks to <a href="https://docs.ropensci.org/drake" class="external-link">drake</a>, the pipeline is highly scalable and reproducible.
<ul>
<li>Want to change some parameter? No problem! Only parts of the pipeline which changed will rerun, while up-to-date ones will be skipped.</li>
<li>Want to reuse the intermediate results for your own analyses? No problem! The pipeline has smartly defined checkpoints which can be loaded from a <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cache.</li>
<li>Want to extend the pipeline? No problem! The pipeline definition is just an R object which can be arbitrarily extended.</li>
</ul>
</li>
</ul>
<p><code>{scdrake}</code> is aimed at both non-technical and bioinformatic public: both will benefit from reports and visualizations, and the latter also from the possibility to utilize all benefits of <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> for custom analyses.</p>
<p>Huge thanks go to the authors of the <a href="https://bioconductor.org/books/release/OSCA" class="external-link">Orchestrating Single-Cell Analysis with Bioconductor</a> book on whose methods and recommendations is <code>{scdrake}</code> largely based.</p>
<div class="section level2">
<h2 id="installation-instructions">Installation instructions<a class="anchor" aria-label="anchor" href="#installation-instructions"></a>
</h2>
<p><code>{scdrake}</code> is currently not released on <a href="https://cran.r-project.org" class="external-link">CRAN</a>, however, you can install its latest stable version from GitHub using the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.14"</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"bioinfocz/scdrake@v1.0.1"</span><span class="op">)</span></code></pre></div>
<p>For development version use</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.14"</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"bioinfocz/scdrake"</span><span class="op">)</span></code></pre></div>
<div class="section level3">
<h3 id="parallelization">Parallelization<a class="anchor" aria-label="anchor" href="#parallelization"></a>
</h3>
<p>To fully utilize the parallelization, you should consider installing <a href="https://zeromq.org" class="external-link">ZeroMQ</a> and <a href="https://mschubert.github.io/clustermq/" class="external-link">clustermq</a> R package (probably only usable on Linux), which is used by <a href="https://docs.ropensci.org/drake" class="external-link">drake</a>. However, it is also possible to use the <a href="https://future.futureverse.org" class="external-link">future</a> package, which doesn‚Äôt require any additional system dependencies, but it‚Äôs a bit slower.</p>
<p>More reading about this topic can be found at <a href="https://books.ropensci.org/drake/hpc.html" class="external-link uri">https://books.ropensci.org/drake/hpc.html</a>.</p>
</div>
<div class="section level3">
<h3 id="using-renv">Using <code>renv</code>
<a class="anchor" aria-label="anchor" href="#using-renv"></a>
</h3>
<p>Because <code>{scdrake}</code> uses a lot of packages, we also encourage users to use the <a href="https://rstudio.github.io/renv/" class="external-link">renv</a> package, which will install all dependencies to a local, private, project library. In case you have already initialized a new <code>{scdrake}</code> project (as described in <code><a href="articles/scdrake.html">vignette("scdrake")</a></code>) and your working directory is in the project‚Äôs root, you can initialize a new <a href="https://rstudio.github.io/renv/" class="external-link">renv</a> private library with</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">renv</span><span class="fu">::</span><span class="fu">restore</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>It will use the <code>renv.lock</code> file (which is bundled with <code>{scdrake}</code> package) holding all dependencies and their versions.</p>
<p>To save time and space, you can also symlink the <code>renv/library</code> directory to multiple <code>{scdrake}</code> projects.</p>
</div>
</div>
<div class="section level2">
<h2 id="vignettes-and-other-readings">Vignettes and other readings<a class="anchor" aria-label="anchor" href="#vignettes-and-other-readings"></a>
</h2>
<p>See <a href="https://bioinfocz.github.io/scdrake" class="uri">https://bioinfocz.github.io/scdrake</a> for a documentation website, where links to vignettes below become real :-)</p>
<ul>
<li>Get started: <code><a href="articles/scdrake.html">vignette("scdrake")</a></code>
</li>
<li>Pipeline overview: <code><a href="articles/pipeline_overview.html">vignette("pipeline_overview")</a></code>
<ul>
<li>Cluster markers: <code><a href="articles/cluster_markers.html">vignette("cluster_markers")</a></code>
</li>
</ul>
</li>
<li>Running the pipeline, environment variables: <code><a href="articles/scdrake_run.html">vignette("scdrake_run")</a></code>
</li>
<li>Config files (basic concept): <code><a href="articles/scdrake_config.html">vignette("scdrake_config")</a></code>
</li>
<li>Targets and config parameters for each stage:
<ul>
<li>Common:
<ul>
<li>Pipeline config: <code><a href="articles/config_pipeline.html">vignette("config_pipeline")</a></code>
</li>
<li>Main config: <code><a href="articles/config_main.html">vignette("config_main")</a></code>
</li>
<li>Cluster markers stage: <code><a href="articles/stage_cluster_markers.html">vignette("stage_cluster_markers")</a></code>
</li>
<li>Contrasts stage: <code><a href="articles/stage_contrasts.html">vignette("stage_contrasts")</a></code>
</li>
</ul>
</li>
<li>Single-sample pipeline:
<ul>
<li>Reading in data, filtering, quality control (<code>01_input_qc</code>): <code><a href="articles/stage_input_qc.html">vignette("stage_input_qc")</a></code>
</li>
<li>Normalization, HVG selection, dimensionality reduction, clustering (<code>02_norm_clustering</code>): <code><a href="articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code>
</li>
</ul>
</li>
<li>Integration pipeline:
<ul>
<li>Reading in data and integration (<code>01_integration</code>): <code><a href="articles/stage_integration.html">vignette("stage_integration")</a></code>
</li>
<li>Clustering (<code>02_int_clustering</code>): <code><a href="articles/stage_int_clustering.html">vignette("stage_int_clustering")</a></code>
</li>
</ul>
</li>
</ul>
</li>
<li>
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> basics: <code><a href="articles/drake_basics.html">vignette("drake_basics")</a></code>
<ul>
<li>Or the official <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> book: <a href="https://books.ropensci.org/drake/" class="external-link uri">https://books.ropensci.org/drake/</a>
</li>
</ul>
</li>
</ul>
<p>We encourage all users to read <a href="https://books.ropensci.org/drake" class="external-link">basics</a> of the <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> package. While it is not necessary to know all <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> internals to successfully run the <code>{scdrake}</code> pipeline, its knowledge is a plus. You can read the minimum basics in <code><a href="articles/drake_basics.html">vignette("drake_basics")</a></code>.</p>
<p>Also, the prior knowledge of Bioconductor and its classes (especially the <a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>) is considerable.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Below is the citation output from using <code>citation("scdrake")</code> in R. Please run this yourself to check for any updates on how to cite <strong>scdrake</strong>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"scdrake"</span><span class="op">)</span>, bibtex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>To cite package ‚Äòscdrake‚Äô <span class="cf">in</span> publications use<span class="op">:</span></span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="st">  </span>Jiri Novotny and Jan <span class="kw">Kubovciak</span> (<span class="dv">2021</span>). scdrake<span class="op">:</span><span class="st"> </span>A Pipeline For 10x Chromium Single<span class="op">-</span>Cell RNA<span class="op">-</span>seq Data Analysis.</span>
<span id="cb5-4"><a href="#cb5-4"></a>  https<span class="op">:</span><span class="er">//</span>github.com<span class="op">/</span>bioinfocz<span class="op">/</span>scdrake, https<span class="op">:</span><span class="er">//</span>bioinfocz.github.io<span class="op">/</span>scdrake.</span>
<span id="cb5-5"><a href="#cb5-5"></a></span>
<span id="cb5-6"><a href="#cb5-6"></a>A BibTeX entry <span class="cf">for</span> LaTeX users is</span>
<span id="cb5-7"><a href="#cb5-7"></a></span>
<span id="cb5-8"><a href="#cb5-8"></a>  <span class="op">@</span>Manual{,</span>
<span id="cb5-9"><a href="#cb5-9"></a>    title =<span class="st"> </span>{scdrake<span class="op">:</span><span class="st"> </span>A Pipeline For 10x Chromium Single<span class="op">-</span>Cell RNA<span class="op">-</span>seq Data Analysis},</span>
<span id="cb5-10"><a href="#cb5-10"></a>    author =<span class="st"> </span>{Jiri Novotny and Jan Kubovciak},</span>
<span id="cb5-11"><a href="#cb5-11"></a>    year =<span class="st"> </span>{<span class="dv">2021</span>},</span>
<span id="cb5-12"><a href="#cb5-12"></a>    note =<span class="st"> </span>{https<span class="op">:</span><span class="er">//</span>github.com<span class="op">/</span>bioinfocz<span class="op">/</span>scdrake, https<span class="op">:</span><span class="er">//</span>bioinfocz.github.io<span class="op">/</span>scdrake},</span>
<span id="cb5-13"><a href="#cb5-13"></a>  }</span></code></pre></div>
<p>Please note that the <code>{scdrake}</code> was only made possible thanks to many other R and bioinformatics software authors, which are cited either in the vignettes and/or the paper(s) describing this package.</p>
</div>
<div class="section level2">
<h2 id="help-and-support">Help and support<a class="anchor" aria-label="anchor" href="#help-and-support"></a>
</h2>
<p>In case of any problems or suggestions, please, open a new <a href="https://github.com/bioinfocz/scdrake/issues" class="external-link">issue</a>. We will be happy to answer your questions, integrate new ideas, or resolve any problems <span class="emoji" data-emoji="blush">üòä</span></p>
<p>You can also use <a href="https://github.com/bioinfocz/scdrake/discussions" class="external-link">GitHub Discussions</a>, mainly for topics <strong>not</strong> related to development (bugs, feature requests etc.), but if you need e.g.¬†a general help.</p>
</div>
<div class="section level2">
<h2 id="contribution">Contribution<a class="anchor" aria-label="anchor" href="#contribution"></a>
</h2>
<p>If you want to contribute to <code>{scdrake}</code>, read the <a href="CONTRIBUTING.html">contribution guide</a>, please. All pull requests are welcome! :slightly_smiling_face:</p>
</div>
<div class="section level2">
<h2 id="code-of-conduct">Code of Conduct<a class="anchor" aria-label="anchor" href="#code-of-conduct"></a>
</h2>
<p>Please note that the <code>{scdrake}</code> project is released with a <a href="https://bioinfocz.github.io/scdrake/CODE_OF_CONDUCT.html">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<div class="section level3">
<h3 id="funding">Funding<a class="anchor" aria-label="anchor" href="#funding"></a>
</h3>
<p>This work was supported by <a href="https://www.elixir-czech.cz" class="external-link">ELIXIR CZ</a> research infrastructure project (MEYS Grant No: LM2018131) including access to computing and storage facilities.</p>
</div>
<div class="section level3">
<h3 id="software-and-methods-used-by-scdrake">Software and methods used by <code>{scdrake}</code>
<a class="anchor" aria-label="anchor" href="#software-and-methods-used-by-scdrake"></a>
</h3>
<p>Many things are used by <code>{scdrake}</code>, but these are really worth mentioning:</p>
<ul>
<li>The <a href="https://www.bioconductor.org" class="external-link">Bioconductor</a> ecosystem.</li>
<li>The <a href="https://bioconductor.org/books/release/OSCA" class="external-link"><em>Orchestrating Single-Cell Analysis with Bioconductor</em></a> book.</li>
<li>The <a href="https://bioconductor.org/packages/release/bioc/html/scran.html" class="external-link">scran</a>, <a href="https://bioconductor.org/packages/release/bioc/html/scater.html" class="external-link">scater</a>, and other great packages from <a href="https://orcid.org/0000-0002-3564-4813" class="external-link">Aaron Lun</a> et al.</li>
<li>The <a href="https://github.com/ropensci/drake" class="external-link">drake</a> package.</li>
<li>The <a href="https://github.com/rstudio/rmarkdown" class="external-link">rmarkdown</a> package, and other ones from the <a href="https://www.tidyverse.org" class="external-link">tidyverse</a> ecosystem.</li>
</ul>
</div>
<div class="section level3">
<h3 id="development-tools">Development tools<a class="anchor" aria-label="anchor" href="#development-tools"></a>
</h3>
<ul>
<li>Continuous code testing is possible thanks to <a href="https://github.com/features/actions" class="external-link">GitHub actions</a> through <a href="https://usethis.r-lib.org" class="external-link">usethis</a>, <a href="https://remotes.r-lib.org" class="external-link">remotes</a>, and <a href="https://r-lib.github.io/rcmdcheck/" class="external-link">rcmdcheck</a>. Customized to use <a href="https://www.bioconductor.org/help/docker" class="external-link">Bioconductor‚Äôs docker containers</a> and <del><code>{BiocCheck}</code></del> (in the future, probably).</li>
<li>Code coverage assessment is possible thanks to <a href="https://codecov.io/gh" class="external-link">codecov</a> and <a href="https://covr.r-lib.org" class="external-link">covr</a>.</li>
<li>The <a href="https://bioinfocz.github.io/scdrake">documentation website</a> is generated by <a href="https://pkgdown.r-lib.org" class="external-link">pkgdown</a>.</li>
<li>The code is styled automatically thanks to <a href="https://github.com/r-lib/styler" class="external-link">styler</a>.</li>
<li>The documentation is formatted thanks to <a href="https://devtools.r-lib.org/" class="external-link">devtools</a> and <a href="https://roxygen2.r-lib.org/" class="external-link">roxygen2</a>.</li>
</ul>
<p>This package was developed using <code>{biocthis}</code>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/bioinfocz/scdrake" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/bioinfocz/scdrake/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scdrake</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jiri Novotny <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1338-638X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jan Kubovciak <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="NEWS.html"><img src="https://img.shields.io/badge/NEWS-updates-informational" alt="NEWS: updates"></a></li>
<li><a href="https://bioinfocz.github.io/scdrake"><img src="https://img.shields.io/badge/Documentation%20&amp;%20vignettes-bioinfocz.github.io/scdrake-informational" alt="Documentation and vignettes"></a></li>
<li><a href="https://bioinfocz.github.io/scdrake/articles/pipeline_overview.html"><img src="https://img.shields.io/badge/Overview%20&amp;%20outputs-vignette(%22pipeline_overview%22)-informational" alt="Overview and outputs"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://github.com/bioinfocz/scdrake/actions/workflows/check-bioc.yaml" class="external-link"><img src="https://github.com/bioinfocz/scdrake/actions/workflows/check-bioc.yaml/badge.svg?branch=main" alt="R-CMD-check-bioc"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Jiri Novotny, Jan Kubovciak.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '937cedcd467c204b19fc155c5f35019b',
    indexName: 'scdrake',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
