<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<style type="text/css">
@font-face {
font-family: octicons-link;
src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff');
}
body {
-webkit-text-size-adjust: 100%;
text-size-adjust: 100%;
color: #333;
font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 16px;
line-height: 1.6;
word-wrap: break-word;
}
a {
background-color: transparent;
}
a:active,
a:hover {
outline: 0;
}
strong {
font-weight: bold;
}
h1 {
font-size: 2em;
margin: 0.67em 0;
}
img {
border: 0;
}
hr {
box-sizing: content-box;
height: 0;
}
pre {
overflow: auto;
}
code,
kbd,
pre {
font-family: monospace, monospace;
font-size: 1em;
}
input {
color: inherit;
font: inherit;
margin: 0;
}
html input[disabled] {
cursor: default;
}
input {
line-height: normal;
}
input[type="checkbox"] {
box-sizing: border-box;
padding: 0;
}
table {
border-collapse: collapse;
border-spacing: 0;
}
td,
th {
padding: 0;
}
* {
box-sizing: border-box;
}
input {
font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}
a {
color: #4078c0;
text-decoration: none;
}
a:hover,
a:active {
text-decoration: underline;
}
hr {
height: 0;
margin: 15px 0;
overflow: hidden;
background: transparent;
border: 0;
border-bottom: 1px solid #ddd;
}
hr:before {
display: table;
content: "";
}
hr:after {
display: table;
clear: both;
content: "";
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 15px;
margin-bottom: 15px;
line-height: 1.1;
}
h1 {
font-size: 30px;
}
h2 {
font-size: 21px;
}
h3 {
font-size: 16px;
}
h4 {
font-size: 14px;
}
h5 {
font-size: 12px;
}
h6 {
font-size: 11px;
}
blockquote {
margin: 0;
}
ul,
ol {
padding: 0;
margin-top: 0;
margin-bottom: 0;
}
ol ol,
ul ol {
list-style-type: lower-roman;
}
ul ul ol,
ul ol ol,
ol ul ol,
ol ol ol {
list-style-type: lower-alpha;
}
dd {
margin-left: 0;
}
code {
font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
font-size: 12px;
}
pre {
margin-top: 0;
margin-bottom: 0;
font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}
.select::-ms-expand {
opacity: 0;
}
.octicon {
font: normal normal normal 16px/1 octicons-link;
display: inline-block;
text-decoration: none;
text-rendering: auto;
-webkit-font-smoothing: antialiased;
-moz-osx-font-smoothing: grayscale;
-webkit-user-select: none;
-moz-user-select: none;
-ms-user-select: none;
user-select: none;
}
.octicon-link:before {
content: '\f05c';
}
.markdown-body:before {
display: table;
content: "";
}
.markdown-body:after {
display: table;
clear: both;
content: "";
}
.markdown-body>*:first-child {
margin-top: 0 !important;
}
.markdown-body>*:last-child {
margin-bottom: 0 !important;
}
a:not([href]) {
color: inherit;
text-decoration: none;
}
.anchor {
display: inline-block;
padding-right: 2px;
margin-left: -18px;
}
.anchor:focus {
outline: none;
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 1em;
margin-bottom: 16px;
font-weight: bold;
line-height: 1.4;
}
h1 .octicon-link,
h2 .octicon-link,
h3 .octicon-link,
h4 .octicon-link,
h5 .octicon-link,
h6 .octicon-link {
color: #000;
vertical-align: middle;
visibility: hidden;
}
h1:hover .anchor,
h2:hover .anchor,
h3:hover .anchor,
h4:hover .anchor,
h5:hover .anchor,
h6:hover .anchor {
text-decoration: none;
}
h1:hover .anchor .octicon-link,
h2:hover .anchor .octicon-link,
h3:hover .anchor .octicon-link,
h4:hover .anchor .octicon-link,
h5:hover .anchor .octicon-link,
h6:hover .anchor .octicon-link {
visibility: visible;
}
h1 {
padding-bottom: 0.3em;
font-size: 2.25em;
line-height: 1.2;
border-bottom: 1px solid #eee;
}
h1 .anchor {
line-height: 1;
}
h2 {
padding-bottom: 0.3em;
font-size: 1.75em;
line-height: 1.225;
border-bottom: 1px solid #eee;
}
h2 .anchor {
line-height: 1;
}
h3 {
font-size: 1.5em;
line-height: 1.43;
}
h3 .anchor {
line-height: 1.2;
}
h4 {
font-size: 1.25em;
}
h4 .anchor {
line-height: 1.2;
}
h5 {
font-size: 1em;
}
h5 .anchor {
line-height: 1.1;
}
h6 {
font-size: 1em;
color: #777;
}
h6 .anchor {
line-height: 1.1;
}
p,
blockquote,
ul,
ol,
dl,
table,
pre {
margin-top: 0;
margin-bottom: 16px;
}
hr {
height: 4px;
padding: 0;
margin: 16px 0;
background-color: #e7e7e7;
border: 0 none;
}
ul,
ol {
padding-left: 2em;
}
ul ul,
ul ol,
ol ol,
ol ul {
margin-top: 0;
margin-bottom: 0;
}
li>p {
margin-top: 16px;
}
dl {
padding: 0;
}
dl dt {
padding: 0;
margin-top: 16px;
font-size: 1em;
font-style: italic;
font-weight: bold;
}
dl dd {
padding: 0 16px;
margin-bottom: 16px;
}
blockquote {
padding: 0 15px;
color: #777;
border-left: 4px solid #ddd;
}
blockquote>:first-child {
margin-top: 0;
}
blockquote>:last-child {
margin-bottom: 0;
}
table {
display: block;
width: 100%;
overflow: auto;
word-break: normal;
word-break: keep-all;
}
table th {
font-weight: bold;
}
table th,
table td {
padding: 6px 13px;
border: 1px solid #ddd;
}
table tr {
background-color: #fff;
border-top: 1px solid #ccc;
}
table tr:nth-child(2n) {
background-color: #f8f8f8;
}
img {
max-width: 100%;
box-sizing: content-box;
background-color: #fff;
}
code {
padding: 0;
padding-top: 0.2em;
padding-bottom: 0.2em;
margin: 0;
font-size: 85%;
background-color: rgba(0,0,0,0.04);
border-radius: 3px;
}
code:before,
code:after {
letter-spacing: -0.2em;
content: "\00a0";
}
pre>code {
padding: 0;
margin: 0;
font-size: 100%;
word-break: normal;
white-space: pre;
background: transparent;
border: 0;
}
.highlight {
margin-bottom: 16px;
}
.highlight pre,
pre {
padding: 16px;
overflow: auto;
font-size: 85%;
line-height: 1.45;
background-color: #f7f7f7;
border-radius: 3px;
}
.highlight pre {
margin-bottom: 0;
word-break: normal;
}
pre {
word-wrap: normal;
}
pre code {
display: inline;
max-width: initial;
padding: 0;
margin: 0;
overflow: initial;
line-height: inherit;
word-wrap: normal;
background-color: transparent;
border: 0;
}
pre code:before,
pre code:after {
content: normal;
}
kbd {
display: inline-block;
padding: 3px 5px;
font-size: 11px;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.pl-c {
color: #969896;
}
.pl-c1,
.pl-s .pl-v {
color: #0086b3;
}
.pl-e,
.pl-en {
color: #795da3;
}
.pl-s .pl-s1,
.pl-smi {
color: #333;
}
.pl-ent {
color: #63a35c;
}
.pl-k {
color: #a71d5d;
}
.pl-pds,
.pl-s,
.pl-s .pl-pse .pl-s1,
.pl-sr,
.pl-sr .pl-cce,
.pl-sr .pl-sra,
.pl-sr .pl-sre {
color: #183691;
}
.pl-v {
color: #ed6a43;
}
.pl-id {
color: #b52a1d;
}
.pl-ii {
background-color: #b52a1d;
color: #f8f8f8;
}
.pl-sr .pl-cce {
color: #63a35c;
font-weight: bold;
}
.pl-ml {
color: #693a17;
}
.pl-mh,
.pl-mh .pl-en,
.pl-ms {
color: #1d3e81;
font-weight: bold;
}
.pl-mq {
color: #008080;
}
.pl-mi {
color: #333;
font-style: italic;
}
.pl-mb {
color: #333;
font-weight: bold;
}
.pl-md {
background-color: #ffecec;
color: #bd2c00;
}
.pl-mi1 {
background-color: #eaffea;
color: #55a532;
}
.pl-mdr {
color: #795da3;
font-weight: bold;
}
.pl-mo {
color: #1d3e81;
}
kbd {
display: inline-block;
padding: 3px 5px;
font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.task-list-item {
list-style-type: none;
}
.task-list-item+.task-list-item {
margin-top: 3px;
}
.task-list-item input {
margin: 0 0.35em 0.25em -1.6em;
vertical-align: middle;
}
:checked+.radio-label {
z-index: 1;
position: relative;
border-color: #4078c0;
}
.sourceLine {
display: inline-block;
}
code .kw { color: #000000; }
code .dt { color: #ed6a43; }
code .dv { color: #009999; }
code .bn { color: #009999; }
code .fl { color: #009999; }
code .ch { color: #009999; }
code .st { color: #183691; }
code .co { color: #969896; }
code .ot { color: #0086b3; }
code .al { color: #a61717; }
code .fu { color: #63a35c; }
code .er { color: #a61717; background-color: #e3d2d2; }
code .wa { color: #000000; }
code .cn { color: #008080; }
code .sc { color: #008080; }
code .vs { color: #183691; }
code .ss { color: #183691; }
code .im { color: #000000; }
code .va {color: #008080; }
code .cf { color: #000000; }
code .op { color: #000000; }
code .bu { color: #000000; }
code .ex { color: #000000; }
code .pp { color: #999999; }
code .at { color: #008080; }
code .do { color: #969896; }
code .an { color: #008080; }
code .cv { color: #008080; }
code .in { color: #008080; }
</style>
<style>
body {
  box-sizing: border-box;
  min-width: 200px;
  max-width: 980px;
  margin: 0 auto;
  padding: 45px;
  padding-top: 0px;
}
</style>


</head>

<body>

<h1 id="scdrake">scdrake</h1>
<p><a href="NEWS.md"><img src="https://img.shields.io/badge/NEWS-updates-informational" alt="NEWS: updates" /></a> <a href="https://bioinfocz.github.io/scdrake"><img src="https://img.shields.io/badge/Documentation%20&amp;%20vignettes-bioinfocz.github.io/scdrake-informational" alt="Documentation and vignettes (stable version)" /></a> <a href="https://bioinfocz.github.io/scdrake/dev"><img src="https://img.shields.io/badge/Documentation%20&amp;%20vignettes-bioinfocz.github.io/scdrake-informational" alt="Documentation and vignettes (devel version)" /></a> <a href="https://bioinfocz.github.io/scdrake/articles/pipeline_overview.html"><img src="https://img.shields.io/badge/Overview%20&amp;%20outputs-vignette(%22pipeline_overview%22)-informational" alt="Overview and outputs" /></a> <a href="https://github.com/bioinfocz/scdrake/blob/main/diagrams/README.md"><img src="https://img.shields.io/badge/Pipeline%20diagram-Show-informational" alt="Pipeline diagram" /></a> <img src="https://img.shields.io/github/license/bioinfocz/scdrake" alt="License" /> <a href="https://lifecycle.r-lib.org/articles/stages.html#stable"><img src="https://img.shields.io/badge/lifecycle-stable-green.svg" alt="Lifecycle: experimental" /></a> <a href="https://github.com/bioinfocz/scdrake/actions/workflows/docker-ci.yml"><img src="https://github.com/bioinfocz/scdrake/actions/workflows/docker-ci.yml/badge.svg?branch=main" alt="Docker Image CI" /></a></p>
<p><code>{scdrake}</code> is a scalable and reproducible pipeline for
secondary analysis of droplet-based single-cell RNA-seq data (scRNA-seq)
and spot-based spatial transcriptomics data (SRT).
<code>{scdrake}</code> is an R package built on top of the
<code>{drake}</code> package, a <a href="https://www.gnu.org/software/make">Make</a>-like pipeline toolkit
for <a href="https://www.r-project.org">R language</a>.</p>
<p>The main features of the <code>{scdrake}</code> pipeline are:</p>
<ul>
<li>Import of scRNA-seq data: <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">10x
Genomics Cell Ranger</a> output, delimited table, or
<code>SingleCellExperiment</code> object.</li>
<li>Import of SRT data: <a href="https://www.10xgenomics.com/support/software/space-ranger/latest/getting-started/what-is-space-ranger">10x
Genomics Space Ranger</a> output, delimited table, or
<code>SingleCellExperiment</code> object, and tissue positions file as
in Space ranger.</li>
<li>Quality control and filtering of cells/spots and genes, removal of
empty droplets.</li>
<li>Higly variable genes detection, cell cycle scoring, normalization,
clustering, and dimensionality reduction.</li>
<li>Spatially variable genes detection (for SRT data)</li>
<li>Cell type annotation using reference sets, cell type annotation
using user-provided marker genes.</li>
<li>Integration of multiple datasets.</li>
<li>Computation of cluster markers and differentially expressed genes
between clusters (denoted as “contrasts”).</li>
<li>Rich graphical and HTML outputs based on customizable RMarkdown
documents.
<ul>
<li>You can find links to example outputs <a href="https://bioinfocz.github.io/scdrake/articles/pipeline_overview.html">here</a>.</li>
</ul></li>
<li>Thanks to <code>{drake}</code>, the pipeline is highly efficient,
scalable and reproducible, and also extendable.
<ul>
<li>Want to change some parameter? No problem! Only parts of the
pipeline which changed will rerun, while up-to-date ones will be
skipped.</li>
<li>Want to reuse the intermediate results for your own analyses? No
problem! The pipeline has smartly defined checkpoints which can be
loaded from a <code>{drake}</code> cache.</li>
<li>Want to extend the pipeline? No problem! The pipeline definition is
just an R object which can be arbitrarily extended.</li>
</ul></li>
</ul>
<p>For whom is <code>{scdrake}</code> purposed? It is primarily intended
for tech-savvy users (bioinformaticians), who pass on the results
(reports, images) to non-technical persons (biologists). At the same
time, bioinformaticians can quickly react to biologists’ needs by
changing the parameters of the pipeline, which then efficiently skips
already finished parts. This dialogue between the biologist and the
bioinformatician is indispensable during scRNA-seq data analysis.
<code>{scdrake}</code> ensures that this communication is performed in
an effective and reproducible manner.</p>
<p>The pipeline structure along with <a href="https://github.com/bioinfocz/scdrake/blob/main/diagrams/README.md">diagrams</a>
and links to outputs is described in
<code>vignette(&quot;pipeline_overview&quot;)</code> (<a href="https://bioinfocz.github.io/scdrake/articles/pipeline_overview.html">link</a>).</p>
<p>If you use <code>{scdrake}</code> in your research, please, consider
citing</p>
<blockquote>
<p>Kubovciak J, Kolar M, Novotny J (2023). “Scdrake: a reproducible and
scalable pipeline for scRNA-seq data analysis.” <em>Bioinformatics
Advances</em>, <strong>3</strong>(1). <a href="https://doi.org/10.1093/bioadv/vbad089">doi:10.1093/bioadv/vbad089</a>.</p>
</blockquote>
<p>Huge thanks go to the authors of the <a href="https://bioconductor.org/books/3.15/OSCA">Orchestrating
Single-Cell Analysis with Bioconductor</a> book on whose methods and
recommendations is <code>{scdrake}</code> largely based.</p>
<hr />
<h1 id="installation-instructions">Installation instructions</h1>
<h2 id="using-a-docker-image-recommended">Using a Docker image
(recommended)</h2>
<p>A Docker image based on the <a href="https://bioconductor.org/help/docker/">official Bioconductor
image</a> (version 3.15) is available. This is the most handy and
reproducible way how to use <code>{scdrake}</code> as all the
dependencies are already installed and their versions are fixed. In
addition, the parent Bioconductor image comes bundled with RStudio
Server.</p>
<p>The complete guide to the usage of <code>{scdrake}</code>’s Docker
image can be found in the <a href="https://bioinfocz.github.io/scdrake/articles/scdrake_docker.html">Docker
vignette</a>. <strong>We strongly recommend to go through even if you
are an experienced Docker user.</strong> Below you can find just the
basic command to download the image and to run a detached container with
RStudio in Docker or to run <code>{scdrake}</code> in Singularity.</p>
<p>You can also run the image in <a href="https://docs.sylabs.io/guides/latest/user-guide/quick_start.html">SingularityCE</a>
(without RStudio) - see the Singularity section in the Docker vignette
above. If the image is already downloaded in the local Docker storage,
you can use
<code>singularity pull docker-daemon:&lt;image&gt;</code></p>
<p>You can pull the Docker image with the latest stable
<code>{scdrake}</code> version using</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull jirinovo/scdrake:1.6.0</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">singularity</span> pull docker:jirinovo/scdrake:1.6.0</span></code></pre></div>
<p>or list available versions in <a href="https://hub.docker.com/r/jirinovo/scdrake/tags">our Docker Hub
repository</a>.</p>
<p>For the latest development version use</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull jirinovo/scdrake:latest</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">singularity</span> pull docker:jirinovo/scdrake:latest</span></code></pre></div>
<p><strong>Note for Mac users with M1/M2 chipsets</strong>: until
version 1.5.0 (inclusive), <code>arm64</code> images are available.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull jirinovo/scdrake:1.5.0-bioc3.15-arm64</span></code></pre></div>
<h3 id="running-the-container">Running the container</h3>
<p>For the most common cases of host machines: Linux running Docker
Engine, and Windows or MacOS running Docker Desktop.</p>
<p>First make a shared directory that will be mounted to the
container:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> ~/scdrake_projects</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/scdrake_projects</span></code></pre></div>
<p>And run the image that will expose RStudio Server on port 8787 on
your host:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-d</span> <span class="dt">\</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">-v</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>:/home/rstudio/scdrake_projects <span class="dt">\</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-p</span> 8787:8787 <span class="dt">\</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-e</span> USERID=<span class="va">$(</span><span class="fu">id</span> <span class="at">-u</span><span class="va">)</span> <span class="dt">\</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-e</span> GROUPID=<span class="va">$(</span><span class="fu">id</span> <span class="at">-g</span><span class="va">)</span> <span class="dt">\</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">-e</span> PASSWORD=1234 <span class="dt">\</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  jirinovo/scdrake:1.6.0</span></code></pre></div>
<p>For Singularity, also make shared directories and execute the
container (“run and forget”):</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/scdrake_singularity</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/scdrake_singularity</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> home/<span class="va">${USER}</span> scdrake_projects</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="ex">singularity</span> exec <span class="dt">\</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">-e</span> <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--no-home</span> <span class="dt">\</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--bind</span> <span class="st">&quot;home/</span><span class="va">${USER}</span><span class="st">/:/home/</span><span class="va">${USER}</span><span class="st">,scdrake_projects/:/home/</span><span class="va">${USER}</span><span class="st">/scdrake_projects&quot;</span> <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--pwd</span> <span class="st">&quot;/home/</span><span class="va">${USER}</span><span class="st">/scdrake_projects&quot;</span> <span class="dt">\</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    path/to/scdrake_image.sif <span class="dt">\</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    scdrake <span class="op">&lt;</span>args<span class="op">&gt;</span> <span class="op">&lt;</span>command<span class="op">&gt;</span></span></code></pre></div>
<h2 id="installing-scdrake-manually-not-recommended">Installing
<code>{scdrake}</code> manually (not recommended)</h2>
<details>
<summary>
Click for details
</summary>

<h3 id="install-the-required-system-packages">Install the required
system packages</h3>
<ul>
<li>For Linux, follow the commands for your distribution <a href="required_libs_linux.md">here</a>.</li>
<li>For MacOS:
<code>$ brew install libxml2 imagemagick@6 harfbuzz fribidi libgit2 geos pandoc</code></li>
</ul>
<h3 id="install-r--42">Install R &gt;= 4.2</h3>
<p>See <a href="https://cloud.r-project.org/">https://cloud.r-project.org/</a></p>
<p>From now on, all commands are for R.</p>
<h3 id="install-renv">Install <code>{renv}</code></h3>
<p><a href="https://rstudio.github.io/renv/"><code>{renv}</code></a> is
an R package for management of local R libraries. It is intended to be
used on a per-project basis, i.e. each project should use its own
library of R packages.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;renv&quot;</span>)</span></code></pre></div>
<h3 id="initialize-a-new-renv-library">Initialize a new
<code>{renv}</code> library</h3>
<p>Switch to directory where you will analyze data and initialize a new
<code>{renv}</code> library:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>renv<span class="sc">::</span><span class="fu">consent</span>(<span class="cn">TRUE</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>renv<span class="sc">::</span><span class="fu">init</span>()</span></code></pre></div>
<p>Now exit and run again R. You should see a message that renv library
has been activated.</p>
<h3 id="install-biocmanager">Install BiocManager</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>renv<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;BiocManager&quot;</span>)</span></code></pre></div>
<h3 id="install-bioconductor-315">Install Bioconductor 3.15</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="at">version =</span> <span class="st">&quot;3.15&quot;</span>)</span></code></pre></div>
<h3 id="restore-scdrake-dependencies-from-lockfile">Restore
<code>{scdrake}</code> dependencies from lockfile</h3>
<p><code>{renv}</code> also allows to export the current installed
versions of R packages (and other things) into a lockfile. Such lockfile
is available for <code>{scdrake}</code> and you can use it to install
all dependencies by</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="do">## -- This is a lockfile for the latest stable version of scdrake.</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">&quot;https://raw.githubusercontent.com/bioinfocz/scdrake/1.6.0/renv.lock&quot;</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="do">## -- You can increase the number of CPU cores to speed up the installation.</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">Ncpus =</span> <span class="dv">2</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>renv<span class="sc">::</span><span class="fu">restore</span>(<span class="at">lockfile =</span> <span class="st">&quot;renv.lock&quot;</span>, <span class="at">repos =</span> BiocManager<span class="sc">::</span><span class="fu">repositories</span>())</span></code></pre></div>
<p>For the lockfile for the latest development version use</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">&quot;https://raw.githubusercontent.com/bioinfocz/scdrake/main/renv.lock&quot;</span>)</span></code></pre></div>
<h3 id="install-the-scdrake-package">Install the <code>{scdrake}</code>
package</h3>
<p>Now we can finally install the <code>{scdrake}</code> package, but
using a non-standard approach - without its dependencies (which are
already installed from the lockfile).</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;bioinfocz/scdrake@1.6.0&quot;</span>,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">dependencies =</span> <span class="cn">FALSE</span>, <span class="at">upgrade =</span> <span class="cn">FALSE</span>,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">keep_source =</span> <span class="cn">TRUE</span>, <span class="at">build_vignettes =</span> <span class="cn">TRUE</span>,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">repos =</span> BiocManager<span class="sc">::</span><span class="fu">repositories</span>()</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>For the latest development version use
<code>&quot;bioinfocz/scdrake&quot;</code>.</p>
<h3 id="install-the-command-line-interface-cli">Install the command line
interface (CLI)</h3>
<p>Optionally, you can install <code>{scdrake}</code>’s CLI scripts
with</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>scdrake<span class="sc">::</span><span class="fu">install_cli</span>()</span></code></pre></div>
<p>CLI should be now accessible as a <code>scdrake</code> command. By
default, the CLI is installed into <code>~/.local/bin</code>, which is
usually present in the <code>PATH</code> environment variable. In case
it isn’t, just add to your <code>~/.bashrc</code>:
<code>export PATH=&quot;${HOME}/.local/bin:${PATH}&quot;</code></p>
<p><strong>Every time you will be using the CLI make sure your current
working directory is inside an <code>{renv}</code> project.</strong> You
can read the reasons below.</p>
<details>
<summary>
Show details
</summary>

<p>You might notice that a per-project <code>{renv}</code> library and
an installed CLI are “disconnected” and if you install
<code>{scdrake}</code> and its CLI within multiple projects
(<code>{renv}</code> libraries), then the CLI scripts in
<code>~/.local/bin</code> will be overwritten each time. But when you
run the <code>scdrake</code> command inside an <code>{renv}</code>
project, the <code>renv</code> directory is automatically detected and
the <code>{renv}</code> library is activated by
<code>renv::load()</code>, so the proper, locally installed
<code>{scdrake}</code> package is then used.</p>
<p>Also, there is a built-in guard: the version of the CLI must match
the version of the bundled CLI scripts inside the installed
<code>{scdrake}</code> package. Anyway, we think changes in the CLI
won’t be very frequent, so this shouldn’t be a problem most of the
time.</p>
</details>

<blockquote>
<p>TIP: To save time and space, you can symlink the
<code>renv/library</code> directory to multiple <code>{scdrake}</code>
projects.</p>
</blockquote>
</details>

<hr />
<h2 id="quickstart">Quickstart</h2>
<p>First run the <code>scdrake</code> image in Docker or Singularity -
see the <a href="https://bioinfocz.github.io/scdrake/articles/scdrake_docker.html">Docker
vignette</a></p>
<p>Then you can go through the <a href="https://bioinfocz.github.io/scdrake/articles/scdrake.html">Get
Started vignette</a></p>
<hr />
<h2 id="vignettes-and-other-readings">Vignettes and other readings</h2>
<p>See <a href="https://bioinfocz.github.io/scdrake">https://bioinfocz.github.io/scdrake</a>
for a documentation website of the latest stable version (1.6.0) where
links to vignettes below become real :-)</p>
<p>See <a href="https://bioinfocz.github.io/scdrake/dev">https://bioinfocz.github.io/scdrake/dev</a>
for a documentation website of the current development version.</p>
<ul>
<li>Guides:
<ul>
<li>Using the Docker image: <a href="https://bioinfocz.github.io/scdrake/articles/scdrake_docker.html">https://bioinfocz.github.io/scdrake/articles/scdrake_docker.html</a>
(or <code>vignette(&quot;scdrake_docker&quot;)</code>)</li>
<li>01 Quick start (single-sample pipeline):
<code>vignette(&quot;scdrake&quot;)</code></li>
<li>02 Integration pipeline guide:
<code>vignette(&quot;scdrake_integration&quot;)</code></li>
<li>Advanced topics: <code>vignette(&quot;scdrake_advanced&quot;)</code></li>
<li>Extending the pipeline: <code>vignette(&quot;scdrake_extend&quot;)</code></li>
<li><code>{drake}</code> basics: <code>vignette(&quot;drake_basics&quot;)</code>
<ul>
<li>Or the official <code>{drake}</code> book: <a href="https://books.ropensci.org/drake/">https://books.ropensci.org/drake/</a></li>
</ul></li>
</ul></li>
<li>General information:
<ul>
<li>Pipeline overview: <code>vignette(&quot;pipeline_overview&quot;)</code></li>
<li>FAQ &amp; Howtos: <code>vignette(&quot;scdrake_faq&quot;)</code></li>
<li>Spatial extension: <code>vignette(&quot;scdrake_spatial&quot;)</code></li>
<li>Command line interface (CLI):
<code>vignette(&quot;scdrake_cli&quot;)</code></li>
<li>Config files (internals):
<code>vignette(&quot;scdrake_config&quot;)</code></li>
<li>Environment variables: <code>vignette(&quot;scdrake_envvars&quot;)</code></li>
</ul></li>
<li>General configs:
<ul>
<li>Pipeline config -&gt; <code>vignette(&quot;config_pipeline&quot;)</code></li>
<li>Main config -&gt; <code>vignette(&quot;config_main&quot;)</code></li>
</ul></li>
<li>Pipelines and stages:
<ul>
<li>Single-sample pipeline:
<ul>
<li>Stage <code>01_input_qc</code>: reading in data, filtering, quality
control -&gt; <code>vignette(&quot;stage_input_qc&quot;)</code></li>
<li>Stage <code>02_norm_clustering</code>: normalization, HVG selection,
SVG selection, dimensionality reduction, clustering, (marker-based) cell
type annotation -&gt;
<code>vignette(&quot;stage_norm_clustering&quot;)</code></li>
</ul></li>
<li>Integration pipeline:
<ul>
<li>Stage <code>01_integration</code>: reading in data and integration
-&gt; <code>vignette(&quot;stage_integration&quot;)</code></li>
<li>Stage <code>02_int_clustering</code>: post-integration clustering
and cell annotation -&gt;
<code>vignette(&quot;stage_int_clustering&quot;)</code></li>
</ul></li>
<li>Common stages:
<ul>
<li>Stage <code>cluster_markers</code> -&gt;
<code>vignette(&quot;stage_cluster_markers&quot;)</code></li>
<li>Stage <code>contrasts</code> (differential expression) -&gt;
<code>vignette(&quot;stage_contrasts&quot;)</code></li>
</ul></li>
</ul></li>
</ul>
<p>We encourage all users to read <a href="https://books.ropensci.org/drake">basics</a> of the
<code>{drake}</code> package. While it is not necessary to know all
<code>{drake}</code> internals to successfully run the
<code>{scdrake}</code> pipeline, its knowledge is a plus. You can read
the minimum basics in <code>vignette(&quot;drake_basics&quot;)</code>.</p>
<p>Also, the prior knowledge of Bioconductor and its classes (especially
the <a href="https://bioconductor.org/packages/3.15/bioc/html/SingleCellExperiment.html">SingleCellExperiment</a>)
is considerable.</p>
<hr />
<h2 id="citation">Citation</h2>
<p>Below is the citation output from using
<code>citation(&quot;scdrake&quot;)</code> in R. Please run this yourself to check
for any updates on how to cite <strong>scdrake</strong>.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">citation</span>(<span class="st">&quot;scdrake&quot;</span>), <span class="at">bibtex =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>To cite package ‘scdrake’ in publications use:

  Jiri Novotny and Jan Kubovciak (2021). scdrake: A Pipeline For 10x Chromium Single-Cell RNA-seq Data Analysis.
  https://github.com/bioinfocz/scdrake, https://bioinfocz.github.io/scdrake.

A BibTeX entry for LaTeX users is

  @Manual{,
    title = {scdrake: A Pipeline For 10x Chromium Single-Cell RNA-seq Data Analysis},
    author = {Jiri Novotny and Jan Kubovciak},
    year = {2021},
    note = {https://github.com/bioinfocz/scdrake, https://bioinfocz.github.io/scdrake},
  }</code></pre>
<p>Please note that the <code>{scdrake}</code> was only made possible
thanks to many other R and bioinformatics software authors, which are
cited either in the vignettes and/or the paper(s) describing this
package.</p>
<h2 id="help-and-support">Help and support</h2>
<p>In case of any problems or suggestions, please, open a new <a href="https://github.com/bioinfocz/scdrake/issues">issue</a>. We will be
happy to answer your questions, integrate new ideas, or resolve any
problems <span class="emoji" data-emoji="blush">😊</span></p>
<p>You can also use <a href="https://github.com/bioinfocz/scdrake/discussions">GitHub
Discussions</a>, mainly for topics <strong>not</strong> related to
development (bugs, feature requests etc.), but if you need e.g. a
general help.</p>
<h2 id="contribution">Contribution</h2>
<p>If you want to contribute to <code>{scdrake}</code>, read the <a href=".github/CONTRIBUTING.md">contribution guide</a>, please. All pull
requests are welcome! <span class="emoji" data-emoji="slightly_smiling_face">🙂</span></p>
<h2 id="code-of-conduct">Code of Conduct</h2>
<p>Please note that the <code>{scdrake}</code> project is released with
a <a href="https://bioinfocz.github.io/scdrake/CODE_OF_CONDUCT.html">Contributor
Code of Conduct</a>. By contributing to this project, you agree to abide
by its terms.</p>
<h2 id="acknowledgements">Acknowledgements</h2>
<h3 id="funding">Funding</h3>
<p>This work was supported by <a href="https://www.elixir-czech.cz">ELIXIR CZ</a> research infrastructure
project (MEYS Grant No: LM2018131 and LM2023055) including access to
computing and storage facilities.</p>
<h3 id="software-and-methods-used-by-scdrake">Software and methods used
by <code>{scdrake}</code></h3>
<p>Many things are used by <code>{scdrake}</code>, but these are really
worth mentioning:</p>
<ul>
<li>The <a href="https://www.bioconductor.org">Bioconductor</a>
ecosystem.</li>
<li>The <a href="https://bioconductor.org/books/3.15/OSCA"><em>Orchestrating
Single-Cell Analysis with Bioconductor</em></a> book.</li>
<li>The <a href="https://bioconductor.org/packages/3.15/bioc/html/scran.html">scran</a>,
<a href="https://bioconductor.org/packages/3.15/bioc/html/scater.html">scater</a>,
and other great packages from <a href="https://orcid.org/0000-0002-3564-4813">Aaron Lun</a> et al.</li>
<li>The <a href="https://github.com/ropensci/drake">drake</a>
package.</li>
<li>The <a href="https://github.com/rstudio/rmarkdown">rmarkdown</a>
package, and other ones from the <a href="https://www.tidyverse.org">tidyverse</a> ecosystem.</li>
</ul>
<h3 id="development-tools">Development tools</h3>
<ul>
<li>Continuous code testing is possible thanks to <a href="https://github.com/features/actions">GitHub Actions</a> through
<code>{usethis}</code>, <code>{remotes}</code>, and
<code>{rcmdcheck}</code>. Customized to use <a href="https://www.bioconductor.org/help/docker">Bioconductor’s docker
containers</a>.</li>
<li>The <a href="https://bioinfocz.github.io/scdrake">documentation
website</a> is generated by <code>{pkgdown}</code>.</li>
<li>The code is styled automatically thanks to
<code>{styler}</code>.</li>
<li>The documentation is formatted thanks to <code>{devtools}</code> and
<code>{roxygen2}</code>.</li>
</ul>
<p>This package was developed using <code>{biocthis}</code>.</p>

</body>
</html>
