<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light"><script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/5/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>scdrake TODO list • scdrake</title><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="scdrake TODO list"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/5/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)"><!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/5/darkly/bootstrap.css" media="(prefers-color-scheme: dark)"><!-- preferably CSS --><link rel="stylesheet" href="preferably.css"><link id="css-code-light" rel="stylesheet" href="code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)"><link id="css-code-dark" rel="stylesheet" href="code-color-scheme-dark.css" media="(prefers-color-scheme: dark)"><script src="darkswitch.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">scdrake</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="articles/scdrake.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="articles/scdrake_integration.html">Integration pipeline guide</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="articles/scdrake_faq.html">FAQ &amp; Howtos</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Guides</h6>
    <a class="dropdown-item" href="articles/scdrake_docker.html">Using the Docker image</a>
    <a class="dropdown-item" href="articles/scdrake_integration.html">02 Integration pipeline guide</a>
    <a class="dropdown-item" href="articles/scdrake_advanced.html">Advanced topics</a>
    <a class="dropdown-item" href="articles/scdrake_extend.html">Extending the pipeline</a>
    <a class="dropdown-item" href="articles/drake_basics.html">drake basics</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General information</h6>
    <a class="dropdown-item" href="articles/pipeline_overview.html">Pipeline overview</a>
    <a class="dropdown-item" href="articles/scdrake_faq.html">FAQ &amp; Howtos</a>
    <a class="dropdown-item" href="articles/scdrake_cli.html">Command line interface</a>
    <a class="dropdown-item" href="articles/scdrake_config.html">scdrake config internals</a>
    <a class="dropdown-item" href="articles/scdrake_envvars.html">Environment variables</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General configs</h6>
    <a class="dropdown-item" href="articles/config_pipeline.html">Pipeline config</a>
    <a class="dropdown-item" href="articles/config_main.html">Main config</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single-sample pipeline stages</h6>
    <a class="dropdown-item" href="articles/stage_input_qc.html">Input and quality control stage (`01_input_qc`)</a>
    <a class="dropdown-item" href="articles/stage_norm_clustering.html">Normalization and clustering stage (`02_norm_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Integration pipeline stages</h6>
    <a class="dropdown-item" href="articles/stage_integration.html">Integration stage (`01_integration`)</a>
    <a class="dropdown-item" href="articles/stage_int_clustering.html">Post-integration clustering stage (`02_int_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Common stages</h6>
    <a class="dropdown-item" href="articles/stage_cluster_markers.html">Cluster markers stage (`cluster_markers`)</a>
    <a class="dropdown-item" href="articles/stage_contrasts.html">Contrasts stage (`contrasts`)</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bioinfocz/scdrake" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>

        
        <li>
          <a class="external-link nav-link" id="css-toggle-btn" aria-label="github">
            <span class="fas fa fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        
        
      </ul></div>

    
  </div>
</nav><div class="container template-title-body">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>scdrake TODO list</h1>
      <small class="dont-index">Source: <a href="https://github.com/bioinfocz/scdrake/blob/masterTODO.md" class="external-link"><code>TODO.md</code></a></small>
    </div>


<div id="scdrake-todo-list" class="section level1">

</div>
<div class="section level1">
<h1 id="quick-fixes">Quick fixes<a class="anchor" aria-label="anchor" href="#quick-fixes"></a></h1>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
Set <code>SC3_N_CORES</code> to 1 by default.</li>
<li>
<input type="checkbox" disabled checked>
Download <code>yq_linux_arm64</code> for Mac M1.</li>
<li>
<input type="checkbox" disabled checked>
Fix <a href="https://github.com/bioinfocz/scdrake/blob/main/R/project.R#L127" class="external-link uri">https://github.com/bioinfocz/scdrake/blob/main/R/project.R#L127</a>
</li>
<li>
<input type="checkbox" disabled checked>
Use <code>rds</code> as the default storage format.</li>
<li>
<input type="checkbox" disabled checked>
More options for drake (e.g. for memory to support low end machines).</li>
<li>
<input type="checkbox" disabled checked>
Rewrite docs for Docker:<ul class="task-list"><li>
<input type="checkbox" disabled checked>
Docker Desktop vs. Docker Engine</li>
<li>
<input type="checkbox" disabled checked>
Linux vs. Mac and Windows</li>
<li>
<input type="checkbox" disabled checked>
Running scdrake without starting a detached container (<code>docker run -it -v ...</code>).</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
FAQ docs:<ul class="task-list"><li>
<input type="checkbox" disabled checked>
Add notes about:</li>
</ul><pre><code>Error in `get_result(output = out, options)`:
! callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed
ℹ See `$stdout` and `$stderr` for standard output and error.
Type .Last.error to see the more details.</code></pre>
In DD, increase memory (Settings -&gt; Resources).</li>
<li>
<input type="checkbox" disabled checked>
Test with <code>SCTransform</code>.<ul><li>It is failing during tests, don’t know why.</li>
</ul></li>
<li>
<input type="checkbox" disabled>
Use package version in <code>DESCRIPTION</code> for the latest commit as <code>x.x.9000</code> (until the next release).<ul><li>This should also trigger <code>pkgdown</code> build of the development version -&gt; <a href="https://pkgdown.r-lib.org/reference/build_site.html?#setting-development-mode" class="external-link uri">https://pkgdown.r-lib.org/reference/build_site.html?#setting-development-mode</a>
</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
Suppress warnings from <code>DT</code>.</li>
<li>
<input type="checkbox" disabled>
Reorganize <code>inst/</code> into <code>project/</code> etc.</li>
<li>
<input type="checkbox" disabled checked><del>Update <code><a href="reference/init_project.html">init_project()</a></code> and <code><a href="reference/update_project.html">update_project()</a></code></del><ul class="task-list"><li>
<input type="checkbox" disabled checked>
The script will be distributed through a package function <code><a href="reference/cli.html">install_cli()</a></code>.</li>
</ul></li>
</ul><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/copy.html" class="external-link">file_copy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"scdrake"</span>, package <span class="op">=</span> <span class="st">"scdrake"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="va">dir</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/file_chmod.html" class="external-link">file_chmod</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"scdrake"</span><span class="op">)</span>, <span class="st">"+x"</span><span class="op">)</span></span>
<span><span class="va">verbose</span> <span class="op"><a href="reference/grapes-and-and-grapes.html">%&amp;&amp;%</a></span> <span class="fu">cli_alert_success</span><span class="op">(</span><span class="st">"Copied the {.file scdrake} script file."</span><span class="op">)</span></span></code></pre></div>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
Lockfile: only store scdrake deps, not scdrake itself.<ul class="task-list"><li>
<input type="checkbox" disabled>
Local install as:</li>
</ul></li>
</ul><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## -- Replace &lt;version&gt; with scdrake version.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/bioinfocz/scdrake/&lt;version&gt;/renv.lock"</span><span class="op">)</span></span>
<span><span class="co">## -- For Rhtslib this error appears during renv::restore(): /usr/bin/tar: Unexpected EOF in archive</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"Rhtslib"</span><span class="op">)</span></span>
<span><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/consent.html" class="external-link">consent</a></span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/restore.html" class="external-link">restore</a></span><span class="op">(</span>lockfile <span class="op">=</span> <span class="st">"renv.lock"</span>, prompt <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"bioinfocz/scdrake@&lt;version&gt;"</span>, dependencies <span class="op">=</span> <span class="cn">FALSE</span>, upgrade <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
Add to installation instructions the required libs.</li>
<li>
<input type="checkbox" disabled checked>
Report 01_input_qc Rmd -&gt; fix line 96: <code>h = cfg$EMPTY_DROPLETS_LOWER</code>
</li>
<li>
<input type="checkbox" disabled>
Change <code>.data$var</code> to <code>"var"</code> in tidyverse functions.</li>
<li>
<input type="checkbox" disabled>
Fix compatibility with SingleR 2.0: remove lines 186 and 204 in <code>R/cell_annotation.R</code><ul><li>Not yet - errors in Bioc 3.16 (Matrix package).</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
Remove exclamation marks from scdrake load messages (env variables) -&gt; change to “i”.</li>
<li>
<input type="checkbox" disabled checked>
Test relative paths for <code><a href="reference/init_project.html">init_project()</a></code>
</li>
<li>
<input type="checkbox" disabled checked>
More checks of config parameters:<ul class="task-list"><li>
<input type="checkbox" disabled checked>
pandoc in $PATH</li>
<li>
<input type="checkbox" disabled checked>
input file for 01_input_qc</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
Catch error in emptyDroplets() and display an info message.</li>
<li>
<input type="checkbox" disabled checked>
CLI R script:<ul class="task-list"><li>
<input type="checkbox" disabled checked>
Add <code>SCDRAKE_CLI_VERSION</code> to scdrake package (<code>zzz.R</code>) and check the match inside the script.</li>
<li>
<input type="checkbox" disabled checked>
Rename <code>scdrake-cli.R</code> to <code>scdrake.R</code>.</li>
<li>
<input type="checkbox" disabled checked>
Add confirmations to <code><a href="reference/cli.html">install_cli()</a></code>.</li>
</ul></li>
<li>
<input type="checkbox" disabled>
Docker container<ul class="task-list"><li>
<input type="checkbox" disabled>
Can we also make an image based on Bioc without Rstudio?</li>
<li>
<input type="checkbox" disabled>
Docs:<ul class="task-list"><li>
<input type="checkbox" disabled checked>
How to mount the current directory and use it as the project directory inside the container.</li>
<li>
<input type="checkbox" disabled>
Using Singularity/Apptainer - but without RStudio, which seems to be problematic.</li>
<li>
<input type="checkbox" disabled>
(?) How to permanently store the cache in <code>/home/rstudio/.cache/R</code>.</li>
<li>
<input type="checkbox" disabled checked>
Replace all Bioc versions with variable.</li>
<li>
<input type="checkbox" disabled checked>
In <code>scdrake.Rmd</code> use <code>download_example_data = TRUE</code>.</li>
<li>
<input type="checkbox" disabled checked>
Instruct to use WSL2 on Windows.</li>
</ul></li>
<li>
<input type="checkbox" disabled>
Script to export the analysis as a Docker image:<ul><li>(?) Docker command to save image.</li>
<li>(?) Dockerfile inheriting from the scdrake image + copy the analysis results (scdrake project directory) inside.</li>
</ul></li>
</ul></li>
<li>
<input type="checkbox" disabled>
GitHub Actions:<ul class="task-list"><li>
<input type="checkbox" disabled>
New action to build the Docker image, run R CMD CHECK (CLI script) and push the image to Docker Hub.</li>
<li>
<input type="checkbox" disabled>
Modified current <code>check-bioc.yml</code> action to install deps from <code>renv.lock</code> (basically following the installation instructions using a clean Ubuntu container).</li>
</ul></li>
<li>
<input type="checkbox" disabled>
(?) Add Cerebro.<ul><li>Function <code>start_cerebro(sce = ...)</code> - loads and prepares a SCE object from a stage.</li>
<li>Include results from cluster markers - parameter <code>markers_in</code> or <code>contrast</code> in <code>start_cerebro()</code>.
<ul><li>Or utilize custom tables: <a href="https://romanhaa.github.io/cerebroApp/articles/export_and_visualize_custom_tables_and_plots.html" class="external-link uri">https://romanhaa.github.io/cerebroApp/articles/export_and_visualize_custom_tables_and_plots.html</a>
</li>
</ul></li>
</ul></li>
<li>
<input type="checkbox" disabled>
(?) Export to <a href="https://www.bioconductor.org/packages/release/bioc/html/iSEE.html" class="external-link">iSEE</a>
</li>
</ul></div>
<div class="section level1">
<h1 id="general">General<a class="anchor" aria-label="anchor" href="#general"></a></h1>
<ul class="task-list"><li>
<input type="checkbox" disabled>
check <a href="https://github.com/ktrns/scrnaseq" class="external-link uri">https://github.com/ktrns/scrnaseq</a> - there might be some interesting features worth implementing (mainly in sense of output - plots, statistics, etc.)</li>
<li>
<input type="checkbox" disabled checked>
use functions instead of long target expressions - more reproducible<ul><li>use function names in the format <code>&lt;target_name&gt;_fn()</code> for better readability (in case of single target specific function)</li>
</ul></li>
<li>
<input type="checkbox" disabled>
(?) shorten the names of stages:<ul><li>
<code>input_qc</code> -&gt; <code>qc</code>
</li>
<li>
<code>norm_clustering</code> -&gt; <code>normclust</code>
</li>
<li>
<code>cluster_markers</code> -&gt; <code>markers</code>
</li>
</ul></li>
<li>
<input type="checkbox" disabled>
(?) Rewrite to <a href="https://github.com/ropensci/targets" class="external-link">targets</a>. Unfortunately, “drake is superseded. The targets R package is the long-term successor of drake, and it is more robust and easier to use. Please visit <a href="https://books.ropensci.org/targets/drake.html" class="external-link uri">https://books.ropensci.org/targets/drake.html</a> for full context and advice on transitioning.”</li>
<li>
<input type="checkbox" disabled><a href="https://rmarkdown.rstudio.com/flexdashboard/examples.html" class="external-link">flexdashboard</a> seems like a nice alternative to traditional layout of Rmd documents</li>
</ul></div>
<div class="section level1">
<h1 id="open-source-project">Open-source project<a class="anchor" aria-label="anchor" href="#open-source-project"></a></h1>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
rewrite to R package - see <a href="https://github.com/tiernanmartin/drakepkg" class="external-link">this</a> and <a href="https://github.com/januz/drakepkg" class="external-link">this</a> example, and also <a href="https://books.ropensci.org/drake/projects.html#workflows-as-r-packages" class="external-link">a chapter</a> from <code>drake</code>s manual</li>
<li>
<input type="checkbox" disabled checked>
write documentation<ul><li>
<input type="checkbox" disabled checked>
installation</li>
<li>
<input type="checkbox" disabled checked>
basic usage</li>
<li>
<input type="checkbox" disabled checked>
config variables with examples</li>
<li>
<input type="checkbox" disabled><del>better documentation inside YAML config files</del><ul><li>Configs are now documented in pkgdown.</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
vignettes</li>
<li>
<input type="checkbox" disabled checked>
Roxygen documentation</li>
<li>
<input type="checkbox" disabled checked>
documentation of targets</li>
<li><a href="https://martinctc.github.io/blog/make-package-even-more-awesome/" class="external-link">Top 10 tips to make your R package even more awesome</a></li>
<li>
<input type="checkbox" disabled checked>
GitHub actions<ul class="task-list"><li>
<input type="checkbox" disabled checked><a href="https://github.com/r-lib/actions" class="external-link uri">https://github.com/r-lib/actions</a><ul class="task-list"><li>
<input type="checkbox" disabled checked><a href="https://lcolladotor.github.io/biocthis/" class="external-link uri">https://lcolladotor.github.io/biocthis/</a> - specially for Bioconductor packages</li>
</ul></li>
<li>
<input type="checkbox" disabled><a href="https://github.com/nvuillam/mega-linter" class="external-link uri">https://github.com/nvuillam/mega-linter</a>
</li>
<li>
<input type="checkbox" disabled checked>
automatic <code>pkgdown</code>
</li>
</ul></li>
</ul></li>
<li>
<input type="checkbox" disabled><a href="https://docs.github.com/en/github/building-a-strong-community" class="external-link">Building a strong community</a>
</li>
</ul></div>
<div class="section level1">
<h1 id="docs">Docs<a class="anchor" aria-label="anchor" href="#docs"></a></h1>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
Quick start - only the most important steps.</li>
<li>
<input type="checkbox" disabled checked><del>“Longer start” guide?</del>
</li>
<li>
<input type="checkbox" disabled checked>
Running the pipeline, environment variables<ul class="task-list"><li>
<input type="checkbox" disabled checked>
Make “environment variables” a separate guide</li>
<li>
<input type="checkbox" disabled checked><del>Do we need the rest (“Running the pipeline”)? Maybe just tell differences between <code><a href="reference/run_pipeline.html">run_single_sample()</a></code> and <code><a href="reference/run_pipeline.html">run_single_sample_r()</a></code>.</del> No. :)<ul class="task-list"><li>
<input type="checkbox" disabled checked>
Put the section with plan modifications in a separate vignette (or to FAQ &amp; Howto below).</li>
</ul></li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
FAQ &amp; Howto<ul class="task-list"><li>
<input type="checkbox" disabled checked>
Installation problems -&gt; refer to README.</li>
<li>
<input type="checkbox" disabled checked>
Change the output directory. Base out dir and individual stage’s dirs.</li>
<li>
<input type="checkbox" disabled checked>
Load intermediate results.</li>
<li>
<input type="checkbox" disabled checked>
Modify a plan or use targets in your own plan -&gt; link to vignette.</li>
<li>
<input type="checkbox" disabled checked>
Manual cell annotation: <code>CELL_GROUPINGS</code> or modify <code>SCE</code> object (e.g. Seurat’s <code>CellSelector</code>).</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
Fix links to vignettes.</li>
<li>
<input type="checkbox" disabled checked>
In HTML reports, add package prefixes and use <code><a href="https://downlit.r-lib.org/reference/autolink.html" class="external-link">downlit::autolink()</a></code>.</li>
<li>
<input type="checkbox" disabled checked>
Elaborate pipeline diagram.</li>
<li>
<input type="checkbox" disabled checked><code>scdrake_extend</code> vignette: using stage report Rmds and generate_stage_report()</li>
<li>
<input type="checkbox" disabled checked>
Change <code>DRAKE_REBUILD</code> to either <code>all</code> or <code>current</code> to rebuild all targets (<code>force = TRUE</code>) or invalidate the current ones (<code><a href="https://docs.ropensci.org/drake/reference/clean.html" class="external-link">drake::clean()</a></code>).<ul class="task-list"><li>
<input type="checkbox" disabled checked>
Add to <code>config_pipeline.Rmd</code>.</li>
</ul></li>
<li>
<input type="checkbox" disabled>
New feature: filter input (<code>sce_raw</code>) by barcodes.</li>
</ul></div>
<div class="section level1">
<h1 id="config">Config<a class="anchor" aria-label="anchor" href="#config"></a></h1>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
use YAML format for configs<ul class="task-list"><li>
<input type="checkbox" disabled checked>
save them as <code>&lt;name&gt;.yaml.default</code> and add script/command to merge them with existing configs named <code>&lt;name&gt;.yaml</code>
</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
find the right way to update configs - arrays should be overwritten by local config<ul><li>
<code>yq merge -a overwrite -x *.default *.yml</code> is used</li>
</ul></li>
<li>
<input type="checkbox" disabled checked><del>add parameter to config for not updating configs (prevent overwrite)</del><ul><li><del>warn user when this parameter is False (or interactive y/n to overwrite)</del></li>
<li>comments in local configs are now preserved unless <code>USE_DEFAULT_CONFIG_STRUCTURE: True</code>, which replaces the structure of local configs by default ones (i.e. only values are overwritten by local configs)</li>
</ul></li>
<li>
<input type="checkbox" disabled>
more checks of config variables, similar to those written for <code>CLUSTER_MARKER_SOURCES</code>
</li>
<li>
<input type="checkbox" disabled>
YAML validation:<ul><li>
<a href="http://json-schema.org/" class="external-link uri">http://json-schema.org/</a> and <a href="https://json-schema-everywhere.github.io/" class="external-link uri">https://json-schema-everywhere.github.io/</a>
</li>
<li><a href="https://github.com/neilpa/yajsv" class="external-link uri">https://github.com/neilpa/yajsv</a></li>
<li><a href="https://stackoverflow.com/questions/3262569/validating-a-yaml-document-in-python" class="external-link uri">https://stackoverflow.com/questions/3262569/validating-a-yaml-document-in-python</a></li>
<li><a href="https://docs.python-cerberus.org/en/stable/" class="external-link uri">https://docs.python-cerberus.org/en/stable/</a></li>
<li><a href="http://www.kuwata-lab.com/kwalify/" class="external-link uri">http://www.kuwata-lab.com/kwalify/</a></li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
environment variables with path to pipeline config directory, and single sample or integration config directory</li>
</ul></div>
<div class="section level1">
<h1 id="drake">drake<a class="anchor" aria-label="anchor" href="#drake"></a></h1>
<ul class="task-list"><li>
<input type="checkbox" disabled checked><del>create <code>make.R</code> and <code>_drake.R</code> for main pipeline stages (<code>01_input_qc</code>, etc.)</del><ul><li><del>i.e. <code>make(plan, targets = "report_input_qc", ...)</code></del></li>
<li>targets are now specified in the <code>00_pipeline.yaml</code> config file</li>
</ul></li>
<li>
<input type="checkbox" disabled checked><del>add plan graph widgets to Rmds</del><ul><li>
<code>htmlwidgets</code> are somehow modifying the global env, so <code>drake</code> throws error</li>
<li>actually not useful at all, so there is a separate <code>make_graphs.R</code> for saving graphs to HTML files</li>
</ul></li>
<li>
<input type="checkbox" disabled>
(?) remove the config dependency main_config from report Rmds, so they won’t be generated each time you change the main config<ul class="task-list"><li>
<input type="checkbox" disabled checked>
from main_config move variables which have a direct impact on analysis to separate file (e.g. number of cores is negligible, while seed is not)</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
(?) use bang-bang operator (<code>!!</code>) to force evaluation of config parameters inside target commands</li>
<li>
<input type="checkbox" disabled checked>
remove plan graphs from the main plan or create a separate <code>make_plans.R</code>
</li>
<li>
<input type="checkbox" disabled checked>
use <code><a href="reference/with_plan.html">with_plan()</a></code> with <code><a href="https://future.futureverse.org/reference/sequential.html" class="external-link">future::sequential</a></code> for functions known to use the <code>future</code> parallelism (e.g. some <code>Seurat</code> functions such as <code>FindMarkers()</code>) - <code>drake</code> doesn’t work well for within-target paralellism<ul class="task-list"><li>
<input type="checkbox" disabled>
(?) in config allow to specify <code><a href="https://future.futureverse.org/reference/plan.html" class="external-link">future::plan</a></code> both for drake and within-target parallelism</li>
</ul></li>
<li>
<input type="checkbox" disabled>
more parameters in config for <code>make()</code> and <code>drake_plan()</code>
</li>
</ul></div>
<div class="section level1">
<h1 id="misc">Misc<a class="anchor" aria-label="anchor" href="#misc"></a></h1>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
put Cleanup section in Rmds to collapsible block</li>
<li>
<input type="checkbox" disabled checked>
replace <code>clean_tmpfiles()</code> in RMarkdown (possibly causing failures of parallel rendering targets)</li>
<li>
<input type="checkbox" disabled><del>Consider modifying <code>CLUSTER_MARKERS_SOURCES</code> directly in <code>process_config.R</code> such as parameter groups (for <code>FindAllMarkers()</code> etc.) will be under separate lists in each source (now all params are under <code>params</code> list). This would be clearer than doing that in drake plan.</del><ul><li>
<code>CLUSTER_MARKERS_SOURCES</code> are now parsed to a tibble and defined as a target in the plan.</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
use <a href="https://here.r-lib.org/" class="external-link">here</a> package to refer to files relative to project root (important mainly for Rmds, which call <code><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd()</a></code> to their locations)</li>
<li>
<input type="checkbox" disabled checked>
to generated HTMLs, add footer with link to this pipeline<ul><li>Added as subtitle.</li>
</ul></li>
<li>
<input type="checkbox" disabled>
function to export to <a href="https://github.com/gorgitko/ShinyCell" class="external-link">ShinyCell</a><ul><li>probably not worth to be a target in the pipeline</li>
</ul></li>
<li>
<input type="checkbox" disabled><del>(?) remove namespace prefixes (<code>dplyr::</code> etc.) and manage conflicts with <code>conflicted</code></del>
</li>
<li>
<input type="checkbox" disabled>
instead of <code><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist()</a></code> use proper <code>purrr</code> functions such as <code>map_chr()</code>
</li>
</ul></div>
<div class="section level1">
<h1 id="single-sample-analysis">Single-sample analysis<a class="anchor" aria-label="anchor" href="#single-sample-analysis"></a></h1>
<div class="section level2">
<h2 id="id_01_input_qc">01_input_qc<a class="anchor" aria-label="anchor" href="#id_01_input_qc"></a></h2>
<ul class="task-list"><li>
<input type="checkbox" disabled>
general rewrite<ul><li>create separate pipelines for different input data scenarios:
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
single-sample</li>
<li>
<input type="checkbox" disabled checked>
integration</li>
<li>
<input type="checkbox" disabled checked>
(?) already processed single-sample with barcode filtering (e.g. data are already normalized and we want to only do gene filtering and skip normalization -&gt; directly go to HVG selection, PCA, dimreds, clustering, etc.)<ul><li>Now not a separate pipeline, but a type of input datat.</li>
</ul></li>
</ul></li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
use <code>readd()</code> instead of <code>loadd()</code> of some variables</li>
<li>
<input type="checkbox" disabled>
filtering of cells by barcodes imported from a file<ul><li>in different stages: before QC, before normalization, before clustering, …</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
parameter to skip cell and/or gene filtering</li>
<li>
<input type="checkbox" disabled>
QC plots - color by passed filter</li>
<li>
<input type="checkbox" disabled checked>
configurable usage of specific MAD stats used for dataset-sensitive filtering (e.g. only turn on filtering on total UMI)<ul><li>Can be done by setting e.g. -Inf</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
do not filter cells by fraction of rRNA genes</li>
<li>
<input type="checkbox" disabled checked>
by default do not filter genes by fraction of their expression in cells (set to 0)<ul><li>i.e. genes with captured expression in less than 5 % of cells used to be discarded</li>
</ul></li>
<li>
<input type="checkbox" disabled>
allow more flexibility for annotation<ul class="task-list"><li>
<input type="checkbox" disabled>
import of GTF/GFF3 files</li>
<li>
<input type="checkbox" disabled>
load annotation DB from file</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
import of other formats<ul class="task-list"><li>
<input type="checkbox" disabled checked>
feature-barcode matrix in delimited format, one column for ENSEMBL ID</li>
<li>
<input type="checkbox" disabled checked>
SCE object in Rds format</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_02_norm_clustering">02_norm_clustering<a class="anchor" aria-label="anchor" href="#id_02_norm_clustering"></a></h2>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
use only one selected type of normalization in plan (scran or sctransform specified in config)</li>
<li>
<input type="checkbox" disabled checked>
save dimred plots in PDF and in Rmd output PNG with link to PDF<ul><li>easier handling of dynamic chunk figure size</li>
<li>faster Rmd rendering</li>
<li>handy - people can download PDFs</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
make a simplified Rmd - probably only dimreds</li>
<li>
<input type="checkbox" disabled checked>
add selected marker plots</li>
<li>
<input type="checkbox" disabled>
export to Loupe, see <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/visualization/latest/tutorial-interoperability" class="external-link">this</a> and <a href="https://github.com/satijalab/seurat/issues/1391" class="external-link">this</a><ul><li>make a function for that, bud don’t make a target for that</li>
</ul></li>
<li>
<input type="checkbox" disabled>
import of (user defined) clusters from Loupe - but only if we implement the export to Loupe</li>
<li>
<input type="checkbox" disabled checked><code>Seurat</code> is using natural logarithm for normalized data (but <code>Seurat v4</code> is using log2 for <code>FindMarkers()</code>). It would be good to have log2 scale when plotting expression in <code>seu</code> (scatterplot, violin).<ul class="task-list"><li>
<input type="checkbox" disabled>
(!) check in <code><a href="reference/sce_norm.html">sctransform_normalization()</a></code> whether the transformation to log2 is done correctly</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
allow cluster columns to be character factors, not just integers<ul class="task-list"><li>
<input type="checkbox" disabled checked>
add variable to config to recode clusters, e.g. 1 -&gt; “CD8+”</li>
<li>
<input type="checkbox" disabled checked>
keep original cluster numbers in “<clustering_name>_orig” column</clustering_name></li>
<li>
<input type="checkbox" disabled>
rename clusters based on dimred filters, e.g. UMAP1 &gt; 5 -&gt; “CD8+”, add new clustering column for each filter<ul class="task-list"><li>
<input type="checkbox" disabled><code>CellSelector()</code> could be used instead, see <a href="https://satijalab.org/seurat/v3.2/visualization_vignette.html" class="external-link">this</a>
</li>
</ul></li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
if <code>FORCE_PCS_SCE</code> is set, still save PCs computed by <code><a href="https://rdrr.io/pkg/scran/man/denoisePCA.html" class="external-link">scran::getDenoisedPCs</a></code> to show it later in plot</li>
<li>
<input type="checkbox" disabled checked>
add total UMI and doublet score dimred plots for each dimred (target <code>dimred_plot</code>)<ul class="task-list"><li>
<input type="checkbox" disabled checked>
generalize <code>dimred_plot</code> target (the underlying function) to plot any column in object.<ul><li><del><code>CLUSTERING_NAMES</code> could be then “Phase”, “total”, “batch”, etc.</del></li>
<li>added <code>DIMRED_PLOTS_OTHER</code> dict variable to config and created separated targets for clustering and other variables dimplots (<code>dimred_plot_clustering</code> and <code>dimred_plots_other_vars</code>)</li>
</ul></li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
rewrite <code>dimred_plot</code> to dynamic target, improve dependency handling</li>
<li>
<input type="checkbox" disabled checked>
When cluster renaming is used, <code>seu_final_norm_clustering</code> and <code>sce_final_norm_clustering</code> are updated, which will cause <code>cluster_markers</code> stage targets to be outdated (i.e. cluster markers will be recomputed). The solution would be not to save renamed clusters to objects, but keep them in dynamic targets, which will be passed to marker functions.</li>
<li>
<input type="checkbox" disabled checked>
to remove duplicated code in <code>Rmd/02_norm_clustering_simple</code>, it would be good to isolate this code and use the include chunk</li>
<li>
<input type="checkbox" disabled checked>
remove the addition of defined number of PCs</li>
<li>
<input type="checkbox" disabled checked>
optionally skip normalization</li>
</ul><div class="section level3">
<h3 id="cc-regression">(CC) regression<a class="anchor" aria-label="anchor" href="#cc-regression"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled><del>if <code>CC_REGRESS</code> or <code>VARS_TO_REGRESS</code>, compute dimreds also when they are not used</del>
</li>
<li>
<input type="checkbox" disabled checked>
consider a change of cell-cycle regression for SCE object to CC genes removal strategy described <a href="http://bioconductor.org/books/release/OSCA/cell-cycle-assignment.html#removing-cell-cycle-related-genes" class="external-link">here</a> or modified <a href="https://support.bioconductor.org/p/133154/#133250" class="external-link">here</a><ul class="task-list"><li>
<input type="checkbox" disabled>
(?) remove CC-related genes prior to sctransform instead of their regression out</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="hvg-selection">HVG selection<a class="anchor" aria-label="anchor" href="#hvg-selection"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled>
(?) allow to use blocking on cell cycle in <code>modelGeneVar()</code> and <code>modelGeneCV2()</code>
</li>
<li>
<input type="checkbox" disabled>
add parameter to exclude certain genes from HVG selection</li>
</ul></div>
<div class="section level3">
<h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
revision of implemented clustering algorithms<ul class="task-list"><li>
<input type="checkbox" disabled>
are we using the right algorithms? E.g. is k-means necessary when we have SC3? (But which is also much slower.)</li>
<li>
<input type="checkbox" disabled>
are there any new algorithms?</li>
</ul></li>
<li>
<input type="checkbox" disabled>
check SC3 gene filter param and function for optimal number of clusters (k)</li>
<li>
<input type="checkbox" disabled>
(?) add <a href="http://bioconductor.org/books/release/OSCA/clustering.html#general-purpose-cluster-diagnostics" class="external-link">clustering diagnostics</a><ul class="task-list"><li>
<input type="checkbox" disabled><a href="https://cran.r-project.org/web/packages/clustree/vignettes/clustree.html" class="external-link">clustree</a>
</li>
<li>
<input type="checkbox" disabled><a href="https://github.com/NHLBI-BCB/IKAP" class="external-link">IKAP – Identifying K mAjor cell Population groups in scRNA-seq analysis</a>
</li>
</ul></li>
<li>
<input type="checkbox" disabled><a href="http://bioconductor.org/books/release/OSCA/clustering.html#subclustering" class="external-link">subclustering</a>
</li>
<li>
<input type="checkbox" disabled>
(?) <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9743492/" class="external-link">SC3s</a><ul><li>Python implementation…</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="cell-type-recognition">Cell type recognition<a class="anchor" aria-label="anchor" href="#cell-type-recognition"></a></h3>
<ul><li>
<a href="https://github.com/seandavi/awesome-single-cell#cell-type-identification-and-classification" class="external-link">List of cell type annotation tools</a>
<ul><li>
<a href="https://bioconductor.org/packages/release/bioc/html/SingleR.html" class="external-link">x</a>
<ul><li><a href="http://bioconductor.org/books/release/SingleRBook/" class="external-link uri">http://bioconductor.org/books/release/SingleRBook/</a></li>
<li><a href="https://github.com/LTLA/SingleR" class="external-link uri">https://github.com/LTLA/SingleR</a></li>
</ul></li>
<li><a href="https://github.com/irrationone/cellassign/" class="external-link">cellassign</a></li>
<li><a href="https://github.com/ZJUFanLab/scCATCH" class="external-link">ssCATCH</a></li>
</ul></li>
<li>
<a href="https://gitlab.com/sysbiobig/scseqcomm" class="external-link">scSeqComm</a> - analysis of intercellular and intracellular signaling.</li>
</ul></div>
<div class="section level3">
<h3 id="ligand-receptor-interactions">Ligand-receptor interactions<a class="anchor" aria-label="anchor" href="#ligand-receptor-interactions"></a></h3>
<ul><li>
<a href="https://www.cellphonedb.org/" class="external-link">CellPhoneDB</a> - interactions between clusters</li>
</ul></div>
</div>
</div>
<div class="section level1">
<h1 id="markers">Markers<a class="anchor" aria-label="anchor" href="#markers"></a></h1>
<ul class="task-list"><li>
<input type="checkbox" disabled><a href="https://github.com/theislab/scCODA" class="external-link">scCODA</a> - identification of compositional changes in high-throughput sequencing count data, especially cell compositions from scRNA-seq</li>
</ul><div class="section level2">
<h2 id="cluster_markers">cluster_markers<a class="anchor" aria-label="anchor" href="#cluster_markers"></a></h2>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
rewrite to scran, reusable for DEA</li>
<li>
<input type="checkbox" disabled>
global markers<ul class="task-list"><li>
<input type="checkbox" disabled checked>
cluster markers computation</li>
<li>
<input type="checkbox" disabled checked>
heatmaps</li>
<li>
<input type="checkbox" disabled checked>
marker plots</li>
<li>
<input type="checkbox" disabled checked>
marker table reports</li>
<li>
<input type="checkbox" disabled checked>
summary report (index page)</li>
<li>
<input type="checkbox" disabled>
GSEA</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
add marker method parameters to output Rmd (e.g. <code>cluster_sce_kmeans_kc_3</code> -&gt; <code>lfc_cutoff: 0.5</code>)<ul class="task-list"><li>
<input type="checkbox" disabled checked>
added to marker table report</li>
</ul></li>
<li>
<input type="checkbox" disabled>
add filtering to <code><a href="reference/filter_markers.html">filter_markers()</a></code> and config also by other variables (<code>self.average</code> etc.)</li>
<li>
<input type="checkbox" disabled checked>
separate global markers and contrasts</li>
</ul></div>
<div class="section level2">
<h2 id="contrasts">contrasts<a class="anchor" aria-label="anchor" href="#contrasts"></a></h2>
<ul class="task-list"><li>
<input type="checkbox" disabled>
contrasts<ul class="task-list"><li>
<input type="checkbox" disabled checked>
contrasts computation</li>
<li>
<input type="checkbox" disabled checked>
heatmaps</li>
<li>
<input type="checkbox" disabled checked>
marker plots</li>
<li>
<input type="checkbox" disabled checked>
marker table reports</li>
<li>
<input type="checkbox" disabled checked>
summary report (index page)</li>
<li>
<input type="checkbox" disabled>
GSEA</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="gsea-alternative">GSEA alternative<a class="anchor" aria-label="anchor" href="#gsea-alternative"></a></h2>
<ul class="task-list"><li>
<input type="checkbox" disabled><a href="https://github.com/yoseflab/VISION" class="external-link">Vision</a> - Signature Analysis and Visualization for Single-Cell RNA-seq</li>
</ul></div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a></h2>
<ul class="task-list"><li>
<input type="checkbox" disabled>
additional plots: plot alongside with clustering plot</li>
<li>
<input type="checkbox" disabled>
allow to specify cluster specific colors<ul class="task-list"><li>
<input type="checkbox" disabled>
maybe add it to <code>CELL_GROUPINGS</code>
</li>
</ul></li>
<li>
<input type="checkbox" disabled>
cluster marker genes heatmap:<ul class="task-list"><li>
<input type="checkbox" disabled>
(?) rewrite to <code>ComplexHeatmap</code>
</li>
<li>
<input type="checkbox" disabled>
add gaps to rows (to see each top N genes from a cluster)</li>
</ul></li>
<li>
<input type="checkbox" disabled>
look at hexagonal plots: <a href="https://github.com/SaskiaFreytag/schex" class="external-link uri">https://github.com/SaskiaFreytag/schex</a>
</li>
<li>
<input type="checkbox" disabled>
Use <a href="https://github.com/VPetukhov/ggrastr" class="external-link"><code>ggrastr</code></a> for large data (&gt; 10 000 cells). It converts points to raster -&gt; better PDF performance.</li>
<li>
<input type="checkbox" disabled checked><del>fix header tooltips in <code>cluster_markers_table_template.Rmd</code></del><ul><li>header tooltips are not used anymore</li>
</ul></li>
<li>
<input type="checkbox" disabled checked>
dotplots (similar to heatmaps): <code><a href="https://rdrr.io/pkg/scater/man/plotDots.html" class="external-link">scater::plotDots()</a></code>
</li>
<li>
<input type="checkbox" disabled>
alternative plots:<ul class="task-list"><li>
<input type="checkbox" disabled>
barplot heatmap (<a href="https://pasteboard.co/JNneVHK.png" class="external-link">here</a> and <a href="https://pasteboard.co/JNnfE4V.png" class="external-link">here</a>), <a href="https://doi.org/10.1016/j.celrep.2017.11.050" class="external-link">paper</a>
</li>
<li>
<input type="checkbox" disabled><a href="https://pasteboard.co/JNngyug.png" class="external-link">cluster correlation</a> (<a href="https://www.nature.com/articles/s41593-017-0056-2" class="external-link">paper</a>)</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="marker-set-comparisons">Marker set comparisons<a class="anchor" aria-label="anchor" href="#marker-set-comparisons"></a></h2>
<ul class="task-list"><li>
<input type="checkbox" disabled>
Venn diagrams of significant markers.</li>
</ul></div>
</div>
<div class="section level1">
<h1 id="gene-set-analysis">Gene set analysis<a class="anchor" aria-label="anchor" href="#gene-set-analysis"></a></h1>
<ul><li>some other methods:
<ul><li>
<a href="https://bioconductor.org/packages/release/bioc/html/AUCell.html" class="external-link">AUCell</a> - analysis of “gene set” activity in single-cell RNA-seq data (e.g. identify cells with specific gene signatures)</li>
<li>
<a href="https://github.com/xzhoulab/iDEA" class="external-link">iDEA</a> - Integrative Differential expression and gene set Enrichment Analysis using summary statistics for scRNA-seq</li>
<li>
<a href="https://www.bioconductor.org/packages/release/bioc/html/GSVA.html" class="external-link">GSVA</a> - Gene Set Variation Analysis
<ul><li><a href="https://rdrr.io/bioc/singleCellTK/man/gsvaSCE.html" class="external-link"><code>singleCellTK::gsvaSCE()</code></a></li>
</ul></li>
<li>
<a href="https://github.com/broadinstitute/ssGSEA2.0" class="external-link">ssGSEA</a> - single sample Gene Set Enrichment Analysis</li>
<li>
<a href="http://bioconductor.org/packages/release/bioc/html/progeny.html" class="external-link">progeny</a> - common pathways in cancer</li>
<li>
<a href="https://bioconductor.org/packages/release/data/experiment/html/dorothea.html" class="external-link">dorothea</a> - transcription factor regulons</li>
<li>
<a href="https://www.bioconductor.org/packages/release/bioc/html/viper.html" class="external-link">viper</a> - inference of protein activity from gene expression data</li>
<li>
<a href="https://github.com/aertslab/SCENIC" class="external-link">SCENIC</a> - Single-Cell rEgulatory Network Inference and Clustering</li>
<li>
<a href="https://bioconductor.org/packages/release/bioc/html/GENIE3.html" class="external-link">GENIE3</a> - GEne Network Inference with Ensemble of trees</li>
<li>
<a href="https://bioconductor.org/packages/release/bioc/html/hipathia.html" class="external-link">Hipathia</a> - High-throughput Pathway Analysis</li>
</ul></li>
<li>
<input type="checkbox" disabled>
add <a href="https://yulab-smu.github.io/clusterProfiler-book/chapter5.html#go-classification" class="external-link">GeneOntology</a><ul><li>
<a href="https://guangchuangyu.github.io/2015/10/use-simplify-to-remove-redundancy-of-enriched-go-terms/" class="external-link">simplify</a> the GO terms (remove redundant ones)</li>
</ul></li>
<li>
<input type="checkbox" disabled>
more MSigDB genesets - hallmark, …</li>
<li>
<input type="checkbox" disabled>
(?) To output datatable add child rows with gene IDs present in a term. See <a href="https://rstudio.github.io/DT/002-rowdetails.html" class="external-link">DT child rows</a> and its <a href="http://www.reigo.eu/2018/04/extending-dt-child-row-example/" class="external-link">extension</a><ul><li>Or create separate HTML reports for them?</li>
</ul></li>
<li>
<input type="checkbox" disabled>
(?) ORA separately for up- and down-regulated markers</li>
<li>
<input type="checkbox" disabled>
add description of GSEA and ORA methods to Rmd</li>
<li>
<input type="checkbox" disabled>
add export of cnetplot from <code>clusterProfiler</code> to graph file importable to <code>Cytoscape</code> (already done in another project: Jirka - BLF and CHDF spheroids)</li>
</ul></div>
<div class="section level1">
<h1 id="trajectory-inference">trajectory inference<a class="anchor" aria-label="anchor" href="#trajectory-inference"></a></h1>
<ul class="task-list"><li>
<input type="checkbox" disabled>
TBD</li>
<li>
<input type="checkbox" disabled><a href="https://dynverse.org/" class="external-link">dynverse</a><ul><li>it was used in one of projects (Jirka - BLF and CHDF spheroids), could be easy to implement</li>
</ul></li>
<li>
<input type="checkbox" disabled><a href="http://velocyto.org/" class="external-link">velocyto</a>
</li>
</ul></div>
<div class="section level1">
<h1 id="integration">Integration<a class="anchor" aria-label="anchor" href="#integration"></a></h1>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>
look at <a href="https://bioconductor.org/packages/release/bioc/vignettes/batchelor/inst/doc/correction.html" class="external-link">batchelor vignette</a><ul><li>
<code>correctExperiments()</code> seems easier to use (“combining the assay data and column metadata for easy downstream use”)</li>
</ul></li>
<li>
<input type="checkbox" disabled checked><a href="http://bioconductor.org/books/release/OSCA/integrating-datasets.html#correction-diagnostics" class="external-link">correction diagnostics</a>
</li>
<li>
<input type="checkbox" disabled>
(?) <a href="http://bioconductor.org/books/release/OSCA/integrating-datasets.html#integration-with-markers" class="external-link">integration using known marker genes of each batch</a>
</li>
<li>
<input type="checkbox" disabled>
(?) use separate HVG parameters for each integration method</li>
</ul></div>


  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="copyright">
  <p></p><p>Developed by Jiri Novotny, Jan Kubovciak, Michal Kolar.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

    </footer></div>

  

  

  </body></html>

