<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="
    This pipeline is implemented in the drake Make-like toolkit (https://github.com/ropensci/drake) and
    provides the following steps of scRNA-seq secondary analysis: input quality control, filtering of cells/genes,
    normalization, clustering, cell type annotation, cluster markers, differential expression analysis
    (constrasts -&gt; cluster vs. cluster), integration of multiple samples.
    All main parts are accompanied by rich graphical outputs and HTML reports
    (RMarkdown). Thanks to the drake package, the pipeline is highly reproducible, scalable and efficient, provides easy
    access to all intermediate results, and can be arbitrarily extended as pipeline definitions are abstracted as
    R objects. A command line interface is available to perform the most important actions directly in shell terminal.">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/5/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>A pipeline for droplet-based single-cell RNA-seq data secondary analysis implemented in the drake Make-like toolkit for R language ‚Ä¢ scdrake</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A pipeline for droplet-based single-cell RNA-seq data secondary analysis implemented in the drake Make-like toolkit for R language">
<meta property="og:description" content="
    This pipeline is implemented in the drake Make-like toolkit (https://github.com/ropensci/drake) and
    provides the following steps of scRNA-seq secondary analysis: input quality control, filtering of cells/genes,
    normalization, clustering, cell type annotation, cluster markers, differential expression analysis
    (constrasts -&gt; cluster vs. cluster), integration of multiple samples.
    All main parts are accompanied by rich graphical outputs and HTML reports
    (RMarkdown). Thanks to the drake package, the pipeline is highly reproducible, scalable and efficient, provides easy
    access to all intermediate results, and can be arbitrarily extended as pipeline definitions are abstracted as
    R objects. A command line interface is available to perform the most important actions directly in shell terminal.">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/5/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/5/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="preferably.css">
<link id="css-code-light" rel="stylesheet" href="code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="darkswitch.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">scdrake</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.5.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/scdrake.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="articles/scdrake_integration.html">Integration pipeline guide</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="articles/scdrake_faq.html">FAQ &amp; Howtos</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Guides</h6>
    <a class="dropdown-item" href="articles/scdrake_docker.html">Using the Docker image</a>
    <a class="dropdown-item" href="articles/scdrake_integration.html">02 Integration pipeline guide</a>
    <a class="dropdown-item" href="articles/scdrake_advanced.html">Advanced topics</a>
    <a class="dropdown-item" href="articles/scdrake_extend.html">Extending the pipeline</a>
    <a class="dropdown-item" href="articles/drake_basics.html">drake basics</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General information</h6>
    <a class="dropdown-item" href="articles/pipeline_overview.html">Pipeline overview</a>
    <a class="dropdown-item" href="articles/scdrake_faq.html">FAQ &amp; Howtos</a>
    <a class="dropdown-item" href="articles/scdrake_cli.html">Command line interface</a>
    <a class="dropdown-item" href="articles/scdrake_config.html">scdrake config internals</a>
    <a class="dropdown-item" href="articles/scdrake_envvars.html">Environment variables</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General configs</h6>
    <a class="dropdown-item" href="articles/config_pipeline.html">Pipeline config</a>
    <a class="dropdown-item" href="articles/config_main.html">Main config</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single-sample pipeline stages</h6>
    <a class="dropdown-item" href="articles/stage_input_qc.html">Input and quality control stage (`01_input_qc`)</a>
    <a class="dropdown-item" href="articles/stage_norm_clustering.html">Normalization and clustering stage (`02_norm_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Integration pipeline stages</h6>
    <a class="dropdown-item" href="articles/stage_integration.html">Integration stage (`01_integration`)</a>
    <a class="dropdown-item" href="articles/stage_int_clustering.html">Post-integration clustering stage (`02_int_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Common stages</h6>
    <a class="dropdown-item" href="articles/stage_cluster_markers.html">Cluster markers stage (`cluster_markers`)</a>
    <a class="dropdown-item" href="articles/stage_contrasts.html">Contrasts stage (`contrasts`)</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bioinfocz/scdrake" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>

        
        <li>
          <a class="external-link nav-link" id="css-toggle-btn" aria-label="github">
            <span class="fas fa fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        
        
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="scdrake">scdrake<a class="anchor" aria-label="anchor" href="#scdrake"></a>
</h1></div>

<p><a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> is a scalable and reproducible pipeline for secondary analysis of droplet-based single-cell RNA-seq data. <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> is an R package built on top of the <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> package, a <a href="https://www.gnu.org/software/make" class="external-link">Make</a>-like pipeline toolkit for <a href="https://www.r-project.org" class="external-link">R language</a>.</p>
<p>The main features of the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> pipeline are:</p>
<ul>
<li>Import of scRNA-seq data: <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">10x Genomics Cell Ranger</a> output, delimited table, or <code>SingleCellExperiment</code> object.</li>
<li>Quality control and filtering of cells and genes, removal of empty droplets.</li>
<li>Higly variable genes detection, cell cycle scoring, normalization, clustering, and dimensionality reduction.</li>
<li>Cell type annotation.</li>
<li>Integration of multiple datasets.</li>
<li>Computation of cluster markers and differentially expressed genes between clusters (denoted as ‚Äúcontrasts‚Äù).</li>
<li>Rich graphical and HTML outputs based on customizable RMarkdown documents.
<ul>
<li>You can find links to example outputs <a href="https://bioinfocz.github.io/scdrake/articles/pipeline_overview.html">here</a>.</li>
</ul>
</li>
<li>Thanks to <a href="https://docs.ropensci.org/drake" class="external-link">drake</a>, the pipeline is highly efficient, scalable and reproducible, and also extendable.
<ul>
<li>Want to change some parameter? No problem! Only parts of the pipeline which changed will rerun, while up-to-date ones will be skipped.</li>
<li>Want to reuse the intermediate results for your own analyses? No problem! The pipeline has smartly defined checkpoints which can be loaded from a <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cache.</li>
<li>Want to extend the pipeline? No problem! The pipeline definition is just an R object which can be arbitrarily extended.</li>
</ul>
</li>
</ul>
<p>For whom is <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> purposed? It is primarily intended for tech-savvy users (bioinformaticians), who pass on the results (reports, images) to non-technical persons (biologists). At the same time, bioinformaticians can quickly react to biologists‚Äô needs by changing the parameters of the pipeline, which then efficiently skips already finished parts. This dialogue between the biologist and the bioinformatician is indispensable during scRNA-seq data analysis. <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> ensures that this communication is performed in an effective and reproducible manner.</p>
<p>The pipeline structure along with <a href="https://github.com/bioinfocz/scdrake/blob/main/diagrams/README.md" class="external-link">diagrams</a> and links to outputs is described in <code><a href="articles/pipeline_overview.html">vignette("pipeline_overview")</a></code> (<a href="https://bioinfocz.github.io/scdrake/articles/pipeline_overview.html">link</a>).</p>
<p>If you use <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> in your research, please, consider citing</p>
<blockquote>
<p>Kubovciak J, Kolar M, Novotny J (2023). ‚ÄúScdrake: a reproducible and scalable pipeline for scRNA-seq data analysis.‚Äù <em>Bioinformatics Advances</em>, <strong>3</strong>(1). <a href="https://doi.org/10.1093/bioadv/vbad089" class="external-link">doi:10.1093/bioadv/vbad089</a>.</p>
</blockquote>
<p>Huge thanks go to the authors of the <a href="https://bioconductor.org/books/3.15/OSCA" class="external-link">Orchestrating Single-Cell Analysis with Bioconductor</a> book on whose methods and recommendations is <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> largely based.</p>
<hr>
</div>
<div class="section level1">
<h1 id="installation-instructions">Installation instructions<a class="anchor" aria-label="anchor" href="#installation-instructions"></a>
</h1>
<div class="section level2">
<h2 id="using-a-docker-image-recommended">Using a Docker image (recommended)<a class="anchor" aria-label="anchor" href="#using-a-docker-image-recommended"></a>
</h2>
<p>A Docker image based on the <a href="https://bioconductor.org/help/docker/" class="external-link">official Bioconductor image</a> (version 3.15) is available. This is the most handy and reproducible way how to use <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> as all the dependencies are already installed and their versions are fixed. In addition, the parent Bioconductor image comes bundled with RStudio Server.</p>
<p>The complete guide to the usage of <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>‚Äôs Docker image can be found in the <a href="https://bioinfocz.github.io/scdrake/articles/scdrake_docker.html">Docker vignette</a>. <strong>We strongly recommend to go through even if you are an experienced Docker user.</strong> Below you can find just the basic command to download the image and to run a detached container with RStudio in Docker or to run <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> in Singularity.</p>
<p>You can also run the image in <a href="https://docs.sylabs.io/guides/latest/user-guide/quick_start.html" class="external-link">SingularityCE</a> (without RStudio) - see the Singularity section in the Docker vignette above. If the image is already downloaded in the local Docker storage, you can use <code>singularity pull docker-daemon:&lt;image&gt;</code></p>
<p>You can pull the Docker image with the latest stable <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> version using</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull jirinovo/scdrake:1.5.2</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">singularity</span> pull docker:jirinovo/scdrake:1.5.2</span></code></pre></div>
<p>or list available versions in <a href="https://hub.docker.com/r/jirinovo/scdrake/tags" class="external-link">our Docker Hub repository</a>.</p>
<p>For the latest development version use</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull jirinovo/scdrake:latest</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">singularity</span> pull docker:jirinovo/scdrake:latest</span></code></pre></div>
<p><strong>Note for Mac users with M1/M2 chipsets</strong>: until version 1.5.0 (inclusive), <code>arm64</code> images are available.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull jirinovo/scdrake:1.5.0-bioc3.15-arm64</span></code></pre></div>
<div class="section level3">
<h3 id="running-the-container">Running the container<a class="anchor" aria-label="anchor" href="#running-the-container"></a>
</h3>
<p>For the most common cases of host machines: Linux running Docker Engine, and Windows or MacOS running Docker Desktop.</p>
<p>First make a shared directory that will be mounted to the container:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> ~/scdrake_projects</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/scdrake_projects</span></code></pre></div>
<p>And run the image that will expose RStudio Server on port 8787 on your host:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-d</span> <span class="dt">\</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">-v</span> <span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span>:/home/rstudio/scdrake_projects <span class="dt">\</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-p</span> 8787:8787 <span class="dt">\</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-e</span> USERID=<span class="va">$(</span><span class="fu">id</span> <span class="at">-u</span><span class="va">)</span> <span class="dt">\</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-e</span> GROUPID=<span class="va">$(</span><span class="fu">id</span> <span class="at">-g</span><span class="va">)</span> <span class="dt">\</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">-e</span> PASSWORD=1234 <span class="dt">\</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  jirinovo/scdrake:1.5.2</span></code></pre></div>
<p>For Singularity, also make shared directories and execute the container (‚Äúrun and forget‚Äù):</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> ~/scdrake_singularity</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/scdrake_singularity</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> home/<span class="va">${USER}</span> scdrake_projects</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="ex">singularity</span> exec <span class="dt">\</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">-e</span> <span class="dt">\</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--no-home</span> <span class="dt">\</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--bind</span> <span class="st">"home/</span><span class="va">${USER}</span><span class="st">/:/home/</span><span class="va">${USER}</span><span class="st">,scdrake_projects/:/home/</span><span class="va">${USER}</span><span class="st">/scdrake_projects"</span> <span class="dt">\</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--pwd</span> <span class="st">"/home/</span><span class="va">${USER}</span><span class="st">/scdrake_projects"</span> <span class="dt">\</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    path/to/scdrake_image.sif <span class="dt">\</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    scdrake <span class="op">&lt;</span>args<span class="op">&gt;</span> <span class="op">&lt;</span>command<span class="op">&gt;</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="installing-scdrake-manually-not-recommended">Installing <code>{scdrake}</code> manually (not recommended)<a class="anchor" aria-label="anchor" href="#installing-scdrake-manually-not-recommended"></a>
</h2>
<details><summary>
Click for details
</summary><div class="section level3">
<h3 id="install-the-required-system-packages">Install the required system packages<a class="anchor" aria-label="anchor" href="#install-the-required-system-packages"></a>
</h3>
<ul>
<li>For Linux, follow the commands for your distribution <a href="required_libs_linux.html">here</a>.</li>
<li>For MacOS: <code>$ brew install libxml2 imagemagick@6 harfbuzz fribidi libgit2 geos pandoc</code>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="install-r--42">Install R &gt;= 4.2<a class="anchor" aria-label="anchor" href="#install-r--42"></a>
</h3>
<p>See <a href="https://cloud.r-project.org/" class="external-link uri">https://cloud.r-project.org/</a></p>
<p>From now on, all commands are for R.</p>
</div>
<div class="section level3">
<h3 id="install-renv">Install <code>{renv}</code>
<a class="anchor" aria-label="anchor" href="#install-renv"></a>
</h3>
<p><a href="https://rstudio.github.io/renv/" class="external-link"><code>{renv}</code></a> is an R package for management of local R libraries. It is intended to be used on a per-project basis, i.e.¬†each project should use its own library of R packages.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"renv"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="initialize-a-new-renv-library">Initialize a new <code>{renv}</code> library<a class="anchor" aria-label="anchor" href="#initialize-a-new-renv-library"></a>
</h3>
<p>Switch to directory where you will analyze data and initialize a new <a href="https://rstudio.github.io/renv/" class="external-link">renv</a> library:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/consent.html" class="external-link">consent</a></span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/init.html" class="external-link">init</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Now exit and run again R. You should see a message that renv library has been activated.</p>
</div>
<div class="section level3">
<h3 id="install-biocmanager">Install BiocManager<a class="anchor" aria-label="anchor" href="#install-biocmanager"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="install-bioconductor-315">Install Bioconductor 3.15<a class="anchor" aria-label="anchor" href="#install-bioconductor-315"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.15"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="restore-scdrake-dependencies-from-lockfile">Restore <code>{scdrake}</code> dependencies from lockfile<a class="anchor" aria-label="anchor" href="#restore-scdrake-dependencies-from-lockfile"></a>
</h3>
<p><a href="https://rstudio.github.io/renv/" class="external-link">renv</a> also allows to export the current installed versions of R packages (and other things) into a lockfile. Such lockfile is available for <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> and you can use it to install all dependencies by</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## -- This is a lockfile for the latest stable version of scdrake.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/bioinfocz/scdrake/1.5.2/renv.lock"</span><span class="op">)</span></span>
<span><span class="co">## -- You can increase the number of CPU cores to speed up the installation.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>Ncpus <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">renv</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/renv/reference/restore.html" class="external-link">restore</a></span><span class="op">(</span>lockfile <span class="op">=</span> <span class="st">"renv.lock"</span>, repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For the lockfile for the latest development version use</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/bioinfocz/scdrake/main/renv.lock"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="install-the-scdrake-package">Install the <code>{scdrake}</code> package<a class="anchor" aria-label="anchor" href="#install-the-scdrake-package"></a>
</h3>
<p>Now we can finally install the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> package, but using a non-standard approach - without its dependencies (which are already installed from the lockfile).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span></span>
<span>  <span class="st">"bioinfocz/scdrake@1.5.2"</span>,</span>
<span>  dependencies <span class="op">=</span> <span class="cn">FALSE</span>, upgrade <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  keep_source <span class="op">=</span> <span class="cn">TRUE</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For the latest development version use <code>"bioinfocz/scdrake"</code>.</p>
</div>
<div class="section level3">
<h3 id="install-the-command-line-interface-cli">Install the command line interface (CLI)<a class="anchor" aria-label="anchor" href="#install-the-command-line-interface-cli"></a>
</h3>
<p>Optionally, you can install <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>‚Äôs CLI scripts with</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scdrake</span><span class="fu">::</span><span class="fu"><a href="reference/cli.html">install_cli</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>CLI should be now accessible as a <code>scdrake</code> command. By default, the CLI is installed into <code>~/.local/bin</code>, which is usually present in the <code>PATH</code> environment variable. In case it isn‚Äôt, just add to your <code>~/.bashrc</code>: <code>export PATH="${HOME}/.local/bin:${PATH}"</code></p>
<p><strong>Every time you will be using the CLI make sure your current working directory is inside an <a href="https://rstudio.github.io/renv/" class="external-link">renv</a> project.</strong> You can read the reasons below.</p>
<details><summary>
Show details
</summary><p>You might notice that a per-project <a href="https://rstudio.github.io/renv/" class="external-link">renv</a> library and an installed CLI are ‚Äúdisconnected‚Äù and if you install <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> and its CLI within multiple projects (<a href="https://rstudio.github.io/renv/" class="external-link">renv</a> libraries), then the CLI scripts in <code>~/.local/bin</code> will be overwritten each time. But when you run the <code>scdrake</code> command inside an <a href="https://rstudio.github.io/renv/" class="external-link">renv</a> project, the <code>renv</code> directory is automatically detected and the <a href="https://rstudio.github.io/renv/" class="external-link">renv</a> library is activated by <code><a href="https://rstudio.github.io/renv/reference/load.html" class="external-link">renv::load()</a></code>, so the proper, locally installed <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> package is then used.</p>
<p>Also, there is a built-in guard: the version of the CLI must match the version of the bundled CLI scripts inside the installed <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> package. Anyway, we think changes in the CLI won‚Äôt be very frequent, so this shouldn‚Äôt be a problem most of the time.</p>
</details><blockquote>
<p>TIP: To save time and space, you can symlink the <code>renv/library</code> directory to multiple <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> projects.</p>
</blockquote>

<hr>
</div>
</details>
</div>
<div class="section level2">
<h2 id="quickstart">Quickstart<a class="anchor" aria-label="anchor" href="#quickstart"></a>
</h2>
<p>First run the <code>scdrake</code> image in Docker or Singularity - see the <a href="https://bioinfocz.github.io/scdrake/articles/scdrake_docker.html">Docker vignette</a></p>
<p>Then you can go through the <a href="https://bioinfocz.github.io/scdrake/articles/scdrake.html">Get Started vignette</a></p>
<hr>
</div>
<div class="section level2">
<h2 id="vignettes-and-other-readings">Vignettes and other readings<a class="anchor" aria-label="anchor" href="#vignettes-and-other-readings"></a>
</h2>
<p>See <a href="https://bioinfocz.github.io/scdrake" class="uri">https://bioinfocz.github.io/scdrake</a> for a documentation website of the latest stable version (1.5.2) where links to vignettes below become real :-)</p>
<p>See <a href="https://bioinfocz.github.io/scdrake/dev" class="uri">https://bioinfocz.github.io/scdrake/dev</a> for a documentation website of the current development version.</p>
<ul>
<li>Guides:
<ul>
<li>Using the Docker image: <a href="https://bioinfocz.github.io/scdrake/articles/scdrake_docker.html" class="uri">https://bioinfocz.github.io/scdrake/articles/scdrake_docker.html</a> (or <code><a href="articles/scdrake_docker.html">vignette("scdrake_docker")</a></code>)</li>
<li>01 Quick start (single-sample pipeline): <code><a href="articles/scdrake.html">vignette("scdrake")</a></code>
</li>
<li>02 Integration pipeline guide: <code><a href="articles/scdrake_integration.html">vignette("scdrake_integration")</a></code>
</li>
<li>Advanced topics: <code><a href="articles/scdrake_advanced.html">vignette("scdrake_advanced")</a></code>
</li>
<li>Extending the pipeline: <code><a href="articles/scdrake_extend.html">vignette("scdrake_extend")</a></code>
</li>
<li>
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> basics: <code><a href="articles/drake_basics.html">vignette("drake_basics")</a></code>
<ul>
<li>Or the official <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> book: <a href="https://books.ropensci.org/drake/" class="external-link uri">https://books.ropensci.org/drake/</a>
</li>
</ul>
</li>
</ul>
</li>
<li>General information:
<ul>
<li>Pipeline overview: <code><a href="articles/pipeline_overview.html">vignette("pipeline_overview")</a></code>
</li>
<li>FAQ &amp; Howtos: <code><a href="articles/scdrake_faq.html">vignette("scdrake_faq")</a></code>
</li>
<li>Command line interface (CLI): <code><a href="articles/scdrake_cli.html">vignette("scdrake_cli")</a></code>
</li>
<li>Config files (internals): <code><a href="articles/scdrake_config.html">vignette("scdrake_config")</a></code>
</li>
<li>Environment variables: <code><a href="articles/scdrake_envvars.html">vignette("scdrake_envvars")</a></code>
</li>
</ul>
</li>
<li>General configs:
<ul>
<li>Pipeline config -&gt; <code><a href="articles/config_pipeline.html">vignette("config_pipeline")</a></code>
</li>
<li>Main config -&gt; <code><a href="articles/config_main.html">vignette("config_main")</a></code>
</li>
</ul>
</li>
<li>Pipelines and stages:
<ul>
<li>Single-sample pipeline:
<ul>
<li>Stage <code>01_input_qc</code>: reading in data, filtering, quality control -&gt; <code><a href="articles/stage_input_qc.html">vignette("stage_input_qc")</a></code>
</li>
<li>Stage <code>02_norm_clustering</code>: normalization, HVG selection, dimensionality reduction, clustering, cell type annotation -&gt; <code><a href="articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code>
</li>
</ul>
</li>
<li>Integration pipeline:
<ul>
<li>Stage <code>01_integration</code>: reading in data and integration -&gt; <code><a href="articles/stage_integration.html">vignette("stage_integration")</a></code>
</li>
<li>Stage <code>02_int_clustering</code>: post-integration clustering and cell annotation -&gt; <code><a href="articles/stage_int_clustering.html">vignette("stage_int_clustering")</a></code>
</li>
</ul>
</li>
<li>Common stages:
<ul>
<li>Stage <code>cluster_markers</code> -&gt; <code><a href="articles/stage_cluster_markers.html">vignette("stage_cluster_markers")</a></code>
</li>
<li>Stage <code>contrasts</code> (differential expression) -&gt; <code><a href="articles/stage_contrasts.html">vignette("stage_contrasts")</a></code>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>We encourage all users to read <a href="https://books.ropensci.org/drake" class="external-link">basics</a> of the <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> package. While it is not necessary to know all <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> internals to successfully run the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> pipeline, its knowledge is a plus. You can read the minimum basics in <code><a href="articles/drake_basics.html">vignette("drake_basics")</a></code>.</p>
<p>Also, the prior knowledge of Bioconductor and its classes (especially the <a href="https://bioconductor.org/packages/3.15/bioc/html/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a>) is considerable.</p>
<hr>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Below is the citation output from using <code>citation("scdrake")</code> in R. Please run this yourself to check for any updates on how to cite <strong>scdrake</strong>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"scdrake"</span><span class="op">)</span>, bibtex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>To cite package ‚Äòscdrake‚Äô <span class="cf">in</span> publications use<span class="sc">:</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  Jiri Novotny and Jan <span class="fu">Kubovciak</span> (<span class="dv">2021</span>). scdrake<span class="sc">:</span> A Pipeline For 10x Chromium Single<span class="sc">-</span>Cell RNA<span class="sc">-</span>seq Data Analysis.</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  https<span class="sc">:</span><span class="er">//</span>github.com<span class="sc">/</span>bioinfocz<span class="sc">/</span>scdrake, https<span class="sc">:</span><span class="er">//</span>bioinfocz.github.io<span class="sc">/</span>scdrake.</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>A BibTeX entry <span class="cf">for</span> LaTeX users is</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="sc">@</span>Manual{,</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    title <span class="ot">=</span> {scdrake<span class="sc">:</span> A Pipeline For 10x Chromium Single<span class="sc">-</span>Cell RNA<span class="sc">-</span>seq Data Analysis},</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    author <span class="ot">=</span> {Jiri Novotny and Jan Kubovciak},</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>    year <span class="ot">=</span> {<span class="dv">2021</span>},</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>    note <span class="ot">=</span> {https<span class="sc">:</span><span class="er">//</span>github.com<span class="sc">/</span>bioinfocz<span class="sc">/</span>scdrake, https<span class="sc">:</span><span class="er">//</span>bioinfocz.github.io<span class="sc">/</span>scdrake},</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>  }</span></code></pre></div>
<p>Please note that the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> was only made possible thanks to many other R and bioinformatics software authors, which are cited either in the vignettes and/or the paper(s) describing this package.</p>
</div>
<div class="section level2">
<h2 id="help-and-support">Help and support<a class="anchor" aria-label="anchor" href="#help-and-support"></a>
</h2>
<p>In case of any problems or suggestions, please, open a new <a href="https://github.com/bioinfocz/scdrake/issues" class="external-link">issue</a>. We will be happy to answer your questions, integrate new ideas, or resolve any problems <span class="emoji" data-emoji="blush">üòä</span></p>
<p>You can also use <a href="https://github.com/bioinfocz/scdrake/discussions" class="external-link">GitHub Discussions</a>, mainly for topics <strong>not</strong> related to development (bugs, feature requests etc.), but if you need e.g.¬†a general help.</p>
</div>
<div class="section level2">
<h2 id="contribution">Contribution<a class="anchor" aria-label="anchor" href="#contribution"></a>
</h2>
<p>If you want to contribute to <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>, read the <a href="CONTRIBUTING.html">contribution guide</a>, please. All pull requests are welcome! <span class="emoji" data-emoji="slightly_smiling_face">üôÇ</span></p>
</div>
<div class="section level2">
<h2 id="code-of-conduct">Code of Conduct<a class="anchor" aria-label="anchor" href="#code-of-conduct"></a>
</h2>
<p>Please note that the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> project is released with a <a href="https://bioinfocz.github.io/scdrake/CODE_OF_CONDUCT.html">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<div class="section level3">
<h3 id="funding">Funding<a class="anchor" aria-label="anchor" href="#funding"></a>
</h3>
<p>This work was supported by <a href="https://www.elixir-czech.cz" class="external-link">ELIXIR CZ</a> research infrastructure project (MEYS Grant No: LM2018131) including access to computing and storage facilities.</p>
</div>
<div class="section level3">
<h3 id="software-and-methods-used-by-scdrake">Software and methods used by <code>{scdrake}</code>
<a class="anchor" aria-label="anchor" href="#software-and-methods-used-by-scdrake"></a>
</h3>
<p>Many things are used by <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>, but these are really worth mentioning:</p>
<ul>
<li>The <a href="https://www.bioconductor.org" class="external-link">Bioconductor</a> ecosystem.</li>
<li>The <a href="https://bioconductor.org/books/3.15/OSCA" class="external-link"><em>Orchestrating Single-Cell Analysis with Bioconductor</em></a> book.</li>
<li>The <a href="https://bioconductor.org/packages/3.15/bioc/html/scran.html" class="external-link">scran</a>, <a href="https://bioconductor.org/packages/3.15/bioc/html/scater.html" class="external-link">scater</a>, and other great packages from <a href="https://orcid.org/0000-0002-3564-4813" class="external-link">Aaron Lun</a> et al.</li>
<li>The <a href="https://github.com/ropensci/drake" class="external-link">drake</a> package.</li>
<li>The <a href="https://github.com/rstudio/rmarkdown" class="external-link">rmarkdown</a> package, and other ones from the <a href="https://www.tidyverse.org" class="external-link">tidyverse</a> ecosystem.</li>
</ul>
</div>
<div class="section level3">
<h3 id="development-tools">Development tools<a class="anchor" aria-label="anchor" href="#development-tools"></a>
</h3>
<ul>
<li>Continuous code testing is possible thanks to <a href="https://github.com/features/actions" class="external-link">GitHub Actions</a> through <a href="https://usethis.r-lib.org" class="external-link">usethis</a>, <a href="https://remotes.r-lib.org" class="external-link">remotes</a>, and <a href="https://r-lib.github.io/rcmdcheck/" class="external-link">rcmdcheck</a>. Customized to use <a href="https://www.bioconductor.org/help/docker" class="external-link">Bioconductor‚Äôs docker containers</a>.</li>
<li>The <a href="https://bioinfocz.github.io/scdrake">documentation website</a> is generated by <a href="https://pkgdown.r-lib.org" class="external-link">pkgdown</a>.</li>
<li>The code is styled automatically thanks to <a href="https://github.com/r-lib/styler" class="external-link">styler</a>.</li>
<li>The documentation is formatted thanks to <a href="https://devtools.r-lib.org/" class="external-link">devtools</a> and <a href="https://roxygen2.r-lib.org/" class="external-link">roxygen2</a>.</li>
</ul>
<p>This package was developed using <code>{biocthis}</code>.</p>
</div>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/bioinfocz/scdrake" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/bioinfocz/scdrake/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scdrake</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Jiri Novotny <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1338-638X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jan Kubovciak <br><small class="roles"> Author </small>  </li>
<li>Michal Kolar <br><small class="roles"> Funder </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="NEWS.html"><img src="https://img.shields.io/badge/NEWS-updates-informational" alt="NEWS: updates"></a></li>
<li><a href="https://bioinfocz.github.io/scdrake"><img src="https://img.shields.io/badge/Documentation%20&amp;%20vignettes-bioinfocz.github.io/scdrake-informational" alt="Documentation and vignettes (stable version)"></a></li>
<li><a href="https://bioinfocz.github.io/scdrake/dev"><img src="https://img.shields.io/badge/Documentation%20&amp;%20vignettes-bioinfocz.github.io/scdrake-informational" alt="Documentation and vignettes (devel version)"></a></li>
<li><a href="https://bioinfocz.github.io/scdrake/articles/pipeline_overview.html"><img src="https://img.shields.io/badge/Overview%20&amp;%20outputs-vignette(%22pipeline_overview%22)-informational" alt="Overview and outputs"></a></li>
<li><a href="https://github.com/bioinfocz/scdrake/blob/main/diagrams/README.md" class="external-link"><img src="https://img.shields.io/badge/Pipeline%20diagram-Show-informational" alt="Pipeline diagram"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://github.com/bioinfocz/scdrake/actions/workflows/docker-ci.yml" class="external-link"><img src="https://github.com/bioinfocz/scdrake/actions/workflows/docker-ci.yml/badge.svg?branch=main" alt="Docker Image CI"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="copyright">
  <p></p>
<p>Developed by Jiri Novotny, Jan Kubovciak, Michal Kolar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
