---
title: "scdrake | 01 - data load, QC"
author: "Jiri Novotny, Jan Kubovciak"
institute: "Laboratory of Genomics and Bioinformatics @ Institute of Molecular Genetics of the ASCR"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    number_sections: false
    theme: "flatly"
    self_contained: true
    code_download: true
    df_print: "paged"
params:
  css_file: !expr here::here("Rmd/common/stylesheet.css")
  drake_cache_dir: !expr here::here(".drake")
date: "`r glue('Document generated: {Sys.Date()}')`"
css: "`r params$css_file`"
---

```{r, message = FALSE, warning = FALSE}
library(drake)
if (rlang::is_true(getOption("knitr.in.progress"))) {
  params_ <- scdrake_list(params)
}
drake_cache_dir <- params_$drake_cache_dir

loadd(
  config_main, config_input_qc, empty_droplets, sce_valid_cells_info, barcode_ranks,
  qc_filter, custom_filter, sce_qc_filter_rowSums, sce_custom_filter_rowSums,
  path = drake_cache_dir
)

cfg <- config_input_qc

input_type <- cfg$INPUT_DATA$type
filtering_type <- ifelse(cfg$SAVE_DATASET_SENSITIVE_FILTERING, "dataset-sensitive", "custom")
```

***

```{r, child = here::here("Rmd/common/_header.Rmd")}
```

***

```{r, results = "asis"}
if (input_type == "cellranger") {
  md_header("Input data: 10x Genomics Cell Ranger data", 1)
  cat("The feature-barcode matrix was imported from Cell Ranger output (the official quantification tool from 10x Genomics).")
} else if (input_type == "table") {
  md_header("Input data: delimited text (table)", 1)
  cat("The feature-barcode matrix was imported from delimited file.")
} else if (input_type == "sce") {
  md_header("Input data: `SingleCellExperiment` object", 1)
  cat("The object holding experimental data (feature-barcode matrix, gene annotation etc.) was imported from Rds file.")
}
```

Each row of feature-barcode matrix corresponds to a gene, while each column corresponds to a cell barcode.
Summary of imported data:

```{r}
cat(readd(sce_raw_info, path = drake_cache_dir)$str)
```

***

# Empty droplets

Empty droplets often contain RNA from the ambient solution, resulting in non-zero counts after debarcoding.
It is desired to discard such droplets.

## Barcode rank plot

A useful diagnostic for droplet-based data is the barcode rank plot, which shows the total UMI (log-)count for each
barcode on the x-axis and the (log-)rank on the y-axis.
This is effectively a transposed empirical cumulative density plot with log-transformed axes. 
It is useful as it allows examine the distribution of total UMI counts across barcodes, focusing on those with the largest counts.

```{r, message = FALSE, warning = FALSE, results = "hold"}
uniq <- !duplicated(barcode_ranks$rank)
plot(barcode_ranks$rank[uniq], barcode_ranks$total[uniq], log = "xy", xlab = "Rank", ylab = "Total")
o <- order(barcode_ranks$rank)
lines(barcode_ranks$rank[o], barcode_ranks$fitted[o], col = "red")

abline(h = metadata(barcode_ranks)$knee, col = "dodgerblue", lty = 2)
abline(h = metadata(barcode_ranks)$inflection, col = "forestgreen", lty = 2)
legend("bottomleft", lty = 2, col = c("dodgerblue", "forestgreen"), legend = c("knee", "inflection"))
```

The knee and inflection points on the curve mark the transition between two components of the total UMI count distribution.
This is assumed to represent the difference between empty droplets with little RNA and cell-containing droplets with much more RNA.

```{r, results = "asis", eval = !cfg$EMPTY_DROPLETS_ENABLED}
cat("Removal of empty droplets was disabled. You can enable it by setting `EMPTY_DROPLETS_ENABLED` parameter to `TRUE`.")
```

```{r, child = here::here("Rmd/single_sample/01_input_qc_empty_droplets.Rmd"), eval = cfg$EMPTY_DROPLETS_ENABLED}
```

***

# Gene + Cell quality filtering 

## Pre-filtering QC

Given sets of mitochondrial and ribosomal genes in the data, the `scater` package automatically calculates
several per-cell QC metrics:

- Number of UMI.
- Number of detected genes (non-zero UMI count).
- Percentage of expressed mitochondrial genes ($\frac {UMI_{mitochondrial}} {UMI_{sum}} * 100$).

A standard approach is to filter cells with low amount of reads as well as genes that are present in
at least a certain amount of cells, using fixed thresholds (here referred to as **custom filtering**).
While simple, using fixed thresholds requires knowledge of the experiment and of the experimental protocol.

An alternative approach is to use adaptive, data-driven thresholds to identify outlying cells,
based on the set of QC metrics just calculated, here referred to as **dataset-sensitive filtering**.
We identify cells that are outliers for the various QC metrics, based on the median absolute deviation (MAD)
from the median value of each QC metric across all cells. Specifically, a value is considered an outlier if it
is more than `r cfg$MAD_THRESHOLD` MADs from the median in the "problematic" direction.

Additionaly, extremely high number of detected genes could indicate doublets (more sensitive doublet detection is
done after library normalization). However, depending on the cell type composition in your sample,
you may have cells with higher number of genes (and also higher counts) from one cell type.

Now we can plot some of the QC features. Cells are colored by `discard_qc`, meaning if a cell would be discarded by
MAD thresholding on a QC metric.

```{r, fig.height = 10, fig.width = 8}
cowplot::plot_grid(plotlist = readd(sce_unfiltered_plotlist, path = drake_cache_dir), ncol = 2)
```

## Filtering {.tabset}

### Dataset-sensitive filters

#### Cell filtering

Filter cells based on QC metrics and MAD threshold (`r cfg$MAD_THRESHOLD`):

- Low number of UMI (lower tail).
- Low number of detected genes (lower tail).
- High expression of mitochondrial genes (upper tail).

**Removing `r sum(qc_filter)` low quality cells based on MAD.**

#### Gene filtering

We excluded genes that are not expressed in our system and don't contribute any information to our experiment.
Very lowly expressed genes may only contribute noise.

Table of zero-expression genes count:

```{r}
janitor::tabyl(sce_qc_filter_rowSums == 0) %>%
  dplyr::rename(zero_expression = `sce_qc_filter_rowSums == 0`) %>%
  dplyr::mutate(percent = scales::percent(percent)) %>%
  render_bootstrap_table(full_width = FALSE, position = "left")
```

**Removing `r sum(!readd(sce_qc_gene_filter, path = drake_cache_dir))` genes with UMI per cell less than
`r cfg$MIN_UMI` and expressed in less than `r cfg$MIN_RATIO_CELLS * 100` % of all cells.**

Info on filtered dataset:

```{r}
cat(readd(sce_qc_filter_genes_info, path = drake_cache_dir)$str)
```

### Custom filters

#### Cell filtering

Filter dataset based on custom (fixed) thresholds of QC metrics:

- &lt;Min; Max&gt; UMI per cell: &lt;`r cfg$MIN_UMI_CF`; `r as.character(cfg$MAX_UMI_CF)`&gt;
- Min. number of features (genes) detected: `r cfg$MIN_FEATURES`
- Max. ratio of mitochondrial genes expression: `r cfg$MAX_MITO_RATIO`

**Removing `r sum(custom_filter)` low quality cells based on custom thresholds.**

#### Gene filtering

Gene filtering is equal to dataset-sensitive filtering, except for the cell count and proportion may change.
We excluded genes that are not expressed in our system and don't contribute any information to our experiment.
Very lowly expressed genes may only contribute noise.

Table of zero-expression genes count:

```{r}
janitor::tabyl(sce_custom_filter_rowSums == 0) %>%
  dplyr::rename(zero_expression = `sce_custom_filter_rowSums == 0`) %>%
  dplyr::mutate(percent = scales::percent(percent)) %>%
  render_bootstrap_table(full_width = FALSE, position = "left")
```

**Removing `r sum(!readd(sce_qc_gene_filter, path = drake_cache_dir))` genes with UMI per cell less than
`r cfg$MIN_UMI` and expressed in less than `r cfg$MIN_RATIO_CELLS * 100` % of all cells.**

Info on custom filtered dataset:

```{r}
cat(readd(sce_custom_filter_genes_info, path = drake_cache_dir)$str)
```

***

## Post-filtering QC 

**Final filtering selection: using <span style='color: green;'>`r filtering_type`</span> filtering.**

```{r}
cat(readd(sce_final_input_qc_info, path = drake_cache_dir)$str)
```

### Cell and gene number history

```{r}
render_bootstrap_table(readd(sce_history, path = drake_cache_dir), full_width = FALSE, position = "left")
```

```{r}
print(readd(sce_history_plot, path = drake_cache_dir))
```

### Dataset-sensitive filtering

Plots of QC metrics after dataset-sensitive filtering.
`discard_custom` means if given cell was discarded in **custom filtering**.

```{r, fig.height = 10, fig.width = 8}
cowplot::plot_grid(plotlist = readd(sce_qc_filter_genes_plotlist, path = drake_cache_dir), ncol = 2)
```

### Filtering based on custom filters

Plots of QC metrics after custom filtering.
`discard_qc` means if given cell was discarded in **dataset-sensitive filtering**.

```{r, fig.height = 10, fig.width = 8}
cowplot::plot_grid(plotlist = readd(sce_custom_filter_genes_plotlist, path = drake_cache_dir), ncol = 2)
```

***

# Gene annotation

- Used annotation package: `r config_main$ANNOTATION_PKG`
  (v`r sessioninfo::package_info(config_main$ANNOTATION_PKG, dependencies = FALSE)$loadedversion`)
- If a single ENSEMBL ID has multiple symbols, gene descriptions, or ENTREZ IDs, they are collapsed by comma (`,`).
- ENSEMBL ID is used as a symbol for ENSEMBL IDs with unknown symbols.
- ENSEMBL ID is appended to symbols having multiple ENSEMBL IDs (e.g. TBCE has both ENSG00000285053 and ENSG00000284770
  ENSEMBL IDs assigned -> its symbol is changed to TBCE_ENSG00000285053 and TBCE_ENSG00000284770).

```{r}
readd(gene_annotation, path = drake_cache_dir) %>%
  head() %>%
  render_bootstrap_table()
```

#

***

<details>
  <summary class="config">Show input parameters</summary>
  <hr />
  <h4>Main config</h4>

```{r}
print(config_main)
```
  
  <hr />
  <h4>Input QC config</h4>

```{r}
print(cfg)
```
  <hr />
</details>

```{r, child = here::here("Rmd/common/_footer.Rmd")}
```
