---
title: "scdrake | 02 - Normalization and clustering"
author: "Jiri Novotny, Jan Kubovciak"
institute: "Laboratory of Genomics and Bioinformatics @ Institute of Molecular Genetics of the ASCR"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    number_sections: false
    theme: "flatly"
    self_contained: true
    code_download: true
    df_print: "paged"
params:
  css_file: !expr here::here("Rmd/common/stylesheet.css")
  drake_cache_dir: !expr here::here(".drake")
date: "`r glue('Document generated: {Sys.Date()}')`"
css: "`r params$css_file`"
---

```{r, include = FALSE, message = FALSE, warning = FALSE}
library(drake)
if (rlang::is_true(getOption("knitr.in.progress"))) {
  params_ <- scdrake_list(params)
}
drake_cache_dir <- params_$drake_cache_dir

loadd(
  config_main, config_norm_clustering, doublet_score,
  cluster_kmeans_kc_tables, cluster_sc3_tables, cluster_sc3_stability_plots,
  dimred_plots_clustering, dimred_plots_other_vars,
  selected_markers_plots, selected_markers_plots_files,
  path = drake_cache_dir
)

cfg <- config_norm_clustering

sce <- readd(sce_final_norm_clustering, path = drake_cache_dir)
sce_metadata <- metadata(sce)
sce_colData <- colData(sce)
sce_rowData <- rowData(sce)
normalization_method <- sce_metadata$normalization_type
hvg_metric <- sce_metadata$hvg_metric
hvg_selection <- sce_metadata$hvg_selection
hvg_selection_value <- sce_metadata$hvg_selection_value

selected_markers_plots_final <- dplyr::bind_cols(
  selected_markers_plots,
  selected_markers_plots_files = selected_markers_plots_files
)
```

***

```{r, child = here::here("Rmd/common/_header.Rmd")}
```

***

# Input data overview

Just to review data from the preceding pipeline step (`01 - quality control`):

```{r}
cat(readd(sce_final_input_qc_info, path = drake_cache_dir)$str)
```

***

# Cell cycle phase assignment

Assign each cell a score, based on its expression of G2/M and S phase markers.
These marker sets should be anticorrelated in their expression levels, and cells expressing neither are likely not cycling and in G1 phase.

# Normalization

Normalization method: "`r cfg$NORMALIZATION_TYPE`"

```{r, results = "asis"}
if (normalization_method == "scran") {
  cat("**`scran`: normalization by deconvolution**\n\n")
  cat(
    "Cell-specific biases are normalized using the `computeSumFactors()` method,",
    "which implements the deconvolution strategy for scaling normalization (*A. T. Lun, Bach, and Marioni 2016*).",
    "This computes size factors that are used to scale the counts in each cell.",
    "The assumption is that most genes are not differentially expressed (DE) between cells,",
    "such that any differences in expression across the majority of genes represents some technical bias that should be removed."
  )
} else if (normalization_method == "sctransform") {
  cat("**`sctransform`: regularized negative binomial regression to normalize UMI count data.**\n\n")
}
```

# Highly variable genes (HVGs) selection

```{r, results = "asis"}
catg0('**HVG metric: "{hvg_metric}"**\n\n')

if (hvg_metric == "gene_var") {
  var_field <- "bio"
  
  cat(
    "`modelGeneVar()` models the variance of the log-expression profiles for each gene,",
    "decomposing it into technical and biological components based on a fitted mean-variance trend.\n\n"
  )
  
  plot_hvg_fit(sce, "var")
} else if (hvg_metric == "gene_cv2") {
  var_field <- "ratio"
  
  cat(
    "`modelGeneCV2()` models the squared coefficient of variation (CV2) of the normalized expression profiles for each gene,",
    "fitting a trend to account for the mean-variance relationship across genes.\n\n"
  )
  
  plot_hvg_fit(sce, "cv2")
} else if (hvg_metric == "sctransform") {
  catg0("HVGs (n = {cfg$SCT_N_HVG}) were selected by the `sctransform` method.")
}

if (!is_null(sce_metadata$hvg_rm_cc_genes) && sce_metadata$hvg_rm_cc_genes) {
  catg0(
    "Using the percentage of variance explained by the cell cycle phase in the expression profile for each gene, ",
    "we removed the following genes with percentage > {sce_metadata$hvg_cc_genes_var_expl_threshold} prior to HVG selection:"
  )
  
  render_bootstrap_table(sce_rowData[sce_rowData$is_cc_related, c("ENSEMBL", "SYMBOL")], row.names = FALSE)
}

if (hvg_metric %in% c("gene_var", "gene_cv2")) {
  catg0('Based on "{hvg_metric}", HVGs were selected by: ')
  
  if (hvg_selection == "top") {
    catg0("top {hvg_selection_value} HVGs.\n\n")
  } else if (hvg_selection == "significance") {
    catg0("FDR < {hvg_selection_value}\n\n")
  } else if (hvg_selection == "threshold") {
    catg0("variance or CV2 > {hvg_selection_value}\n\n")
  }
  
  catg0("**Found {length(sce_metadata$hvg_ids)} HVGs.**\n\n")
}
```

Variance explained plot:

```{r}
scater::plotExplanatoryVariables(readd(variance_explained, path = drake_cache_dir))
```

Plot of HVGs:

```{r}
readd(hvg_plot, path = drake_cache_dir)
```

***

# Doublet score assignment

The `doubletCluster()` function identifes clusters with expression profiles lying between two other clusters.
Considering every possible triplet of clusters, the method uses the number of DE genes, the median library size,
and the proporion of cells in the cluster to mark clusters as possible doublets.

```{r, results = "asis"}
if (cfg$NORMALIZATION_TYPE == "scran" && cfg$SCRAN_USE_QUICKCLUSTER) {
  cat("Prior to normalization, quick clustering was performed. We can use these cluster to look at doublet score within them:\n\n")
  boxplot(doublet_score ~ cluster_quickcluster, data = sce_colData)
}
```

```{r, results = "asis"}
if (sce_metadata$has_filtered_doublets) {
  n_doublets <- sum(sce_colData$is_doublet)
  doublets_pct <- (n_doublets / ncol(sce)) * 100
  catn(glue("Discarded {n_doublets} cells ({doublets_pct} % of all cells) with doublet score above {sce_metadata$max_doublet_score}"))
} else {
  catn(glue("Cells were not filtered by doublet score."))
}
```

***

# PCA {.tabset}

## PCA plots {.tabset}

### By cell phase

```{r}
patchwork::wrap_plots(readd(pca_phase_plots, path = drake_cache_dir), nrow = 1, guides = "collect")
```

### By doublet score

```{r}
readd(pca_doublet_plot, path = drake_cache_dir)
```

### By total UMI

```{r}
readd(pca_total_plot, path = drake_cache_dir)
```

## PCs selection

- Elbow point method: a simple heuristic for choosing PCs involves identifying the elbow point in the percentage of
  variance explained by successive PCs. This refers to the "elbow" in the curve of a scree plot as shown.
- Technical variance method: use the technical component estimates to determine the proportion of variance that should be retained.
  This is implemented in `denoisePCA()`, which takes the estimates returned by `modelGeneVar()`.
- Forced: use a predefined number of PCs.

```{r}
readd(pca_selected_pcs_plot, path = drake_cache_dir)
```

**`r sce_metadata$pca_selected_pcs` PCs were selected using the "`r sce_metadata$pca_selection_method`" method**

***

# Clustering {.tabset}

## Graph-based

Graph-based clustering is commonly used for scRNA-seq, and often shows a good performance.

First, we will use `scran` to generate the shared nearest neighbor graph,
which will then be subjected to community detection using algorithms implemented in the `igraph` package.

Once the SNN graph is generated, we can use any of the community detection algorithms in `igraph` to find the clusters.
Here, we illustrate two of these methods; the walktrap algorithm (*Pons and Latapy 2005*) and the Louvain method (*Blondel et al. 2008*).
We can visualise similarity of both assignments with heatmap and PCA.

```{r}
pheatmap::pheatmap(table(readd(cluster_graph_walktrap, path = drake_cache_dir)[[1]], readd(cluster_graph_louvain, path = drake_cache_dir)[[1]]))
```

**Walktrap algorithm: found `r readd(cluster_graph_walktrap_n, path = drake_cache_dir)` clusters.**

```{r, results = "asis"}
cells_per_cluster_table_collapsed_html(readd(cluster_graph_walktrap_table, path = drake_cache_dir))
```

**Louvain algorithm: found `r readd(cluster_graph_louvain_n, path = drake_cache_dir)` clusters.**

```{r, results = "asis"}
cells_per_cluster_table_collapsed_html(readd(cluster_graph_louvain_table, path = drake_cache_dir))
```

## k-means

k-means is a generic clustering algorithm that has been used in many application areas.
In R, it can be applied via the `kmeans()` function.
Typically, it is applied to a reduced dimension representation of the expression data
(most often PCA, because of the interpretability of the low-dimensional distances).
We need to define the number of clusters in advance.

### Best K

Determining the number of clusters from the gap statistics and their standard deviations.

```{r}
readd(kmeans_gaps_plot, path = drake_cache_dir)
```

**The best K selected: `r readd(kmeans_best_k, path = drake_cache_dir)`**

```{r, results = "asis"}
cells_per_cluster_table_collapsed_html(readd(cluster_kmeans_kbest_table, path = drake_cache_dir))
```

### Custom K {.tabset}

- **Custom K (`KMEANS_K`): `r str_comma(cfg$KMEANS_K)`**

```{r, results = "asis"}
for (k in cfg$KMEANS_K) {
  cluster_col <- glue("cluster_kmeans_kc_{k}")
  md_header(glue("k = {k}"), 4)
  cells_per_cluster_table_collapsed_html(cluster_kmeans_kc_tables[[cluster_col]])
}
```

## SC3 {.tabset}

Single-Cell Consensus Clustering (`SC3`) is a tool for unsupervised clustering of scRNA-seq data.
SC3 achieves high accuracy and robustness by consistently integrating different clustering solutions through a consensus approach.

*Cluster stability index* shows how stable each cluster is accross the selected range of `k`s.
The stability index varies between 0 and 1, where 1 means that the same cluster appears in every solution for different `k`.

```{r, results = "asis"}
for (k in cfg$SC3_K) {
  cluster_col <- glue("cluster_sc3_{k}")
  md_header(glue("k = {k}"), 3)
  cells_per_cluster_table_collapsed_html(cluster_sc3_tables[[cluster_col]])
  print(cluster_sc3_stability_plots[[cluster_col]])
}
```

#

***

```{r, results = "asis"}
generate_dimred_plots_section(
  dimred_names = cfg$NORM_CLUSTERING_REPORT_DIMRED_NAMES,
  clustering_names = cfg$NORM_CLUSTERING_REPORT_CLUSTERING_NAMES,
  selected_markers = selected_markers_plots_final,
  dimred_plots_clustering = dimred_plots_clustering,
  dimred_plots_other_vars = dimred_plots_other_vars,
  selected_markers_files_rel_start = fs::path_dir(cfg$NORM_CLUSTERING_REPORT_HTML_FILE)
)
```

#

***

<details>
  <summary class="config">Show input parameters</summary>
  <hr />
  <h4>Main config</h4>

```{r}
print(config_main)
```

  <hr />
  <h4>Normalization and clustering config</h4>

```{r}
print(cfg)
```
  <hr />
</details>

```{r, child = here::here("Rmd/common/_footer.Rmd")}
```
