<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="01_input_qc stage of the single-sample pipeline.">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/5/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Input and quality control stage (`01_input_qc`) • scdrake</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Input and quality control stage (`01_input_qc`)">
<meta property="og:description" content="01_input_qc stage of the single-sample pipeline.">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/5/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/5/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="../darkswitch.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scdrake</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.5.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake_integration.html">Integration pipeline guide</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Guides</h6>
    <a class="dropdown-item" href="../articles/scdrake_docker.html">Using the Docker image</a>
    <a class="dropdown-item" href="../articles/scdrake_integration.html">02 Integration pipeline guide</a>
    <a class="dropdown-item" href="../articles/scdrake_advanced.html">Advanced topics</a>
    <a class="dropdown-item" href="../articles/scdrake_extend.html">Extending the pipeline</a>
    <a class="dropdown-item" href="../articles/drake_basics.html">drake basics</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General information</h6>
    <a class="dropdown-item" href="../articles/pipeline_overview.html">Pipeline overview</a>
    <a class="dropdown-item" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
    <a class="dropdown-item" href="../articles/scdrake_cli.html">Command line interface</a>
    <a class="dropdown-item" href="../articles/scdrake_config.html">scdrake config internals</a>
    <a class="dropdown-item" href="../articles/scdrake_envvars.html">Environment variables</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General configs</h6>
    <a class="dropdown-item" href="../articles/config_pipeline.html">Pipeline config</a>
    <a class="dropdown-item" href="../articles/config_main.html">Main config</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single-sample pipeline stages</h6>
    <a class="dropdown-item" href="../articles/stage_input_qc.html">Input and quality control stage (`01_input_qc`)</a>
    <a class="dropdown-item" href="../articles/stage_norm_clustering.html">Normalization and clustering stage (`02_norm_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Integration pipeline stages</h6>
    <a class="dropdown-item" href="../articles/stage_integration.html">Integration stage (`01_integration`)</a>
    <a class="dropdown-item" href="../articles/stage_int_clustering.html">Post-integration clustering stage (`02_int_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Common stages</h6>
    <a class="dropdown-item" href="../articles/stage_cluster_markers.html">Cluster markers stage (`cluster_markers`)</a>
    <a class="dropdown-item" href="../articles/stage_contrasts.html">Contrasts stage (`contrasts`)</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bioinfocz/scdrake" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>

        
        <li>
          <a class="external-link nav-link" id="css-toggle-btn" aria-label="github">
            <span class="fas fa fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        
        
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Input and quality control stage (`01_input_qc`)</h1>
            
            <h4 data-toc-skip class="date"><sup>Document generated:
2023-12-19 06:17:14 UTC+0000</sup></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bioinfocz/scdrake/blob/mastervignettes/stage_input_qc.Rmd" class="external-link"><code>vignettes/stage_input_qc.Rmd</code></a></small>
      <div class="d-none name"><code>stage_input_qc.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2 tabset">
<h2 class="tabset" id="section"></h2>



<ul class="nav nav-tabs" id="section" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#overview" id="overview-tab" type="button" role="tab" aria-controls="overview" aria-selected="true" class="active nav-link">Overview</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#config-parameters" id="config-parameters-tab" type="button" role="tab" aria-controls="config-parameters" aria-selected="false" class="nav-link">Config parameters</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#outputs" id="outputs-tab" type="button" role="tab" aria-controls="outputs" aria-selected="false" class="nav-link">Outputs</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="overview" role="tabpanel" aria-labelledby="overview-tab">

<p>An initial stage to read in scRNA-seq data, remove empty droplets
(optionally), and perform quality control, and cell and gene
filtering.</p>
<p>More information about scRNA-seq quality control can be found in <a href="https://bioconductor.org/books/3.15/OSCA.basic/quality-control.html" class="external-link">OSCA</a>.</p>
<p>⚙️ Config file:
<code>config/single_sample/01_input_qc.yaml</code></p>
<p>📋 HTML report target (in <code>config/pipeline.yaml</code>):
<code>DRAKE_TARGETS: ["report_input_qc"]</code></p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/single_sample/pbmc1k/01_input_qc/01_input_qc.html" class="external-link">📜
Example report</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/single_sample_pbmc1k_full/01_input_qc.default.yaml" class="external-link">used
config</a>)</p>
<p>🪜 Structure</p>
<ul>
<li>Reading in feature-barcode matrix either from:
<ul>
<li>
<code>cellranger</code> output</li>
<li>
<code>SingleCellExperiment</code> object as a Rds file or as a
target from an existing <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cache (i.e. from a
different <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> project)</li>
<li>Delimited table</li>
</ul>
</li>
<li>Subsetting of the imported <code>SingleCellExperiment</code> object
to cells of interest based on their metadata values, e.g. a set of
clusters from a particular clustering method (optional)</li>
<li>Removal of <a href="https://bioconductor.org/books/3.15/OSCA.advanced/droplet-processing.html" class="external-link">empty
droplets</a> (<code><a href="https://rdrr.io/pkg/DropletUtils/man/emptyDrops.html" class="external-link">DropletUtils::emptyDrops()</a></code>) (optional)</li>
<li>Calculation of per-cell QC metrics
(<code><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">scater::perCellQCMetrics()</a></code>, <a href="https://bioconductor.org/books/3.15/OSCA.basic/quality-control.html#common-choices-of-qc-metrics" class="external-link">theory</a>):
<ul>
<li>Number of UMI</li>
<li>Number of detected genes (non-zero UMI count)</li>
<li>Percentage of expressed mitochondrial genes (<span class="math inline">\(\frac {UMI_{mitochondrial}} {UMI_{sum}} *
100\)</span>)</li>
</ul>
</li>
<li>Two methods for cell filtering (optional):
<ul>
<li>Dataset-sensitive filtering, based on the median absolute deviation
(MAD) from the median value of each QC metric across all cells (<a href="https://bioconductor.org/books/3.15/OSCA.basic/quality-control.html#quality-control-outlier" class="external-link">theory</a>).
For “number of UMI” metric, only a lower tail is used.</li>
<li>Custom filtering, based on fixed thresholds. For “number of UMI”
metric, both upper and lower bounds are used (<a href="https://bioconductor.org/books/3.15/OSCA.basic/quality-control.html#fixed-qc" class="external-link">theory</a>.</li>
<li>For both filtering types you can choose how to join the filters:
either jointly (using the <em>AND</em> operator) or individually (using
the <em>OR</em> operator).</li>
</ul>
</li>
<li>Gene filtering (optional) based on a minimum number of UMI per cell
and a minimum ratio of cells expressing a gene.
<ul>
<li>A gene is considered expressed when number of its UMI across all
cells is greater than X and at the same time it is expressed in at least
Y ratio of cells.</li>
</ul>
</li>
<li>Plots of QC metrics with visualization of differences between
dataset-sensitive and custom filtering</li>
<li>Preparation of gene annotation</li>
<li>Final selection of either dataset-sensitive or custom filtered
dataset. This selection is upon the user.</li>
</ul>
</div>
<div class="tab-pane" id="config-parameters" role="tabpanel" aria-labelledby="config-parameters-tab">

<p>Config for this stage is stored in
<code>config/single_sample/01_input_qc.yaml</code> file. Directory with
this file is read from the <code>SCDRAKE_SINGLE_SAMPLE_CONFIG_DIR</code>
environment variable upon <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> load or attach, and
saved as <code>scdrake_single_sample_config_dir</code> option. This
option is used as the default argument value in several
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> functions.</p>
<hr>
<div class="section level4">
<h4 id="input-files">Input files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h4>
<div class="sourceCode" id="cb1"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_DATA</span><span class="kw">:</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">type</span><span class="kw">:</span><span class="at"> </span><span class="st">"cellranger"</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">path</span><span class="kw">:</span><span class="at"> </span><span class="st">"/path/to/dir"</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">delimiter</span><span class="kw">:</span><span class="at"> </span><span class="st">","</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">target_name</span><span class="kw">:</span><span class="at"> </span><span class="st">"target_name"</span></span></code></pre></div>
<p><strong>Type:</strong> list of named character scalars</p>
<p>This parameter is specifying input data type and path. There are four
possible types of input data:</p>
<ul>
<li>
<p><code>type: "cellranger"</code>: an output from <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">cellranger</a>,
that is, files of a raw feature-barcode matrix -
<code>barcodes.tsv</code>, <code>features.tsv</code>, and
<code>matrix.mtx</code> (can be also gzipped having a <code>.gz</code>
extension). Internally, BC matrix is imported via
<code><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html" class="external-link">DropletUtils::read10xCounts()</a></code></p>
<ul>
<li>You can download example PBMC data from 10x Genomics using the
<code><a href="../reference/download_example_data.html">download_pbmc1k()</a></code> and <code><a href="../reference/download_example_data.html">download_pbmc1k()</a></code>
functions.</li>
</ul>
</li>
<li><p><code>type: "table"</code>: a delimited file (table) representing
a feature-barcode matrix, with an additional column named
<code>ENSEMBL</code> containing Ensembl IDs of features (rows).</p></li>
<li>
<p><code>type: "sce"</code>: a saved
<code>SingleCellExperiment</code> in Rds format (i.e. saved by the
<code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS()</a></code> function). The object must contain
feature-barcode matrix with raw UMI counts in the <code>counts</code>
slot of <code>assays</code> (such that this BC matrix can be retrieved
by <code>counts(sce)</code>).</p>
<ul>
<li>Additionaly, you can include whatever you want in
<code>colData()</code> and use it later (e.g. cell grouping/clusters to
compute markers or to include batch effect during that).</li>
</ul>
</li>
<li><p><code>type: "sce_drake_cache"</code>: a saved
<code>SingleCellExperiment</code> in a <code>drake</code> cache,
e.g. from an another <code>scdrake</code> project</p></li>
<li>
<p><code>path: "path/to/file/or/dir"</code>: a path to input file or
directory (can be relative to project root directory).</p>
<ul>
<li>For <code>type: "cellranger"</code>, this is a path to directory
containing the <code>barcodes.tsv</code>, <code>features.tsv</code>, and
<code>matrix.mtx</code> files (can be also gzipped having a
<code>.gz</code> extension)</li>
<li>For <code>type: "sce_drake_cache"</code>, a path to
<code>drake</code> cache directory (usually named as
<code>".drake"</code>)</li>
<li>For <code>type: "sce"</code> or <code>"table"</code>, a path to Rds
or text file</li>
</ul>
</li>
<li><p><code>delimiter: ","</code>: used when
<code>type: "table"</code>. Specifies the field delimiter in the
table.</p></li>
<li><p><code>target_name: "target_name"</code>: used when
<code>type: "sce_drake_cache"</code>. Specifies a name of target in
<code>drake</code> cache to be imported,
e.g. <code>"sce_final_input_qc"</code></p></li>
</ul>
<hr>
<div class="sourceCode" id="cb2"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_QC_REPORT_RMD_FILE</span><span class="kw">:</span><span class="at"> </span><span class="st">"Rmd/single_sample/01_input_qc.Rmd"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>A path to RMarkdown file used for HTML report of this pipeline
stage.</p>
<hr>
</div>
<div class="section level4">
<h4 id="subsetting-of-imported-data">Subsetting of imported data<a class="anchor" aria-label="anchor" href="#subsetting-of-imported-data"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_DATA_SUBSET</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co">## Example of simple subsetting by cluster numbers</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_DATA_SUBSET</span><span class="kw">:</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">subset_by</span><span class="kw">:</span><span class="at"> </span><span class="st">"cluster_kmeans_k6"</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">values</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"3"</span><span class="kw">,</span><span class="at"> </span><span class="st">"4"</span><span class="kw">]</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">negate</span><span class="kw">:</span><span class="at"> </span><span class="ch">false</span></span></code></pre></div>
<p><strong>Type:</strong> <code>null</code> (default) or named list</p>
<p>The imported data can be subsetted by a simple selection of values
present in a column of cell metadata (<code>colData()</code>). In the
example above, cells assigned to clusters <code>"3"</code> and
<code>"4"</code> of the <code>cluster_kmeans_k6</code> columns will be
kept. You can also negate the selection by specifying
<code>negate: true</code>.</p>
<hr>
</div>
<div class="section level4">
<h4 id="removal-of-empty-droplets">Removal of empty droplets<a class="anchor" aria-label="anchor" href="#removal-of-empty-droplets"></a>
</h4>
<p>See <code><a href="https://rdrr.io/pkg/DropletUtils/man/emptyDrops.html" class="external-link">?DropletUtils::emptyDrops</a></code> for more details.</p>
<hr>
<div class="sourceCode" id="cb4"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">EMPTY_DROPLETS_ENABLED</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>False</code>, skip calculation and removal of empty
droplets.</p>
<p>You might consider turning off this procedure if the input is not a
raw feature-barcode matrix from <code>cellranger</code>, but an already
processed dataset (table or <code>SingleCellExperiment</code> object).
Otherwise <code><a href="https://rdrr.io/pkg/DropletUtils/man/emptyDrops.html" class="external-link">DropletUtils::emptyDrops()</a></code> will fail as there are
not enough empty droplets having the total UMI count &lt;
<code>EMPTY_DROPLETS_LOWER</code> (100 by default).</p>
<hr>
<div class="sourceCode" id="cb5"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">EMPTY_DROPLETS_LOWER</span><span class="kw">:</span><span class="at"> </span><span class="dv">100</span></span></code></pre></div>
<p><strong>Type:</strong> positive integer scalar</p>
<p>A lower bound on the total UMI count at or below which all barcodes
are assumed to correspond to empty droplets.</p>
<hr>
<div class="sourceCode" id="cb6"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">EMPTY_DROPLETS_FDR_THRESHOLD</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.01</span></span></code></pre></div>
<p><strong>Type:</strong> numeric scalar &lt;0; 1&gt;</p>
<p>A threshold for FDR adjusted p-values for null hypothesis that
barcode comes from ambient environment. In other words, probability that
a droplet was empty and contained ambient RNA.</p>
</div>
<div class="section level4">
<h4 id="cell-filtering">Cell filtering<a class="anchor" aria-label="anchor" href="#cell-filtering"></a>
</h4>
<p>Two types of cell filtering are performed: <em>dataset-sensitive</em>
(using MAD threshold), and <em>custom</em> (using custom
thresholds).</p>
<p>In both filtering types, violation of only one metric threshold leads
to removal of a cell.</p>
<p>The choice of cell filtering thresholds depends very much on your
dataset. What works for an ordinary high-quality dataset might not work
well for a low-quality or special dataset with rare cell types. The <a href="https://bioconductor.org/books/3.15/OSCA.basic/quality-control.html" class="external-link">quality
control</a> chapter in OSCA gives an excellent background to the QC of
scRNA-seq data.</p>
<hr>
<div class="sourceCode" id="cb7"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ENABLE_CELL_FILTERING</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>False</code>, both types of cell filtering will be
disabled.</p>
<p>The filtering thresholds in both filtering types will be overridden
such that every cell will pass the filtering
(e.g. <code>MAD_THRESHOLD: .inf</code> or
<code>MIN_UMI_CF: -.inf</code>). In the end,
<code>SAVE_DATASET_SENSITIVE_FILTERING</code> will have no effect as
both filtered SCE targets will be identical. All QC metrics and plots
will still be calculated and created.</p>
<hr>
<div class="sourceCode" id="cb8"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SAVE_DATASET_SENSITIVE_FILTERING</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>True</code>, proceed to other stages with dataset filtered
by dataset-sensitive filtering (target
<code>sce_qc_filter_genes</code>), otherwise with dataset filtered by
custom filtering (target <code>sce_custom_filter_genes</code>). Note
that the selected SCE target will be referred to as
<code>sce_final_input_qc</code> in subsequent stages.</p>
<div class="section level5">
<h5 id="dataset-sensitive-cell-filtering">Dataset-sensitive cell filtering<a class="anchor" aria-label="anchor" href="#dataset-sensitive-cell-filtering"></a>
</h5>
<div class="sourceCode" id="cb9"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MAD_THRESHOLD</span><span class="kw">:</span><span class="at"> </span><span class="dv">3</span></span></code></pre></div>
<p><strong>Type:</strong> positive numeric scalar</p>
<p>A threshold for maximum MAD (median absolute deviation) for QC cell
metrics:</p>
<ul>
<li>Number of UMI (lower tail).</li>
<li>Number of detected genes = non-zero UMI count (lower tail).</li>
<li>Percentage of expressed mitochondrial genes = <span class="math inline">\(\frac {UMI_{mitochondrial}} {UMI_{sum}} *
100\)</span> (upper tail).</li>
</ul>
<p>MAD threshold of 3 will retain 99% of non-outlier values that follow
a normal distribution.</p>
<p>“Lower tail” means cells having a metric value less than
<code>-MAD_THRESHOLD</code> MAD will be discarded. For “upper tail”, it
is <code>+MAD_THRESHOLD</code> MAD.</p>
<p>Violation of only one metric threshold leads to removal of a
cell.</p>
<p>To disable the dataset-sensitive filtering, set
<code>MAD_THRESHOLD: .inf</code>. That will force passing of each cell
as every QC metric will be always lower than positive infinity MAD.</p>
<hr>
<div class="sourceCode" id="cb10"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DATASET_SENSITIVE_FILTERS_OPERATOR</span><span class="kw">:</span><span class="at"> </span><span class="st">"&amp;"</span></span></code></pre></div>
<p><strong>Type:</strong> a character scalar
(<code>"&amp;" | "|"</code>)</p>
<p>How to join the QC filters:</p>
<ul>
<li>Jointly (<em>AND</em>/<code>&amp;</code> operator), i.e., remove
only cells that violate <strong>ALL</strong> filters (permissive)</li>
<li>Individually (<em>OR</em>/<code>|</code> operator), i.e., remove
cells that violate <strong>AT LEAST ONE</strong> filter (strict)</li>
</ul>
</div>
<div class="section level5">
<h5 id="custom-cell-filtering">Custom cell filtering<a class="anchor" aria-label="anchor" href="#custom-cell-filtering"></a>
</h5>
<div class="sourceCode" id="cb11"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MIN_UMI_CF</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span></code></pre></div>
<p><strong>Type:</strong> positive integer scalar</p>
<p>A threshold for minimum number of UMI per cell, i.e. cells with UMI
less than <code>MIN_UMI_CF</code> will be removed.</p>
<p>To disable this filter, set <code>MIN_UMI_CF: -.inf</code></p>
<hr>
<div class="sourceCode" id="cb12"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MAX_UMI_CF</span><span class="kw">:</span><span class="at"> </span><span class="dv">50000</span></span></code></pre></div>
<p><strong>Type:</strong> positive integer scalar</p>
<p>A threshold for maximum number of UMI per cell, i.e. cells with UMI
greater than <code>MAX_UMI_CF</code> will be removed.</p>
<p>To disable this filter, set <code>MAX_UMI_CF: .inf</code></p>
<hr>
<div class="sourceCode" id="cb13"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MIN_FEATURES</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span></code></pre></div>
<p><strong>Type:</strong> positive integer scalar</p>
<p>A threshold for minimum number of features (genes) detected per cell,
i.e. cells with detected features less than <code>MIN_FEATURES</code>
will be removed.</p>
<p>To disable this filter, set <code>MIN_FEATURES: -.inf</code></p>
<hr>
<div class="sourceCode" id="cb14"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MAX_MITO_RATIO</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.2</span></span></code></pre></div>
<p><strong>Type:</strong> numeric scalar &lt;0; 1&gt;</p>
<p>A threshold for maximum ratio of expressed mitochondrial genes per
cell, i.e. cells with mitochondrial genes detected in more than
(<code>MAX_MITO_RATIO</code> * 100)% of all genes will be removed.</p>
<p>To disable this filter, set <code>MAX_MITO_RATIO: .inf</code></p>
<hr>
<div class="sourceCode" id="cb15"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">CUSTOM_FILTERS_OPERATOR</span><span class="kw">:</span><span class="at"> </span><span class="st">"&amp;"</span></span></code></pre></div>
<p><strong>Type:</strong> a character scalar
(<code>"&amp;" | "|"</code>)</p>
<p>How to join the QC filters:</p>
<ul>
<li>Jointly (<em>AND</em>/<code>&amp;</code> operator), i.e., remove
only cells that violate <strong>ALL</strong> filters (permissive)</li>
<li>Individually (<em>OR</em>/<code>|</code> operator), i.e., remove
cells that violate <strong>AT LEAST ONE</strong> filter (strict)</li>
</ul>
</div>
</div>
<div class="section level4">
<h4 id="gene-filtering">Gene filtering<a class="anchor" aria-label="anchor" href="#gene-filtering"></a>
</h4>
<p>Gene filtering thresholds are applied in both types of cell filtering
(after it is performed). The filter is computed by the
<code><a href="../reference/get_gene_filter.html">get_gene_filter()</a></code> function as following:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">num_cells</span> <span class="op">&lt;-</span> <span class="va">min_ratio_cells</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">is_expressed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu">counts</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">min_umi</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">num_cells</span></span></code></pre></div>
<p>A gene is considered expressed when number of its UMIs across all
cells is greater than <code>min_umi</code> and at the same time it is
expressed in at least <code>min_ratio_cells</code> ratio of cells.</p>
<hr>
<div class="sourceCode" id="cb17"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ENABLE_GENE_FILTERING</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>False</code>, gene filtering will be disabled.</p>
<p>The filtering thresholds will be overridden such that every gene will
pass the filtering (<code>MIN_UMI: 0</code> and
<code>MIN_RATIO_CELLS: 0</code>).</p>
<hr>
<div class="sourceCode" id="cb18"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MITO_REGEX</span><span class="kw">:</span><span class="at"> </span><span class="st">"^MT-"</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">RIBO_REGEX</span><span class="kw">:</span><span class="at"> </span><span class="st">"^RP[SL]"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>Regexes used to match and count occurences of mitochondrial and
ribosomal features. The latter is currently not used for gene
filtering.</p>
<hr>
<div class="sourceCode" id="cb19"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MIN_UMI</span><span class="kw">:</span><span class="at"> </span><span class="dv">1</span></span></code></pre></div>
<p><strong>Type:</strong> zero or positive integer scalar</p>
<p>A threshold for minimum number of UMI per cell, i.e. genes with UMI
&lt; <code>MIN_UMI</code> will be removed.</p>
<hr>
<div class="sourceCode" id="cb20"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MIN_RATIO_CELLS</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.01</span></span></code></pre></div>
<p><strong>Type:</strong> numeric scalar &lt;0; 1&gt;</p>
<p>A minimum ratio of cells expressing a gene, i.e. genes expressed in
less than (<code>MIN_RATIO_CELLS</code> * 100)% of cells will be
removed.</p>
<hr>
</div>
<div class="section level4">
<h4 id="output-files">Output files<a class="anchor" aria-label="anchor" href="#output-files"></a>
</h4>
<div class="sourceCode" id="cb21"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_QC_BASE_OUT_DIR</span><span class="kw">:</span><span class="at"> </span><span class="st">"01_input_qc"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>A path to base output directory for this stage. It will be created
under <code>BASE_OUT_DIR</code> specified in <code>00_main.yaml</code>
config.</p>
<hr>
<div class="sourceCode" id="cb22"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_QC_REPORT_HTML_FILE</span><span class="kw">:</span><span class="at"> </span><span class="st">"01_input_qc.html"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>A name of HTML report file for this stage. Created in
<code>INPUT_QC_BASE_OUT_DIR</code>. Subdirectories are not allowed.</p>
</div>
<div class="section level4">
<h4 id="html-output-parameters">HTML output parameters<a class="anchor" aria-label="anchor" href="#html-output-parameters"></a>
</h4>
<div class="sourceCode" id="cb23"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_QC_KNITR_MESSAGE</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_QC_KNITR_WARNING</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_QC_KNITR_ECHO</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>These are passed to <code><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">knitr::opts_chunk()</a></code> and used for
rendering of stage’s HTML report.</p>
</div>
</div>
<div class="tab-pane" id="outputs" role="tabpanel" aria-labelledby="outputs-tab">

<p>Here you can find description of the most important targets for this
stage. However, for a full overview, you have to inspect the <a href="https://github.com/bioinfocz/scdrake/blob/main/R/plans_single_sample.R" class="external-link">source
code</a> of the <code><a href="../reference/get_subplan_single_sample.html">get_input_qc_subplan()</a></code> function.</p>
<p>Targets can be loaded from the <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cache
(<code>.drake</code> directory by default) using the
<code><a href="https://docs.ropensci.org/drake/reference/readd.html" class="external-link">drake::loadd()</a></code> or <code><a href="https://docs.ropensci.org/drake/reference/readd.html" class="external-link">drake::readd()</a></code>
functions.</p>
<p>HTML report target name: <code>report_input_qc</code></p>
<div class="section level4">
<h4 id="singlecellexperiment-objects">
<code>SingleCellExperiment</code> objects<a class="anchor" aria-label="anchor" href="#singlecellexperiment-objects"></a>
</h4>
<p><code>sce_raw</code>: an untouched SCE object as loaded from:</p>
<ul>
<li>
<code><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html" class="external-link">DropletUtils::read10xCounts()</a></code> (<code>cellranger</code>
input)</li>
<li>
<code><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">readr::read_delim()</a></code> (delimited table input)</li>
<li>
<code>SingleCellExperiment</code> object (Rds file or
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cache)</li>
</ul>
<hr>
<p><code>sce_valid_cells</code>: a SCE object with empty droplets
removed</p>
<hr>
<p><code>sce_unfiltered</code>: <code>sce_valid_cells</code> with added
columns:</p>
<ul>
<li>Columns from the <code>cell_qc</code> target containing DataFrame
with cell QC metrics</li>
<li>
<code>discard_qc</code>: dataset-sensitive filter</li>
<li>
<code>discard_custom</code>: custom filter</li>
</ul>
<hr>
<p><code>sce_qc_filter</code>, <code>sce_qc_filter_genes</code>: a SCE
object with cells filtered by dataset-sensitive filtering. The latter
with filtered genes.</p>
<p><code>sce_custom_filter</code>, <code>sce_custom_filter_genes</code>:
a SCE object with cells filtered by custom filtering. The latter with
filtered genes.</p>
<hr>
<p><code>sce_selected</code>: a SCE object selected from either
<code>sce_qc_filter_genes</code> or <code>sce_custom_filter_genes</code>
- depends on the <code>SAVE_DATASET_SENSITIVE_FILTERING</code>
parameter</p>
<p><code>sce_final_input_qc</code>: <code>sce_selected</code> with added
gene annotation (<code>gene_annotation</code> target) to
<code>rowData()</code>. This is the final SCE object which will be used
in the next stage <code>norm_clustering</code>.</p>
</div>
<div class="section level4">
<h4 id="cell-filters">Cell filters<a class="anchor" aria-label="anchor" href="#cell-filters"></a>
</h4>
<p><code>cell_qc</code>: a DataFrame with cell QC metrics</p>
<hr>
<p>The following list of filters are computed by
<code><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">scater::isOutlier()</a></code> from <code>cell_qc</code> columns,
which are shown for individual filters (e.g. <code>qc_lib</code>:
<code>total</code> column). In those, <code>FALSE</code> refers to cells
<strong>not passing</strong> a filter.</p>
<p><code>qc_filters</code>: a list of named lists with logical values
for dataset-sensitive cell filtering:</p>
<ul>
<li>
<code>qc_lib</code> (<code>total</code>): low number of UMI (lower
tail)</li>
<li>
<code>qc_nexprs</code> (<code>detected</code>): low number of
detected genes = non-zero UMI count (lower tail)</li>
<li>
<code>qc_mito</code> (<code>subsets_mito_percent</code>): high
percentage of expressed mitochondrial genes expression (upper tail)</li>
</ul>
<p><code>custom_filters</code>: a list of named lists with logical
values for custom cell filtering:</p>
<ul>
<li>
<code>low_count</code> (<code>total</code>): low number of
UMIs.</li>
<li>
<code>high_count</code> (<code>total</code>): high number of
UMIs.</li>
<li>
<code>low_expression</code> (<code>detected</code>): low number of
detected genes.</li>
<li>
<code>high_mito</code> (<code>subsets_mito_percent</code>): high
percentage of expressed mitochondrial genes.</li>
</ul>
<p><code>qc_filter</code>, <code>custom_filter</code>: a summarized
filter (by “or” operation) of <code>qc_filters</code> and
<code>custom_filters</code>, respectively</p>
</div>
<div class="section level4">
<h4 id="gene-filters">Gene filters<a class="anchor" aria-label="anchor" href="#gene-filters"></a>
</h4>
<p><code>sce_qc_gene_filter</code>, <code>sce_custom_gene_filter</code>:
gene filters (logical vectors) obtained from
<code><a href="../reference/get_gene_filter.html">get_gene_filter()</a></code> for SCE objects with cells filtered by
dataset-sensitive and custom filter, respectively. In those,
<code>FALSE</code> refers to genes <strong>not passing</strong> the
filter.</p>
</div>
<div class="section level4">
<h4 id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a>
</h4>
<p><code>sce_unfiltered_plotlist</code>: a list of various violin plots
for cell QC metrics (<a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a> objects), generated using
<code>sce_unfiltered</code> target. Cells are colored by
<code>discard_qc</code> column - blue points pass the dataset-sensitive
filter, while the orange ones do not.</p>
<p>For example, the following code is used to plot total UMI counts:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scdrake</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_colData.html">plot_colData</a></span><span class="op">(</span></span>
<span>  <span class="va">sce_unfiltered</span>,</span>
<span>  y <span class="op">=</span> <span class="st">"total"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"discard_qc"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Total count"</span>,</span>
<span>  scale_y <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<hr>
<p><code>sce_qc_filter_genes_plotlist</code>,
<code>sce_custom_filter_genes_plotlist</code>: similar to
<code>sce_unfiltered_plotlist</code>, but in each of them, cells are
colored by the other filter. That is, cells in
<code>sce_qc_filter_genes_plotlist</code> are colored by
<code>discard_custom</code> column, and thus, showing which cells would
be discarded in custom-filtered dataset. Vice versa for
<code>sce_custom_filter_genes_plotlist</code>.</p>
</div>
<div class="section level4">
<h4 id="other-targets">Other targets<a class="anchor" aria-label="anchor" href="#other-targets"></a>
</h4>
<p><code>config_input_qc</code>: a list holding parameters for this
stage</p>
<hr>
<p><code>barcode_ranks</code>: a DataFrame with barcode ranks for <a href="https://bioconductor.org/books/3.15/OSCA.advanced/droplet-processing.html" class="external-link">knee
plot</a>, computed by <code><a href="https://rdrr.io/pkg/DropletUtils/man/barcodeRanks.html" class="external-link">DropletUtils::barcodeRanks()</a></code> on UMI
counts from <code>sce_raw</code> target</p>
<p><code>empty_droplets</code>: a DataFrame of statistics for empty
droplets, computed <code><a href="https://rdrr.io/pkg/DropletUtils/man/emptyDrops.html" class="external-link">DropletUtils::emptyDrops()</a></code> on UMI counts
from <code>sce_raw</code> target</p>
<hr>
<p><code>sce_history</code>: a tibble with cell and gene “history”,
showing their numbers in unfiltered and filtered data.</p>
<p><code>sce_history_plot</code>: a <a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a> object
summarizing information from <code>sce_history</code> target.</p>
<hr>
<p><code>gene_annotation</code>: a dataframe with gene annotation,
processed such that:</p>
<ul>
<li>If a single ENSEMBL ID has multiple symbols, gene descriptions, or
ENTREZ IDs, they are collapsed by comma (<code>,</code>).</li>
<li>ENSEMBL ID is used as a symbol for ENSEMBL IDs with unknown
symbols.</li>
<li>ENSEMBL ID is appended to symbols having multiple ENSEMBL IDs.
<ul>
<li>For example, <em>TBCE</em> has both ENSG00000285053 and
ENSG00000284770 ENSEMBL IDs assigned -&gt; its symbol is changed to
TBCE_ENSG00000285053 and TBCE_ENSG00000284770.</li>
</ul>
</li>
</ul>
<p><code>gene_annotation</code> can be also retrieved from
<code>rowData(sce_final_input_qc)</code></p>
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="copyright">
  <p></p>
<p>Developed by Jiri Novotny, Jan Kubovciak, Michal Kolar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
