<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light"><script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/3/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Function reference â€¢ scdrake</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="../docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/3/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)"><!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/3/darkly/bootstrap.css" media="(prefers-color-scheme: dark)"><!-- preferably CSS --><link rel="stylesheet" href="../preferably.css"><link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)"><link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)"><script src="../darkswitch.js"></script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scdrake</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/scdrake.html">Get started</a>
</li>
<li>
  <a href="../articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="divider">
    <li class="dropdown-header">General information</li>
    <li>
      <a href="../articles/pipeline_overview.html">Pipeline overview</a>
    </li>
    <li>
      <a href="../articles/scdrake_run.html">Running the pipeline, environment variables</a>
    </li>
    <li>
      <a href="../articles/cluster_markers.html">Cluster markers</a>
    </li>
    <li>
      <a href="../articles/drake_basics.html">drake basics</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Configs</li>
    <li>
      <a href="../articles/scdrake_config.html">scdrake configs: a general overview</a>
    </li>
    <li>
      <a href="../articles/config_pipeline.html">Pipeline config</a>
    </li>
    <li>
      <a href="../articles/config_main.html">Main config</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Single-sample pipeline stages</li>
    <li>
      <a href="../articles/stage_input_qc.html">Input and quality control stage</a>
    </li>
    <li>
      <a href="../articles/stage_norm_clustering.html">Normalization and clustering stage</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Integration pipeline stages</li>
    <li>
      <a href="../articles/stage_integration.html">Integration stage</a>
    </li>
    <li>
      <a href="../articles/stage_int_clustering.html">Post-integration clustering stage</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Common stages</li>
    <li>
      <a href="../articles/stage_cluster_markers.html">Cluster markers stage</a>
    </li>
    <li>
      <a href="../articles/stage_contrasts.html">Contrasts stage</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
          <a href="#" id="css-toggle-btn">
            <span class="fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        <li>
  <a href="https://github.com/bioinfocz/scdrake" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
        
        


      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="project">Project <a href="#project" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Initialize or update a <code>scdrake</code> project.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="init_project.html">init_project()</a></code> </p>
        </td>
        <td><p>Initialize a new <code>scdrake</code> project.</p></td>
      </tr><tr><td>
          <p><code><a href="update_project.html">update_project()</a></code> </p>
        </td>
        <td><p>Update <code>scdrake</code> project files.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="config">Config <a href="#config" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="load-or-update-config-files">Load or update config files <a href="#load-or-update-config-files" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="load_config.html">load_config()</a></code> <code><a href="load_config.html">load_pipeline_config()</a></code> <code><a href="load_config.html">load_single_sample_configs()</a></code> <code><a href="load_config.html">load_integration_configs()</a></code> </p>
        </td>
        <td><p>Load a YAML config file as <code>scdrake_list</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="update_config.html">update_config()</a></code> <code><a href="update_config.html">update_pipeline_config()</a></code> <code><a href="update_config.html">update_single_sample_configs()</a></code> <code><a href="update_config.html">update_integration_configs()</a></code> <code><a href="update_config.html">update_configs()</a></code> </p>
        </td>
        <td><p>Update a local YAML config file using a default one.</p></td>
      </tr><tr><td>
          <p><code><a href="scdrake_list.html">scdrake_list()</a></code> <code><a href="scdrake_list.html">`$`(<i>&lt;scdrake_list&gt;</i>)</a></code> <code><a href="scdrake_list.html">`[`(<i>&lt;scdrake_list&gt;</i>)</a></code> <code><a href="scdrake_list.html">`[[`(<i>&lt;scdrake_list&gt;</i>)</a></code> </p>
        </td>
        <td><p>A <code>scdrake</code>'s list with overloaded, strict access operators.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="yq-tool">yq tool <a href="#yq-tool" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>Functions related to the <a href="https://github.com/mikefarah/yq" class="external-link">yq</a> tool.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="check_yq.html">check_yq()</a></code> </p>
        </td>
        <td><p>Check the availability and version of the <code>yq</code> tool.</p></td>
      </tr><tr><td>
          <p><code><a href="download_yq.html">download_yq()</a></code> </p>
        </td>
        <td><p>Download the binary of the <code>yq</code> tool (version 3.4.1).</p></td>
      </tr><tr><td>
          <p><code><a href="get_yq_default_path.html">get_yq_default_path()</a></code> </p>
        </td>
        <td><p>Return a default download path for the <code>yq</code> tool's binary.</p></td>
      </tr><tr><td>
          <p><code><a href="yq_merge_cmd.html">yq_merge_cmd()</a></code> </p>
        </td>
        <td><p>Merge two YAML files using the <code>yq</code> tool.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="run-pipeline">Run pipeline <a href="#run-pipeline" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="run_pipeline.html">run_single_sample()</a></code> <code><a href="run_pipeline.html">run_integration()</a></code> </p>
        </td>
        <td><p>A shortcut to run the <code>scdrake</code> pipeline in the current R session.</p></td>
      </tr><tr><td>
          <p><code><a href="create_dirs.html">create_single_sample_dirs()</a></code> <code><a href="create_dirs.html">create_integration_dirs()</a></code> </p>
        </td>
        <td><p>Create a basic directory structure based on paths in config.</p></td>
      </tr><tr><td>
          <p><code><a href="scdrake_make.html">scdrake_make()</a></code> <code><a href="scdrake_make.html">scdrake_r_make()</a></code> </p>
        </td>
        <td><p>Execute a <code>scdrake</code>'s pipeline plan.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="get-pipeline-plans">Get pipeline plans <a href="#get-pipeline-plans" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="full-plans-of-the-single-sample-and-integration-pipelines">Full plans of the single-sample and integration pipelines <a href="#full-plans-of-the-single-sample-and-integration-pipelines" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="get_plan.html">get_single_sample_plan()</a></code> <code><a href="get_plan.html">get_integration_plan()</a></code> </p>
        </td>
        <td><p>Get a <code>drake</code> plan for a specific analysis type.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="subplans-common-to-the-single-sample-and-integration-pipelines">Subplans common to the single-sample and integration pipelines <a href="#subplans-common-to-the-single-sample-and-integration-pipelines" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="get_subplan_common.html">get_common_subplan()</a></code> <code><a href="get_subplan_common.html">get_cluster_markers_subplan()</a></code> <code><a href="get_subplan_common.html">get_contrasts_subplan()</a></code> </p>
        </td>
        <td><p>Get a <code>drake</code> plan common to all pipelines.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="single-sample-pipeline-subplans">Single-sample pipeline subplans <a href="#single-sample-pipeline-subplans" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="get_subplan_single_sample.html">get_input_qc_subplan()</a></code> <code><a href="get_subplan_single_sample.html">get_norm_clustering_subplan()</a></code> </p>
        </td>
        <td><p>Get a <code>drake</code> plan for a stage of single-sample pipeline.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="integration-pipeline-subplans">Integration pipeline subplans <a href="#integration-pipeline-subplans" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="get_subplan_integration.html">get_integration_subplan()</a></code> <code><a href="get_subplan_integration.html">get_int_clustering_subplan()</a></code> </p>
        </td>
        <td><p>Get a <code>drake</code> plan for a stage of integration pipeline.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="download-example-data">Download example data <a href="#download-example-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="download_example_data.html">download_pbmc1k()</a></code> <code><a href="download_example_data.html">download_pbmc3k()</a></code> </p>
        </td>
        <td><p>Download PBMC example data from 10x Genomics.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="options-used-by-scdrake">Options used by scdrake <a href="#options-used-by-scdrake" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="scdrake_options.html">get_scdrake_default_options()</a></code> <code><a href="scdrake_options.html">get_scdrake_options()</a></code> </p>
        </td>
        <td><p>Options used by <code>scdrake</code>.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="target-related-functions">Target-related functions <a href="#target-related-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Those functions are mostly used for individual targets within pipeline plans. They are usually named as <code>&lt;target_name&gt;_fn()</code>.</p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="single-sample-pipeline-stage-input-qc">Single-sample pipeline / stage 01_input_qc <a href="#single-sample-pipeline-stage-input-qc" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="empty_droplets_fn.html">empty_droplets_fn()</a></code> </p>
        </td>
        <td><p>Calculate empty droplet statistics for each cell.</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_filter.html">get_gene_filter()</a></code> </p>
        </td>
        <td><p>Get a logical filter for genes not passing a ratio of cells expressing a gene and a minimum number of UMI per gene.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_final_input_qc_fn.html">sce_final_input_qc_fn()</a></code> </p>
        </td>
        <td><p>Finalize a <code>SingleCellExperiment</code> object which will proceed to the <code>02_norm_clustering</code> stage.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_history_fn.html">sce_history_fn()</a></code> </p>
        </td>
        <td><p>Create a tibble with history of cell and gene filtering.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_history_plot_fn.html">sce_history_plot_fn()</a></code> </p>
        </td>
        <td><p>Plot history of cell and gene filtering.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_raw_fn.html">sce_raw_fn()</a></code> </p>
        </td>
        <td><p>Import scRNA-seq data.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_selected_fn.html">sce_selected_fn()</a></code> </p>
        </td>
        <td><p>Select a <code>SingleCellExperiment</code> object which will proceed to the <code>02_norm_clustering</code> stage.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_valid_cells_fn.html">sce_valid_cells_fn()</a></code> </p>
        </td>
        <td><p>Subset cells in a <code>SingleCellExperiment</code> object to non-empty ones and add corresponding statistics.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="single-sample-pipeline-stage-norm-clustering">Single-sample pipeline / stage 02_norm_clustering <a href="#single-sample-pipeline-stage-norm-clustering" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="cc_genes_fn.html">cc_genes_fn()</a></code> </p>
        </td>
        <td><p>Get dataframe of cell cycle genes.</p></td>
      </tr><tr><td>
          <p><code><a href="pca_phase_plots_fn.html">pca_phase_plots_fn()</a></code> </p>
        </td>
        <td><p>Make dimred plots of PCA colored by cell cycle phase.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_cc_fn.html">sce_cc_fn()</a></code> </p>
        </td>
        <td><p>Assign cell cycle phase to cells.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_norm.html">sce_norm_fn()</a></code> <code><a href="sce_norm.html">scran_normalization()</a></code> <code><a href="sce_norm.html">sctransform_normalization()</a></code> </p>
        </td>
        <td><p>Normalize counts either by <code>scran</code> or <code>sctransform</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_norm_hvg_fn.html">sce_norm_hvg_fn()</a></code> </p>
        </td>
        <td><p>Find highly variable genes (HVGs).</p></td>
      </tr><tr><td>
          <p><code><a href="sce_rm_doublets_fn.html">sce_rm_doublets_fn()</a></code> </p>
        </td>
        <td><p>Remove cell doublets from a <code>SingleCellExperiment</code> object.</p></td>
      </tr><tr><td>
          <p><code><a href="selected_markers_plots_fn.html">selected_markers_plots_fn()</a></code> </p>
        </td>
        <td><p>Make a grid of feature plots for selected genes.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="integration-pipeline-stage-integration">Integration pipeline / stage 01_integration <a href="#integration-pipeline-stage-integration" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="get_int_method_description.html">get_int_method_description()</a></code> </p>
        </td>
        <td><p>Return description for an integration method.</p></td>
      </tr><tr><td>
          <p><code><a href="hvg_int_list_fn.html">hvg_int_list_fn()</a></code> </p>
        </td>
        <td><p>Get a named list of HVG data (gene IDs, fit, metadata).</p></td>
      </tr><tr><td>
          <p><code><a href="hvg_plot_int_fn.html">hvg_plot_int_fn()</a></code> </p>
        </td>
        <td><p>Make a HVG plot for uncorrected <code>SingleCellExperiment</code> object (processed by <code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchNorm.html" class="external-link">batchelor::multiBatchNorm()</a></code>).</p></td>
      </tr><tr><td>
          <p><code><a href="int_diagnostics_df_fn.html">int_diagnostics_df_fn()</a></code> </p>
        </td>
        <td><p>Compute and make integration diagnostics and plots for each integration method.</p></td>
      </tr><tr><td>
          <p><code><a href="integration_methods_df_fn.html">integration_methods_df_fn()</a></code> </p>
        </td>
        <td><p>Make a tibble with integration methods and their parameters.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_int_clustering_df_fn.html">sce_int_clustering_df_fn()</a></code> </p>
        </td>
        <td><p>Compute a quick graph-based clustering for each integration method.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_int_combine_hvgs.html">sce_int_combine_hvgs()</a></code> </p>
        </td>
        <td><p>Combine HVG data from a list of <code>SingleCellExperiment</code> objects.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_int_df_fn.html">sce_int_df_fn()</a></code> </p>
        </td>
        <td><p>Perform integration of single-samples.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_int_dimred_df_fn.html">sce_int_dimred_df_fn()</a></code> </p>
        </td>
        <td><p>Calculate t-SNE and UMAP for each integration method result (<code>SingleCellExperiment</code> object).</p></td>
      </tr><tr><td>
          <p><code><a href="sce_int_dimred_plots_df_fn.html">sce_int_dimred_plots_df_fn()</a></code> </p>
        </td>
        <td><p>Make dimred plot for each combination of integration method, dimred method, and coloring by
batch (single-sample) and cell cycle phase.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_int_import_fn.html">sce_int_import_fn()</a></code> </p>
        </td>
        <td><p>Import <code>sce_final_norm_clustering</code> targets from <code>drake</code> caches.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_int_pca_df_fn.html">sce_int_pca_df_fn()</a></code> </p>
        </td>
        <td><p>Compute PCA for each integration method result (<code>SingleCellExperiment</code> object).</p></td>
      </tr><tr><td>
          <p><code><a href="sce_int_processed_fn.html">sce_int_processed_fn()</a></code> </p>
        </td>
        <td><p>Subset each object in a list of <code>SingleCellExperiment</code> objects to common data and their corresponding
metadata (e.g. HVGs).</p></td>
      </tr><tr><td>
          <p><code><a href="sce_int_raw_snn_clustering_fn.html">sce_int_raw_snn_clustering_fn()</a></code> </p>
        </td>
        <td><p>Perform a fast shared nearest neighbor clustering of each sample.</p></td>
      </tr><tr><td>
          <p><code><a href="selected_markers_int_df_fn.html">selected_markers_int_df_fn()</a></code> </p>
        </td>
        <td><p>Prepare parameters for expression plots of selected markers.</p></td>
      </tr><tr><td>
          <p><code><a href="selected_markers_plots_int_df_fn.html">selected_markers_plots_int_df_fn()</a></code> </p>
        </td>
        <td><p>Make expression plots of selected markers.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="single-cell-data-related-functions">Single-cell data-related functions <a href="#single-cell-data-related-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="annotation">Annotation <a href="#annotation" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="collapse_ensembl_multivals.html">collapse_ensembl_multivals()</a></code> </p>
        </td>
        <td><p>Collapse a character vector by <code>","</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="make_gene_annotation.html">make_gene_annotation()</a></code> </p>
        </td>
        <td><p>Create a <code>dataframe</code> with annotation of SCE object's genes using an <code>AnnotationDbi</code> package.</p></td>
      </tr><tr><td>
          <p><code><a href="with_dbi.html">with_dbi()</a></code> </p>
        </td>
        <td><p>Load a SQL database file and run a function from the <code>AnnotationDbi</code> package.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="cell-type-annotation">Cell type annotation <a href="#cell-type-annotation" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="cell_annotation_diagnostic_plots_files_fn.html">cell_annotation_diagnostic_plots_files_fn()</a></code> </p>
        </td>
        <td><p>Save cell annotation diagnostic plots to PDF files.</p></td>
      </tr><tr><td>
          <p><code><a href="cell_annotation_diagnostic_plots_fn.html">cell_annotation_diagnostic_plots_fn()</a></code> </p>
        </td>
        <td><p>Generate diagnostic plots for cell annotation.</p></td>
      </tr><tr><td>
          <p><code><a href="cell_annotation_fn.html">cell_annotation_fn()</a></code> </p>
        </td>
        <td><p>Perform cell annotation via <code><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR::SingleR()</a></code>.</p></td>
      </tr><tr><td>
          <p><code><a href="cell_annotation_labels_fn.html">cell_annotation_labels_fn()</a></code> </p>
        </td>
        <td><p>Create a named list of cell labels returned from <code><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR::SingleR()</a></code>.</p></td>
      </tr><tr><td>
          <p><code><a href="cell_annotation_params_fn.html">cell_annotation_params_fn()</a></code> </p>
        </td>
        <td><p>Load a list of cell annotation references into a <code>tibble</code>.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="cluster-markers">Cluster markers <a href="#cluster-markers" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="cluster_markers_fn.html">cluster_markers_fn()</a></code> </p>
        </td>
        <td><p>Add LFC summaries for Wilcox tests, obtained from t-test results.</p></td>
      </tr><tr><td>
          <p><code><a href="cluster_markers_heatmaps_df_fn.html">cluster_markers_heatmaps_df_fn()</a></code> </p>
        </td>
        <td><p>Create a final <code>tibble</code> holding parameters for cluster markers heatmaps generation.</p></td>
      </tr><tr><td>
          <p><code><a href="cluster_markers_out_fn.html">cluster_markers_out_fn()</a></code> </p>
        </td>
        <td><p>Create a final <code>dataframe</code> of cluster markers.</p></td>
      </tr><tr><td>
          <p><code><a href="cluster_markers_params.html">cluster_markers_test_params_fn()</a></code> <code><a href="cluster_markers_params.html">cluster_markers_plot_params_fn()</a></code> <code><a href="cluster_markers_params.html">cluster_markers_heatmap_params_fn()</a></code> </p>
        </td>
        <td><p>Extract columns with certain parameters from cluster markers sources <code>tibble</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="cluster_markers_params_fn.html">cluster_markers_params_fn()</a></code> </p>
        </td>
        <td><p>Load a list of cluster markers sources into a <code>tibble</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="cluster_markers_processed_fn.html">cluster_markers_processed_fn()</a></code> </p>
        </td>
        <td><p>Add additional summary columns of cluster markers for comparisons with other group levels.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="clustering">Clustering <a href="#clustering" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="calc_sc3.html">calc_sc3()</a></code> </p>
        </td>
        <td><p>Calculate SC3 clustering.</p></td>
      </tr><tr><td>
          <p><code><a href="cells_per_cluster_table.html">cells_per_cluster_table()</a></code> <code><a href="cells_per_cluster_table.html">cells_per_cluster_table()</a></code> </p>
        </td>
        <td><p>Get a frequency table of cell-cluster assignments.</p></td>
      </tr><tr><td>
          <p><code><a href="cluster_kmeans_kc_fn.html">cluster_kmeans_kc_fn()</a></code> </p>
        </td>
        <td><p>Create a named list of k-means clusters for a selected number of clusters.</p></td>
      </tr><tr><td>
          <p><code><a href="cluster_sc3_fn.html">cluster_sc3_fn()</a></code> </p>
        </td>
        <td><p>Create a named list of SC3 clusters for a selected number of clusters.</p></td>
      </tr><tr><td>
          <p><code><a href="clustering_helpers.html">get_n_clusters()</a></code> </p>
        </td>
        <td><p>Get a number of unique clusters.</p></td>
      </tr><tr><td>
          <p><code><a href="get_pca_gene_var_pcs.html">get_pca_gene_var_pcs()</a></code> </p>
        </td>
        <td><p>Get a number of PCs corresponding to biological variation.</p></td>
      </tr><tr><td>
          <p><code><a href="get_pca_selected_pcs.html">get_pca_selected_pcs()</a></code> </p>
        </td>
        <td><p>Get a final selection of number of PCs.</p></td>
      </tr><tr><td>
          <p><code><a href="make_kmeans_gaps_plot.html">make_kmeans_gaps_plot()</a></code> </p>
        </td>
        <td><p>Make a plot of k-means gaps.</p></td>
      </tr><tr><td>
          <p><code><a href="make_pca_selected_pcs_plot.html">make_pca_selected_pcs_plot()</a></code> </p>
        </td>
        <td><p>Make a plot showing numbers of selected PCs from all strategies.</p></td>
      </tr><tr><td>
          <p><code><a href="make_sc3_stability_plots.html">make_sc3_stability_plots()</a></code> </p>
        </td>
        <td><p>Create a list of plots with SC3 clustering stability.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_calc_pca.html">sce_calc_pca()</a></code> </p>
        </td>
        <td><p>Calculate PCA of a <code>SingleCellExperiment</code> object.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_compute_dimreds.html">sce_compute_dimreds()</a></code> </p>
        </td>
        <td><p>Compute t-SNE and UMAP dimreds on a <code>SingleCellExperiment</code> object.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="contrasts">Contrasts <a href="#contrasts" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="contrasts_fn.html">contrasts_fn()</a></code> </p>
        </td>
        <td><p>Extract contrast statistics from cluster markers tests.</p></td>
      </tr><tr><td>
          <p><code><a href="contrasts_heatmaps_df_fn.html">contrasts_heatmaps_df_fn()</a></code> </p>
        </td>
        <td><p>Create a <code>tibble</code> holding contrast results and heatmap parameters.</p></td>
      </tr><tr><td>
          <p><code><a href="contrasts_out_fn.html">contrasts_out_fn()</a></code> </p>
        </td>
        <td><p>Format a <code>tibble</code> with contrast results.</p></td>
      </tr><tr><td>
          <p><code><a href="contrasts_params.html">contrasts_params_fn()</a></code> <code><a href="contrasts_params.html">contrasts_heatmap_params_fn()</a></code> <code><a href="contrasts_params.html">contrasts_plot_params_fn()</a></code> </p>
        </td>
        <td><p>Prepare a <code>tibble</code> with parameters for contrasts tests.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="highly-variable-genes-hvgs-">Highly variable genes (HVGs) <a href="#highly-variable-genes-hvgs-" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="get_top_hvgs.html">get_top_hvgs()</a></code> </p>
        </td>
        <td><p>Get top N highly variable genes (HVGs) by a specified metric.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_hvg.html">plot_hvg()</a></code> </p>
        </td>
        <td><p>Plot average expression vs. HVG metrics (total, bio, and technical variance) and highlight HVGs.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_hvg_fit.html">plot_hvg_fit()</a></code> </p>
        </td>
        <td><p>Plot fit of a HVG metric and highlight HVGs.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_get_cc_genes.html">sce_get_cc_genes()</a></code> </p>
        </td>
        <td><p>Identify cell cycle-related genes by using variance explained threshold.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_remove_cc_genes.html">sce_remove_cc_genes()</a></code> </p>
        </td>
        <td><p>Remove cell cycle-related genes from HVGs.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="markers">Markers <a href="#markers" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p><p>These functions are used in both cluster markers and contrasts stages.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add_marker_table_links.html">add_marker_table_links()</a></code> </p>
        </td>
        <td><p>Convert specified columns of a marker table to HTML links.</p></td>
      </tr><tr><td>
          <p><code><a href="filter_markers.html">filter_markers()</a></code> </p>
        </td>
        <td><p>Filter a dataframe with cluster markers.</p></td>
      </tr><tr><td>
          <p><code><a href="generate_markers_results_section.html">generate_markers_results_section()</a></code> </p>
        </td>
        <td><p>Generate a table with links to marker results.</p></td>
      </tr><tr><td>
          <p><code><a href="marker_heatmap.html">marker_heatmap()</a></code> </p>
        </td>
        <td><p>Make a heatmap of cell clusters.</p></td>
      </tr><tr><td>
          <p><code><a href="marker_heatmaps_wrapper.html">marker_heatmaps_wrapper()</a></code> </p>
        </td>
        <td><p>Make a heatmap of global or contrast markers.</p></td>
      </tr><tr><td>
          <p><code><a href="marker_plot.html">marker_plot()</a></code> </p>
        </td>
        <td><p>Make a marker plot.</p></td>
      </tr><tr><td>
          <p><code><a href="markers_dimred_plots.html">markers_dimred_plots()</a></code> </p>
        </td>
        <td><p>Make dimred plots for each marker source defined in <code>CLUSTER_MARKERS_SOURCES</code> (<code>cluster_markers.yaml</code> config)
or <code>CONTRASTS_SOURCES</code> (<code>contrasts.yaml</code> config).</p></td>
      </tr><tr><td>
          <p><code><a href="markers_dimred_plots_files.html">markers_dimred_plots_files()</a></code> </p>
        </td>
        <td><p>Save dimred plots.</p></td>
      </tr><tr><td>
          <p><code><a href="markers_for_tables.html">markers_for_tables()</a></code> </p>
        </td>
        <td><p>Make a table used in HTML report of marker results.</p></td>
      </tr><tr><td>
          <p><code><a href="markers_plots_files.html">markers_plots_files()</a></code> </p>
        </td>
        <td><p>Make plots of top markers.</p></td>
      </tr><tr><td>
          <p><code><a href="markers_plots_top.html">markers_plots_top()</a></code> </p>
        </td>
        <td><p>Get a tibble of top markers extracted from each test result.</p></td>
      </tr><tr><td>
          <p><code><a href="markers_table_files.html">markers_table_files()</a></code> </p>
        </td>
        <td><p>Prepare a table with markers and render a HTML report from RMarkdown template.</p></td>
      </tr><tr><td>
          <p><code><a href="scran_markers.html">scran_markers()</a></code> </p>
        </td>
        <td><p>Compute cell cluster markers.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="singlecellexperiment-object-related-functions">SingleCellExperiment object-related functions <a href="#singlecellexperiment-object-related-functions" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="as_seurat.html">as_seurat()</a></code> </p>
        </td>
        <td><p>Convert a <code>SingleCellExperiment</code> to <code>Seurat</code> object.</p></td>
      </tr><tr><td>
          <p><code><a href="cell_data.html">make_cell_groupings()</a></code> <code><a href="cell_data.html">cell_data_fn()</a></code> </p>
        </td>
        <td><p>Create new cell groups based on existing ones.</p></td>
      </tr><tr><td>
          <p><code><a href="cells_per_cluster_table.html">cells_per_cluster_table()</a></code> <code><a href="cells_per_cluster_table.html">cells_per_cluster_table()</a></code> </p>
        </td>
        <td><p>Get a frequency table of cell-cluster assignments.</p></td>
      </tr><tr><td>
          <p><code><a href="create_seu_for_heatmaps.html">create_seu_for_heatmaps()</a></code> </p>
        </td>
        <td><p>Create a <code>Seurat</code> object used for heatmap generation.</p></td>
      </tr><tr><td>
          <p><code><a href="merge_sce_metadata.html">merge_sce_metadata()</a></code> </p>
        </td>
        <td><p>Merge <code>metadata()</code> of multiple <code>SingleCellExperiment</code> object.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_add_cell_data.html">sce_add_cell_data()</a></code> </p>
        </td>
        <td><p>Add columns to <code>colData()</code> of a <code>SingleCellExperiment</code> object.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_add_colData.html">sce_add_colData()</a></code> </p>
        </td>
        <td><p>Append new columns to <code>colData</code> of a <code>SingleCellExperiment</code> object.</p></td>
      </tr><tr><td>
          <p><code><a href="sce_add_metadata.html">sce_add_metadata()</a></code> </p>
        </td>
        <td><p>Append data to <code>metadata()</code> list of a <code>SingleCellExperiment</code> object.</p></td>
      </tr><tr><td>
          <p><code><a href="seu_add_metadata.html">seu_add_metadata()</a></code> </p>
        </td>
        <td><p>Append data to metadata o a <code>Seurat</code> object's assay.</p></td>
      </tr><tr><td>
          <p><code><a href="which_genes_regex.html">which_genes_regex()</a></code> </p>
        </td>
        <td><p>Get indices of genes whose annotation matches a regex.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="visualization">Visualization <a href="#visualization" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="dimred_plots_clustering_fn.html">dimred_plots_clustering_fn()</a></code> </p>
        </td>
        <td><p>Make dimred plots for each clustering method.</p></td>
      </tr><tr><td>
          <p><code><a href="dimred_plots_other_vars_fn.html">dimred_plots_other_vars_fn()</a></code> </p>
        </td>
        <td><p>Make a dimred plot of selected variables.</p></td>
      </tr><tr><td>
          <p><code><a href="dimred_plots_other_vars_params_fn.html">dimred_plots_other_vars_params_fn()</a></code> </p>
        </td>
        <td><p>Make a tibble with parameters for dimred plots of selected variables.</p></td>
      </tr><tr><td>
          <p><code><a href="highlight_points.html">highlight_points()</a></code> </p>
        </td>
        <td><p>Highlight points belonging to certain levels.</p></td>
      </tr><tr><td>
          <p><code><a href="plotReducedDim_mod.html">plotReducedDim_mod()</a></code> </p>
        </td>
        <td><p>A wrapper around <code><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">scater::plotReducedDim()</a></code>.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_colData.html">plot_colData()</a></code> </p>
        </td>
        <td><p>A wrapper for <code><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">scater::plotColData()</a></code>.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_vln.html">plot_vln()</a></code> </p>
        </td>
        <td><p>Make a combined violin-boxplot plot.</p></td>
      </tr><tr><td>
          <p><code><a href="save_selected_markers_plots_files.html">save_selected_markers_plots_files()</a></code> </p>
        </td>
        <td><p>Save plots of selected markers.</p></td>
      </tr><tr><td>
          <p><code><a href="selected_markers_dimplot.html">selected_markers_dimplot()</a></code> </p>
        </td>
        <td><p>Make a grid of feature plots for selected markers.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="misc-functions">Misc functions <a href="#misc-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="html">HTML <a href="#html" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="cells_per_cluster_table_collapsed_html.html">cells_per_cluster_table_collapsed_html()</a></code> </p>
        </td>
        <td><p>Print a HTML of table collapsible by button.</p></td>
      </tr><tr><td>
          <p><code><a href="generate_dimred_plots_section.html">generate_dimred_plots_section()</a></code> </p>
        </td>
        <td><p>Generate a section with dimred plots used in some RMarkdown files.</p></td>
      </tr><tr><td>
          <p><code><a href="generate_stage_report.html">generate_stage_report()</a></code> </p>
        </td>
        <td><p>Render RMarkdown document for a stage of the pipeline.</p></td>
      </tr><tr><td>
          <p><code><a href="html_links.html">create_a_link()</a></code> <code><a href="html_links.html">create_img_link()</a></code> </p>
        </td>
        <td><p>Generate a HTML link as <code>&lt;a&gt;&lt;/a&gt;</code> or in the form of image as <code>&lt;a&gt;&lt;img /&gt;&lt;/a&gt;</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="md_header.html">md_header()</a></code> </p>
        </td>
        <td><p>Generate a Markdown header.</p></td>
      </tr><tr><td>
          <p><code><a href="render_bootstrap_table.html">render_bootstrap_table()</a></code> </p>
        </td>
        <td><p>Render a dataframe-like object using <code><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">knitr::kable()</a></code> and <code><a href="https://rdrr.io/pkg/kableExtra/man/kable_styling.html" class="external-link">kableExtra::kable_styling()</a></code>.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="text">Text <a href="#text" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="text_utils_glue.html">glue0()</a></code> <code><a href="text_utils_glue.html">catg0()</a></code> <code><a href="text_utils_glue.html">gluec()</a></code> <code><a href="text_utils_glue.html">glue0c()</a></code> <code><a href="text_utils_glue.html">catn()</a></code> </p>
        </td>
        <td><p>Various wrappers around <code><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue::glue()</a></code> and <code><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat()</a></code>.</p></td>
      </tr><tr><td>
          <p><code><a href="text_utils_joining.html">str_comma()</a></code> <code><a href="text_utils_joining.html">str_line()</a></code> <code><a href="text_utils_joining.html">str_space()</a></code> </p>
        </td>
        <td><p>Various utils for joining of text.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h3 id="utils">Utils <a href="#utils" class="anchor" aria-hidden="true"></a></h3>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add_item_to_list.html">add_item_to_list()</a></code> </p>
        </td>
        <td><p>Add named item (including <code>NULL</code>) to a list.</p></td>
      </tr><tr><td>
          <p><code><a href="assert_that_.html">assert_that_()</a></code> </p>
        </td>
        <td><p>A wrapper around <code><a href="https://rdrr.io/pkg/assertthat/man/assert_that.html" class="external-link">assertthat::assert_that()</a></code>.</p></td>
      </tr><tr><td>
          <p><code><a href="check_pkg_installed.html">check_pkg_installed()</a></code> <code><a href="check_pkg_installed.html">check_qs_installed()</a></code> <code><a href="check_pkg_installed.html">check_future_installed()</a></code> <code><a href="check_pkg_installed.html">check_clustermq_installed()</a></code> <code><a href="check_pkg_installed.html">check_future.callr_installed()</a></code> </p>
        </td>
        <td><p>Check if a package is installed and display an informative message.</p></td>
      </tr><tr><td>
          <p><code><a href="filter_nulls.html">filter_nulls()</a></code> </p>
        </td>
        <td><p>Return list with removed <code>NULL</code> elements.</p></td>
      </tr><tr><td>
          <p><code><a href="get_sys_env.html">get_sys_env()</a></code> </p>
        </td>
        <td><p>Read an environment variable and if needed, assign a default value.</p></td>
      </tr><tr><td>
          <p><code><a href="grapes-and-and-grapes.html">`%&amp;&amp;%`</a></code> </p>
        </td>
        <td><p>A short-circuit evaluation-like function similar to <code>bash</code>'s <code>&amp;&amp;</code> operator.</p></td>
      </tr><tr><td>
          <p><code><a href="lapply_rows.html">lapply_rows()</a></code> </p>
        </td>
        <td><p>Apply a function over rows of a <code>data.frame</code>-like object and concatenate the results back to <code>tibble</code> or <code>data.frame</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="list_names_to_values.html">list_names_to_values()</a></code> </p>
        </td>
        <td><p>Append names of a list of lists as values in each sublist.</p></td>
      </tr><tr><td>
          <p><code><a href="lists_to_tibble.html">lists_to_tibble()</a></code> </p>
        </td>
        <td><p>Convert list of lists to tibble.</p></td>
      </tr><tr><td>
          <p><code><a href="na_empty.html">na_empty()</a></code> </p>
        </td>
        <td><p>Return <code>NA</code> if an object is empty.</p></td>
      </tr><tr><td>
          <p><code><a href="replace_list_nas_with_nulls.html">replace_list_nas_with_nulls()</a></code> </p>
        </td>
        <td><p>Replace <code>NA</code>s in a list by <code>NULL</code>s.</p></td>
      </tr><tr><td>
          <p><code><a href="replace_list_nulls.html">replace_list_nulls()</a></code> </p>
        </td>
        <td><p>Replace <code>NULL</code> items in a defined depth of a list.</p></td>
      </tr><tr><td>
          <p><code><a href="save_pdf.html">save_pdf()</a></code> </p>
        </td>
        <td><p>Save a list of plots to multipage PDF.</p></td>
      </tr><tr><td>
          <p><code><a href="save_print.html">save_print()</a></code> <code><a href="save_print.html">save_object_info()</a></code> </p>
        </td>
        <td><p>Capture output from <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> of an object.</p></td>
      </tr><tr><td>
          <p><code><a href="set_rstudio_drake_cache.html">set_rstudio_drake_cache()</a></code> </p>
        </td>
        <td><p>Set <code>rstudio_drake_cache</code> option.</p></td>
      </tr><tr><td>
          <p><code><a href="with_plan.html">with_plan()</a></code> </p>
        </td>
        <td><p>Locally change a <code>future</code> plan.</p></td>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="dot-apply_config_patches.html">.apply_config_patches()</a></code> </p>
        </td>
        <td><p>Apply YAML config patches to package-bundled configs.</p></td>
      </tr><tr><td>
          <p><code><a href="dot-check_integration_methods.html">.check_integration_methods()</a></code> </p>
        </td>
        <td><p>Check for valid <code>INTEGRATION_SOURCES</code> parameter.</p></td>
      </tr><tr><td>
          <p><code><a href="dot-check_marker_sources.html">.check_marker_sources()</a></code> </p>
        </td>
        <td><p>Check for correct structure of <code>CLUSTER_MARKERS_SOURCES</code> or <code>CONTRASTS_SOURCES</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="dot-get_dict_param.html">.get_dict_param()</a></code> </p>
        </td>
        <td><p>Unlist a list of length one containing a named list.</p></td>
      </tr><tr><td>
          <p><code><a href="dot-hereize_paths.html">.hereize_paths()</a></code> </p>
        </td>
        <td><p>Using the <code>here</code> package, contruct paths relative to project's root directory.</p></td>
      </tr><tr><td>
          <p><code><a href="dot-load_config_group.html">.load_config_group()</a></code> </p>
        </td>
        <td><p>Load a group of YAML config files.</p></td>
      </tr><tr><td>
          <p><code><a href="dot-load_configs.html">.load_configs()</a></code> </p>
        </td>
        <td><p>Load a list of YAML config files.</p></td>
      </tr><tr><td>
          <p><code><a href="dot-paths_to_base_dir.html">.paths_to_base_dir()</a></code> </p>
        </td>
        <td><p>Construct paths relative to base directory.</p></td>
      </tr><tr><td>
          <p><code><a href="dot-update_configs_recursive.html">.update_configs_recursive()</a></code> </p>
        </td>
        <td><p>Recursively update local configs using a file glob <code>*.default.yaml</code>.</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_marker_source_params.html">.prepare_marker_source_params()</a></code> <code><a href="prepare_marker_source_params.html">.prepare_marker_sources_params()</a></code> </p>
        </td>
        <td><p>Prepare a <code>tibble</code> with parameters for cluster marker tests.</p></td>
      </tr><tr><td>
          <p><code><a href="process_config.html">.process_pipeline_config()</a></code> <code><a href="process_config.html">.process_main_config()</a></code> <code><a href="process_config.html">.process_cluster_markers_config()</a></code> <code><a href="process_config.html">.process_contrasts_config()</a></code> <code><a href="process_config.html">.prepare_cell_annotation_sources_params()</a></code> <code><a href="process_config.html">.process_integration_config()</a></code> <code><a href="process_config.html">.process_int_clustering_config()</a></code> <code><a href="process_config.html">.process_input_qc_config()</a></code> <code><a href="process_config.html">.process_norm_clustering_config()</a></code> </p>
        </td>
        <td><p>Config processing.</p></td>
      </tr><tr><td>
          <p><code><a href="update_config_group.html">update_config_group</a></code> </p>
        </td>
        <td><p>Update a group of configs, i.e. pipeline, single-sample, or integration.</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jiri Novotny, Jan Kubovciak.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.9000.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

      </footer></div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '937cedcd467c204b19fc155c5f35019b',
    indexName: 'scdrake',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script></body></html>

