<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light"><script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/5/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Function reference â€¢ scdrake</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/5/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)"><!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/5/darkly/bootstrap.css" media="(prefers-color-scheme: dark)"><!-- preferably CSS --><link rel="stylesheet" href="../preferably.css"><link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)"><link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)"><script src="../darkswitch.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scdrake</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/scdrake.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake_integration.html">Integration pipeline guide</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Guides</h6>
    <a class="dropdown-item" href="../articles/scdrake_docker.html">Using the Docker image</a>
    <a class="dropdown-item" href="../articles/scdrake_integration.html">02 Integration pipeline guide</a>
    <a class="dropdown-item" href="../articles/scdrake_extend.html">Extending the pipeline</a>
    <a class="dropdown-item" href="../articles/drake_basics.html">drake basics</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General information</h6>
    <a class="dropdown-item" href="../articles/pipeline_overview.html">Pipeline overview</a>
    <a class="dropdown-item" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
    <a class="dropdown-item" href="../articles/scdrake_cli.html">Command line interface</a>
    <a class="dropdown-item" href="../articles/cluster_markers.html">Cluster markers</a>
    <a class="dropdown-item" href="../articles/scdrake_config.html">scdrake config internals</a>
    <a class="dropdown-item" href="../articles/scdrake_envvars.html">Environment variables</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General configs</h6>
    <a class="dropdown-item" href="../articles/config_pipeline.html">Pipeline config</a>
    <a class="dropdown-item" href="../articles/config_main.html">Main config</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single-sample pipeline stages</h6>
    <a class="dropdown-item" href="../articles/stage_input_qc.html">Input and quality control stage (`01_input_qc`)</a>
    <a class="dropdown-item" href="../articles/stage_norm_clustering.html">Normalization and clustering stage (`02_norm_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Integration pipeline stages</h6>
    <a class="dropdown-item" href="../articles/stage_integration.html">Integration stage (`01_integration`)</a>
    <a class="dropdown-item" href="../articles/stage_int_clustering.html">Post-integration clustering stage (`02_int_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Common stages</h6>
    <a class="dropdown-item" href="../articles/stage_cluster_markers.html">Cluster markers stage (`cluster_markers`)</a>
    <a class="dropdown-item" href="../articles/stage_contrasts.html">Contrasts stage (`contrasts`)</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bioinfocz/scdrake" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>

        
        <li>
          <a class="external-link nav-link" id="css-toggle-btn" aria-label="github">
            <span class="fas fa fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        
        
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="command-line-interface">Command line interface<a class="anchor" aria-label="anchor" href="#command-line-interface"></a></h2>
      
      <p class="section-desc"></p><p>Install or check the command line interface.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="SCDRAKE_CLI_VERSION.html">SCDRAKE_CLI_VERSION</a></code> 
        </dt>
        <dd>Matching CLI version for this package version.</dd>
      </dl><dl><dt>
          
          <code><a href="cli.html">install_cli()</a></code> <code><a href="cli.html">check_cli()</a></code> 
        </dt>
        <dd>Install or check the command line interface scripts.</dd>
      </dl></div><div class="section level2">
      <h2 id="project">Project<a class="anchor" aria-label="anchor" href="#project"></a></h2>
      
      <p class="section-desc"></p><p>Initialize or update a <code>scdrake</code> project.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="init_project.html">init_project()</a></code> 
        </dt>
        <dd>Initialize a new <code>scdrake</code> project.</dd>
      </dl><dl><dt>
          
          <code><a href="update_project.html">update_project()</a></code> 
        </dt>
        <dd>Update <code>scdrake</code> project files.</dd>
      </dl></div><div class="section level2">
      <h2 id="config">Config<a class="anchor" aria-label="anchor" href="#config"></a></h2>
      
      

      
    </div><div class="section level2">
      
      <h3 id="load-or-update-config-files">Load or update config files<a class="anchor" aria-label="anchor" href="#load-or-update-config-files"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="load_config.html">load_config()</a></code> <code><a href="load_config.html">load_pipeline_config()</a></code> <code><a href="load_config.html">load_single_sample_configs()</a></code> <code><a href="load_config.html">load_integration_configs()</a></code> 
        </dt>
        <dd>Load a YAML config file as <code>scdrake_list</code>.</dd>
      </dl><dl><dt>
          
          <code><a href="update_config.html">update_config()</a></code> <code><a href="update_config.html">update_pipeline_config()</a></code> <code><a href="update_config.html">update_single_sample_configs()</a></code> <code><a href="update_config.html">update_integration_configs()</a></code> <code><a href="update_config.html">update_configs()</a></code> 
        </dt>
        <dd>Update a local YAML config file using a default one.</dd>
      </dl><dl><dt>
          
          <code><a href="scdrake_list.html">scdrake_list()</a></code> <code><a href="scdrake_list.html">`$`(<i>&lt;scdrake_list&gt;</i>)</a></code> <code><a href="scdrake_list.html">`[`(<i>&lt;scdrake_list&gt;</i>)</a></code> <code><a href="scdrake_list.html">`[[`(<i>&lt;scdrake_list&gt;</i>)</a></code> 
        </dt>
        <dd>A <code>scdrake</code>'s list with overloaded, strict access operators.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="yq-tool">yq tool<a class="anchor" aria-label="anchor" href="#yq-tool"></a></h3>
      <p class="section-desc"></p><p>Functions related to the <a href="https://github.com/mikefarah/yq" class="external-link">yq</a> tool.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="check_yq.html">check_yq()</a></code> 
        </dt>
        <dd>Check the availability and version of the <code>yq</code> tool.</dd>
      </dl><dl><dt>
          
          <code><a href="download_yq.html">download_yq()</a></code> 
        </dt>
        <dd>Download the binary of the <code>yq</code> tool (version 3.4.1).</dd>
      </dl><dl><dt>
          
          <code><a href="get_yq_default_path.html">get_yq_default_path()</a></code> 
        </dt>
        <dd>Return a default download path for the <code>yq</code> tool's binary.</dd>
      </dl><dl><dt>
          
          <code><a href="yq_merge_cmd.html">yq_merge_cmd()</a></code> 
        </dt>
        <dd>Merge two YAML files using the <code>yq</code> tool.</dd>
      </dl></div><div class="section level2">
      <h2 id="run-pipeline">Run pipeline<a class="anchor" aria-label="anchor" href="#run-pipeline"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="run_pipeline.html">run_single_sample()</a></code> <code><a href="run_pipeline.html">run_integration()</a></code> <code><a href="run_pipeline.html">run_single_sample_r()</a></code> <code><a href="run_pipeline.html">run_integration_r()</a></code> 
        </dt>
        <dd>Run the <code>scdrake</code> pipeline.</dd>
      </dl><dl><dt>
          
          <code><a href="create_dirs.html">create_single_sample_dirs()</a></code> <code><a href="create_dirs.html">create_integration_dirs()</a></code> 
        </dt>
        <dd>Create a basic directory structure based on paths in config.</dd>
      </dl><dl><dt>
          
          <code><a href="scdrake_make.html">scdrake_make()</a></code> <code><a href="scdrake_make.html">scdrake_r_make()</a></code> 
        </dt>
        <dd>Execute a <code>scdrake</code>'s pipeline plan.</dd>
      </dl></div><div class="section level2">
      <h2 id="check-for-scdrake-dependencies">Check for scdrake dependencies<a class="anchor" aria-label="anchor" href="#check-for-scdrake-dependencies"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="check_pandoc.html">check_pandoc()</a></code> 
        </dt>
        <dd>Check for <code>pandoc</code>'s binary.</dd>
      </dl><dl><dt>
          
          <code><a href="check_pkg_installed.html">check_pkg_installed()</a></code> <code><a href="check_pkg_installed.html">check_qs_installed()</a></code> <code><a href="check_pkg_installed.html">check_future_installed()</a></code> <code><a href="check_pkg_installed.html">check_clustermq_installed()</a></code> <code><a href="check_pkg_installed.html">check_future.callr_installed()</a></code> <code><a href="check_pkg_installed.html">check_sc3_version()</a></code> <code><a href="check_pkg_installed.html">check_scdrake_packages()</a></code> 
        </dt>
        <dd>Check if a package is installed and display an informative message.</dd>
      </dl><dl><dt>
          
          <code><a href="check_scdrake.html">check_scdrake()</a></code> 
        </dt>
        <dd>Check for selected <code>scdrake</code> dependencies.</dd>
      </dl></div><div class="section level2">
      <h2 id="get-pipeline-plans">Get pipeline plans<a class="anchor" aria-label="anchor" href="#get-pipeline-plans"></a></h2>
      
      

      
    </div><div class="section level2">
      
      <h3 id="full-plans-of-the-single-sample-and-integration-pipelines">Full plans of the single-sample and integration pipelines<a class="anchor" aria-label="anchor" href="#full-plans-of-the-single-sample-and-integration-pipelines"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="get_plan.html">get_single_sample_plan()</a></code> <code><a href="get_plan.html">get_integration_plan()</a></code> 
        </dt>
        <dd>Get a <code>drake</code> plan for a specific analysis type.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="subplans-common-to-the-single-sample-and-integration-pipelines">Subplans common to the single-sample and integration pipelines<a class="anchor" aria-label="anchor" href="#subplans-common-to-the-single-sample-and-integration-pipelines"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="get_subplan_common.html">get_common_subplan()</a></code> <code><a href="get_subplan_common.html">get_cluster_markers_subplan()</a></code> <code><a href="get_subplan_common.html">get_contrasts_subplan()</a></code> 
        </dt>
        <dd>Get a <code>drake</code> plan common to all pipelines.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="single-sample-pipeline-subplans">Single-sample pipeline subplans<a class="anchor" aria-label="anchor" href="#single-sample-pipeline-subplans"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="get_subplan_single_sample.html">get_input_qc_subplan()</a></code> <code><a href="get_subplan_single_sample.html">get_norm_clustering_subplan()</a></code> 
        </dt>
        <dd>Get a <code>drake</code> plan for a stage of single-sample pipeline.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="integration-pipeline-subplans">Integration pipeline subplans<a class="anchor" aria-label="anchor" href="#integration-pipeline-subplans"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="get_subplan_integration.html">get_integration_subplan()</a></code> <code><a href="get_subplan_integration.html">get_int_clustering_subplan()</a></code> 
        </dt>
        <dd>Get a <code>drake</code> plan for a stage of integration pipeline.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="custom-plan">Custom plan<a class="anchor" aria-label="anchor" href="#custom-plan"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="load_custom_plan.html">load_custom_plan()</a></code> 
        </dt>
        <dd>Source a file returning a custom drake plan.</dd>
      </dl></div><div class="section level2">
      <h2 id="download-example-data">Download example data<a class="anchor" aria-label="anchor" href="#download-example-data"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="download_example_data.html">download_pbmc1k()</a></code> <code><a href="download_example_data.html">download_pbmc3k()</a></code> 
        </dt>
        <dd>Download PBMC example data from 10x Genomics.</dd>
      </dl></div><div class="section level2">
      <h2 id="options-used-by-scdrake">Options used by scdrake<a class="anchor" aria-label="anchor" href="#options-used-by-scdrake"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="scdrake_options.html">get_scdrake_default_options()</a></code> <code><a href="scdrake_options.html">get_scdrake_options()</a></code> 
        </dt>
        <dd>Options used by <code>scdrake</code>.</dd>
      </dl></div><div class="section level2">
      <h2 id="target-related-functions">Target-related functions<a class="anchor" aria-label="anchor" href="#target-related-functions"></a></h2>
      
      <p class="section-desc"></p><p>Those functions are mostly used for individual targets within pipeline plans. They are usually named as <code>&lt;target_name&gt;_fn()</code>.</p>

      
    </div><div class="section level2">
      
      <h3 id="single-sample-pipeline-stage-input-qc">Single-sample pipeline / stage 01_input_qc<a class="anchor" aria-label="anchor" href="#single-sample-pipeline-stage-input-qc"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="empty_droplets_fn.html">empty_droplets_fn()</a></code> 
        </dt>
        <dd>Calculate empty droplet statistics for each cell.</dd>
      </dl><dl><dt>
          
          <code><a href="get_gene_filter.html">get_gene_filter()</a></code> 
        </dt>
        <dd>Get a logical filter for genes not passing a ratio of cells expressing a gene and a minimum number of UMI per gene.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_final_input_qc_fn.html">sce_final_input_qc_fn()</a></code> 
        </dt>
        <dd>Finalize a <code>SingleCellExperiment</code> object which will proceed to the <code>02_norm_clustering</code> stage.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_history_fn.html">sce_history_fn()</a></code> 
        </dt>
        <dd>Create a tibble with history of cell and gene filtering.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_history_plot_fn.html">sce_history_plot_fn()</a></code> 
        </dt>
        <dd>Plot history of cell and gene filtering.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_raw_fn.html">sce_raw_fn()</a></code> 
        </dt>
        <dd>Import scRNA-seq data.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_selected_fn.html">sce_selected_fn()</a></code> 
        </dt>
        <dd>Select a <code>SingleCellExperiment</code> object which will proceed to the <code>02_norm_clustering</code> stage.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_valid_cells_fn.html">sce_valid_cells_fn()</a></code> 
        </dt>
        <dd>Subset cells in a <code>SingleCellExperiment</code> object to non-empty ones and add corresponding statistics.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="single-sample-pipeline-stage-norm-clustering">Single-sample pipeline / stage 02_norm_clustering<a class="anchor" aria-label="anchor" href="#single-sample-pipeline-stage-norm-clustering"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="cc_genes_fn.html">cc_genes_fn()</a></code> 
        </dt>
        <dd>Get dataframe of cell cycle genes.</dd>
      </dl><dl><dt>
          
          <code><a href="pca_phase_plots_fn.html">pca_phase_plots_fn()</a></code> 
        </dt>
        <dd>Make dimred plots of PCA colored by cell cycle phase.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_cc_fn.html">sce_cc_fn()</a></code> 
        </dt>
        <dd>Assign cell cycle phase to cells.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_norm.html">sce_norm_fn()</a></code> <code><a href="sce_norm.html">scran_normalization()</a></code> <code><a href="sce_norm.html">sctransform_normalization()</a></code> 
        </dt>
        <dd>Normalize counts either by <code>scran</code> or <code>sctransform</code>.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_norm_hvg_fn.html">sce_norm_hvg_fn()</a></code> 
        </dt>
        <dd>Find highly variable genes (HVGs).</dd>
      </dl><dl><dt>
          
          <code><a href="sce_rm_doublets_fn.html">sce_rm_doublets_fn()</a></code> 
        </dt>
        <dd>Remove cell doublets from a <code>SingleCellExperiment</code> object.</dd>
      </dl><dl><dt>
          
          <code><a href="selected_markers_plots_fn.html">selected_markers_plots_fn()</a></code> 
        </dt>
        <dd>Make a grid of feature plots for selected genes.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="integration-pipeline-stage-integration">Integration pipeline / stage 01_integration<a class="anchor" aria-label="anchor" href="#integration-pipeline-stage-integration"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="get_int_method_description.html">get_int_method_description()</a></code> 
        </dt>
        <dd>Return description for an integration method.</dd>
      </dl><dl><dt>
          
          <code><a href="hvg_int_list_fn.html">hvg_int_list_fn()</a></code> 
        </dt>
        <dd>Get a named list of HVG data (gene IDs, fit, metadata).</dd>
      </dl><dl><dt>
          
          <code><a href="hvg_plot_int_fn.html">hvg_plot_int_fn()</a></code> 
        </dt>
        <dd>Make a HVG plot for uncorrected <code>SingleCellExperiment</code> object (processed by <code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchNorm.html" class="external-link">batchelor::multiBatchNorm()</a></code>).</dd>
      </dl><dl><dt>
          
          <code><a href="int_diagnostics_df_fn.html">int_diagnostics_df_fn()</a></code> 
        </dt>
        <dd>Compute and make integration diagnostics and plots for each integration method.</dd>
      </dl><dl><dt>
          
          <code><a href="integration_methods_df_fn.html">integration_methods_df_fn()</a></code> 
        </dt>
        <dd>Make a tibble with integration methods and their parameters.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_int_clustering_df_fn.html">sce_int_clustering_df_fn()</a></code> 
        </dt>
        <dd>Compute a quick graph-based clustering for each integration method.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_int_combine_hvgs.html">sce_int_combine_hvgs()</a></code> 
        </dt>
        <dd>Combine HVG data from a list of <code>SingleCellExperiment</code> objects.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_int_df_fn.html">sce_int_df_fn()</a></code> 
        </dt>
        <dd>Perform integration of single-samples.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_int_dimred_df_fn.html">sce_int_dimred_df_fn()</a></code> 
        </dt>
        <dd>Calculate t-SNE and UMAP for each integration method result (<code>SingleCellExperiment</code> object).</dd>
      </dl><dl><dt>
          
          <code><a href="sce_int_dimred_plots_df_fn.html">sce_int_dimred_plots_df_fn()</a></code> 
        </dt>
        <dd>Make dimred plot for each combination of integration method, dimred method, and coloring by
batch (single-sample) and cell cycle phase.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_int_import_fn.html">sce_int_import_fn()</a></code> 
        </dt>
        <dd>Import <code>sce_final_norm_clustering</code> targets from <code>drake</code> caches.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_int_pca_df_fn.html">sce_int_pca_df_fn()</a></code> 
        </dt>
        <dd>Compute PCA for each integration method result (<code>SingleCellExperiment</code> object).</dd>
      </dl><dl><dt>
          
          <code><a href="sce_int_processed_fn.html">sce_int_processed_fn()</a></code> 
        </dt>
        <dd>Subset each object in a list of <code>SingleCellExperiment</code> objects to common data and their corresponding
metadata (e.g. HVGs).</dd>
      </dl><dl><dt>
          
          <code><a href="sce_int_raw_snn_clustering_fn.html">sce_int_raw_snn_clustering_fn()</a></code> 
        </dt>
        <dd>Perform a fast shared nearest neighbor clustering of each sample.</dd>
      </dl><dl><dt>
          
          <code><a href="selected_markers_int_df_fn.html">selected_markers_int_df_fn()</a></code> 
        </dt>
        <dd>Prepare parameters for expression plots of selected markers.</dd>
      </dl><dl><dt>
          
          <code><a href="selected_markers_plots_int_df_fn.html">selected_markers_plots_int_df_fn()</a></code> 
        </dt>
        <dd>Make expression plots of selected markers.</dd>
      </dl></div><div class="section level2">
      <h2 id="single-cell-data-related-functions">Single-cell data-related functions<a class="anchor" aria-label="anchor" href="#single-cell-data-related-functions"></a></h2>
      
      

      
    </div><div class="section level2">
      
      <h3 id="annotation">Annotation<a class="anchor" aria-label="anchor" href="#annotation"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="collapse_ensembl_multivals.html">collapse_ensembl_multivals()</a></code> 
        </dt>
        <dd>Collapse a character vector by <code>","</code>.</dd>
      </dl><dl><dt>
          
          <code><a href="make_gene_annotation.html">make_gene_annotation()</a></code> 
        </dt>
        <dd>Create a <code>dataframe</code> with annotation of SCE object's genes using an <code>AnnotationDbi</code> package.</dd>
      </dl><dl><dt>
          
          <code><a href="with_dbi.html">with_dbi()</a></code> 
        </dt>
        <dd>Load a SQL database file and run a function from the <code>AnnotationDbi</code> package.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="cell-type-annotation">Cell type annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="cell_annotation_diagnostic_plots_files_fn.html">cell_annotation_diagnostic_plots_files_fn()</a></code> 
        </dt>
        <dd>Save cell annotation diagnostic plots to PDF files.</dd>
      </dl><dl><dt>
          
          <code><a href="cell_annotation_diagnostic_plots_fn.html">cell_annotation_diagnostic_plots_fn()</a></code> 
        </dt>
        <dd>Generate diagnostic plots for cell annotation.</dd>
      </dl><dl><dt>
          
          <code><a href="cell_annotation_fn.html">cell_annotation_fn()</a></code> 
        </dt>
        <dd>Perform cell annotation via <code><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR::SingleR()</a></code>.</dd>
      </dl><dl><dt>
          
          <code><a href="cell_annotation_labels_fn.html">cell_annotation_labels_fn()</a></code> 
        </dt>
        <dd>Create a named list of cell labels returned from <code><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR::SingleR()</a></code>.</dd>
      </dl><dl><dt>
          
          <code><a href="cell_annotation_params_fn.html">cell_annotation_params_fn()</a></code> 
        </dt>
        <dd>Load a list of cell annotation references into a <code>tibble</code>.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="cluster-markers">Cluster markers<a class="anchor" aria-label="anchor" href="#cluster-markers"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="cluster_markers_fn.html">cluster_markers_fn()</a></code> 
        </dt>
        <dd>Add LFC summaries for Wilcox tests, obtained from t-test results.</dd>
      </dl><dl><dt>
          
          <code><a href="cluster_markers_heatmaps_df_fn.html">cluster_markers_heatmaps_df_fn()</a></code> 
        </dt>
        <dd>Create a final <code>tibble</code> holding parameters for cluster markers heatmaps generation.</dd>
      </dl><dl><dt>
          
          <code><a href="cluster_markers_out_fn.html">cluster_markers_out_fn()</a></code> 
        </dt>
        <dd>Create a final <code>dataframe</code> of cluster markers.</dd>
      </dl><dl><dt>
          
          <code><a href="cluster_markers_params.html">cluster_markers_test_params_fn()</a></code> <code><a href="cluster_markers_params.html">cluster_markers_plot_params_fn()</a></code> <code><a href="cluster_markers_params.html">cluster_markers_heatmap_params_fn()</a></code> 
        </dt>
        <dd>Extract columns with certain parameters from cluster markers sources <code>tibble</code>.</dd>
      </dl><dl><dt>
          
          <code><a href="cluster_markers_params_fn.html">cluster_markers_params_fn()</a></code> 
        </dt>
        <dd>Load a list of cluster markers sources into a <code>tibble</code>.</dd>
      </dl><dl><dt>
          
          <code><a href="cluster_markers_processed_fn.html">cluster_markers_processed_fn()</a></code> 
        </dt>
        <dd>Add additional summary columns of cluster markers for comparisons with other group levels.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="calc_sc3.html">calc_sc3()</a></code> 
        </dt>
        <dd>Calculate SC3 clustering.</dd>
      </dl><dl><dt>
          
          <code><a href="cells_per_cluster_table.html">cells_per_cluster_table()</a></code> <code><a href="cells_per_cluster_table.html">cells_per_cluster_table()</a></code> 
        </dt>
        <dd>Get a frequency table of cell-cluster assignments.</dd>
      </dl><dl><dt>
          
          <code><a href="cluster_kmeans_kc_fn.html">cluster_kmeans_kc_fn()</a></code> 
        </dt>
        <dd>Create a named list of k-means clusters for a selected number of clusters.</dd>
      </dl><dl><dt>
          
          <code><a href="cluster_sc3_fn.html">cluster_sc3_fn()</a></code> 
        </dt>
        <dd>Create a named list of SC3 clusters for a selected number of clusters.</dd>
      </dl><dl><dt>
          
          <code><a href="clustering_helpers.html">get_n_clusters()</a></code> 
        </dt>
        <dd>Get a number of unique clusters.</dd>
      </dl><dl><dt>
          
          <code><a href="get_pca_gene_var_pcs.html">get_pca_gene_var_pcs()</a></code> 
        </dt>
        <dd>Get a number of PCs corresponding to biological variation.</dd>
      </dl><dl><dt>
          
          <code><a href="get_pca_selected_pcs.html">get_pca_selected_pcs()</a></code> 
        </dt>
        <dd>Get a final selection of number of PCs.</dd>
      </dl><dl><dt>
          
          <code><a href="make_kmeans_gaps_plot.html">make_kmeans_gaps_plot()</a></code> 
        </dt>
        <dd>Make a plot of k-means gaps.</dd>
      </dl><dl><dt>
          
          <code><a href="make_pca_selected_pcs_plot.html">make_pca_selected_pcs_plot()</a></code> 
        </dt>
        <dd>Make a plot showing numbers of selected PCs from all strategies.</dd>
      </dl><dl><dt>
          
          <code><a href="make_sc3_stability_plots.html">make_sc3_stability_plots()</a></code> 
        </dt>
        <dd>Create a list of plots with SC3 clustering stability.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_calc_pca.html">sce_calc_pca()</a></code> 
        </dt>
        <dd>Calculate PCA of a <code>SingleCellExperiment</code> object.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_compute_dimreds.html">sce_compute_dimreds()</a></code> 
        </dt>
        <dd>Compute t-SNE and UMAP dimreds on a <code>SingleCellExperiment</code> object.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="contrasts">Contrasts<a class="anchor" aria-label="anchor" href="#contrasts"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="contrasts_fn.html">contrasts_fn()</a></code> 
        </dt>
        <dd>Extract contrast statistics from cluster markers tests.</dd>
      </dl><dl><dt>
          
          <code><a href="contrasts_heatmaps_df_fn.html">contrasts_heatmaps_df_fn()</a></code> 
        </dt>
        <dd>Create a <code>tibble</code> holding contrast results and heatmap parameters.</dd>
      </dl><dl><dt>
          
          <code><a href="contrasts_out_fn.html">contrasts_out_fn()</a></code> 
        </dt>
        <dd>Format a <code>tibble</code> with contrast results.</dd>
      </dl><dl><dt>
          
          <code><a href="contrasts_params.html">contrasts_params_fn()</a></code> <code><a href="contrasts_params.html">contrasts_heatmap_params_fn()</a></code> <code><a href="contrasts_params.html">contrasts_plot_params_fn()</a></code> 
        </dt>
        <dd>Prepare a <code>tibble</code> with parameters for contrasts tests.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="highly-variable-genes-hvgs-">Highly variable genes (HVGs)<a class="anchor" aria-label="anchor" href="#highly-variable-genes-hvgs-"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="get_top_hvgs.html">get_top_hvgs()</a></code> 
        </dt>
        <dd>Get top N highly variable genes (HVGs) by a specified metric.</dd>
      </dl><dl><dt>
          
          <code><a href="plot_hvg.html">plot_hvg()</a></code> 
        </dt>
        <dd>Plot average expression vs. HVG metrics (total, bio, and technical variance) and highlight HVGs.</dd>
      </dl><dl><dt>
          
          <code><a href="plot_hvg_fit.html">plot_hvg_fit()</a></code> 
        </dt>
        <dd>Plot fit of a HVG metric and highlight HVGs.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_get_cc_genes.html">sce_get_cc_genes()</a></code> 
        </dt>
        <dd>Identify cell cycle-related genes by using variance explained threshold.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_remove_cc_genes.html">sce_remove_cc_genes()</a></code> 
        </dt>
        <dd>Remove cell cycle-related genes from HVGs.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="markers">Markers<a class="anchor" aria-label="anchor" href="#markers"></a></h3>
      <p class="section-desc"></p><p>These functions are used in both cluster markers and contrasts stages.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="add_marker_table_links.html">add_marker_table_links()</a></code> 
        </dt>
        <dd>Convert specified columns of a marker table to HTML links.</dd>
      </dl><dl><dt>
          
          <code><a href="filter_markers.html">filter_markers()</a></code> 
        </dt>
        <dd>Filter a dataframe with cluster markers.</dd>
      </dl><dl><dt>
          
          <code><a href="generate_markers_results_section.html">generate_markers_results_section()</a></code> 
        </dt>
        <dd>Generate a table with links to marker results.</dd>
      </dl><dl><dt>
          
          <code><a href="marker_heatmap.html">marker_heatmap()</a></code> 
        </dt>
        <dd>Make a heatmap of cell clusters.</dd>
      </dl><dl><dt>
          
          <code><a href="marker_heatmaps_wrapper.html">marker_heatmaps_wrapper()</a></code> 
        </dt>
        <dd>Make a heatmap of global or contrast markers.</dd>
      </dl><dl><dt>
          
          <code><a href="marker_plot.html">marker_plot()</a></code> 
        </dt>
        <dd>Make a marker plot.</dd>
      </dl><dl><dt>
          
          <code><a href="markers_dimred_plots.html">markers_dimred_plots()</a></code> 
        </dt>
        <dd>Make dimred plots for each marker source defined in <code>CLUSTER_MARKERS_SOURCES</code> (<code>cluster_markers.yaml</code> config)
or <code>CONTRASTS_SOURCES</code> (<code>contrasts.yaml</code> config).</dd>
      </dl><dl><dt>
          
          <code><a href="markers_dimred_plots_files.html">markers_dimred_plots_files()</a></code> 
        </dt>
        <dd>Save dimred plots.</dd>
      </dl><dl><dt>
          
          <code><a href="markers_for_tables.html">markers_for_tables()</a></code> 
        </dt>
        <dd>Make a table used in HTML report of marker results.</dd>
      </dl><dl><dt>
          
          <code><a href="markers_plots_files.html">markers_plots_files()</a></code> 
        </dt>
        <dd>Make plots of top markers.</dd>
      </dl><dl><dt>
          
          <code><a href="markers_plots_top.html">markers_plots_top()</a></code> 
        </dt>
        <dd>Get a tibble of top markers extracted from each test result.</dd>
      </dl><dl><dt>
          
          <code><a href="markers_table_files.html">markers_table_files()</a></code> 
        </dt>
        <dd>Prepare a table with markers and render a HTML report from RMarkdown template.</dd>
      </dl><dl><dt>
          
          <code><a href="scran_markers.html">scran_markers()</a></code> 
        </dt>
        <dd>Compute cell cluster markers.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="singlecellexperiment-object-related-functions">SingleCellExperiment object-related functions<a class="anchor" aria-label="anchor" href="#singlecellexperiment-object-related-functions"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="additional_cell_data_fn.html">additional_cell_data_fn()</a></code> 
        </dt>
        <dd>Load additional cell data from a CSV or Rds (dataframe) file.</dd>
      </dl><dl><dt>
          
          <code><a href="as_seurat.html">as_seurat()</a></code> 
        </dt>
        <dd>Convert a <code>SingleCellExperiment</code> to <code>Seurat</code> object.</dd>
      </dl><dl><dt>
          
          <code><a href="cell_data.html">make_cell_groupings()</a></code> 
        </dt>
        <dd>Create new cell groups based on existing ones.</dd>
      </dl><dl><dt>
          
          <code><a href="cell_data_fn.html">cell_data_fn()</a></code> 
        </dt>
        <dd>Merge all cell-related data to a single DataFrame.</dd>
      </dl><dl><dt>
          
          <code><a href="cells_per_cluster_table.html">cells_per_cluster_table()</a></code> <code><a href="cells_per_cluster_table.html">cells_per_cluster_table()</a></code> 
        </dt>
        <dd>Get a frequency table of cell-cluster assignments.</dd>
      </dl><dl><dt>
          
          <code><a href="create_seu_for_heatmaps.html">create_seu_for_heatmaps()</a></code> 
        </dt>
        <dd>Create a <code>Seurat</code> object used for heatmap generation.</dd>
      </dl><dl><dt>
          
          <code><a href="merge_sce_metadata.html">merge_sce_metadata()</a></code> 
        </dt>
        <dd>Merge <code>metadata()</code> of multiple <code>SingleCellExperiment</code> object.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_add_cell_data.html">sce_add_cell_data()</a></code> 
        </dt>
        <dd>Add columns to <code>colData()</code> of a <code>SingleCellExperiment</code> object.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_add_colData.html">sce_add_colData()</a></code> 
        </dt>
        <dd>Append new columns to <code>colData</code> of a <code>SingleCellExperiment</code> object.</dd>
      </dl><dl><dt>
          
          <code><a href="sce_add_metadata.html">sce_add_metadata()</a></code> 
        </dt>
        <dd>Append data to <code>metadata()</code> list of a <code>SingleCellExperiment</code> object.</dd>
      </dl><dl><dt>
          
          <code><a href="seu_add_metadata.html">seu_add_metadata()</a></code> 
        </dt>
        <dd>Append data to metadata o a <code>Seurat</code> object's assay.</dd>
      </dl><dl><dt>
          
          <code><a href="which_genes_regex.html">which_genes_regex()</a></code> 
        </dt>
        <dd>Get indices of genes whose annotation matches a regex.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="dimred_plots_clustering_fn.html">dimred_plots_clustering_fn()</a></code> 
        </dt>
        <dd>Make dimred plots for each clustering method.</dd>
      </dl><dl><dt>
          
          <code><a href="dimred_plots_other_vars_fn.html">dimred_plots_other_vars_fn()</a></code> 
        </dt>
        <dd>Make a dimred plot of selected variables.</dd>
      </dl><dl><dt>
          
          <code><a href="dimred_plots_other_vars_params_fn.html">dimred_plots_other_vars_params_fn()</a></code> 
        </dt>
        <dd>Make a tibble with parameters for dimred plots of selected variables.</dd>
      </dl><dl><dt>
          
          <code><a href="highlight_points.html">highlight_points()</a></code> 
        </dt>
        <dd>Highlight points belonging to certain levels.</dd>
      </dl><dl><dt>
          
          <code><a href="plotReducedDim_mod.html">plotReducedDim_mod()</a></code> 
        </dt>
        <dd>A wrapper around <code><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">scater::plotReducedDim()</a></code>.</dd>
      </dl><dl><dt>
          
          <code><a href="plot_colData.html">plot_colData()</a></code> 
        </dt>
        <dd>A wrapper for <code><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">scater::plotColData()</a></code>.</dd>
      </dl><dl><dt>
          
          <code><a href="plot_vln.html">plot_vln()</a></code> 
        </dt>
        <dd>Make a combined violin-boxplot plot.</dd>
      </dl><dl><dt>
          
          <code><a href="save_selected_markers_plots_files.html">save_selected_markers_plots_files()</a></code> 
        </dt>
        <dd>Save plots of selected markers.</dd>
      </dl><dl><dt>
          
          <code><a href="selected_markers_dimplot.html">selected_markers_dimplot()</a></code> 
        </dt>
        <dd>Make a grid of feature plots for selected markers.</dd>
      </dl></div><div class="section level2">
      <h2 id="misc-functions">Misc functions<a class="anchor" aria-label="anchor" href="#misc-functions"></a></h2>
      
      

      
    </div><div class="section level2">
      
      <h3 id="html">HTML<a class="anchor" aria-label="anchor" href="#html"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="cells_per_cluster_table_collapsed_html.html">cells_per_cluster_table_collapsed_html()</a></code> 
        </dt>
        <dd>Print a HTML of table collapsible by button.</dd>
      </dl><dl><dt>
          
          <code><a href="format_used_functions.html">format_used_functions()</a></code> 
        </dt>
        <dd>Use the <a href="https://downlit.r-lib.org/" class="external-link">downlit</a> package to generate a Markdown list of autolinked functions wrapped inside <code>&lt;details&gt;</code> tags.</dd>
      </dl><dl><dt>
          
          <code><a href="generate_dimred_plots_section.html">generate_dimred_plots_section()</a></code> 
        </dt>
        <dd>Generate a section with dimred plots used in some RMarkdown files.</dd>
      </dl><dl><dt>
          
          <code><a href="generate_stage_report.html">generate_stage_report()</a></code> 
        </dt>
        <dd>Render RMarkdown document for a stage of the pipeline.</dd>
      </dl><dl><dt>
          
          <code><a href="html_links.html">create_a_link()</a></code> <code><a href="html_links.html">create_img_link()</a></code> 
        </dt>
        <dd>Generate a HTML link as <code>&lt;a&gt;&lt;/a&gt;</code> or in the form of image as <code>&lt;a&gt;&lt;img /&gt;&lt;/a&gt;</code>.</dd>
      </dl><dl><dt>
          
          <code><a href="md_header.html">md_header()</a></code> 
        </dt>
        <dd>Generate a Markdown header.</dd>
      </dl><dl><dt>
          
          <code><a href="render_bootstrap_table.html">render_bootstrap_table()</a></code> 
        </dt>
        <dd>Render a dataframe-like object using <code><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">knitr::kable()</a></code> and <code><a href="https://rdrr.io/pkg/kableExtra/man/kable_styling.html" class="external-link">kableExtra::kable_styling()</a></code>.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="text">Text<a class="anchor" aria-label="anchor" href="#text"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="format_shell_command.html">format_shell_command()</a></code> 
        </dt>
        <dd>Format a shell command as a Markdown codeblock.</dd>
      </dl><dl><dt>
          
          <code><a href="get_random_strings.html">get_random_strings()</a></code> 
        </dt>
        <dd>Generate random strings compounded from alphabetical characters by default.</dd>
      </dl><dl><dt>
          
          <code><a href="text_utils_glue.html">glue0()</a></code> <code><a href="text_utils_glue.html">catg0()</a></code> <code><a href="text_utils_glue.html">gluec()</a></code> <code><a href="text_utils_glue.html">glue0c()</a></code> <code><a href="text_utils_glue.html">catn()</a></code> 
        </dt>
        <dd>Various wrappers around <code><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue::glue()</a></code> and <code><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat()</a></code>.</dd>
      </dl><dl><dt>
          
          <code><a href="text_utils_joining.html">str_comma()</a></code> <code><a href="text_utils_joining.html">str_line()</a></code> <code><a href="text_utils_joining.html">str_space()</a></code> 
        </dt>
        <dd>Various utils for joining of text.</dd>
      </dl></div><div class="section level2">
      
      <h3 id="utils">Utils<a class="anchor" aria-label="anchor" href="#utils"></a></h3>
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="add_item_to_list.html">add_item_to_list()</a></code> 
        </dt>
        <dd>Add named item (including <code>NULL</code>) to a list.</dd>
      </dl><dl><dt>
          
          <code><a href="assert_that_.html">assert_that_()</a></code> 
        </dt>
        <dd>A wrapper around <code><a href="https://rdrr.io/pkg/assertthat/man/assert_that.html" class="external-link">assertthat::assert_that()</a></code>.</dd>
      </dl><dl><dt>
          
          <code><a href="create_dummy_plot.html">create_dummy_plot()</a></code> 
        </dt>
        <dd>Create a blank <code>ggplot</code> with label.</dd>
      </dl><dl><dt>
          
          <code><a href="filter_nulls.html">filter_nulls()</a></code> 
        </dt>
        <dd>Return list with removed <code>NULL</code> elements.</dd>
      </dl><dl><dt>
          
          <code><a href="get_sys_env.html">get_sys_env()</a></code> 
        </dt>
        <dd>Read an environment variable and if needed, assign a default value.</dd>
      </dl><dl><dt>
          
          <code><a href="get_tmp_dir.html">get_tmp_dir()</a></code> 
        </dt>
        <dd>Get a path to the temporary directory according to the OS.</dd>
      </dl><dl><dt>
          
          <code><a href="grapes-and-and-grapes.html">`%&amp;&amp;%`</a></code> 
        </dt>
        <dd>A short-circuit evaluation-like function similar to <code>bash</code>'s <code>&amp;&amp;</code> operator.</dd>
      </dl><dl><dt>
          
          <code><a href="lapply_rows.html">lapply_rows()</a></code> 
        </dt>
        <dd>Apply a function over rows of a <code>data.frame</code>-like object and concatenate the results back to <code>tibble</code> or <code>data.frame</code>.</dd>
      </dl><dl><dt>
          
          <code><a href="list_names_to_values.html">list_names_to_values()</a></code> 
        </dt>
        <dd>Append names of a list of lists as values in each sublist.</dd>
      </dl><dl><dt>
          
          <code><a href="lists_to_tibble.html">lists_to_tibble()</a></code> 
        </dt>
        <dd>Convert list of lists to tibble.</dd>
      </dl><dl><dt>
          
          <code><a href="na_empty.html">na_empty()</a></code> 
        </dt>
        <dd>Return <code>NA</code> if an object is empty.</dd>
      </dl><dl><dt>
          
          <code><a href="replace_list_nas_with_nulls.html">replace_list_nas_with_nulls()</a></code> 
        </dt>
        <dd>Replace <code>NA</code>s in a list by <code>NULL</code>s.</dd>
      </dl><dl><dt>
          
          <code><a href="replace_list_nulls.html">replace_list_nulls()</a></code> 
        </dt>
        <dd>Replace <code>NULL</code> items in a defined depth of a list.</dd>
      </dl><dl><dt>
          
          <code><a href="save_pdf.html">save_pdf()</a></code> 
        </dt>
        <dd>Save a list of plots to multipage PDF.</dd>
      </dl><dl><dt>
          
          <code><a href="save_print.html">save_print()</a></code> <code><a href="save_print.html">save_object_info()</a></code> 
        </dt>
        <dd>Capture output from <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> of an object.</dd>
      </dl><dl><dt>
          
          <code><a href="set_rstudio_drake_cache.html">set_rstudio_drake_cache()</a></code> 
        </dt>
        <dd>Set <code>rstudio_drake_cache</code> option.</dd>
      </dl><dl><dt>
          
          <code><a href="with_plan.html">with_plan()</a></code> 
        </dt>
        <dd>Locally change a <code>future</code> plan.</dd>
      </dl></div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="dot-apply_config_patches.html">.apply_config_patches()</a></code> 
        </dt>
        <dd>Apply YAML config patches to package-bundled configs.</dd>
      </dl><dl><dt>
          
          <code><a href="dot-check_integration_methods.html">.check_integration_methods()</a></code> 
        </dt>
        <dd>Check for valid <code>INTEGRATION_SOURCES</code> parameter.</dd>
      </dl><dl><dt>
          
          <code><a href="dot-check_marker_sources.html">.check_marker_sources()</a></code> 
        </dt>
        <dd>Check for correct structure of <code>CLUSTER_MARKERS_SOURCES</code> or <code>CONTRASTS_SOURCES</code>.</dd>
      </dl><dl><dt>
          
          <code><a href="dot-get_dict_param.html">.get_dict_param()</a></code> 
        </dt>
        <dd>Unlist a list of length one containing a named list.</dd>
      </dl><dl><dt>
          
          <code><a href="dot-hereize_paths.html">.hereize_paths()</a></code> 
        </dt>
        <dd>Using the <code>here</code> package, contruct paths relative to project's root directory.</dd>
      </dl><dl><dt>
          
          <code><a href="dot-load_config_group.html">.load_config_group()</a></code> 
        </dt>
        <dd>Load a group of YAML config files.</dd>
      </dl><dl><dt>
          
          <code><a href="dot-load_configs.html">.load_configs()</a></code> 
        </dt>
        <dd>Load a list of YAML config files.</dd>
      </dl><dl><dt>
          
          <code><a href="dot-paths_to_base_dir.html">.paths_to_base_dir()</a></code> 
        </dt>
        <dd>Construct paths relative to base directory.</dd>
      </dl><dl><dt>
          
          <code><a href="dot-update_configs_recursive.html">.update_configs_recursive()</a></code> 
        </dt>
        <dd>Recursively update local configs using a file glob <code>*.default.yaml</code>.</dd>
      </dl><dl><dt>
          
          <code><a href="prepare_marker_source_params.html">.prepare_marker_source_params()</a></code> <code><a href="prepare_marker_source_params.html">.prepare_marker_sources_params()</a></code> 
        </dt>
        <dd>Prepare a <code>tibble</code> with parameters for cluster marker tests.</dd>
      </dl><dl><dt>
          
          <code><a href="process_config.html">.process_pipeline_config()</a></code> <code><a href="process_config.html">.process_main_config()</a></code> <code><a href="process_config.html">.process_cluster_markers_config()</a></code> <code><a href="process_config.html">.process_contrasts_config()</a></code> <code><a href="process_config.html">.prepare_cell_annotation_sources_params()</a></code> <code><a href="process_config.html">.process_integration_config()</a></code> <code><a href="process_config.html">.process_int_clustering_config()</a></code> <code><a href="process_config.html">.process_input_qc_config()</a></code> <code><a href="process_config.html">.process_norm_clustering_config()</a></code> 
        </dt>
        <dd>Config processing.</dd>
      </dl><dl><dt>
          
          <code><a href="update_config_group.html">update_config_group</a></code> 
        </dt>
        <dd>Update a group of configs, i.e. pipeline, single-sample, or integration.</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="copyright">
  <p></p><p>Developed by Jiri Novotny, Jan Kubovciak.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

    </footer></div>

  

  

  </body></html>

