---
author: "Made by the [scdrake pipeline](https://bioinfocz.github.io/scdrake)"
institute: |
  Laboratory of Genomics and Bioinformatics
  Institute of Molecular Genetics of the ASCR
  https://img.cas.cz
date: "`r glue::glue('Document generated: {format(Sys.time(), \"%Y-%m-%d %H:%M:%S %Z%z\")}')`"
output:
  html_document:
    toc: false
    number_sections: false
    theme: "flatly"
    self_contained: true
    code_download: true
    df_print: "paged"
params:
  title: "Cluster markers table"
  drake_cache_dir: !expr here::here(".drake")
  other_params: "`r list()`"
title: "`r params$title`"
css: "`r here::here('Rmd/common/stylesheet.css')`"
---

***

```{r, include = FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
suppressPackageStartupMessages(library(magrittr))
DT::datatable(matrix())
```

```{js}
$(document).ready(function() {
  $("a[role='button']").unwrap();
});
```

```{css}
body .main-container {
  max-width: 1440px !important;
}
```

```{r}
other_params <- scdrake::scdrake_list(params$other_params)

```

<details style="display: inline !important;">

<summary style="list-style: none !important;">
<p class="btn btn-primary" type="button" role="link">Show details</p>
</summary>

- ID: `r other_params$id`
- Name: `r other_params$name`
- Description: `r other_params$description`
- Source column / level: `r other_params$source_column` / `r other_params$group_level`
- Block column: `r other_params$block_column`
- Test type: `r other_params$test_type`
- p-value type: `r other_params$pval_type`
- LFC direction: `r other_params$lfc_direction`
- LFC test: |LFC| > `r other_params$lfc_test`
- Min. proportion: `r other_params$min_prop`
- Standardized LFC: `r other_params$std_lfc`

</details>

***

# {.tabset}

## Spatial plots {.tabset}


```{r, results = "asis"}
par <- other_params$dimred_plots[1, ]  # or iterate via for-loop
p <- par$plot_highlighted[[1]]
scdrake::md_header(stringr::str_to_upper("spatial"), 3)
print(p)
```


## Marker table

```{r, results = "asis"}
cat(other_params$heatmap_file_rel, "\n\n")

cat("
<button class='btn btn-success' type='button' data-toggle='collapse' data-target='#table_info'
aria-expanded='false' aria-controls='table_info' style='margin-bottom: 0;'>
  Show table info
</button>
")

cat("
<div class='collapse' id='table_info'>
<hr />
<div class='card card-body'>

Genes in the current cell group (cluster) were tested against each other cell group using the selected test
(all pairwise comparisons).

Table columns:

<ul>
<li>
top - top genes from each pairwise comparison, e.g. top <= 1 will contain the top gene from each pairwise comparison to every other cluster.
If T is instead, say, 5, the set will consist of the union of the top 5 genes from each pairwise comparison.
</li>
<li>
lfc_summary or auc_summary - the summary effect size is defined as the effect size from the pairwise comparison
with the lowest p-value. The effect size can be either log2 fold-change or area under the curve (for Wilcox test).
</li>
<li>pval - combined p-value from all pairwise comparisons</li>
<li>fdr - p-value adjusted for multiple testing issue (false discovery rate)</li>
<li>avg_self - average log2 counts in the current group</li>
<li>avg_other - average log2 counts in the other groups</li>
<li>pct_self - percentage of cells with detected expression (counts > 0) in the current group</li>
<li>pct_other - percentage of cells with detected expression (counts > 0) in the other groups</li>
<li>lfc_&lt;cell_group&gt; or auc_&lt;cell_group&gt; - effect size of the current group versus some other one</li>
</ul>

For more details see the documentation of the used methods from scran package:
<a href='https://rdrr.io/bioc/scran/man/combineMarkers.html' target='_blank'>combineMarkers()</a> and
<a href='https://rdrr.io/bioc/scran/man/findMarkers.html' target='_blank'>findMarkers()</a>

</div>
</div>
")
```

***

```{r, include = FALSE}
markers <- other_params$markers
assert_that(is.data.frame(markers))
```

```{r}
if (other_params$test_type == "wilcox") {
  effect_colnames <- str_subset(colnames(markers), "^auc_")
} else {
  effect_colnames <- str_subset(colnames(markers), "^lfc_")
}

if ("top" %in% colnames(markers)) {
  order_col <- "top"
} else {
  order_col <- "fdr"
}

download_title <- other_params$id

table_dt <- DT::datatable(
  markers,
  filter = "top",
  rownames = FALSE,
  escape = FALSE,
  width = "100%",
  class = "display compact",
  extensions = c("Buttons", "FixedHeader"),
  options = list(
    pageLength = 100,
    order = list(which(colnames(markers) == order_col) - 1, "asc"),
    autoWidth = TRUE,
    dom = "Bipfrtip",
    fixedHeader = TRUE,
    lengthMenu = list(list(10, 25, 50, -1), list(10, 25, 50, "All")),
    buttons =
      list("copy", "print", list(
        extend = "collection",
        text = "Download",
        buttons = list(
          list(extend = "excel", title = download_title),
          list(extend = "pdf", title = download_title),
          list(extend = "csv", title = download_title)
        )
      ))
  )
) %>%
  DT::formatSignif(columns = c("lfc_summary", "pval", "fdr", "avg_self", "avg_other", effect_colnames), digits = 3)

table_dt
```

# {-}

***

```{r, child = here::here("Rmd/common/_footer.Rmd")}
```
