<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light"><script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/5/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Changelog • scdrake</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/5/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)"><!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/5/darkly/bootstrap.css" media="(prefers-color-scheme: dark)"><!-- preferably CSS --><link rel="stylesheet" href="../preferably.css"><link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)"><link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)"><script src="../darkswitch.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scdrake</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">1.7.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/scdrake.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/scdrake_integration.html">Integration pipeline guide</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/pipeline_overview.html">Pipeline overview</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/scdrake_spatial.html">Spatial extension</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/scdrake_docker.html">Using the Docker image</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_integration.html">02 Integration pipeline guide</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_advanced.html">Advanced topics</a></li>
    <li><a class="dropdown-item" href="../articles/interaction_analysis.html">Interaction analysis</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_extend.html">Extending the pipeline</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_development.html">Developing scdrake</a></li>
    <li><a class="dropdown-item" href="../articles/drake_basics.html">drake basics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>General information</h6></li>
    <li><a class="dropdown-item" href="../articles/pipeline_overview.html">Pipeline overview</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_cli.html">Command line interface</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_config.html">scdrake config internals</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_envvars.html">Environment variables</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_spatial.html">Spatial extension</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>General configs</h6></li>
    <li><a class="dropdown-item" href="../articles/config_pipeline.html">Pipeline config</a></li>
    <li><a class="dropdown-item" href="../articles/config_main.html">Main config</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Single-sample pipeline stages</h6></li>
    <li><a class="dropdown-item" href="../articles/stage_input_qc.html">Input and quality control stage (`01_input_qc`)</a></li>
    <li><a class="dropdown-item" href="../articles/stage_norm_clustering.html">Normalization and clustering stage (`02_norm_clustering`)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Integration pipeline stages</h6></li>
    <li><a class="dropdown-item" href="../articles/stage_integration.html">Integration stage (`01_integration`)</a></li>
    <li><a class="dropdown-item" href="../articles/stage_int_clustering.html">Post-integration clustering stage (`02_int_clustering`)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Common stages</h6></li>
    <li><a class="dropdown-item" href="../articles/stage_cluster_markers.html">Cluster markers stage (`cluster_markers`)</a></li>
    <li><a class="dropdown-item" href="../articles/stage_contrasts.html">Contrasts stage (`contrasts`)</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/bioinfocz/scdrake" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>


        <li>
          <a class="external-link nav-link" id="css-toggle-btn" aria-label="github">
            <span class="fas fa fas fa-adjust fa-lg"></span>
          </a>
        </li>



      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Changelog</h1>
      <small>Source: <a href="https://github.com/bioinfocz/scdrake/blob/master/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.7.0" id="scdrake-170">scdrake 1.7.0<a class="anchor" aria-label="anchor" href="#scdrake-170"></a></h2>
<ul><li>Added a new <strong>deconvolution module</strong> for secondary analysis.
<ul><li>Supports deconvolution of spatial transcriptomics data using single-cell references.</li>
<li>See <code><a href="../articles/scdrake_spatial.html">vignette("scdrake_spatial")</a></code> for details.</li>
</ul></li>
<li>Added <strong>spatial artifact detection</strong> for spot-based technologies.</li>
<li>Added <a href="https://www.bioconductor.org/packages/release/bioc/html/SpatialExperiment.html" class="external-link">SpatialExperiment</a> for working with spatial data</li>
<li>Updated visualization:
<ul><li>Switched from self-made functions to <a href="https://github.com/lmweber/ggspavis" class="external-link">ggspavis</a>-based plotting functions for more consistent and customizable outputs.</li>
<li>Updated writing plots functions for dimension-reduction plots</li>
</ul></li>
<li>Updated Spatially variable genes selections</li>
<li>Spatial part is now up-to-date with <a href="https://lmweber.org/OSTA/" class="external-link">Orchestrating Spatial Transcriptomics Analysis with Bioconductor</a> suggestions</li>
<li>HTML reports:
<ul><li>Improved layout and structure for better readability.</li>
<li>Minor visualization enhancements in plots and tables.</li>
</ul></li>
<li>Environment updates:
<ul><li>
<strong>R 4.5.1</strong> and <strong>Bioconductor 3.21</strong>.</li>
<li>Official Docker image updated to match.</li>
</ul></li>
<li>Fixes:
<ul><li>Minor fixes in cell annotation workflows and html report generating functions</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.6.0" id="scdrake-160">scdrake 1.6.0<a class="anchor" aria-label="anchor" href="#scdrake-160"></a></h2>
<ul><li>
<code>scdrake</code> now allows processing of spatial transcriptomics data from spot-based technologies (Visium).
<ul><li>See <code><a href="../articles/scdrake_spatial.html">vignette("scdrake_spatial")</a></code>.</li>
</ul></li>
<li>Added annotation using user-defined marker genes.</li>
<li>Updated <code>stage_input_qc</code> and <code>stage_norm_clustering</code> vignettes.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.5.0" id="scdrake-150">scdrake 1.5.0<a class="anchor" aria-label="anchor" href="#scdrake-150"></a></h2>
<ul><li>Major refactoring:
<ul><li>Divided some sections into smaller subplans (e.g. dimred plots or clustering). This also applies to child RMarkdown documents.</li>
<li>Minor config polishing.</li>
</ul></li>
<li>Changes in clustering:
<ul><li>All clustering types can be now enabled or disabled.</li>
<li>Added Leiden algorithm for graph-based clustering.</li>
<li>Added multiple resolutions for Leiden and Louvain algorithms.</li>
<li>Polished clustering names in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>.</li>
<li>Added <a href="https://lazappi.github.io/clustree/" class="external-link">clustree</a> to visualize cluster abundances under different <code>k</code>s or resolutions.</li>
</ul></li>
<li>Added <a href="https://github.com/immunogenomics/harmony" class="external-link">Harmony</a> integration.</li>
<li>Allow to load data in the <code>01_input_qc</code> stage also as a SCE object from a <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cache.</li>
<li>Subsetting of imported SCE object in the <code>01_input_qc</code> stage.</li>
<li>Major refactoring and update of documentation (vignettes).</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.4.1" id="scdrake-141">scdrake 1.4.1<a class="anchor" aria-label="anchor" href="#scdrake-141"></a></h2>
<ul><li>Fixed line ending option in the autogenerated RStudio file for new projects.</li>
<li>
<code>"rds"</code> is now the default storage format for <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cache.</li>
<li>Added more <a href="https://docs.ropensci.org/drake" class="external-link">drake</a>-specific config parameters to <code>config/pipeline.yaml</code>.
<ul><li>See <code>DRAKE_MEMORY_STRATEGY</code> which can help on machines with low memory.</li>
</ul></li>
<li>Added support of <code>arm64</code> architecture to <code><a href="../reference/download_yq.html">download_yq()</a></code>.</li>
<li>Added a Dockerfile for <code>arm64</code> platform.</li>
<li>Updated <code>scdrake_docker</code> vignette.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.4.0" id="scdrake-140">scdrake 1.4.0<a class="anchor" aria-label="anchor" href="#scdrake-140"></a></h2>
<ul><li>
<code>scdrake</code> has been ported to a Docker image and this is now the recommended usage.
<ul><li>See <code><a href="../articles/scdrake_docker.html">vignette("scdrake_docker")</a></code>.</li>
</ul></li>
<li>A command line interface was implemented. It wraps several main <code>scdrake</code> functions.
<ul><li>See <code><a href="../articles/scdrake_cli.html">vignette("scdrake_cli")</a></code>.</li>
</ul></li>
<li>
<code><a href="../reference/run_pipeline.html">run_single_sample()</a></code> and <code><a href="../reference/run_pipeline.html">run_integration()</a></code> are now soft-deprecated.</li>
<li>Several bugs have been fixed.</li>
<li>Text/visual/user-experience refinements.</li>
<li>Dev changes:
<ul><li>
<code>dev/run_check.R</code> and <code>dev/run_tests.R</code> (former <code>run_pipeline_tests.R</code>) now uses R-based CLI and are run through <code>Rscript</code>.</li>
<li>
<code>deps.Dockerfile</code> will install <code>scdrake</code> with all dependencies from <code>DESCRIPTION</code>.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.2" id="scdrake-132">scdrake 1.3.2<a class="anchor" aria-label="anchor" href="#scdrake-132"></a></h2>
<ul><li>Simplified default configs, vignettes updated accordingly.</li>
<li>Minor updates in some vignettes.</li>
<li>Added pipeline <a href="diagrams/README.html">diagrams</a>.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.1" id="scdrake-131">scdrake 1.3.1<a class="anchor" aria-label="anchor" href="#scdrake-131"></a></h2>
<ul><li>New feature: “inject” custom cell data. See the <code>ADDITIONAL_CELL_DATA_FILE</code> parameter in <code>02_norm_clustering</code> stage vignette.</li>
<li>Datasets for integration can be now imported also from Rds files. The structure of the <code>INTEGRATION_SOURCES</code> parameters has changed - please, review your configs.</li>
<li>Updated documentation.</li>
<li>Fixed some rare bugs.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.3.0" id="scdrake-130">scdrake 1.3.0<a class="anchor" aria-label="anchor" href="#scdrake-130"></a></h2>
<ul><li>Added support for a custom plan located (by default) in <code>plan_custom.R</code> script in project directory.</li>
<li>The GitHub version of <a href="https://github.com/SingleR-inc/SingleR" class="external-link">SingleR</a> was removed from <code>DESCRIPTION</code> as the version for the current Bioconductor version is available from the official Bioc repository. The GitHub version is compatible with Bioc 3.16 and caused installation error on Bioc 3.15</li>
<li>Fixed “object cfg not found”.</li>
<li>Documentation has been improved:
<ul><li>Clearer instructions in the Get Started vignette. The part about the integration pipeline was moved to a separate vignette.</li>
<li>New vignettes about extension of the pipeline and FAQ &amp; Howto.</li>
<li>Clearer installation instructions, including commands needed to install shared libraries on different OSs.</li>
<li>Some parts in other vignettes have been rewritten, spellchecked and improved cosmetically.</li>
</ul></li>
<li>Cell or gene filtering can be now disabled by a single parameter (<code>ENABLE_CELL_FILTERING</code>, <code>ENABLE_GENE_FILTERING</code>).</li>
<li>Require <a href="https://mschubert.github.io/clustermq/" class="external-link">clustermq</a> version <code>&gt;= 0.8.8</code>, but show a warning if the version is greater.</li>
<li>Checks for <a href="https://github.com/hemberg-lab/SC3" class="external-link">SC3</a> version if it is not installed from <code>github.com/gorgitko/SC3</code>, but also allow version from Bioconductor.</li>
<li>Show used functions with links in stage reports.</li>
<li>New pipeline config variable <code>DRAKE_REBUILD</code> allowing to force rebuild targets.</li>
<li>Fixed various corner case bugs.</li>
<li>Small enhancements in some functions.</li>
<li>Some vignettes have been extended or updated.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.2.3" id="scdrake-123">scdrake 1.2.3<a class="anchor" aria-label="anchor" href="#scdrake-123"></a></h2>
<ul><li>Fixed a bug in the <code>sce_final_norm_clustering</code> target where already present columns in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> were not updated by new ones in the <code>cell_data</code> target. This could happen when SCE object was imported in the <code>01_input_qc</code> stage and already contained cluster assignments in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> from the previous <code>scdrake</code> run.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.2.2" id="scdrake-122">scdrake 1.2.2<a class="anchor" aria-label="anchor" href="#scdrake-122"></a></h2>
<ul><li>Fixed a typo in <code>01_input_qc</code> stage report, added a line for <code>emptyDroplets</code> lower bound in barcode rank plot.</li>
<li>
<code><a href="../reference/save_pdf.html">save_pdf()</a></code> is now able to treat invalid plots throwing an error on save.
<ul><li>A new helper function <code><a href="../reference/create_dummy_plot.html">create_dummy_plot()</a></code>.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.2.1" id="scdrake-121">scdrake 1.2.1<a class="anchor" aria-label="anchor" href="#scdrake-121"></a></h2>
<ul><li>Dimred plots of clusterings and other variables are now saved as PDF and PNG files.
<ul><li>In <code>02_norm_clustering</code> and <code>02_int_clustering</code> stage reports the PNG files are displayed and act as links to associated multipage PDF files.</li>
<li>New parameters for the stages above: <code>NORM_CLUSTERING_DIMRED_PLOTS_OUT_DIR</code> and <code>INT_CLUSTERING_DIMRED_PLOTS_OUT_DIR</code>, respectively.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.2.0" id="scdrake-120">scdrake 1.2.0<a class="anchor" aria-label="anchor" href="#scdrake-120"></a></h2>
<ul><li>Added cell type annotation via <a href="https://github.com/SingleR-inc/SingleR" class="external-link">SingleR</a> (<a href="https://github.com/LTLA/SingleR" class="external-link uri">https://github.com/LTLA/SingleR</a>).
<ul><li>See the new config parameters <a href="https://bioinfocz.github.io/scdrake/articles/stage_norm_clustering.html#cell-type-annotation">here</a> and targets <a href="https://bioinfocz.github.io/scdrake/articles/stage_norm_clustering.html#cell-type-annotation-1">here</a>.</li>
</ul></li>
<li>Code and documentation refactoring, minor bug fixes.</li>
<li>Updated <code>renv.lock</code>.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.1.0" id="scdrake-110">scdrake 1.1.0<a class="anchor" aria-label="anchor" href="#scdrake-110"></a></h2>
<ul><li>The single-sample pipeline now supports another two inputs:
<ul><li>A delimited text file (table) representing a feature-barcode matrix.</li>
<li>A <code>SingleCellExperiment</code> object saved as a Rds file.</li>
</ul></li>
<li>Input type is controlled by a new <code>INPUT_DATA</code> parameter (replaces <code>INPUT_10X_DIR</code>) in <code>01_input_qc.yaml</code> config. Details can be found in <code><a href="../articles/stage_input_qc.html">vignette("stage_input_qc")</a></code>.</li>
<li>Removal of empty droplets can be now disabled (<code>EMPTY_DROPLETS_ENABLED</code> parameter).</li>
<li>It is now possible to turn off normalization (<code>NORMALIZATION_TYPE: "none"</code>) as the input can be an already normalized <code>SingleCellExperiment</code> object (meaning it went through the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> pipeline before).</li>
<li>Minor updates and bug fixes, updated documentation.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.1" id="scdrake-101">scdrake 1.0.1<a class="anchor" aria-label="anchor" href="#scdrake-101"></a></h2>
<ul><li>Fixed <code><a href="../reference/download_yq.html">download_yq()</a></code>:
<ul><li>
<code><a href="../reference/download_yq.html">download_yq()</a></code> now supports the masOS (<code>darwin</code>).</li>
<li>Fixed URL to Windows version (<code>.exe</code> extension).</li>
<li>Better OS resolution implemented in <code>.get_os()</code>.</li>
</ul></li>
<li>Fixed issues with <code>yq</code> tool’s binary download on Windows.</li>
<li>Modified <code>.confirm_menu()</code> for non-interactive usage.</li>
<li>Activate an RStudio project on the end of <code><a href="../reference/init_project.html">init_project()</a></code> function.</li>
<li>Fixed links to source code on GH.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.0" id="scdrake-100">scdrake 1.0.0<a class="anchor" aria-label="anchor" href="#scdrake-100"></a></h2>
<ul><li>Initial release.</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="copyright">
  <p></p><p>Developed by Jiri Novotny, Jan Kubovciak, Lucie Pfeiferova, Michal Kolar.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

    </footer></div>





  </body></html>

