<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scdrake">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/5/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Pipeline overview • scdrake</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Pipeline overview">
<meta property="og:description" content="scdrake">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/5/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/5/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="../darkswitch.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scdrake</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake_integration.html">Integration pipeline guide</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Guides</h6>
    <a class="dropdown-item" href="../articles/scdrake_docker.html">Using the Docker image</a>
    <a class="dropdown-item" href="../articles/scdrake_integration.html">02 Integration pipeline guide</a>
    <a class="dropdown-item" href="../articles/scdrake_extend.html">Extending the pipeline</a>
    <a class="dropdown-item" href="../articles/drake_basics.html">drake basics</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General information</h6>
    <a class="dropdown-item" href="../articles/pipeline_overview.html">Pipeline overview</a>
    <a class="dropdown-item" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
    <a class="dropdown-item" href="../articles/scdrake_cli.html">Command line interface</a>
    <a class="dropdown-item" href="../articles/cluster_markers.html">Cluster markers</a>
    <a class="dropdown-item" href="../articles/scdrake_config.html">scdrake config internals</a>
    <a class="dropdown-item" href="../articles/scdrake_envvars.html">Environment variables</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General configs</h6>
    <a class="dropdown-item" href="../articles/config_pipeline.html">Pipeline config</a>
    <a class="dropdown-item" href="../articles/config_main.html">Main config</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single-sample pipeline stages</h6>
    <a class="dropdown-item" href="../articles/stage_input_qc.html">Input and quality control stage (`01_input_qc`)</a>
    <a class="dropdown-item" href="../articles/stage_norm_clustering.html">Normalization and clustering stage (`02_norm_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Integration pipeline stages</h6>
    <a class="dropdown-item" href="../articles/stage_integration.html">Integration stage (`01_integration`)</a>
    <a class="dropdown-item" href="../articles/stage_int_clustering.html">Post-integration clustering stage (`02_int_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Common stages</h6>
    <a class="dropdown-item" href="../articles/stage_cluster_markers.html">Cluster markers stage (`cluster_markers`)</a>
    <a class="dropdown-item" href="../articles/stage_contrasts.html">Contrasts stage (`contrasts`)</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bioinfocz/scdrake" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>

        
        <li>
          <a class="external-link nav-link" id="css-toggle-btn" aria-label="github">
            <span class="fas fa fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        
        
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Pipeline overview</h1>
            
            <h4 data-toc-skip class="date"><sup>Document generated:
2022-12-18 21:19:59 UTC+0000</sup></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bioinfocz/scdrake/blob/mastervignettes/pipeline_overview.Rmd" class="external-link"><code>vignettes/pipeline_overview.Rmd</code></a></small>
      <div class="d-none name"><code>pipeline_overview.Rmd</code></div>
    </div>

    
    
<hr>
<p><a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> offers two pipelines - one for single-sample,
and second one for integration of multiple samples (which were processed
by the single-sample pipeline before). As for now, each pipeline
consists of two subpipelines (referred to as <em>stages</em>), and two
stages common to both single-sample and integration pipelines.</p>
<p><a href="https://github.com/bioinfocz/scdrake/blob/main/diagrams/pipeline_overview.pdf" class="external-link"><img src="pipeline_overview.png" alt="A high-level diagram of {scdrake} pipelines."></a></p>
<p>A more detailed diagram with target structure can be found <a href="https://github.com/bioinfocz/scdrake/blob/main/diagrams/pipeline_overview.pdf" class="external-link">here</a>.</p>
<p>Each stage has its own config, plus there is a main config for each
pipeline. You can read more about configs in a separate
<code><a href="../articles/scdrake_config.html">vignette("scdrake_config")</a></code>. Each stage also outputs a
report in HTML format with rich graphics.</p>
<p>Advanced users might be interested in looking into source code of
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>’s <a href="https://github.com/bioinfocz/scdrake/tree/main/R" class="external-link">plans</a> (files
named <code>plans_*.R</code>).</p>
<p>Pipeline steps are mostly based on recommendations given in a great
book <a href="https://bioconductor.org/books/release/OSCA/" class="external-link">Orchestrating
Single-Cell Analysis with Bioconductor</a>.</p>
<div class="section level2">
<h2 id="example-pipeline-output">Example pipeline output<a class="anchor" aria-label="anchor" href="#example-pipeline-output"></a>
</h2>
<p><strong>You can inspect output from the pipeline <a href="https://onco.img.cas.cz/novotnyj/scdrake/" class="external-link">here</a>.</strong></p>
<p>The used datasets are:</p>
<ul>
<li>
<a href="https://www.10xgenomics.com/resources/datasets/1-k-pbm-cs-from-a-healthy-donor-v-3-chemistry-3-standard-3-0-0" class="external-link">PBMC
1k</a> (v3 chemistry, Cell Ranger 3.0.0)</li>
<li>
<a href="https://www.10xgenomics.com/resources/datasets/3-k-pbm-cs-from-a-healthy-donor-1-standard-1-1-0" class="external-link">PBMC
3k</a> (v2 chemistry, Cell Ranger 1.1.0)</li>
</ul>
<p><strong>All credits for these datasets go to 10x Genomics.</strong>
Visit <a href="https://www.10xgenomics.com/resources/datasets" class="external-link uri">https://www.10xgenomics.com/resources/datasets</a> for more
information.</p>
</div>
<div class="section level2">
<h2 id="single-sample-pipeline">Single-sample pipeline<a class="anchor" aria-label="anchor" href="#single-sample-pipeline"></a>
</h2>
<p>This is a pipeline for processing a single-sample, where input is
either:</p>
<ul>
<li>An output from <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">cellranger</a>,
that is, three files of a raw feature-barcode matrix*
(<code>barcodes.tsv</code>, <code>features.tsv</code>, and
<code>matrix.mtx</code>) These files can be also gzipped having a
<code>.gz</code> extension.</li>
<li>A delimited file (table) representing a feature-barcode matrix, with
an additional column named <code>ENSEMBL</code> containing Ensembl IDs
of features.</li>
<li>A saved <code>SingleCellExperiment</code> in Rds format (i.e. saved
by the <code><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS()</a></code> function).</li>
</ul>
<p>Input specification is controlled via the <code>INPUT_DATA</code>
parameter in <code>01_input_qc</code> config, and more details can be
found in <code><a href="../articles/stage_input_qc.html">vignette("stage_input_qc")</a></code>.</p>
<p>Feature-barcode matrix is similar to count matrix in bulk RNA-seq
where columns are samples, rows are genes or transcripts (features), and
values are numbers of reads or UMIs overlapping the features. In
scRNA-seq, columns are individual cells; the rest is same.</p>
<div class="section level3">
<h3 id="stage-01_input_qc">Stage <code>01_input_qc</code><a class="anchor" aria-label="anchor" href="#stage-01_input_qc"></a>
</h3>
<p>An initial stage to read in scRNA-seq data, remove empty droplets
(optionally), and perform quality control, and cell and gene
filtering.</p>
<p>More information about scRNA-seq quality control can be found <a href="https://bioconductor.org/books/release/OSCA.basic/quality-control.html" class="external-link">here</a>.</p>
<p>⚙️ Config and targets: <code><a href="../articles/stage_input_qc.html">vignette("stage_input_qc")</a></code></p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/single_sample/pbmc1k/01_input_qc/01_input_qc.html" class="external-link">📜
Example report</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/single_sample_pbmc1k_full/01_input_qc.default.yaml" class="external-link">used
config</a>)</p>
<p>🪜 Structure</p>
<ul>
<li>Reading in feature-barcode matrix either from:
<ul>
<li>
<code>cellranger</code> output</li>
<li>
<code>SingleCellExperiment</code> object</li>
<li>delimited table</li>
</ul>
</li>
<li>Removal of <a href="https://bioconductor.org/books/release/OSCA.advanced/droplet-processing.html" class="external-link">empty
droplets</a> (<code><a href="https://rdrr.io/pkg/DropletUtils/man/emptyDrops.html" class="external-link">DropletUtils::emptyDrops()</a></code>) (optional).</li>
<li>Calculation of per-cell QC metrics
(<code><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">scater::perCellQCMetrics()</a></code>, <a href="https://bioconductor.org/books/release/OSCA.basic/quality-control.html#common-choices-of-qc-metrics" class="external-link">theory</a>):
<ul>
<li>Number of UMI.</li>
<li>Number of detected genes (non-zero UMI count).</li>
<li>Percentage of expressed mitochondrial genes (<span class="math inline">\(\frac {UMI_{mitochondrial}} {UMI_{sum}} *
100\)</span>).</li>
</ul>
</li>
<li>Two methods for cell filtering (optional):
<ul>
<li>Dataset-sensitive filtering, based on the median absolute deviation
(MAD) from the median value of each QC metric across all cells. For
“number of UMI” metric, only a lower tail is used (<a href="https://bioconductor.org/books/release/OSCA.basic/quality-control.html#quality-control-outlier" class="external-link">theory</a>.</li>
<li>Custom filtering, based on fixed thresholds. For “number of UMI”
metric, both upper and lower bounds are used (<a href="https://bioconductor.org/books/release/OSCA.basic/quality-control.html#fixed-qc" class="external-link">theory</a>.</li>
<li>In both filtering types, violation of only one metric threshold
leads to removal of a cell. More technically, metric filters are reduced
by <strong>or</strong> operator.</li>
</ul>
</li>
<li>Gene filtering (optional) based on a minimum number of UMI per cell
and a minimum ratio of cells expressing a gene.
<ul>
<li>A gene is considered expressed when number of its UMI across all
cells is greater than X and at the same time it is expressed in at least
Y ratio of cells.</li>
</ul>
</li>
<li>Plots of QC metrics with visualization of differences between
dataset-sensitive and custom filtering.</li>
<li>Preparation of gene annotation.</li>
<li>Final selection of either dataset-sensitive or custom filtered
dataset. This selection is upon the user.</li>
</ul>
</div>
<div class="section level3">
<h3 id="stage-02_norm_clustering">Stage <code>02_norm_clustering</code><a class="anchor" aria-label="anchor" href="#stage-02_norm_clustering"></a>
</h3>
<p>A stage for normalization, dimensionality reduction, and cell
clustering.</p>
<p>⚙️ Config and targets:
<code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code></p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/single_sample/pbmc1k/02_norm_clustering/02_norm_clustering.html" class="external-link">📜
Example report</a></p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/single_sample/pbmc1k/02_norm_clustering/02_norm_clustering_simple.html" class="external-link">📜
Example simplified report</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/single_sample_pbmc1k_full/02_norm_clustering.default.yaml" class="external-link">used
config</a>)</p>
<p>🪜 Structure</p>
<ul>
<li>Calculation of cell cycle score and assignment of cells to phases
(G1, G2M, S) using the list of cell cycle genes in
<code><a href="https://satijalab.org/seurat/reference/cc.genes.updated.2019.html" class="external-link">Seurat::cc.genes.updated.2019</a></code> and
<code><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html" class="external-link">Seurat::CellCycleScoring()</a></code> function.</li>
<li>
<a href="https://bioconductor.org/books/release/OSCA.basic/normalization.html" class="external-link">Normalization</a>
of UMI counts by either:
<ul>
<li>
<a href="https://bioconductor.org/books/release/OSCA.basic/normalization.html#normalization-by-deconvolution" class="external-link">Normalization
by deconvolution</a> implemented in the <a href="https://bioconductor.org/packages/release/bioc/html/scran.html" class="external-link">scran</a>
package.</li>
<li>Regularized negative binomial regression to normalize UMI count data
implemented in the <a href="https://github.com/ChristophH/sctransform" class="external-link">sctransform</a> package
and wrapped by the <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>
package. This method also allows to regress out confounding variables
such as cell cycle score or percentage of mitochondrial genes
expression. Note that for cell cycle score, a better method seems to be
removal of CC-related genes prior to detection of highly variable genes
(HVGs, see below). <code>sctransform</code> also returns a specified
number of HVGs.</li>
</ul>
</li>
<li>Selection of <a href="https://bioconductor.org/books/release/OSCA.basic/feature-selection.html" class="external-link">highly
variable genes</a> (HVGs):
<ul>
<li>HVGs are used downstream for dimensionality reduction and
clustering.</li>
<li>Three metrics for HVGs are available:
<ul>
<li>
<a href="https://bioconductor.org/books/release/OSCA.basic/feature-selection.html#quantifying-per-gene-variation" class="external-link">Gene
variance</a>.</li>
<li>
<a href="http://bioconductor.org/books/release/OSCA.advanced/more-hvgs.html#using-the-coefficient-of-variation" class="external-link">Coefficient
of variation</a> (CV2).</li>
<li>HVGs obtained from the <code>sctransform</code> normalization.</li>
</ul>
</li>
<li>Based on the selected one from the first two metrics above, HVGs are
selected either by:
<ul>
<li>Top N genes based on a metric (e.g. top 1000 genes with the highest
variance).</li>
<li>Significance (FDR) threshold.</li>
<li>Threshold on a metric value.</li>
</ul>
</li>
<li>Prior to HVG selection, cell cycle-related genes can be removed. We
are using a method based on the percentage of variance explained by the
cell cycle phase in the expression profile for each gene (<a href="https://bioconductor.org/books/release/OSCA.advanced/cell-cycle-assignment.html#removing-cell-cycle-related-genes" class="external-link">details</a>).</li>
</ul>
</li>
<li>Removal of <a href="https://bioconductor.org/books/release/OSCA.advanced/doublet-detection.html" class="external-link">cell
doublets</a> (<code><a href="https://rdrr.io/pkg/scDblFinder/man/computeDoubletDensity.html" class="external-link">scDblFinder::computeDoubletDensity()</a></code>).</li>
<li>
<a href="https://bioconductor.org/books/release/OSCA.basic/dimensionality-reduction.html#principal-components-analysis" class="external-link">Calculation
of PCA</a> and <a href="https://bioconductor.org/books/release/OSCA.basic/dimensionality-reduction.html#choosing-the-number-of-pcs" class="external-link">selection
of number of principal components</a> (PCs) which will be used
downstream. There are three methods to select a proper number of PCs:
<ul>
<li>Using the <a href="http://bioconductor.org/books/release/OSCA.advanced/dimensionality-reduction-redux.html#more-choices-for-the-number-of-pcs" class="external-link">elbow
point</a>.</li>
<li>Using the <a href="http://bioconductor.org/books/release/OSCA.advanced/dimensionality-reduction-redux.html#using-the-technical-noise" class="external-link">technical
noise</a>.</li>
<li>Using a fixed number of PCs.</li>
</ul>
</li>
<li>
<a href="https://bioconductor.org/books/release/OSCA.basic/clustering.html" class="external-link">Cell
clustering</a> using several algorithms:
<ul>
<li>
<a href="https://bioconductor.org/books/release/OSCA.basic/clustering.html#clustering-graph" class="external-link">Graph-based
clustering</a> (mutual nearest neighbors) using Louvain and Walktrap
algorithms.</li>
<li>
<a href="https://bioconductor.org/books/release/OSCA.basic/clustering.html#vector-quantization-with-k-means" class="external-link">K-means
clustering</a> using both defined numbers of K and best K
selection.</li>
<li>
<a href="https://bioconductor.org/packages/release/bioc/html/SC3.html" class="external-link">Single-Cell
Consensus Clustering</a> (SC3) with defined numbers of clusters.</li>
</ul>
</li>
<li>Cell type annotation using <a href="https://github.com/LTLA/SingleR" class="external-link">SingleR</a>. More details are
given in the accompanying <a href="http://bioconductor.org/books/release/SingleRBook/" class="external-link">book</a> and
the underlying method is described <a href="http://bioconductor.org/books/release/SingleRBook/introduction.html#method-description" class="external-link">here</a>.
<ul>
<li>Reference datasets can be taken from the <a href="https://github.com/LTLA/celldex" class="external-link">celldex</a>
package or an arbitrary <code>SingleCellExperiment</code> /
<code>SummarizedExperiment</code> object can be loaded from a file.</li>
</ul>
</li>
<li>Assignment of cells to predefined groupings based on existing data.
That means you can reuse an assignment of cells from e.g. clustering,
and rename or merge its levels (cluster numbers).
<ul>
<li>Example: you identify a biological meaning of clusters computed by
k-means with <span class="math inline">\(k = 3\)</span>. You can use
this information, and rename and merge the clusters, such that:
<ul>
<li>Cluster 1 -&gt; T cells</li>
<li>Cluster 2 -&gt; B cells</li>
<li>Cluster 3 -&gt; T cells</li>
</ul>
</li>
<li>Now you can use this new cell grouping for e.g. plotting of reduced
dimensions or calculation of cluster markers.</li>
</ul>
</li>
<li>
<a href="https://bioconductor.org/books/release/OSCA.basic/dimensionality-reduction.html#non-linear-methods-for-visualization" class="external-link">Dimensionality
reduction</a>:
<ul>
<li>PCA</li>
<li><a href="https://bioconductor.org/books/release/OSCA.basic/dimensionality-reduction.html#t-stochastic-neighbor-embedding" class="external-link">t-SNE</a></li>
<li><a href="https://bioconductor.org/books/release/OSCA.basic/dimensionality-reduction.html#uniform-manifold-approximation-and-projection" class="external-link">UMAP</a></li>
</ul>
</li>
<li>Expression plots of selected groups of genes (markers) in reduced
dimensions.</li>
<li>A simplified report for non-technical audience which contains only
dimred plots and selected markers.</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="integration-pipeline">Integration pipeline<a class="anchor" aria-label="anchor" href="#integration-pipeline"></a>
</h2>
<p>This is a pipeline to integrate multiple samples processed by the
single-sample pipeline. Just for clarification, an individual sample is
also denoted as <em>batch</em>.</p>
<div class="section level3">
<h3 id="stage-01_integration">Stage <code>01_integration</code><a class="anchor" aria-label="anchor" href="#stage-01_integration"></a>
</h3>
<p>A stage for reading in single-sample data, integration using several
methods, and computing reduced dimensions. Note that currently only
samples which were <strong>all</strong> normalized by <code>scran</code>
method can be integrated.</p>
<p>More information can be found in this <a href="http://bioconductor.org/books/release/OSCA.multisample" class="external-link">OSCA
chapter</a>.</p>
<p>⚙️ Config and targets: <code><a href="../articles/stage_integration.html">vignette("stage_integration")</a></code></p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/integration/01_integration/01_integration.html" class="external-link">📜
Example report</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/integration/01_integration.default.yaml" class="external-link">used
config</a>)</p>
<p>🪜 Structure</p>
<ul>
<li>Reading in data from <code>drake</code> caches of multiple
single-sample pipelines.</li>
<li>Subsetting to common column and row data, and metadata.</li>
<li>Normalization of each sample for inter-batch sequencing depth
(<code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchNorm.html" class="external-link">batchelor::multiBatchNorm()</a></code>, <a href="http://bioconductor.org/books/3.15/OSCA.multisample/integrating-datasets.html#slower-setup" class="external-link">theory</a>).</li>
<li>Combining of HVGs by either (<a href="http://bioconductor.org/books/3.15/OSCA.multisample/integrating-datasets.html#slower-setup" class="external-link">theory</a>):
<ul>
<li>A HVG metric (gene variance or CV2), then apply top HVGs selection
as in the single-sample pipeline. All samples must have the same HVG
metric used before.</li>
<li>Set operations (without duplicates): intersection, union, or take
them all.</li>
</ul>
</li>
<li>Perform integration using one or more methods:
<ul>
<li>Rescaling (<code><a href="https://rdrr.io/pkg/batchelor/man/rescaleBatches.html" class="external-link">batchelor::rescaleBatches()</a></code>, <a href="http://bioconductor.org/books/3.15/OSCA.multisample/integrating-datasets.html#by-rescaling-the-counts" class="external-link">theory</a>):
scale counts so that the average count within each batch is the same for
each gene.</li>
<li>Regression (<code><a href="https://rdrr.io/pkg/batchelor/man/regressBatches.html" class="external-link">batchelor::regressBatches()</a></code>, <a href="http://bioconductor.org/books/3.15/OSCA.multisample/integrating-datasets.html#by-fitting-a-linear-model" class="external-link">theory</a>):
fit a linear model to each gene regress out uninteresting factors of
variation, returning a matrix of residuals.</li>
<li>Fast mutual nearest neighbors (<code><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html" class="external-link">batchelor::fastMNN()</a></code>, <a href="http://bioconductor.org/books/3.15/OSCA.multisample/integrating-datasets.html#mnn-correction" class="external-link">theory</a>).</li>
</ul>
</li>
</ul>
<p>The steps below are performed for each integration method, plus
separately for HVGs with removed cell cycle-related genes (if this
removal was requested).</p>
<ul>
<li>Calculation of PCA and selection of number of principal components
(PCs) which will be used downstream. This is similar to what is done in
the single-sample pipeline (<code>02_norm_clustering</code> stage), but,
in this case, the selection of PCs can be configured for each
integration method.</li>
<li>Dimensionality reduction, which is same as in the single-sample
pipeline, but as PCA can be configured for each integration method.</li>
<li>Expression plots of selected markers, see single-sample pipeline /
stage <code>02_norm_clustering</code>.</li>
<li>Mutual nearest neighbors (MNN) clustering to assess integration
results.</li>
<li>Integration diagnostics:
<ul>
<li>Assignment of cells of each sample to MNN clusters (tables,
plots).</li>
<li>
<a href="https://en.wikipedia.org/wiki/Rand_index" class="external-link">Rand indices</a>:
used to evaluate biological heterogeneity preservation by summarizing
the agreement between clusterings.</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="stage-02_int_clustering">Stage <code>02_int_clustering</code><a class="anchor" aria-label="anchor" href="#stage-02_int_clustering"></a>
</h3>
<p>This stage basically reproduces the clustering and cell type
annotation steps in the <code>02_norm_clustering</code> stage of the
single-sample pipeline. The only difference is the <strong>user
selection of a final integration method which will be used
downstream.</strong> HVGs, reduced dimensions, and selected markers are
already computed in the previous stage
(<code>01_integration</code>).</p>
<p>⚙️ Config and targets:
<code><a href="../articles/stage_int_clustering.html">vignette("stage_int_clustering")</a></code></p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/integration/02_int_clustering/02_int_clustering.html" class="external-link">📜
Example report</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/integration/02_int_clustering.default.yaml" class="external-link">used
config</a>)</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="common-stages">Common stages<a class="anchor" aria-label="anchor" href="#common-stages"></a>
</h2>
<p>Some stages are common to both single-sample and integration
pipelines.</p>
<div class="section level3">
<h3 id="stage-cluster_markers">Stage <code>cluster_markers</code><a class="anchor" aria-label="anchor" href="#stage-cluster_markers"></a>
</h3>
<p>A stage for calculation, visualization and reporting of cell cluster
markers (“global markers”).</p>
<p>⚙️ Config and targets:
<code><a href="../articles/stage_cluster_markers.html">vignette("stage_cluster_markers")</a></code></p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/single_sample/pbmc1k/cluster_markers/cluster_markers.html" class="external-link">📜
Example report for PBMC 1k data</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/single_sample_pbmc1k_full/cluster_markers.default.yaml" class="external-link">used
config</a>)</p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/integration/cluster_markers/cluster_markers.html" class="external-link">📜
Example report for integrated data</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/integration/cluster_markers.default.yaml" class="external-link">used
config</a>)</p>
<p>🪜 Structure</p>
<ul>
<li>
<a href="https://bioconductor.org/books/release/OSCA.basic/marker-detection.html" class="external-link">Calculation
of cluster markers</a>: any categorical grouping of cells can be used,
including those manually defined based on the existing groups (from
e.g. clustering), as described in the
<code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code>. More details about
marker computation can be found in
<code><a href="../articles/cluster_markers.html">vignette("cluster_markers")</a></code>.</li>
<li>Generation of heatmaps and plots. To reduce the size of heatmaps and
number of plots, you can control a number of top N markers taken by a
chosen metric (FDR, LFC, etc.). Only these top N markers will be
displayed in heatmaps, and plots made for them. This can be specified
for each cell grouping and test type.
<ul>
<li>Marker plots are composed of:
<ul>
<li>Two plots of a dimred: one colored by a clustering, and second
colored by expression of a marker (feature plot).</li>
<li>A plot of summarized marker expression: proportion of cells
expressing a marker, colored by average expression.</li>
<li>A violin plot: expression of a marker in each cluster.</li>
</ul>
</li>
</ul>
</li>
<li>Generation of marker tables for each cell grouping and each of its
levels.</li>
</ul>
</div>
<div class="section level3">
<h3 id="stage-contrasts">Stage <code>contrasts</code><a class="anchor" aria-label="anchor" href="#stage-contrasts"></a>
</h3>
<p>A stage for calculation, visualization and reporting of
differentially expressed markers (“contrasts”). This stage is basically
the same as the <code>cluster_markers</code> stage, but all output is
related to individual comparisons of levels of cell groupings. Hence
“contrasts”, a term known from bulk RNA-seq where sample groups are
compared -&gt; they are put <em>in contrast</em>.</p>
<p>⚙️ Config and targets: <code><a href="../articles/stage_contrasts.html">vignette("stage_contrasts")</a></code></p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/single_sample/pbmc1k/contrasts/contrasts.html" class="external-link">📜
Example report for PBMC 1k data</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/single_sample_pbmc1k_full/contrasts.default.yaml" class="external-link">used
config</a>)</p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/integration/contrasts/contrasts.html" class="external-link">📜
Example report for integrated data</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/integration/contrasts.default.yaml" class="external-link">used
config</a>)</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="copyright">
  <p></p>
<p>Developed by Jiri Novotny, Jan Kubovciak.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
