<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/3/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>01 Get started • scdrake</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="01 Get started">
<meta property="og:description" content="scdrake">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/3/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/3/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="../darkswitch.js"></script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scdrake</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scdrake.html">Get started</a>
</li>
<li>
  <a href="../articles/scdrake_integration.html">Integration pipeline guide</a>
</li>
<li>
  <a href="../articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li>
  <a href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="divider">
    </li>
<li class="dropdown-header">Guides</li>
    <li>
      <a href="../articles/scdrake_integration.html">02 Integration pipeline guide</a>
    </li>
    <li>
      <a href="../articles/scdrake_extend.html">Extending the pipeline</a>
    </li>
    <li>
      <a href="../articles/drake_basics.html">drake basics</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">General information</li>
    <li>
      <a href="../articles/pipeline_overview.html">Pipeline overview</a>
    </li>
    <li>
      <a href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
    </li>
    <li>
      <a href="../articles/cluster_markers.html">Cluster markers</a>
    </li>
    <li>
      <a href="../articles/scdrake_config.html">scdrake config internals</a>
    </li>
    <li>
      <a href="../articles/scdrake_envvars.html">Environment variables</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">General configs</li>
    <li>
      <a href="../articles/config_pipeline.html">Pipeline config</a>
    </li>
    <li>
      <a href="../articles/config_main.html">Main config</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Single-sample pipeline stages</li>
    <li>
      <a href="../articles/stage_input_qc.html">Input and quality control stage (`01_input_qc`)</a>
    </li>
    <li>
      <a href="../articles/stage_norm_clustering.html">Normalization and clustering stage (`02_norm_clustering`)</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Integration pipeline stages</li>
    <li>
      <a href="../articles/stage_integration.html">Integration stage (`01_integration`)</a>
    </li>
    <li>
      <a href="../articles/stage_int_clustering.html">Post-integration clustering stage (`02_int_clustering`)</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Common stages</li>
    <li>
      <a href="../articles/stage_cluster_markers.html">Cluster markers stage (`cluster_markers`)</a>
    </li>
    <li>
      <a href="../articles/stage_contrasts.html">Contrasts stage (`contrasts`)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
          <a href="#" id="css-toggle-btn">
            <span class="fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        <li>
  <a href="https://github.com/bioinfocz/scdrake" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
        
        


      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>01 Get started</h1>
                        <h4 data-toc-skip class="author">Jiri
Novotny</h4>
            <address class="author_afil">
      Institute of Molecular Genetics of the Czech Academy of
Sciences<br><a class="author_email" href="mailto:#"></a><a href="mailto:jiri.novotny@img.cas.cz" class="email">jiri.novotny@img.cas.cz</a>
      </address>
                              <h4 data-toc-skip class="author">Jan
Kubovciak</h4>
            <address class="author_afil">
      Institute of Molecular Genetics of the Czech Academy of
Sciences<br><a class="author_email" href="mailto:#"></a><a href="mailto:jan.kubovciak@img.cas.cz" class="email">jan.kubovciak@img.cas.cz</a>
      </address>
                  
            <h4 data-toc-skip class="date">2022-09-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bioinfocz/scdrake/blob/mastervignettes/scdrake.Rmd" class="external-link"><code>vignettes/scdrake.Rmd</code></a></small>
      <div class="hidden name"><code>scdrake.Rmd</code></div>

    </div>

    
    
<hr>
<div class="section level2">
<h2 id="quick-start-guide">Quick start guide<a class="anchor" aria-label="anchor" href="#quick-start-guide"></a>
</h2>
<p>In this guide you will see how to quickly setup
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> and run the first stage
(<code>01_input_qc</code>) of its single-sample pipeline using the
provided example data. Once you familiarize yourself with
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> basics, you can continue with guide for the
integration pipeline in
<code><a href="../articles/scdrake_integration.html">vignette("scdrake_integration")</a></code>.</p>
</div>
<div class="section level2">
<h2 id="the-three-steps">The three steps<a class="anchor" aria-label="anchor" href="#the-three-steps"></a>
</h2>
<p>After <a href="index.html#installation-instructions">installation</a>, you
basically need the three steps described below to run the
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> pipeline.</p>
<div class="section level3">
<h3 id="step-1-initialize-a-new-project">Step 1: initialize a new project<a class="anchor" aria-label="anchor" href="#step-1-initialize-a-new-project"></a>
</h3>
<p><a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> is a project-based package, so the first step
is to initialize a new project. <em>Project</em> simply means a
directory in which the analysis of your data will take place. That also
means:</p>
<ul>
<li>Your current working directory is set to the project directory.</li>
<li>Whenever you specify file paths, you specify them relative to the
project directory (e.g. <code>output/plots/figure1.pdf</code>). This way
your project is <strong>transferable</strong> between different
computers and data locations.</li>
</ul>
<p>Now we initialize a new <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> project directory.
First we load the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> package, and then call the
<code><a href="../reference/init_project.html">init_project()</a></code> function:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/init_project.html">init_project</a></span><span class="op">(</span><span class="st">"~/scdrake_single_sample_example"</span>, download_example_data <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>This will:</p>
<ul>
<li>Create a new project directory named
<code>scdrake_single_sample_example</code> in your home.</li>
<li>Switch your current working directory to project’s directory.</li>
<li>Copy all files, which are bundled with the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>
package: default YAML configs, Rmd files, R scripts etc.</li>
<li>Download (if needed) the <a href="https://github.com/mikefarah/yq" class="external-link">yq tool</a> used to manipulate
YAML files.</li>
<li>Create a new RStudio project (<code>.RProj</code>) file and set it
as the active project (can be disabled).</li>
<li>Download the example dataset (10X Genomics PBMC).</li>
</ul>
<blockquote>
<p>Important: whenever you will be running the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>
pipeline, make sure your working directory is set to the project’s
root.</p>
</blockquote>
<p>Let’s inspect files in the project directory:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span>all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #0000BB; font-weight: bold;">/tmp/RtmpWbugOM/file1a2cf4ce21e</span></span></span>
<span><span class="co">## ├── .here</span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">Rmd</span></span></span>
<span><span class="co">## │   ├── <span style="color: #0000BB; font-weight: bold;">common</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">_footer.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">_header.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">cluster_markers.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">cluster_markers_table_template.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">contrasts.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">contrasts_table_template.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">dimred_plots.Rmd</span></span></span>
<span><span class="co">## │   │   └── stylesheet.css</span></span>
<span><span class="co">## │   ├── <span style="color: #0000BB; font-weight: bold;">integration</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">01_integration.Rmd</span></span></span>
<span><span class="co">## │   │   └── <span style="color: #00BB00;">02_int_clustering.Rmd</span></span></span>
<span><span class="co">## │   └── <span style="color: #0000BB; font-weight: bold;">single_sample</span></span></span>
<span><span class="co">## │       ├── <span style="color: #00BB00;">01_input_qc.Rmd</span></span></span>
<span><span class="co">## │       ├── <span style="color: #0000BB; font-weight: bold;">01_input_qc_children</span></span></span>
<span><span class="co">## │       │   ├── <span style="color: #00BB00;">cell_filtering_custom.Rmd</span></span></span>
<span><span class="co">## │       │   ├── <span style="color: #00BB00;">cell_filtering_qc.Rmd</span></span></span>
<span><span class="co">## │       │   ├── <span style="color: #00BB00;">empty_droplets.Rmd</span></span></span>
<span><span class="co">## │       │   ├── <span style="color: #00BB00;">gene_filtering_custom.Rmd</span></span></span>
<span><span class="co">## │       │   └── <span style="color: #00BB00;">gene_filtering_qc.Rmd</span></span></span>
<span><span class="co">## │       ├── <span style="color: #00BB00;">02_norm_clustering.Rmd</span></span></span>
<span><span class="co">## │       └── <span style="color: #00BB00;">02_norm_clustering_simple.Rmd</span></span></span>
<span><span class="co">## ├── <span style="color: #00BB00;">_drake_integration.R</span></span></span>
<span><span class="co">## ├── <span style="color: #00BB00;">_drake_single_sample.R</span></span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">config</span></span></span>
<span><span class="co">## │   ├── <span style="color: #0000BB; font-weight: bold;">integration</span></span></span>
<span><span class="co">## │   │   ├── 00_main.default.yaml</span></span>
<span><span class="co">## │   │   ├── 00_main.yaml</span></span>
<span><span class="co">## │   │   ├── 01_integration.default.yaml</span></span>
<span><span class="co">## │   │   ├── 01_integration.yaml</span></span>
<span><span class="co">## │   │   ├── 02_int_clustering.default.yaml</span></span>
<span><span class="co">## │   │   ├── 02_int_clustering.yaml</span></span>
<span><span class="co">## │   │   ├── cluster_markers.default.yaml</span></span>
<span><span class="co">## │   │   ├── cluster_markers.yaml</span></span>
<span><span class="co">## │   │   ├── contrasts.default.yaml</span></span>
<span><span class="co">## │   │   └── contrasts.yaml</span></span>
<span><span class="co">## │   ├── pipeline.default.yaml</span></span>
<span><span class="co">## │   ├── pipeline.yaml</span></span>
<span><span class="co">## │   └── <span style="color: #0000BB; font-weight: bold;">single_sample</span></span></span>
<span><span class="co">## │       ├── 00_main.default.yaml</span></span>
<span><span class="co">## │       ├── 00_main.yaml</span></span>
<span><span class="co">## │       ├── 01_input_qc.default.yaml</span></span>
<span><span class="co">## │       ├── 01_input_qc.yaml</span></span>
<span><span class="co">## │       ├── 02_norm_clustering.default.yaml</span></span>
<span><span class="co">## │       ├── 02_norm_clustering.yaml</span></span>
<span><span class="co">## │       ├── cluster_markers.default.yaml</span></span>
<span><span class="co">## │       ├── cluster_markers.yaml</span></span>
<span><span class="co">## │       ├── contrasts.default.yaml</span></span>
<span><span class="co">## │       └── contrasts.yaml</span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">example_data</span></span></span>
<span><span class="co">## │   ├── <span style="color: #0000BB; font-weight: bold;">pbmc1k</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #BB0000; font-weight: bold;">barcodes.tsv.gz</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #BB0000; font-weight: bold;">features.tsv.gz</span></span></span>
<span><span class="co">## │   │   └── <span style="color: #BB0000; font-weight: bold;">matrix.mtx.gz</span></span></span>
<span><span class="co">## │   └── <span style="color: #0000BB; font-weight: bold;">pbmc3k</span></span></span>
<span><span class="co">## │       ├── barcodes.tsv</span></span>
<span><span class="co">## │       ├── genes.tsv</span></span>
<span><span class="co">## │       └── matrix.mtx</span></span>
<span><span class="co">## ├── file1a2cf4ce21e.Rproj</span></span>
<span><span class="co">## ├── <span style="color: #00BB00;">plan_custom.R</span></span></span>
<span><span class="co">## └── renv.lock</span></span></code></pre>
<p>The most important files and directories are:</p>
<ul>
<li>
<code>Rmd/</code>: RMarkdown files used for reporting of pipeline
results. You can edit them according to your needs, but keep in mind
that they will be overwritten by package-bundled ones whenever you call
<code><a href="../reference/update_project.html">update_project()</a></code>.</li>
<li>
<code>config/</code>: configuration files in YAML format. The
default configuration files (<code>*.default.yaml</code>) are bundled
with the package and used to supply new parameters to local configs
(<code>*.yaml</code>) when such events appear. See
<code><a href="../articles/scdrake_config.html">vignette("scdrake_config")</a></code> for more details.</li>
<li>
<code>_drake_integration.R</code>,
<code>_drake_single_sample.R</code>: entry scripts for
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> when pipeline is executed through
<code><a href="https://docs.ropensci.org/drake/reference/r_make.html" class="external-link">drake::r_make()</a></code> (the most reproducible way to run the
pipeline).</li>
<li>
<code>plan_custom.R</code>: here you can define your own
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> pipeline (plan) to extend <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>.
More on this topic in <code><a href="../articles/scdrake_extend.html">vignette("scdrake_extend")</a></code>.</li>
<li>
<code>example_data/</code>: example datasets. Those are raw
feature-barcode matrices output by <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">cellranger</a>.</li>
<li>
<code>scdrake_single_sample_example.Rproj</code>: RStudio project
file. If you open this file in RStudio, your working directory will be
automatically set to project’s root directory.</li>
</ul>
<hr>
</div>
<div class="section level3">
<h3 id="step-2-modify-the-configuration-files">Step 2: modify the configuration files<a class="anchor" aria-label="anchor" href="#step-2-modify-the-configuration-files"></a>
</h3>
<p>The configuration files for all subpipelines (stages) are stored in
the <code>config/</code> directory. At this time we only need to modify
a single parameter to run the first stage (<code>01_input_qc</code>) of
the single-sample pipeline.</p>
<p>To do so, open <code>config/single_sample/01_input_qc.yaml</code> and
set value of <code>path</code> inside <code>INPUT_DATA</code> to
<code>"example_data/pbmc1k"</code>, such that:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_DATA</span><span class="kw">:</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">type</span><span class="kw">:</span><span class="at"> </span><span class="st">"cellranger"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">path</span><span class="kw">:</span><span class="at"> </span><span class="st">"example_data/pbmc1k"</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">delimiter</span><span class="kw">:</span><span class="at"> </span><span class="st">","</span></span></code></pre></div>
<blockquote>
<p>Important: all paths in configs must be relative to the project root
directory, or absolute (<strong>not recommended</strong>), unless
otherwise stated.</p>
</blockquote>
<hr>
</div>
<div class="section level3">
<h3 id="step-3-run-the-single-sample-pipeline">Step 3: run the single-sample pipeline<a class="anchor" aria-label="anchor" href="#step-3-run-the-single-sample-pipeline"></a>
</h3>
<p>Before we run the pipeline, we only need one last step, and that is
installation of the annotation package for human genome (because our
example data come from human samples):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"org.Hs.eg.db"</span><span class="op">)</span></span></code></pre></div>
<p>By default, only the final target <code>report_input_qc</code> (and
all targets it depends on) of the input and QC stage
(<code>01_input_qc</code>) will be made, that is, a HTML report. This is
specified by <code>DRAKE_TARGETS</code> parameter in
<code>config/pipeline.yaml</code>. You can look <a href="#how-to-know-which-targets-you-can-load-and-what-each-parameter-do">below</a>
for how to identify other targets which can be made.</p>
<p>Now we can run the single-sample pipeline for the PBMC 1k
dataset:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/run_pipeline.html">run_single_sample_r</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The output is saved in <code>output/single_sample</code> directory,
as specified by <code>BASE_OUT_DIR</code> in
<code>config/single_sample/00_main.yaml</code>. You can find the final
report of the <code>01_input_qc</code> stage in
<code>output/single_sample/01_input_qc/01_input_qc.html</code>.</p>
<p>If you want to run the main target for next stage
(<code>02_norm_clustering</code>), just open
<code>config/pipeline.yaml</code> and change <code>DRAKE_TARGETS</code>
to <code>["report_norm_clustering"]</code> (or comment the current
<code>DRAKE_TARGETS</code> line and uncomment the predefined one).</p>
<p>Just note the default config files are meant for the example PBMC
data, and so before you analyse your own data, you should read the
section <a href="#before-you-analyse-your-own-data">below</a> about
important steps before you do so.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="before-you-analyse-your-own-data">Before you analyse your own data<a class="anchor" aria-label="anchor" href="#before-you-analyse-your-own-data"></a>
</h2>
<p>The default config files are designed to work with provided example
data and contain parameters which are used to test/showcase various
features of the pipeline. Hence, some parameters are example
data-specific, and might lead to pipeline failure when kept as-is for
your data. Also, some parameters are critical for the analysis outcomes
and you should be aware of them.</p>
<p>Below you can find important parameters which you should review
before you run the pipeline on your data for the first time (but it
doesn’t mean you shouldn’t use them later). All parameters are
documented in vignettes for their respective stages.</p>
<details><summary>
▶ Show important parameters
</summary><div class="section level4">
<h4 class="unlisted" id="main-config-00_main-yaml">Main config (<code>00_main.yaml</code>)<a class="anchor" aria-label="anchor" href="#main-config-00_main-yaml"></a>
</h4>
<ul>
<li><code>ORGANISM</code></li>
<li><code>ANNOTATION_LIST</code></li>
<li><code>ENSEMBL_SPECIES</code></li>
<li><code>BASE_OUT_DIR</code></li>
</ul>
</div>
<div class="section level4">
<h4 class="unlisted" id="single-sample-input-qc-stage-01_input_qc-yaml">Single-sample / Input QC stage
(<code>01_input_qc.yaml</code>)<a class="anchor" aria-label="anchor" href="#single-sample-input-qc-stage-01_input_qc-yaml"></a>
</h4>
<ul>
<li><code>INPUT_DATA</code></li>
<li><code>EMPTY_DROPLETS_ENABLED</code></li>
<li><code>SAVE_DATASET_SENSITIVE_FILTERING</code></li>
</ul>
</div>
<div class="section level4">
<h4 class="unlisted" id="single-sample-normalization-and-clustering-stage-02_norm_clustering-yaml">Single-sample / Normalization and clustering stage
(<code>02_norm_clustering.yaml</code>)<a class="anchor" aria-label="anchor" href="#single-sample-normalization-and-clustering-stage-02_norm_clustering-yaml"></a>
</h4>
<ul>
<li><code>NORMALIZATION_TYPE</code></li>
<li><code>HVG_METRIC</code></li>
<li><code>PCA_SELECTION_METHOD</code></li>
<li>
<code>PCA_FORCED_PCS</code> (if <code>PCA_SELECTION_METHOD</code> is
<code>"forced"</code>)</li>
<li><code>KMEANS_K</code></li>
<li><code>SC3_K</code></li>
<li><code>SC3_N_CORES</code></li>
</ul>
</div>
<div class="section level4">
<h4 class="unlisted" id="integration-integration-01_integration-yaml">Integration / Integration
(<code>01_integration.yaml</code>)<a class="anchor" aria-label="anchor" href="#integration-integration-01_integration-yaml"></a>
</h4>
<ul>
<li><code>INTEGRATION_SOURCES</code></li>
</ul>
</div>
<div class="section level4">
<h4 class="unlisted" id="integration-post-integration-clustering-02_int_clustering-yaml">Integration / Post-integration clustering
(<code>02_int_clustering.yaml</code>)<a class="anchor" aria-label="anchor" href="#integration-post-integration-clustering-02_int_clustering-yaml"></a>
</h4>
<ul>
<li><code>KMEANS_K</code></li>
<li><code>SC3_K</code></li>
<li><code>SC3_N_CORES</code></li>
</ul>
</div>
<div class="section level4">
<h4 class="unlisted" id="common-cluster-markers-and-contrasts-stages-cluster_markers-yaml-contrasts-yaml">Common / Cluster markers and contrasts stages
(<code>cluster_markers.yaml</code>, <code>contrasts.yaml</code>)<a class="anchor" aria-label="anchor" href="#common-cluster-markers-and-contrasts-stages-cluster_markers-yaml-contrasts-yaml"></a>
</h4>
<ul>
<li><code>CLUSTER_MARKERS_SOURCES</code></li>
<li><code>CONTRASTS_SOURCES</code></li>
</ul>
<hr>
</div>
</details>
</div>
<div class="section level2">
<h2 id="retrieving-intermediate-results-targets">Retrieving intermediate results (targets)<a class="anchor" aria-label="anchor" href="#retrieving-intermediate-results-targets"></a>
</h2>
<p>In <a href="https://docs.ropensci.org/drake" class="external-link">drake</a>’s terminology, a pipeline is called
<em>plan</em>, and is composed of <em>targets</em>. When a target is
finished, its value (object) is saved to cache (the directory
<code>.drake</code> by default). The cache has two main purposes:</p>
<ul>
<li>If a target is not changed, its value is loaded from the cache.
<em>Change</em> involves e.g. target’s definition (code) or change in
upstream targets on which the target depends. This way
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> is able to skip computation of finished targets and
greatly enhance the running time. More details <a href="https://books.ropensci.org/drake/triggers.html" class="external-link">here</a>.</li>
<li>Users also have access to the cache, and so you can load any
finished target into your R session.</li>
</ul>
<p>Users can access the cache via two <a href="https://docs.ropensci.org/drake" class="external-link">drake</a>’s
functions:</p>
<ul>
<li>
<code><a href="https://docs.ropensci.org/drake/reference/readd.html" class="external-link">drake::loadd()</a></code> loads target’s value to the current
session as a variable of the same name.</li>
<li>
<code><a href="https://docs.ropensci.org/drake/reference/readd.html" class="external-link">drake::readd()</a></code> returns target’s value (so it can be
assigned to variable).</li>
</ul>
<p>Let’s try it and load the filtered <code>SingleCellExperiment</code>
object:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">drake</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/drake/reference/readd.html" class="external-link">loadd</a></span><span class="op">(</span><span class="va">sce_final_input_qc</span><span class="op">)</span></span></code></pre></div>
<p>Value of the target <code>sce_final_input_qc</code> was loaded as a
variable of the same name to your current R session (or more precisely,
to the global environment).</p>
<p>Similarly, we can load this target to a variable of our choice:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">drake</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/drake/reference/readd.html" class="external-link">readd</a></span><span class="op">(</span><span class="va">sce_final_input_qc</span><span class="op">)</span></span></code></pre></div>
<p>And work further with the loaded object, e.g.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"NOC2L"</span>, exprs_values <span class="op">=</span> <span class="st">"counts"</span>, swap_rownames <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="how-to-know-which-targets-you-can-load-and-what-each-parameter-do">How to know which targets you can load and what each parameter
do?<a class="anchor" aria-label="anchor" href="#how-to-know-which-targets-you-can-load-and-what-each-parameter-do"></a>
</h2>
<p>Each pipeline stage has its own vignette describing its configuration
parameters and also the most important targets. At <a href="https://bioinfocz.github.io/scdrake" class="uri">https://bioinfocz.github.io/scdrake</a> simply navigate to
the top dropdown menu Articles and choose a stage you are interested in.
Also, in <code>config/pipeline.yaml</code> there are commented the
summary targets for each stage - HTML reports.</p>
<p>For a more schematic overview of pipelines and stages see
<code><a href="../articles/pipeline_overview.html">vignette("pipeline_overview")</a></code> (also includes <a href="https://github.com/bioinfocz/scdrake/blob/main/diagrams/README.md" class="external-link">diagrams</a>).</p>
<p>Advanced users might be interested in looking into source code of
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>’s <a href="https://github.com/bioinfocz/scdrake/tree/main/R" class="external-link">plans</a> (files
named <code>plans_*.R</code>).</p>
<hr>
</div>
<div class="section level2">
<h2 id="modifying-parameters-and-rerunning-the-pipeline">Modifying parameters and rerunning the pipeline<a class="anchor" aria-label="anchor" href="#modifying-parameters-and-rerunning-the-pipeline"></a>
</h2>
<p>Optionally, to further demonstrate the <a href="https://docs.ropensci.org/drake" class="external-link">drake</a>’s ability
to skip finished targets, let’s assume you decided to be more benevolent
and used less strict values for custom cell filtering in
<code>config/single_sample/01_input_qc.yaml</code>:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MIN_UMI_CF</span><span class="kw">:</span><span class="at"> </span><span class="dv">500</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">MAX_UMI_CF</span><span class="kw">:</span><span class="at"> </span><span class="dv">70000</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">MIN_FEATURES</span><span class="kw">:</span><span class="at"> </span><span class="dv">500</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">MAX_MITO_RATIO</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.3</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_QC_BASE_OUT_DIR</span><span class="kw">:</span><span class="at"> </span><span class="st">"01_input_qc_benevolent"</span></span></code></pre></div>
<p>After <code><a href="../reference/run_pipeline.html">run_single_sample_r()</a></code> you can see the most
time-consuming targets <code>sce_raw</code> and
<code>empty_droplets</code> were skipped. Also, we simply output the
HTML report for new parameters into a different directory, and so we can
easily compare it with the previous report.</p>
<hr>
</div>
<div class="section level2">
<h2 id="running-the-pipeline-in-parallel-mode">Running the pipeline in parallel mode<a class="anchor" aria-label="anchor" href="#running-the-pipeline-in-parallel-mode"></a>
</h2>
<p>Because <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> knows the inner relationships between
targets in your plan, it also knows which targets are independent of
each other, and thus, can be run concurrently. This is called
<em>implicit parallelism</em>, and to fully utilize this important
feature, you just need to modify <code>config/pipeline.yaml</code> by
setting <code>DRAKE_PARALLELISM</code> to either:</p>
<ul>
<li>
<code>"future"</code>: uses the <a href="https://future.futureverse.org" class="external-link">future</a> as the
backend. This backend should work by simply installing the
<a href="https://future.futureverse.org" class="external-link">future</a> package.
<ul>
<li>Install by
<code>BiocManager::install(c("future", "future.callr"))</code>
</li>
</ul>
</li>
<li>
<code>"clustermq"</code>: uses the <a href="https://mschubert.github.io/clustermq/" class="external-link">clustermq</a> as the
backend. This is faster than <code>"future"</code>, but besides the
<a href="https://mschubert.github.io/clustermq/" class="external-link">clustermq</a> package it also requires the <a href="https://zeromq.org/" class="external-link">ZeroMQ</a> library to be <a href="https://zeromq.org/download/" class="external-link">installed</a> on your system.
<ul>
<li>You need to install a specific <a href="https://mschubert.github.io/clustermq/" class="external-link">clustermq</a> version by
<code>remotes::install_version("clustermq", version = "0.8.8")</code>
(you might need <code>BiocManager::install("remotes")</code>).</li>
<li>
<a href="https://mschubert.github.io/clustermq/" class="external-link">clustermq</a> also supports HPC cluster schedulers, see
<a href="https://books.ropensci.org/drake/hpc.html#the-template-argument-for-persistent-workers" class="external-link">here</a>
for more details.</li>
</ul>
</li>
</ul>
<p>For a general overview of <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> parallelism see <a href="https://books.ropensci.org/drake/hpc.html" class="external-link uri">https://books.ropensci.org/drake/hpc.html</a></p>
<hr>
</div>
<div class="section level2">
<h2 id="updating-the-project-files">Updating the project files<a class="anchor" aria-label="anchor" href="#updating-the-project-files"></a>
</h2>
<p>When a new version of <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> is released, you can
update your project with</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/update_project.html">update_project</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This will <strong>overwrite</strong> project files by the
package-bundled ones:</p>
<ul>
<li>RMarkdown documents in <code>Rmd/</code>.</li>
<li>Initial scripts for <code><a href="https://docs.ropensci.org/drake/reference/r_make.html" class="external-link">drake::r_make()</a></code>:
<code>_drake_single_sample.R</code> and
<code>_drake_integration.R</code>.</li>
<li>Default YAML configs in <code>config/</code>
(<code>*.default.yaml</code>).</li>
</ul>
<p>By default, you will be asked if you want to continue, as you might
lose your local modifications.</p>
<hr>
</div>
<div class="section level2">
<h2 id="going-further">Going further<a class="anchor" aria-label="anchor" href="#going-further"></a>
</h2>
<p>Now you should know the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> basics:</p>
<ul>
<li>How to initialize a new project and which files does it
contain.</li>
<li>How to modify pipeline parameters stored in YAML config files.</li>
<li>How to run the pipeline.</li>
</ul>
<p>For the integration pipeline you might be interested in its guide in
<code><a href="../articles/scdrake_integration.html">vignette("scdrake_integration")</a></code>.</p>
<p>For the full insight into <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> you can also read
the following vignettes:</p>
<ul>
<li>Guides:
<ul>
<li>01 Quick start (single-sample pipeline):
<code><a href="../articles/scdrake.html">vignette("scdrake")</a></code>
</li>
<li>02 Integration pipeline guide:
<code><a href="../articles/scdrake_integration.html">vignette("scdrake_integration")</a></code>
</li>
<li>Extending the pipeline: <code><a href="../articles/scdrake_extend.html">vignette("scdrake_extend")</a></code>
</li>
<li>
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> basics: <code><a href="../articles/drake_basics.html">vignette("drake_basics")</a></code>
<ul>
<li>Or the official <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> book: <a href="https://books.ropensci.org/drake/" class="external-link uri">https://books.ropensci.org/drake/</a>
</li>
</ul>
</li>
</ul>
</li>
<li>General information:
<ul>
<li>Pipeline overview: <code><a href="../articles/pipeline_overview.html">vignette("pipeline_overview")</a></code>
</li>
<li>FAQ &amp; Howtos: <code><a href="../articles/scdrake_faq.html">vignette("scdrake_faq")</a></code>
</li>
<li>Config files: <code><a href="../articles/scdrake_config.html">vignette("scdrake_config")</a></code>
</li>
<li>Cluster markers: <code><a href="../articles/cluster_markers.html">vignette("cluster_markers")</a></code>
</li>
<li>Environment variables: <code><a href="../articles/scdrake_envvars.html">vignette("scdrake_envvars")</a></code>
</li>
</ul>
</li>
<li>General configs:
<ul>
<li>Pipeline config: <code><a href="../articles/config_pipeline.html">vignette("config_pipeline")</a></code>
</li>
<li>Main config: <code><a href="../articles/config_main.html">vignette("config_main")</a></code>
</li>
</ul>
</li>
<li>Targets and config parameters for each stage:
<ul>
<li>Single-sample pipeline:
<ul>
<li>Reading in data, filtering, quality control (stage
<code>01_input_qc</code>): <code><a href="../articles/stage_input_qc.html">vignette("stage_input_qc")</a></code>
</li>
<li>Normalization, HVG selection, dimensionality reduction, clustering,
cell type annotation (stage <code>02_norm_clustering</code>):
<code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code>
</li>
</ul>
</li>
<li>Integration pipeline:
<ul>
<li>Reading in data and integration (stage <code>01_integration</code>):
<code><a href="../articles/stage_integration.html">vignette("stage_integration")</a></code>
</li>
<li>Post-integration clustering (stage <code>02_int_clustering</code>):
<code><a href="../articles/stage_int_clustering.html">vignette("stage_int_clustering")</a></code>
</li>
</ul>
</li>
<li>Common:
<ul>
<li>Cluster markers stage:
<code><a href="../articles/stage_cluster_markers.html">vignette("stage_cluster_markers")</a></code>
</li>
<li>Contrasts stage: <code><a href="../articles/stage_contrasts.html">vignette("stage_contrasts")</a></code>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jiri Novotny, Jan Kubovciak.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '937cedcd467c204b19fc155c5f35019b',
    indexName: 'scdrake',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
