<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/3/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Get started â€¢ scdrake</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/3/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/3/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="../darkswitch.js"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Get started">
<meta property="og:description" content="scdrake">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scdrake</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/scdrake.html">Get started</a>
</li>
<li>
  <a href="../articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="divider">
    <li class="dropdown-header">General information</li>
    <li>
      <a href="../articles/pipeline_overview.html">Pipeline overview</a>
    </li>
    <li>
      <a href="../articles/scdrake_run.html">Running the pipeline, environment variables</a>
    </li>
    <li>
      <a href="../articles/cluster_markers.html">Cluster markers</a>
    </li>
    <li>
      <a href="../articles/drake_basics.html">drake basics</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Configs</li>
    <li>
      <a href="../articles/scdrake_config.html">scdrake configs: a general overview</a>
    </li>
    <li>
      <a href="../articles/config_pipeline.html">Pipeline config</a>
    </li>
    <li>
      <a href="../articles/config_main.html">Main config</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Single-sample pipeline stages</li>
    <li>
      <a href="../articles/stage_input_qc.html">Input and quality control stage</a>
    </li>
    <li>
      <a href="../articles/stage_norm_clustering.html">Normalization and clustering stage</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Integration pipeline stages</li>
    <li>
      <a href="../articles/stage_integration.html">Integration stage</a>
    </li>
    <li>
      <a href="../articles/stage_int_clustering.html">Post-integration clustering stage</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Common stages</li>
    <li>
      <a href="../articles/stage_cluster_markers.html">Cluster markers stage</a>
    </li>
    <li>
      <a href="../articles/stage_contrasts.html">Contrasts stage</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
          <a href="#" id="css-toggle-btn">
            <span class="fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        <li>
  <a href="https://github.com/bioinfocz/scdrake">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
        
        


      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="scdrake_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Get started</h1>
                        <h4 class="author">Jiri Novotny</h4>
            <address class="author_afil">
      Institute of Molecular Genetics of the Czech Academy of Sciences<br><a class="author_email" href="mailto:#"></a><a href="mailto:jiri.novotny@img.cas.cz" class="email">jiri.novotny@img.cas.cz</a>
      </address>
                              <h4 class="author">Jan Kubovciak</h4>
            <address class="author_afil">
      Institute of Molecular Genetics of the Czech Academy of Sciences<br><a class="author_email" href="mailto:#"></a><a href="mailto:jan.kubovciak@img.cas.cz" class="email">jan.kubovciak@img.cas.cz</a>
      </address>
                  
            <h4 class="date">2021-11-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bioinfocz/scdrake/blob/mastervignettes/scdrake.Rmd"><code>vignettes/scdrake.Rmd</code></a></small>
      <div class="hidden name"><code>scdrake.Rmd</code></div>

    </div>

    
    
<script>
$(document).ready(function(){
  $('[data-toggle="tooltip"]').tooltip(); 
});
</script><hr>
<div id="project-initialization" class="section level1">
<h1 class="hasAnchor">
<a href="#project-initialization" class="anchor"></a>Project initialization</h1>
<p><a href="https://github.com/bioinfocz/scdrake">scdrake</a> is a project-based package, so the first step is to initialize a new project. However, this RMarkdown document is not possible to use as usual in RStudio, i.e.Â to run the code chunks (<a data-toggle="tooltip" title="Because during the project initialization we will change the current working directory and set new active RStudio project.">why?</a>). Instead, open the rendered version of this document and copy-paste the code to RStudio console - there is not much code here to do so ðŸ™‚</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bioinfocz/scdrake">scdrake</a></span><span class="op">)</span>
<span class="fu"><a href="../reference/init_project.html">init_project</a></span><span class="op">(</span><span class="st">"~/scdrake_example_project"</span><span class="op">)</span></code></pre></div>
<p>This will:</p>
<ul>
<li>Create a new project directory. Existing directory is automatically detected and you will be asked if it can be overwritten (unless <code>ask = FALSE</code>).</li>
<li>Switch your current working directory to <code>~/scdrake_example_project</code> (can be disabled).</li>
<li>Copy all files, which are bundled with the <a href="https://github.com/bioinfocz/scdrake">scdrake</a> package: YAML configs, Rmd files, R scripts etc.</li>
<li>Download (if needed) the <a href="https://github.com/mikefarah/yq">yq tool</a> used to manipulate YAML files.</li>
<li>Create a new RStudio project (<code>.RProj</code>) file and set it as active project (can be disabled).</li>
</ul>
<blockquote>
<p>Whenever you will be running the <a href="https://github.com/bioinfocz/scdrake">scdrake</a> pipeline, make sure your working directory is set to the projectâ€™s root.</p>
</blockquote>
<p>Letâ€™s inspect files in the project directory:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="http://fs.r-lib.org/reference/dir_tree.html">dir_tree</a></span><span class="op">(</span>all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## /tmp/RtmpGrcBxX/file10501603abbb
## â”œâ”€â”€ .here
## â”œâ”€â”€ Rmd
## â”‚   â”œâ”€â”€ common
## â”‚   â”‚   â”œâ”€â”€ _footer.Rmd
## â”‚   â”‚   â”œâ”€â”€ _header.Rmd
## â”‚   â”‚   â”œâ”€â”€ cluster_markers.Rmd
## â”‚   â”‚   â”œâ”€â”€ cluster_markers_table_template.Rmd
## â”‚   â”‚   â”œâ”€â”€ contrasts.Rmd
## â”‚   â”‚   â”œâ”€â”€ contrasts_table_template.Rmd
## â”‚   â”‚   â””â”€â”€ stylesheet.css
## â”‚   â”œâ”€â”€ integration
## â”‚   â”‚   â”œâ”€â”€ 01_integration.Rmd
## â”‚   â”‚   â””â”€â”€ 02_int_clustering.Rmd
## â”‚   â””â”€â”€ single_sample
## â”‚       â”œâ”€â”€ 01_input_qc.Rmd
## â”‚       â”œâ”€â”€ 02_norm_clustering.Rmd
## â”‚       â””â”€â”€ 02_norm_clustering_simple.Rmd
## â”œâ”€â”€ _drake_integration.R
## â”œâ”€â”€ _drake_single_sample.R
## â”œâ”€â”€ config
## â”‚   â”œâ”€â”€ integration
## â”‚   â”‚   â”œâ”€â”€ 00_main.default.yaml
## â”‚   â”‚   â”œâ”€â”€ 00_main.yaml
## â”‚   â”‚   â”œâ”€â”€ 01_integration.default.yaml
## â”‚   â”‚   â”œâ”€â”€ 01_integration.yaml
## â”‚   â”‚   â”œâ”€â”€ 02_int_clustering.default.yaml
## â”‚   â”‚   â”œâ”€â”€ 02_int_clustering.yaml
## â”‚   â”‚   â”œâ”€â”€ cluster_markers.default.yaml
## â”‚   â”‚   â”œâ”€â”€ cluster_markers.yaml
## â”‚   â”‚   â”œâ”€â”€ contrasts.default.yaml
## â”‚   â”‚   â””â”€â”€ contrasts.yaml
## â”‚   â”œâ”€â”€ pipeline.default.yaml
## â”‚   â”œâ”€â”€ pipeline.yaml
## â”‚   â””â”€â”€ single_sample
## â”‚       â”œâ”€â”€ 00_main.default.yaml
## â”‚       â”œâ”€â”€ 00_main.yaml
## â”‚       â”œâ”€â”€ 01_input_qc.default.yaml
## â”‚       â”œâ”€â”€ 01_input_qc.yaml
## â”‚       â”œâ”€â”€ 02_norm_clustering.default.yaml
## â”‚       â”œâ”€â”€ 02_norm_clustering.yaml
## â”‚       â”œâ”€â”€ cluster_markers.default.yaml
## â”‚       â”œâ”€â”€ cluster_markers.yaml
## â”‚       â”œâ”€â”€ contrasts.default.yaml
## â”‚       â””â”€â”€ contrasts.yaml
## â”œâ”€â”€ file10501603abbb.Rproj
## â””â”€â”€ renv.lock</code></pre>
<ul>
<li>
<code>.here</code>: used by the <a href="https://here.r-lib.org/">here</a> package to locate and memorize the projectâ€™s root directory. Only used when RStudio project file (<code>.Rproj</code>) is not found.</li>
<li>
<code>Rmd/</code>: RMarkdown files used for reporting of pipeline results. You can edit them according to your needs, but keep in mind that they will be overwritten by package-bundle ones if you call <code><a href="../reference/update_project.html">update_project()</a></code>.</li>
<li>
<code>config/</code>: config files in YAML format. See <code><a href="../articles/scdrake_config.html">vignette("scdrake_config")</a></code> for more details.</li>
<li>
<code>_drake_integration.R</code>, <code>_drake_single_sample.R</code>: entry points for <a href="https://docs.ropensci.org/drake">drake</a> when pipeline is executed through <code><a href="https://docs.ropensci.org/drake/reference/r_make.html">drake::r_make()</a></code>.</li>
<li>
<code>scdrake_example_project.Rproj</code>: RStudio project file. Created when you call <code><a href="../reference/init_project.html">init_project()</a></code> with <code>use_rstudio = TRUE</code> (default).</li>
</ul>
</div>
<div id="running-the-single-sample-pipeline" class="section level1">
<h1 class="hasAnchor">
<a href="#running-the-single-sample-pipeline" class="anchor"></a>Running the single-sample pipeline</h1>
<p>First we download example datasets from 10x Genomics:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/download_example_data.html">download_pbmc1k</a></span><span class="op">(</span><span class="st">"example_data/pbmc1k"</span><span class="op">)</span></code></pre></div>
<p>The dataset contains data of <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices">unfiltered raw feature-barcode matrix</a>.</p>
<p>We have to modify <code>config/single_sample/01_input_qc.yaml</code> to contain a proper path to the data. Open this file and set value of <code>INPUT_10X_DIR</code> to <code>"example_data/pbmc1k"</code>.</p>
<p>Finally, we can run the single-sample pipeline for the PBMC 1k dataset. By default, only the final target (<code>report_input_qc</code>) of the input and QC stage (<code>01_input_qc</code>) will be made, that is, a HTML report. This is specified by <code>DRAKE_TARGETS</code> parameter in <code>config/pipeline.yaml</code>.</p>
<blockquote>
<p>In case you are not using RStudio, you must have directory with <code>pandoc</code> binary in your <code>PATH</code> or <code>RSTUDIO_PANDOC</code> environment variable, or you have to modify the <code>RSTUDIO_PANDOC</code> parameter in <code>config/pipeline.yaml</code> and supply the directory in which <code>pandoc</code> binary is located.</p>
</blockquote>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_single_sample.html">run_single_sample</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>The output is saved in <code>output/single_sample/pbmc1k</code>, as specified by <code>BASE_OUT_DIR</code> in <code>config/single_sample/00_main.yaml</code>. For <code>01_input_qc</code> stage, you can find its final report in <code>output/single_sample/pbmc1k/01_input_qc/01_input_qc.html</code>.</p>
</div>
<div id="retrieving-pipeline-results" class="section level1">
<h1 class="hasAnchor">
<a href="#retrieving-pipeline-results" class="anchor"></a>Retrieving pipeline results</h1>
<p>In <a href="https://docs.ropensci.org/drake">drake</a>â€™s terminology, a pipeline is called <em>plan</em>, and is composed of <em>targets</em>. When a target is finished, its value (object) is saved to cache (directory <code>.drake</code> by default), and you can anytime load it into your R session. This is done via two <a href="https://docs.ropensci.org/drake">drake</a>â€™s functions:</p>
<ul>
<li>
<code><a href="https://docs.ropensci.org/drake/reference/readd.html">drake::loadd()</a></code> loads targetâ€™s value to the global environment.</li>
<li>
<code><a href="https://docs.ropensci.org/drake/reference/readd.html">drake::readd()</a></code> returns targetâ€™s value (so it can be assigned to variable).</li>
</ul>
<p>Letâ€™s try it and load the filtered <code>SingleCellExperiment</code> object:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">drake</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/drake/reference/readd.html">loadd</a></span><span class="op">(</span><span class="va">sce_final_input_qc</span><span class="op">)</span></code></pre></div>
<p>Value of the target <code>sce_final_input_qc</code> was loaded as a variable of the same name to your current R session (or more precisely, to the global environment). Similarly, we can load this target to variable of our choice:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">drake</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/drake/reference/readd.html">readd</a></span><span class="op">(</span><span class="va">sce_final_input_qc</span><span class="op">)</span></code></pre></div>
</div>
<div id="running-the-integration-pipeline" class="section level1">
<h1 class="hasAnchor">
<a href="#running-the-integration-pipeline" class="anchor"></a>Running the integration pipeline</h1>
<p>The integration pipeline takes as input the <code>sce_final_norm_clustering</code> targets from multiple single-sample pipelines. This target is located in stage <code>02_norm_clustering</code>, but in our example with PBMC 1k dataset, we only run the target <code>report_input_qc</code>, which is basically the final target of the <code>01_input_qc</code> stage. So we need to run the pipeline again, specifying this target: open <code>config/pipeline.yaml</code> and set <code>DRAKE_TARGETS</code> to <code>["sce_final_norm_clustering"]</code>.</p>
<p>Also, to speed things up, consider setting <code>SC3_DRY</code> in <code>config/single_sample/02_norm_clustering.yaml</code> to <code>TRUE</code>. SC3 is a computationally demanding clustering algorithm, and by setting this parameter the SC3 computation will be skipped, and random clusters will be assigned to cells.</p>
<p>Now we can run the pipeline again:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_single_sample.html">run_single_sample</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<blockquote>
<p>Normally, you would put each sample to its own <a href="https://github.com/bioinfocz/scdrake">scdrake</a> project, and also use an another one for the integration pipeline. But to make the examples faster, we will process all of them in the current project directory.</p>
</blockquote>
<p>As the integration pipeline requires two and more samples, we need to download an another one - the PBMC 3k dataset:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/download_example_data.html">download_pbmc3k</a></span><span class="op">(</span><span class="st">"example_data/pbmc3k"</span><span class="op">)</span></code></pre></div>
<p>We need to modify our configs:</p>
<ul>
<li>
<code>config/single_sample/01_input_qc.yaml</code>: set <code>INPUT_10X_DIR</code> to <code>"example_data/pbmc3k"</code>
</li>
<li>
<code>config/pipeline.yaml</code>: set <code>DRAKE_CACHE_DIR</code> to <code>".drake_pbmc3k"</code>. We need to use an another cache directory as we are running the pipeline for both samples in the same project directory.</li>
</ul>
<p>The config modifications for the second sample are ready, so letâ€™s run the pipeline:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_single_sample.html">run_single_sample</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Finally, we modify configs for the integration pipeline:</p>
<ul>
<li>
<code>config/integration/01_integration.yaml</code>: in the <code>INTEGRATION_SOURCES/pbmc1k</code> parameter, set <code>cache_path</code> to <code>".drake"</code>.</li>
<li>
<code>config/pipeline.yaml</code>:
<ul>
<li>Set <code>DRAKE_TARGETS</code> to <code>["report_integration"]</code>. To save time, we only run the final target of the <code>01_integration</code> stage.</li>
<li>Set <code>DRAKE_CACHE_DIR</code> to <code>".drake_integration"</code>.</li>
</ul>
</li>
</ul>
<p>And now we can run the integration pipeline:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/run_single_sample.html">run_integration</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>The output is saved in <code>output/integration</code>, as specified by <code>BASE_OUT_DIR</code> in <code>config/integration/00_main.yaml</code>. For <code>01_integration</code> stage, you can find its final report in <code>output/integration/01_integration/01_integration.html</code>.</p>
<p>You can try to load the target <code>sce_int_dimred_df</code> containing integrated <code>SingleCellExperiment</code> objects with computed reduced dimensions. We have to use other than default path to <code>drake</code> cache:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">drake</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/drake/reference/readd.html">loadd</a></span><span class="op">(</span><span class="va">sce_int_dimred_df</span>, path <span class="op">=</span> <span class="st">".drake_integration"</span><span class="op">)</span></code></pre></div>
<p>Because we are using the single project directory for both samples and their integration, the config modifications for the integration pipeline can look a bit complicated. However, normally, you would:</p>
<ul>
<li>Initialize a new <a href="https://github.com/bioinfocz/scdrake">scdrake</a> project for each of your samples.</li>
<li>In each project, run pipeline for the <code>sce_final_norm_clustering</code> target.</li>
<li>Initialize a new <a href="https://github.com/bioinfocz/scdrake">scdrake</a> project for integration, modify the <code>INTEGRATION_SOURCES</code> parameter in <code>config/integration/01_integration.yaml</code>, and run the integration pipeline.</li>
</ul>
</div>
<div id="running-the-pipeline-reproducibly" class="section level1">
<h1 class="hasAnchor">
<a href="#running-the-pipeline-reproducibly" class="anchor"></a>Running the pipeline reproducibly</h1>
<p>When we called <code><a href="../reference/run_single_sample.html">run_single_sample()</a></code> or <code><a href="../reference/run_single_sample.html">run_integration()</a></code>, the pipeline was executed in your current R session. That is considered a bad practice in the world of reproducibility, as it may bring an unexpected behaviour. The proper way is to run pipeline in a fresh new R session. In the <a href="https://docs.ropensci.org/drake">drake</a>, an independent script and <code><a href="https://docs.ropensci.org/drake/reference/r_make.html">drake::r_make()</a></code> function are used for this purpose:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">drake</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/drake/reference/r_make.html">r_make</a></span><span class="op">(</span><span class="st">"_drake_single_sample.R"</span><span class="op">)</span>
<span class="fu">drake</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/drake/reference/r_make.html">r_make</a></span><span class="op">(</span><span class="st">"_drake_integration.R"</span><span class="op">)</span></code></pre></div>
<p>This function will source the input R script in which <code><a href="https://docs.ropensci.org/drake/reference/drake_config.html">drake::drake_config()</a></code> is the last returned value, and execute the pipeline (plan) in a new R session. More details can be found in the <a href="https://docs.ropensci.org/drake">drake</a> book <a href="https://books.ropensci.org/drake/projects.html#usage">here</a>.</p>
<p>Of course, you can modify <code>_drake_single_sample.R</code> and <code>_drake_integration.R</code> to your needs, but, by default, loading the config files from proper paths in these scripts is done via environment variables. You can read more in <code><a href="../articles/scdrake_run.html">vignette("scdrake_run")</a></code>.</p>
</div>
<div id="updating-the-project-files" class="section level1">
<h1 class="hasAnchor">
<a href="#updating-the-project-files" class="anchor"></a>Updating the project files</h1>
<p>When a new version of <a href="https://github.com/bioinfocz/scdrake">scdrake</a> is released, you can update your project with</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/update_project.html">update_project</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>This will <strong>overwrite</strong> project files by the package-bundled ones:</p>
<ul>
<li>RMarkdown documents in <code>Rmd/</code>.</li>
<li>Initial scripts for <code><a href="https://docs.ropensci.org/drake/reference/r_make.html">drake::r_make()</a></code>: <code>_drake_single_sample.R</code> and <code>_drake_integration.R</code>.</li>
<li>Default YAML configs.</li>
</ul>
<p>By default, you will be asked if you want to continue, as you might lose your local modifications.</p>
</div>
<div id="going-further" class="section level1">
<h1 class="hasAnchor">
<a href="#going-further" class="anchor"></a>Going further</h1>
<p>Now you know the <a href="https://github.com/bioinfocz/scdrake">scdrake</a> basics:</p>
<ul>
<li>How to initialize a new project and which files does it contain.</li>
<li>How to modify pipeline parameters stored in YAML config files.</li>
<li>How to run the pipeline, including the reproducible way.</li>
</ul>
<p>But for the full insight into <a href="https://github.com/bioinfocz/scdrake">scdrake</a> you should also read the following vignettes:</p>
<ul>
<li>Pipeline overview: <code><a href="../articles/pipeline_overview.html">vignette("pipeline_overview")</a></code>
<ul>
<li>Cluster markers: <code><a href="../articles/cluster_markers.html">vignette("cluster_markers")</a></code>
</li>
</ul>
</li>
<li>Running the pipeline, environment variables: <code><a href="../articles/scdrake_run.html">vignette("scdrake_run")</a></code>
</li>
<li>Config files (basic concept): <code><a href="../articles/scdrake_config.html">vignette("scdrake_config")</a></code>
</li>
<li>Targets and config parameters for each stage:
<ul>
<li>Common:
<ul>
<li>Pipeline config: <code><a href="../articles/config_pipeline.html">vignette("config_pipeline")</a></code>
</li>
<li>Main config: <code><a href="../articles/config_main.html">vignette("config_main")</a></code>
</li>
<li>Cluster markers stage: <code><a href="../articles/stage_cluster_markers.html">vignette("stage_cluster_markers")</a></code>
</li>
<li>Contrasts stage: <code><a href="../articles/stage_contrasts.html">vignette("stage_contrasts")</a></code>
</li>
</ul>
</li>
<li>Single-sample pipeline:
<ul>
<li>Reading in data, filtering, quality control (<code>01_input_qc</code>): <code><a href="../articles/stage_input_qc.html">vignette("stage_input_qc")</a></code>
</li>
<li>Normalization, HVG selection, clustering (<code>02_norm_clustering</code>): <code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code>
</li>
</ul>
</li>
<li>Integration pipeline:
<ul>
<li>Reading in data and integration (<code>01_integration</code>): <code><a href="../articles/stage_integration.html">vignette("stage_integration")</a></code>
</li>
<li>Clustering (<code>02_int_clustering</code>): <code><a href="../articles/stage_int_clustering.html">vignette("stage_int_clustering")</a></code>
</li>
</ul>
</li>
</ul>
</li>
<li>
<a href="https://docs.ropensci.org/drake">drake</a> basics: <code><a href="../articles/drake_basics.html">vignette("drake_basics")</a></code>
<ul>
<li>Or the official <a href="https://docs.ropensci.org/drake">drake</a> book: <a href="https://books.ropensci.org/drake/" class="uri">https://books.ropensci.org/drake/</a>
</li>
</ul>
</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jiri Novotny, Jan Kubovciak.</p>
</div>

<div class="pkgdown">
  <p>Made with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1, using <a href="https://preferably.amirmasoudabdol.name/?source=footer">preferably</a> template.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
