<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/5/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Spatial extension â€¢ scdrake</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Spatial extension">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/5/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/5/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="../darkswitch.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scdrake</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scdrake.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/scdrake_integration.html">Integration pipeline guide</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/pipeline_overview.html">Pipeline overview</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/scdrake_spatial.html">Spatial extension</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/scdrake_docker.html">Using the Docker image</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_integration.html">02 Integration pipeline guide</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_advanced.html">Advanced topics</a></li>
    <li><a class="dropdown-item" href="../articles/interaction_analysis.html">Interaction analysis</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_extend.html">Extending the pipeline</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_development.html">Developing scdrake</a></li>
    <li><a class="dropdown-item" href="../articles/drake_basics.html">drake basics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>General information</h6></li>
    <li><a class="dropdown-item" href="../articles/pipeline_overview.html">Pipeline overview</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_cli.html">Command line interface</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_config.html">scdrake config internals</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_envvars.html">Environment variables</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_spatial.html">Spatial extension</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>General configs</h6></li>
    <li><a class="dropdown-item" href="../articles/config_pipeline.html">Pipeline config</a></li>
    <li><a class="dropdown-item" href="../articles/config_main.html">Main config</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Single-sample pipeline stages</h6></li>
    <li><a class="dropdown-item" href="../articles/stage_input_qc.html">Input and quality control stage (`01_input_qc`)</a></li>
    <li><a class="dropdown-item" href="../articles/stage_norm_clustering.html">Normalization and clustering stage (`02_norm_clustering`)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Integration pipeline stages</h6></li>
    <li><a class="dropdown-item" href="../articles/stage_integration.html">Integration stage (`01_integration`)</a></li>
    <li><a class="dropdown-item" href="../articles/stage_int_clustering.html">Post-integration clustering stage (`02_int_clustering`)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Common stages</h6></li>
    <li><a class="dropdown-item" href="../articles/stage_cluster_markers.html">Cluster markers stage (`cluster_markers`)</a></li>
    <li><a class="dropdown-item" href="../articles/stage_contrasts.html">Contrasts stage (`contrasts`)</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bioinfocz/scdrake" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>


        <li>
          <a class="external-link nav-link" id="css-toggle-btn" aria-label="github">
            <span class="fas fa fas fa-adjust fa-lg"></span>
          </a>
        </li>



      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Spatial extension</h1>
            
            <h4 data-toc-skip class="date"><sup>Document generated:
2025-08-26 19:26:19 UTC+0000</sup></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bioinfocz/scdrake/blob/master/vignettes/scdrake_spatial.Rmd" class="external-link"><code>vignettes/scdrake_spatial.Rmd</code></a></small>
      <div class="d-none name"><code>scdrake_spatial.Rmd</code></div>
    </div>

    
    
<hr>
<p>The <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> package now provides <strong>spatial
extensions</strong> for both the first stage (<code>01_input_qc</code>)
and the second stage (<code>02_norm_clustering</code>) of the
single-sample pipeline.<br>
These spatial workflows are currently optimized for
<strong>Visium</strong> data and other <strong>spot-based spatial
transcriptomics technologies</strong>.</p>
<p>The aim is to offer functionality comparable to
<strong>Seurat</strong>, <strong>Giotto</strong> (R), and
<strong>scanpy</strong> (Python), while following the principles of the
<a href="https://lmweber.org/OSTA/" class="external-link"><em>Orchestrating Spatial
Transcriptomics Analysis with Bioconductor</em></a> guide.</p>
<p>For advanced analysis of spatial datasetsâ€”particularly for cellâ€“cell
interaction and communication region identification we recommend: - <a href="https://github.com/SiYangming/CellChat2" class="external-link">CellChat2</a> (R) -
<strong><a href="https://liana-py.readthedocs.io/en/latest/notebooks/bivariate.html" class="external-link">LIANA+</a></strong>
(Python)</p>
<p>The latter is accompanied by the vignette
<code><a href="../articles/interaction_analysis.html">vignette("interaction_analysis")</a></code>, which provides a
step-by-step example of exporting count matrices and metadata from
<code>SpatialExperiment</code> objects in R, and then importing them
into Python to create an <code>AnnData</code> object.</p>
<p>This vignette is a <strong>supplement</strong> to: -
<code><a href="../articles/stage_input_qc.html">vignette("stage_input_qc")</a></code> -
<code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code></p>
<hr>
<hr>
<div class="section level2">
<h2 id="spatial-extention-functions">Spatial extention functions<a class="anchor" aria-label="anchor" href="#spatial-extention-functions"></a>
</h2>
<hr>
<div class="section level3">
<h3 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h3>
<p>If the <code>spaceranger</code> option is selected in the the first
stage (<code>01_input_qc</code>) for reading input data, the data are
loaded as <code>SpatialExperiment.</code></p>
<hr>
</div>
<div class="section level3">
<h3 id="spatial-visualization">Spatial visualization<a class="anchor" aria-label="anchor" href="#spatial-visualization"></a>
</h3>
<p>For (<code>01_input_qc</code>) and (<code>02_norm_clustering</code>)
of the single-sample pipeline we now offer visualization of tissue, as
pseudo tissue spot visualization from <code>ggspavis</code>. For
(<code>01_input_qc</code>) stage, quality control metrics can be
overlaid directly onto the H&amp;E-stained tissue image, enabling visual
inspection of tissue quality and spatial artifacts. In other section
only spots are visualized for greater clarity. For downstream sections,
visualization defaults to pseudo-tissue spot plots (using
<a href="https://github.com/lmweber/ggspavis" class="external-link">ggspavis</a>), where spots are displayed without the
underlying image for greater clarity.</p>
<hr>
</div>
<div class="section level3">
<h3 id="selection-of-spatially-variable-genes">Selection of spatially variable genes<a class="anchor" aria-label="anchor" href="#selection-of-spatially-variable-genes"></a>
</h3>
<p>If the SVG option is enabled in stage <code>02_norm_clustering</code>
(see <code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code>), spatially variable
genes are identified alongside highly variable genes (HVGs). That is
done using nnSVG::nvSVG, with the selection method MoransI. A
straightforward union of SVGs and HVGs is taking to further processing,
see <a href="https://www.biorxiv.org/content/10.1101/2021.08.27.457741v1" class="external-link uri">https://www.biorxiv.org/content/10.1101/2021.08.27.457741v1</a>
for more details.</p>
<hr>
</div>
<div class="section level3">
<h3 id="marker-based-annotation">Marker-based annotation<a class="anchor" aria-label="anchor" href="#marker-based-annotation"></a>
</h3>
<p>Marker-based annotation was implemented for both single-cell and
spatial datasets. In summary, expression profiles and statistical
metrics are computed for each cell/spot, the result is visualized using
a heatmap and dimension reduction plot. For spatial datasets is enabled
to visualized results in tissue coordinates, both enrichment plots for
each annotation label (individual enrichment plots) and for overall
results for each spot. This functionality is adapted from the <a href="https://drieslab.github.io/Giotto_website/" class="external-link">Giotto</a> package,
with methods based on <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-6-144" class="external-link">Kim
et al., 2005</a>.</p>
<hr>
</div>
<div class="section level3">
<h3 id="spot-deconvolution">Spot deconvolution<a class="anchor" aria-label="anchor" href="#spot-deconvolution"></a>
</h3>
<p><a href="https://github.com/YMa-lab/CARD" class="external-link">CARD</a> and <a href="https://github.com/ggrajeda/spacexr" class="external-link">RCTD</a> can be use for spot
deconvolution. CARD uses a Bayesian hierarchical model that incorporates
spatial correlation between spots, while RCTD applies a generalized
linear model for a more direct, assumption-light fit. We include both
methods to give users the choice between a spatially smoothed approach
(CARD) and a faster, model-driven method (RCTD), depending on dataset
characteristics and analysis goals.</p>
<hr>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="copyright">
  <p></p>
<p>Developed by Jiri Novotny, Jan Kubovciak, Lucie Pfeiferova, Michal Kolar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

    </footer>
</div>





  </body>
</html>
