<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scdrake">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/5/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>01 Get started • scdrake</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="01 Get started">
<meta property="og:description" content="scdrake">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/5/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/5/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="../darkswitch.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scdrake</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/scdrake.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake_integration.html">Integration pipeline guide</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Guides</h6>
    <a class="dropdown-item" href="../articles/scdrake_docker.html">Using the Docker image</a>
    <a class="dropdown-item" href="../articles/scdrake_integration.html">02 Integration pipeline guide</a>
    <a class="dropdown-item" href="../articles/scdrake_extend.html">Extending the pipeline</a>
    <a class="dropdown-item" href="../articles/drake_basics.html">drake basics</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General information</h6>
    <a class="dropdown-item" href="../articles/pipeline_overview.html">Pipeline overview</a>
    <a class="dropdown-item" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
    <a class="dropdown-item" href="../articles/scdrake_cli.html">Command line interface</a>
    <a class="dropdown-item" href="../articles/cluster_markers.html">Cluster markers</a>
    <a class="dropdown-item" href="../articles/scdrake_config.html">scdrake config internals</a>
    <a class="dropdown-item" href="../articles/scdrake_envvars.html">Environment variables</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General configs</h6>
    <a class="dropdown-item" href="../articles/config_pipeline.html">Pipeline config</a>
    <a class="dropdown-item" href="../articles/config_main.html">Main config</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single-sample pipeline stages</h6>
    <a class="dropdown-item" href="../articles/stage_input_qc.html">Input and quality control stage (`01_input_qc`)</a>
    <a class="dropdown-item" href="../articles/stage_norm_clustering.html">Normalization and clustering stage (`02_norm_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Integration pipeline stages</h6>
    <a class="dropdown-item" href="../articles/stage_integration.html">Integration stage (`01_integration`)</a>
    <a class="dropdown-item" href="../articles/stage_int_clustering.html">Post-integration clustering stage (`02_int_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Common stages</h6>
    <a class="dropdown-item" href="../articles/stage_cluster_markers.html">Cluster markers stage (`cluster_markers`)</a>
    <a class="dropdown-item" href="../articles/stage_contrasts.html">Contrasts stage (`contrasts`)</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bioinfocz/scdrake" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>

        
        <li>
          <a class="external-link nav-link" id="css-toggle-btn" aria-label="github">
            <span class="fas fa fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        
        
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>01 Get started</h1>
                        <h4 data-toc-skip class="author">Jiri
Novotny</h4>
            <address class="author_afil">
      Institute of Molecular Genetics of the Czech Academy of
Sciences<br><a class="author_email" href="mailto:#"></a><a href="mailto:jiri.novotny@img.cas.cz" class="email">jiri.novotny@img.cas.cz</a>
      </address>
                              <h4 data-toc-skip class="author">Jan
Kubovciak</h4>
            <address class="author_afil">
      Institute of Molecular Genetics of the Czech Academy of
Sciences<br><a class="author_email" href="mailto:#"></a><a href="mailto:jan.kubovciak@img.cas.cz" class="email">jan.kubovciak@img.cas.cz</a>
      </address>
                  
            <h4 data-toc-skip class="date"><sup>Document generated:
2022-12-19 07:25:14 UTC+0000</sup></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bioinfocz/scdrake/blob/mastervignettes/scdrake.Rmd" class="external-link"><code>vignettes/scdrake.Rmd</code></a></small>
      <div class="d-none name"><code>scdrake.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="quick-start-guide">Quick start guide<a class="anchor" aria-label="anchor" href="#quick-start-guide"></a>
</h2>
<p>In this guide you will see how to quickly setup
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> and run the first stage
(<code>01_input_qc</code>) of its single-sample pipeline using the
provided example data. Once you familiarize yourself with
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> basics, you can continue with guide for the
integration pipeline in
<code><a href="../articles/scdrake_integration.html">vignette("scdrake_integration")</a></code>.</p>
<p>If possible, we will show how to follow the steps from within the R
and through the command line interface (CLI), see
<code><a href="../articles/scdrake_cli.html">vignette("scdrake_cli")</a></code> for a short overview.</p>
<p>We <strong>strongly recommend</strong> to use the
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>’s Docker image as it is well tested. If you are
not familiar with Docker, please, refer to
<code><a href="../articles/scdrake_docker.html">vignette("scdrake_docker")</a></code>. We assume that the container
has a shared directory mounted as
<code>/home/rstudio/scdrake_projects</code>, as described in the
aforementioned vignette.</p>
<p>Note to the CLI: although you can specify the project directory using
the <code>-d / --dir</code> parameter, we rather recommend to always use
it as the current working directory before you issue
<code>scdrake</code> commands. Below in the examples you will notice
that the <code>-d</code> parameter is never used.</p>
<hr>
</div>
<div class="section level2">
<h2 id="the-three-steps">The three steps<a class="anchor" aria-label="anchor" href="#the-three-steps"></a>
</h2>
<p>After <a href="../index.html#installation-instructions">installation</a>, you
basically need the three steps described below to run the
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> pipeline.</p>
<div class="section level3 tabset">
<h3 class="tabset" id="step-1-initialize-a-new-project">Step 1: initialize a new project<a class="anchor" aria-label="anchor" href="#step-1-initialize-a-new-project"></a>
</h3>
<p><a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> is a project-based package, so the first step
is to initialize a new project. <em>Project</em> simply means a
directory in which the analysis of your data will take place. That also
means:</p>
<ul>
<li>Your current working directory is set to the project directory.</li>
<li>Whenever you specify file paths in config files, you do so relative
to the project directory (e.g. <code>output/plots/figure1.pdf</code>).
This way your project is <strong>transferable</strong> between different
computers and data locations.</li>
</ul>
<p>Now we initialize a new <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> project directory.</p>



<ul class="nav nav-tabs" id="step-1-initialize-a-new-project" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#in-r" id="in-r-tab" type="button" role="tab" aria-controls="in-r" aria-selected="true" class="active nav-link">In R</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#on-command-line" id="on-command-line-tab" type="button" role="tab" aria-controls="on-command-line" aria-selected="false" class="nav-link">On command line</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#on-command-line-docker" id="on-command-line-docker-tab" type="button" role="tab" aria-controls="on-command-line-docker" aria-selected="false" class="nav-link">On command line (Docker)</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="in-r" role="tabpanel" aria-labelledby="in-r-tab">

<p>First we load the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> package, and then call the
<code><a href="../reference/init_project.html">init_project()</a></code> function:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/init_project.html">init_project</a></span><span class="op">(</span><span class="st">"~/scdrake_projects/pbmc1k"</span>, download_example_data <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="tab-pane" id="on-command-line" role="tabpanel" aria-labelledby="on-command-line-tab">

<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> ~/scdrake_projects/pbmc1k</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/scdrake_projects/pbmc1k</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">scdrake</span> <span class="at">--download-example-data</span> init-project</span></code></pre></div>
</div>
<div class="tab-pane" id="on-command-line-docker" role="tabpanel" aria-labelledby="on-command-line-docker-tab">

<p>We assume that the container has a shared directory mounted as
<code>/home/rstudio/scdrake_projects</code>, as described in
<code><a href="../articles/scdrake_docker.html">vignette("scdrake_docker")</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> ~/scdrake_projects/pbmc1k</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ~/scdrake_projects/pbmc1k</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> exec <span class="at">-it</span> <span class="at">-u</span> rstudio <span class="at">-w</span> /home/rstudio/scdrake_projects/pbmc1k <span class="op">&lt;</span>CONTAINER ID or NAME<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  scdrake <span class="at">--download-example-data</span> init-project</span></code></pre></div>
</div>
</div>
</div>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="section"></h3>
<p>This will:</p>
<ul>
<li>Create a new project directory named <code>pbmc1k</code> in
<code>/home/&lt;user&gt;/scdrake_projects</code>.</li>
<li>Copy all files, which are bundled with the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>
package: default YAML configs, Rmd files, R scripts etc.</li>
<li>Download (if needed) the <a href="https://github.com/mikefarah/yq" class="external-link">yq tool</a> used to manipulate
YAML files.</li>
<li>Create a new RStudio project (<code>.RProj</code>) file and set it
as the active project (can be disabled).</li>
<li>Download the example dataset (10x Genomics PBMC).</li>
<li>(For R) Switch your current R working directory to project’s
directory and if you are using RStudio, also switch the active
project.</li>
</ul>
<blockquote>
<p>Important: whenever you will be running the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>
pipeline, make sure your working directory is set to the project’s
root.</p>
</blockquote>
<p>Let’s inspect files in the project directory:</p>
</div>
<div class="section level3 tabset">
<h3 class="tabset" id="section-1"></h3>



<ul class="nav nav-tabs" id="section-1" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#in-r-1" id="in-r-1-tab" type="button" role="tab" aria-controls="in-r-1" aria-selected="true" class="active nav-link">In R</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#on-command-line-1" id="on-command-line-1-tab" type="button" role="tab" aria-controls="on-command-line-1" aria-selected="false" class="nav-link">On command line</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#on-command-line-docker-1" id="on-command-line-docker-1-tab" type="button" role="tab" aria-controls="on-command-line-docker-1" aria-selected="false" class="nav-link">On command line (Docker)</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="in-r-1" role="tabpanel" aria-labelledby="in-r-1-tab">

<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span>all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="tab-pane" id="on-command-line-1" role="tabpanel" aria-labelledby="on-command-line-1-tab">

<p>If you have the <code>tree</code> tool installed
(<code>sudo apt install tree</code> to fix it):</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">tree</span></span></code></pre></div>
<p>Otherwise:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">'fs::dir_tree("~/scdrake_projects/pbmc1k")'</span></span></code></pre></div>
</div>
<div class="tab-pane" id="on-command-line-docker-1" role="tabpanel" aria-labelledby="on-command-line-docker-1-tab">

<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> exec <span class="at">-it</span> <span class="at">-u</span> rstudio <span class="at">-w</span> /home/rstudio/scdrake_projects/pbmc1k <span class="op">&lt;</span>CONTAINER ID or NAME<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  Rscript <span class="at">-e</span> <span class="st">'fs::dir_tree()'</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="section-2"></h3>
<pre><code><span><span class="co">## <span style="color: #0000BB; font-weight: bold;">/tmp/Rtmpc2CN4W/file9d234442d188</span></span></span>
<span><span class="co">## ├── .here</span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">Rmd</span></span></span>
<span><span class="co">## │   ├── <span style="color: #0000BB; font-weight: bold;">common</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">_footer.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">_header.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">cluster_markers.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">cluster_markers_table_template.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">contrasts.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">contrasts_table_template.Rmd</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">dimred_plots.Rmd</span></span></span>
<span><span class="co">## │   │   └── stylesheet.css</span></span>
<span><span class="co">## │   ├── <span style="color: #0000BB; font-weight: bold;">integration</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #00BB00;">01_integration.Rmd</span></span></span>
<span><span class="co">## │   │   └── <span style="color: #00BB00;">02_int_clustering.Rmd</span></span></span>
<span><span class="co">## │   └── <span style="color: #0000BB; font-weight: bold;">single_sample</span></span></span>
<span><span class="co">## │       ├── <span style="color: #00BB00;">01_input_qc.Rmd</span></span></span>
<span><span class="co">## │       ├── <span style="color: #0000BB; font-weight: bold;">01_input_qc_children</span></span></span>
<span><span class="co">## │       │   ├── <span style="color: #00BB00;">cell_filtering_custom.Rmd</span></span></span>
<span><span class="co">## │       │   ├── <span style="color: #00BB00;">cell_filtering_qc.Rmd</span></span></span>
<span><span class="co">## │       │   ├── <span style="color: #00BB00;">empty_droplets.Rmd</span></span></span>
<span><span class="co">## │       │   ├── <span style="color: #00BB00;">gene_filtering_custom.Rmd</span></span></span>
<span><span class="co">## │       │   └── <span style="color: #00BB00;">gene_filtering_qc.Rmd</span></span></span>
<span><span class="co">## │       ├── <span style="color: #00BB00;">02_norm_clustering.Rmd</span></span></span>
<span><span class="co">## │       └── <span style="color: #00BB00;">02_norm_clustering_simple.Rmd</span></span></span>
<span><span class="co">## ├── <span style="color: #00BB00;">_drake_integration.R</span></span></span>
<span><span class="co">## ├── <span style="color: #00BB00;">_drake_single_sample.R</span></span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">config</span></span></span>
<span><span class="co">## │   ├── <span style="color: #0000BB; font-weight: bold;">integration</span></span></span>
<span><span class="co">## │   │   ├── 00_main.default.yaml</span></span>
<span><span class="co">## │   │   ├── 00_main.yaml</span></span>
<span><span class="co">## │   │   ├── 01_integration.default.yaml</span></span>
<span><span class="co">## │   │   ├── 01_integration.yaml</span></span>
<span><span class="co">## │   │   ├── 02_int_clustering.default.yaml</span></span>
<span><span class="co">## │   │   ├── 02_int_clustering.yaml</span></span>
<span><span class="co">## │   │   ├── cluster_markers.default.yaml</span></span>
<span><span class="co">## │   │   ├── cluster_markers.yaml</span></span>
<span><span class="co">## │   │   ├── contrasts.default.yaml</span></span>
<span><span class="co">## │   │   └── contrasts.yaml</span></span>
<span><span class="co">## │   ├── pipeline.default.yaml</span></span>
<span><span class="co">## │   ├── pipeline.yaml</span></span>
<span><span class="co">## │   └── <span style="color: #0000BB; font-weight: bold;">single_sample</span></span></span>
<span><span class="co">## │       ├── 00_main.default.yaml</span></span>
<span><span class="co">## │       ├── 00_main.yaml</span></span>
<span><span class="co">## │       ├── 01_input_qc.default.yaml</span></span>
<span><span class="co">## │       ├── 01_input_qc.yaml</span></span>
<span><span class="co">## │       ├── 02_norm_clustering.default.yaml</span></span>
<span><span class="co">## │       ├── 02_norm_clustering.yaml</span></span>
<span><span class="co">## │       ├── cluster_markers.default.yaml</span></span>
<span><span class="co">## │       ├── cluster_markers.yaml</span></span>
<span><span class="co">## │       ├── contrasts.default.yaml</span></span>
<span><span class="co">## │       └── contrasts.yaml</span></span>
<span><span class="co">## ├── <span style="color: #0000BB; font-weight: bold;">example_data</span></span></span>
<span><span class="co">## │   ├── <span style="color: #0000BB; font-weight: bold;">pbmc1k</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #BB0000; font-weight: bold;">barcodes.tsv.gz</span></span></span>
<span><span class="co">## │   │   ├── <span style="color: #BB0000; font-weight: bold;">features.tsv.gz</span></span></span>
<span><span class="co">## │   │   └── <span style="color: #BB0000; font-weight: bold;">matrix.mtx.gz</span></span></span>
<span><span class="co">## │   └── <span style="color: #0000BB; font-weight: bold;">pbmc3k</span></span></span>
<span><span class="co">## │       ├── barcodes.tsv</span></span>
<span><span class="co">## │       ├── genes.tsv</span></span>
<span><span class="co">## │       └── matrix.mtx</span></span>
<span><span class="co">## ├── <span style="color: #00BB00;">plan_custom.R</span></span></span>
<span><span class="co">## ├── renv.lock</span></span>
<span><span class="co">## └── scdrake.Rproj</span></span></code></pre>
<p>The most important files and directories are:</p>
<ul>
<li>
<code>config/</code>: configuration files in YAML format. The
default configuration files (<code>*.default.yaml</code>) are bundled
with the package and used to supply new parameters to local configs
(<code>*.yaml</code>) when such events appear. See
<code><a href="../articles/scdrake_config.html">vignette("scdrake_config")</a></code> for more details.</li>
<li>
<code>Rmd/</code>: RMarkdown files used for reporting of pipeline
results. You can edit them according to your needs, but keep in mind
that they will be overwritten by package-bundled ones whenever you call
<code><a href="../reference/update_project.html">update_project()</a></code>.</li>
<li>
<code>_drake_integration.R</code>,
<code>_drake_single_sample.R</code>: entry scripts for
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> when pipeline is executed through
<code><a href="../reference/run_pipeline.html">run_single_sample_r()</a></code> or <code><a href="../reference/run_pipeline.html">run_integration_r()</a></code>
(the most reproducible way to run the pipeline).</li>
<li>
<code>plan_custom.R</code>: here you can define your own
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> pipeline (plan) to extend <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>.
More on this topic in <code><a href="../articles/scdrake_extend.html">vignette("scdrake_extend")</a></code>.</li>
<li>
<code>example_data/</code>: example datasets. Those are raw
feature-barcode matrices output by <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="external-link">cellranger</a>.</li>
<li>
<code>scdrake.Rproj</code>: RStudio project file. If you open this
file in RStudio, your working directory will be automatically set to
project’s root directory.</li>
</ul>
<hr>
</div>
<div class="section level3">
<h3 id="step-2-modify-the-configuration-files">Step 2: modify the configuration files<a class="anchor" aria-label="anchor" href="#step-2-modify-the-configuration-files"></a>
</h3>
<p>The configuration files for all subpipelines (called <em>stages</em>)
are stored in the <code>config/</code> directory. At this time we only
need to modify a single parameter to run the first stage
(<code>01_input_qc</code>) of the single-sample pipeline.</p>
<p>To do so, open <code>config/single_sample/01_input_qc.yaml</code> and
set value of <code>path</code> inside <code>INPUT_DATA</code> to
<code>"example_data/pbmc1k"</code>, such that:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_DATA</span><span class="kw">:</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">type</span><span class="kw">:</span><span class="at"> </span><span class="st">"cellranger"</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">path</span><span class="kw">:</span><span class="at"> </span><span class="st">"example_data/pbmc1k"</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="fu">delimiter</span><span class="kw">:</span><span class="at"> </span><span class="st">","</span></span></code></pre></div>
<blockquote>
<p>Important: all paths in configs must be relative to the project root
directory, or absolute (<strong>not recommended</strong>), unless
otherwise stated.</p>
</blockquote>
<hr>
</div>
<div class="section level3 tabset">
<h3 class="tabset" id="step-3-run-the-single-sample-pipeline">Step 3: run the single-sample pipeline<a class="anchor" aria-label="anchor" href="#step-3-run-the-single-sample-pipeline"></a>
</h3>
<p>By default, only the final target <code>report_input_qc</code> (and
all targets it depends on) of the input and QC stage
(<code>01_input_qc</code>) will be made, that is, a HTML report. This is
specified by <code>DRAKE_TARGETS</code> parameter in
<code>config/pipeline.yaml</code>. You can look <a href="#how-to-know-which-targets-you-can-load-and-what-each-parameter-do">below</a>
for how to identify other targets which can be made.</p>
<p>Now we can run the single-sample pipeline for the PBMC 1k
dataset:</p>



<ul class="nav nav-tabs" id="step-3-run-the-single-sample-pipeline" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#in-r-2" id="in-r-2-tab" type="button" role="tab" aria-controls="in-r-2" aria-selected="true" class="active nav-link">In R</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#on-command-line-2" id="on-command-line-2-tab" type="button" role="tab" aria-controls="on-command-line-2" aria-selected="false" class="nav-link">On command line</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#on-command-line-docker-2" id="on-command-line-docker-2-tab" type="button" role="tab" aria-controls="on-command-line-docker-2" aria-selected="false" class="nav-link">On command line (Docker)</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="in-r-2" role="tabpanel" aria-labelledby="in-r-2-tab">

<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/run_pipeline.html">run_single_sample_r</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="tab-pane" id="on-command-line-2" role="tabpanel" aria-labelledby="on-command-line-2-tab">

<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scdrake</span> <span class="at">--pipeline-type</span> single_sample run</span></code></pre></div>
</div>
<div class="tab-pane" id="on-command-line-docker-2" role="tabpanel" aria-labelledby="on-command-line-docker-2-tab">

<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> exec <span class="at">-it</span> <span class="at">-u</span> rstudio <span class="at">-w</span> /home/rstudio/scdrake_projects/pbmc1k <span class="op">&lt;</span>CONTAINER ID or NAME<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  scdrake <span class="at">--pipeline-type</span> single_sample run</span></code></pre></div>
</div>
</div>
</div>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="section-3"></h3>
<p>The output is saved in <code>output/single_sample</code> directory,
as specified by <code>BASE_OUT_DIR</code> in
<code>config/single_sample/00_main.yaml</code>. You can find the final
report of the <code>01_input_qc</code> stage in
<code>output/single_sample/01_input_qc/01_input_qc.html</code>.</p>
<p>If you want to run the main target for next stage
(<code>02_norm_clustering</code>), just open
<code>config/pipeline.yaml</code> and change <code>DRAKE_TARGETS</code>
to <code>["report_norm_clustering"]</code> (or comment the current
<code>DRAKE_TARGETS</code> line and uncomment the predefined one).</p>
<p>Just note the default config files are meant for the example PBMC
data, and so before you analyse your own data, you should read the
section <a href="#before-you-analyse-your-own-data">below</a> about
important steps before you do so.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="before-you-analyse-your-own-data">Before you analyse your own data<a class="anchor" aria-label="anchor" href="#before-you-analyse-your-own-data"></a>
</h2>
<p>The default config files are designed to work with provided example
data and contain parameters which are used to test/showcase various
features of the pipeline. Hence, some parameters are example
data-specific, and might lead to pipeline failure when kept as-is for
your data. Also, some parameters are critical for the analysis outcomes
and you should be aware of them.</p>
<p>Below you can find important parameters which you should review
before you run the pipeline on your data for the first time (but it
doesn’t mean you shouldn’t use them later). All parameters are
documented in vignettes for their respective stages.</p>
<details><summary>
Show important parameters
</summary><div class="section level3 unlisted">
<h3 class="unlisted" id="main-config-00_main-yaml">Main config (<code>00_main.yaml</code>)<a class="anchor" aria-label="anchor" href="#main-config-00_main-yaml"></a>
</h3>
<ul>
<li><code>ORGANISM</code></li>
<li><code>ANNOTATION_LIST</code></li>
<li><code>ENSEMBL_SPECIES</code></li>
<li><code>BASE_OUT_DIR</code></li>
</ul>
</div>
<div class="section level3 unlisted">
<h3 class="unlisted" id="single-sample-input-qc-stage-01_input_qc-yaml">Single-sample / Input QC stage
(<code>01_input_qc.yaml</code>)<a class="anchor" aria-label="anchor" href="#single-sample-input-qc-stage-01_input_qc-yaml"></a>
</h3>
<ul>
<li><code>INPUT_DATA</code></li>
<li><code>EMPTY_DROPLETS_ENABLED</code></li>
<li><code>SAVE_DATASET_SENSITIVE_FILTERING</code></li>
</ul>
</div>
<div class="section level3 unlisted">
<h3 class="unlisted" id="single-sample-normalization-and-clustering-stage-02_norm_clustering-yaml">Single-sample / Normalization and clustering stage
(<code>02_norm_clustering.yaml</code>)<a class="anchor" aria-label="anchor" href="#single-sample-normalization-and-clustering-stage-02_norm_clustering-yaml"></a>
</h3>
<ul>
<li><code>NORMALIZATION_TYPE</code></li>
<li><code>HVG_METRIC</code></li>
<li><code>PCA_SELECTION_METHOD</code></li>
<li>
<code>PCA_FORCED_PCS</code> (if <code>PCA_SELECTION_METHOD</code> is
<code>"forced"</code>)</li>
<li><code>KMEANS_K</code></li>
<li><code>SC3_K</code></li>
<li><code>SC3_N_CORES</code></li>
</ul>
</div>
<div class="section level3 unlisted">
<h3 class="unlisted" id="integration-integration-01_integration-yaml">Integration / Integration
(<code>01_integration.yaml</code>)<a class="anchor" aria-label="anchor" href="#integration-integration-01_integration-yaml"></a>
</h3>
<ul>
<li><code>INTEGRATION_SOURCES</code></li>
</ul>
</div>
<div class="section level3 unlisted">
<h3 class="unlisted" id="integration-post-integration-clustering-02_int_clustering-yaml">Integration / Post-integration clustering
(<code>02_int_clustering.yaml</code>)<a class="anchor" aria-label="anchor" href="#integration-post-integration-clustering-02_int_clustering-yaml"></a>
</h3>
<ul>
<li><code>KMEANS_K</code></li>
<li><code>SC3_K</code></li>
<li><code>SC3_N_CORES</code></li>
</ul>
</div>
<div class="section level3 unlisted">
<h3 class="unlisted" id="common-cluster-markers-and-contrasts-stages-cluster_markers-yaml-contrasts-yaml">Common / Cluster markers and contrasts stages
(<code>cluster_markers.yaml</code>, <code>contrasts.yaml</code>)<a class="anchor" aria-label="anchor" href="#common-cluster-markers-and-contrasts-stages-cluster_markers-yaml-contrasts-yaml"></a>
</h3>
<ul>
<li><code>CLUSTER_MARKERS_SOURCES</code></li>
<li><code>CONTRASTS_SOURCES</code></li>
</ul>
<hr>
</div>
</details>
</div>
<div class="section level2">
<h2 id="retrieving-intermediate-results-targets">Retrieving intermediate results (targets)<a class="anchor" aria-label="anchor" href="#retrieving-intermediate-results-targets"></a>
</h2>
<p>In <a href="https://docs.ropensci.org/drake" class="external-link">drake</a>’s terminology, a pipeline is called
<em>plan</em>, and is composed of <em>targets</em>. When a target is
finished, its value (object) is saved to cache (the directory
<code>.drake</code> by default). The cache has two main purposes:</p>
<ul>
<li>If a target is not changed, its value is loaded from the cache.
<em>Change</em> involves e.g. target’s definition (code) or change in
upstream targets on which the target depends. This way
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> is able to skip computation of finished targets and
greatly enhance the runtime. More details <a href="https://books.ropensci.org/drake/triggers.html" class="external-link">here</a>.</li>
<li>Users also have access to the cache, and so you can load any
finished target into your R session.</li>
</ul>
<p>Users can access the cache via two <a href="https://docs.ropensci.org/drake" class="external-link">drake</a>’s
functions:</p>
<ul>
<li>
<code><a href="https://docs.ropensci.org/drake/reference/readd.html" class="external-link">drake::loadd()</a></code> loads target’s value to the current
session as a variable of the same name.</li>
<li>
<code><a href="https://docs.ropensci.org/drake/reference/readd.html" class="external-link">drake::readd()</a></code> returns target’s value (so it can be
assigned to variable).</li>
</ul>
<p>Let’s try it and load the filtered <code>SingleCellExperiment</code>
object:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">drake</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/drake/reference/readd.html" class="external-link">loadd</a></span><span class="op">(</span><span class="va">sce_final_input_qc</span><span class="op">)</span></span></code></pre></div>
<p>Value of the target <code>sce_final_input_qc</code> was loaded as a
variable of the same name to your current R session (or more precisely,
to the global environment).</p>
<p>Similarly, we can load this target to a variable of our choice:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">drake</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/drake/reference/readd.html" class="external-link">readd</a></span><span class="op">(</span><span class="va">sce_final_input_qc</span><span class="op">)</span></span></code></pre></div>
<p>And work further with the loaded object, e.g.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"NOC2L"</span>, exprs_values <span class="op">=</span> <span class="st">"counts"</span>, swap_rownames <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="how-to-know-which-targets-you-can-load-and-what-each-parameter-do">How to know which targets you can load and what each parameter
do?<a class="anchor" aria-label="anchor" href="#how-to-know-which-targets-you-can-load-and-what-each-parameter-do"></a>
</h2>
<p>Each pipeline stage has its own vignette describing its configuration
parameters and also the most important targets. At <a href="https://bioinfocz.github.io/scdrake" class="uri">https://bioinfocz.github.io/scdrake</a> simply navigate to
the top dropdown menu Articles and choose a stage you are interested in.
Also, in <code>config/pipeline.yaml</code> there are commented the
summary targets for each stage - HTML reports.</p>
<p>For a more schematic overview of pipelines and stages see
<code><a href="../articles/pipeline_overview.html">vignette("pipeline_overview")</a></code> (also includes <a href="https://github.com/bioinfocz/scdrake/blob/main/diagrams/README.md" class="external-link">diagrams</a>).</p>
<p>Advanced users might be interested in looking into source code of
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>’s <a href="https://github.com/bioinfocz/scdrake/tree/main/R" class="external-link">plans</a> (files
named <code>plans_*.R</code>).</p>
<hr>
</div>
<div class="section level2">
<h2 id="modifying-parameters-and-rerunning-the-pipeline">Modifying parameters and rerunning the pipeline<a class="anchor" aria-label="anchor" href="#modifying-parameters-and-rerunning-the-pipeline"></a>
</h2>
<p>Optionally, to further demonstrate the <a href="https://docs.ropensci.org/drake" class="external-link">drake</a>’s ability
to skip finished targets, let’s assume you decided to be more benevolent
and used less strict values for custom cell filtering in
<code>config/single_sample/01_input_qc.yaml</code>:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">MIN_UMI_CF</span><span class="kw">:</span><span class="at"> </span><span class="dv">500</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">MAX_UMI_CF</span><span class="kw">:</span><span class="at"> </span><span class="dv">70000</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">MIN_FEATURES</span><span class="kw">:</span><span class="at"> </span><span class="dv">500</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">MAX_MITO_RATIO</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.3</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">INPUT_QC_BASE_OUT_DIR</span><span class="kw">:</span><span class="at"> </span><span class="st">"01_input_qc_benevolent"</span></span></code></pre></div>
<p>After repeating the Step 3 above you can see the most time-consuming
targets <code>sce_raw</code> and <code>empty_droplets</code> were
skipped. Also, we simply output the HTML report for new parameters into
a different directory, and so we can easily compare it with the previous
report.</p>
<hr>
</div>
<div class="section level2">
<h2 id="running-the-pipeline-in-parallel-mode">Running the pipeline in parallel mode<a class="anchor" aria-label="anchor" href="#running-the-pipeline-in-parallel-mode"></a>
</h2>
<p>Because <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> knows the inner relationships between
targets in your plan, it also knows which targets are independent of
each other, and thus, can be run concurrently. This is called
<em>implicit parallelism</em>, and to fully utilize this important
feature, you just need to modify <code>config/pipeline.yaml</code> by
setting <code>DRAKE_PARALLELISM</code> to either:</p>
<ul>
<li>
<code>"future"</code>: uses the <a href="https://future.futureverse.org" class="external-link">future</a> as the
backend. This backend should work by simply installing the
<a href="https://future.futureverse.org" class="external-link">future</a> package.
<ul>
<li>Install by
<code>BiocManager::install(c("future", "future.callr"))</code>
</li>
</ul>
</li>
<li>
<code>"clustermq"</code>: uses the <a href="https://github.com/mschubert/clustermq" class="external-link">clustermq</a> as the
backend. This is faster than <code>"future"</code>, but besides the
<a href="https://github.com/mschubert/clustermq" class="external-link">clustermq</a> package it also requires the <a href="https://zeromq.org/" class="external-link">ZeroMQ</a> library to be <a href="https://zeromq.org/download/" class="external-link">installed</a> on your system.
<ul>
<li>A specific version of the <a href="https://github.com/mschubert/clustermq" class="external-link">clustermq</a> package is needed
and can be installed with
<code>remotes::install_version("clustermq", version = "0.8.8")</code>
(you might need <code>BiocManager::install("remotes")</code>).</li>
<li>
<a href="https://github.com/mschubert/clustermq" class="external-link">clustermq</a> also supports HPC cluster schedulers, see
<a href="https://books.ropensci.org/drake/hpc.html#the-template-argument-for-persistent-workers" class="external-link">here</a>
for more details.</li>
</ul>
</li>
</ul>
<p>If you have installed <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> from the
<code>renv.lock</code> file or you are using the Docker image, then
these two packages above are always installed.</p>
<p>For a general overview of <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> parallelism see <a href="https://books.ropensci.org/drake/hpc.html" class="external-link uri">https://books.ropensci.org/drake/hpc.html</a></p>
<hr>
<div class="section level3 tabset">
<h3 class="tabset" id="updating-the-project-files">Updating the project files<a class="anchor" aria-label="anchor" href="#updating-the-project-files"></a>
</h3>
<p>When a new version of <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> is released, you can
update your project with (assuming your working directory is in a
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> project’s root):</p>



<ul class="nav nav-tabs" id="updating-the-project-files" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#in-r-3" id="in-r-3-tab" type="button" role="tab" aria-controls="in-r-3" aria-selected="true" class="active nav-link">In R</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#on-command-line-3" id="on-command-line-3-tab" type="button" role="tab" aria-controls="on-command-line-3" aria-selected="false" class="nav-link">On command line</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#on-command-line-docker-3" id="on-command-line-docker-3-tab" type="button" role="tab" aria-controls="on-command-line-docker-3" aria-selected="false" class="nav-link">On command line (Docker)</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="in-r-3" role="tabpanel" aria-labelledby="in-r-3-tab">

<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/update_project.html">update_project</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="tab-pane" id="on-command-line-3" role="tabpanel" aria-labelledby="on-command-line-3-tab">

<div class="sourceCode" id="cb18"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">scdrake</span> update-project</span></code></pre></div>
</div>
<div class="tab-pane" id="on-command-line-docker-3" role="tabpanel" aria-labelledby="on-command-line-docker-3-tab">

<div class="sourceCode" id="cb19"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> exec <span class="at">-it</span> <span class="at">-u</span> rstudio <span class="at">-w</span> /path/to/scdrake/project <span class="op">&lt;</span>CONTAINER ID or NAME<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  scdrake update-project</span></code></pre></div>
</div>
</div>
</div>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="section-4"></h3>
<p>This will <strong>overwrite</strong> project files by the
package-bundled ones:</p>
<ul>
<li>RMarkdown documents in <code>Rmd/</code>.</li>
<li>Initial scripts for <code><a href="../reference/run_pipeline.html">run_single_sample_r()</a></code> and
<code><a href="../reference/run_pipeline.html">run_integration_r()</a></code> (wrappers around
<code><a href="https://docs.ropensci.org/drake/reference/r_make.html" class="external-link">drake::r_make()</a></code>): <code>_drake_single_sample.R</code> and
<code>_drake_integration.R</code>.</li>
<li>Default YAML configs in <code>config/</code>
(<code>*.default.yaml</code>).</li>
</ul>
<p>By default, you will be asked if you want to continue, as you might
lose your local modifications.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="going-further">Going further<a class="anchor" aria-label="anchor" href="#going-further"></a>
</h2>
<p>Now you should know the <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> basics:</p>
<ul>
<li>How to initialize a new project and which files does it
contain.</li>
<li>How to modify pipeline parameters stored in YAML config files.</li>
<li>How to run the pipeline.</li>
</ul>
<p>For the integration pipeline you might be interested in its guide in
<code><a href="../articles/scdrake_integration.html">vignette("scdrake_integration")</a></code>.</p>
<p>For the full insight into <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> you can also read
the following vignettes:</p>
<ul>
<li>Guides:
<ul>
<li>Using the Docker image: <code><a href="../articles/scdrake_docker.html">vignette("scdrake_docker")</a></code>
</li>
<li>01 Quick start (single-sample pipeline):
<code><a href="../articles/scdrake.html">vignette("scdrake")</a></code>
</li>
<li>02 Integration pipeline guide:
<code><a href="../articles/scdrake_integration.html">vignette("scdrake_integration")</a></code>
</li>
<li>Extending the pipeline: <code><a href="../articles/scdrake_extend.html">vignette("scdrake_extend")</a></code>
</li>
<li>
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> basics: <code><a href="../articles/drake_basics.html">vignette("drake_basics")</a></code>
<ul>
<li>Or the official <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> book: <a href="https://books.ropensci.org/drake/" class="external-link uri">https://books.ropensci.org/drake/</a>
</li>
</ul>
</li>
</ul>
</li>
<li>General information:
<ul>
<li>Pipeline overview: <code><a href="../articles/pipeline_overview.html">vignette("pipeline_overview")</a></code>
</li>
<li>FAQ &amp; Howtos: <code><a href="../articles/scdrake_faq.html">vignette("scdrake_faq")</a></code>
</li>
<li>Command line interface (CLI):
<code><a href="../articles/scdrake_cli.html">vignette("scdrake_cli")</a></code>
</li>
<li>Config files: <code><a href="../articles/scdrake_config.html">vignette("scdrake_config")</a></code>
</li>
<li>Cluster markers: <code><a href="../articles/cluster_markers.html">vignette("cluster_markers")</a></code>
</li>
<li>Environment variables: <code><a href="../articles/scdrake_envvars.html">vignette("scdrake_envvars")</a></code>
</li>
</ul>
</li>
<li>General configs:
<ul>
<li>Pipeline config: <code><a href="../articles/config_pipeline.html">vignette("config_pipeline")</a></code>
</li>
<li>Main config: <code><a href="../articles/config_main.html">vignette("config_main")</a></code>
</li>
</ul>
</li>
<li>Targets and config parameters for each stage:
<ul>
<li>Single-sample pipeline:
<ul>
<li>Reading in data, filtering, quality control (stage
<code>01_input_qc</code>): <code><a href="../articles/stage_input_qc.html">vignette("stage_input_qc")</a></code>
</li>
<li>Normalization, HVG selection, dimensionality reduction, clustering,
cell type annotation (stage <code>02_norm_clustering</code>):
<code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code>
</li>
</ul>
</li>
<li>Integration pipeline:
<ul>
<li>Reading in data and integration (stage <code>01_integration</code>):
<code><a href="../articles/stage_integration.html">vignette("stage_integration")</a></code>
</li>
<li>Post-integration clustering (stage <code>02_int_clustering</code>):
<code><a href="../articles/stage_int_clustering.html">vignette("stage_int_clustering")</a></code>
</li>
</ul>
</li>
<li>Common:
<ul>
<li>Cluster markers stage:
<code><a href="../articles/stage_cluster_markers.html">vignette("stage_cluster_markers")</a></code>
</li>
<li>Contrasts stage: <code><a href="../articles/stage_contrasts.html">vignette("stage_contrasts")</a></code>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="copyright">
  <p></p>
<p>Developed by Jiri Novotny, Jan Kubovciak.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
