<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/3/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Integration stage • scdrake</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Integration stage">
<meta property="og:description" content="01_integration stage of the integration pipeline.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/3/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/3/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="../darkswitch.js"></script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scdrake</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/scdrake.html">Get started</a>
</li>
<li>
  <a href="../articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="divider">
    </li>
<li class="dropdown-header">General information</li>
    <li>
      <a href="../articles/pipeline_overview.html">Pipeline overview</a>
    </li>
    <li>
      <a href="../articles/scdrake_run.html">Running the pipeline, environment variables</a>
    </li>
    <li>
      <a href="../articles/cluster_markers.html">Cluster markers</a>
    </li>
    <li>
      <a href="../articles/drake_basics.html">drake basics</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Configs</li>
    <li>
      <a href="../articles/scdrake_config.html">scdrake configs: a general overview</a>
    </li>
    <li>
      <a href="../articles/config_pipeline.html">Pipeline config</a>
    </li>
    <li>
      <a href="../articles/config_main.html">Main config</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Single-sample pipeline stages</li>
    <li>
      <a href="../articles/stage_input_qc.html">Input and quality control stage</a>
    </li>
    <li>
      <a href="../articles/stage_norm_clustering.html">Normalization and clustering stage</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Integration pipeline stages</li>
    <li>
      <a href="../articles/stage_integration.html">Integration stage</a>
    </li>
    <li>
      <a href="../articles/stage_int_clustering.html">Post-integration clustering stage</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Common stages</li>
    <li>
      <a href="../articles/stage_cluster_markers.html">Cluster markers stage</a>
    </li>
    <li>
      <a href="../articles/stage_contrasts.html">Contrasts stage</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
          <a href="#" id="css-toggle-btn">
            <span class="fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        <li>
  <a href="https://github.com/bioinfocz/scdrake" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
        
        


      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Integration stage</h1>
            
            <h4 data-toc-skip class="date">2022-01-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bioinfocz/scdrake/blob/mastervignettes/stage_integration.Rmd" class="external-link"><code>vignettes/stage_integration.Rmd</code></a></small>
      <div class="hidden name"><code>stage_integration.Rmd</code></div>

    </div>

    
    
<hr>
<p>For a general overview of this integration pipeline stage see <code><a href="../articles/pipeline_overview.html">vignette("pipeline_overview")</a></code>.</p>
<p>Config for this stage is stored in <code>01_integration.yaml</code> file. Directory with this file is read from <code>SCDRAKE_INTEGRATION_CONFIG_DIR</code> environment variable upon <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> load or attach, and saved as <code>scdrake_integration_config_dir</code> option. This option is used as the default argument value in several <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> functions.</p>
<hr>
<div class="section level2">
<h2 id="config-parameters">Config parameters<a class="anchor" aria-label="anchor" href="#config-parameters"></a>
</h2>
<div class="section level3">
<h3 id="integration-sources">Integration sources<a class="anchor" aria-label="anchor" href="#integration-sources"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_SOURCES</span><span class="kw">:</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> </span><span class="fu">pbmc1k</span><span class="kw">:</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">cache_path</span><span class="kw">:</span><span class="at"> </span><span class="st">".drake_pbmc1k"</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">description</span><span class="kw">:</span><span class="at"> </span><span class="st">"10x Genomics PBMC 1k dataset"</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">hvg_rm_cc_genes</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">hvg_cc_genes_var_expl_threshold</span><span class="kw">:</span><span class="at"> </span><span class="dv">5</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">pbmc3k</span><span class="kw">:</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">cache_path</span><span class="kw">:</span><span class="at"> </span><span class="st">".drake_pbmc3k"</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">description</span><span class="kw">:</span><span class="at"> </span><span class="st">"10x Genomics PBMC 3k dataset"</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">hvg_rm_cc_genes</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">hvg_cc_genes_var_expl_threshold</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span></code></pre></div>
<p><strong>Type:</strong> list of named lists</p>
<p>This parameter specifies datasets to be integrated. Those are read in from single-sample pipeline results (the <code>sc_final_norm_clustering</code> targets) taken from corresponding <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> caches.</p>
<p>Each name of a nested list specifies the name of single-sample, e.g. <code>pbmc1k</code>. Parameters for each single-sample are:</p>
<ul>
<li>
<code>cache_path: ".drake_pbmc1k"</code> (character scalar): a path to <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cache.</li>
<li>
<code>description: "10x Genomics PBMC 1k dataset"</code> (character scalar): a description of the single-sample, will appear in reports.</li>
<li>
<code>hvg_rm_cc_genes: True</code> (logical scalar): if <code>True</code>, cell cycle-related genes will be removed prior to selection of highly variable genes (HVGs). This is the same method of cell cycle correction as in the <code>02_norm_clustering</code> stage of the single-sample pipeline (see <code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code> for more details).</li>
<li>
<code>hvg_cc_genes_var_expl_threshold: 5</code> (numeric scalar): a threshold on cell cycle variance explained. Genes with var. expl. greater than this threshold will be marked as CC-related.</li>
</ul>
<p>Currently, there are some limitations of which datasets can be integrated:</p>
<ul>
<li>All single-samples must be normalized by <code>{scran}</code> method.</li>
<li>If the HVG selection in integrated data is based on combined HVG metrics (see the relevant parameters below), all single-samples had to use the same HVG metric (<code>HVG_METRIC</code> of <code>"gene_var"</code> or <code>"gene_cv2"</code>).</li>
</ul>
</div>
<div class="section level3">
<h3 id="selection-of-highly-variable-genes-hvgs">Selection of highly variable genes (HVGs)<a class="anchor" aria-label="anchor" href="#selection-of-highly-variable-genes-hvgs"></a>
</h3>
<p>This is done prior to integration, and is same for all methods specified in the <code>INTEGRATION_METHODS</code> (see below).</p>
<p>If a single-sample has the <code>hvg_rm_cc_genes</code> set to <code>True</code>, cell cycle-related genes will be removed prior to HVG combination and selection.</p>
<hr>
<p><code>HVG_COMBINATION_INT: "hvg_metric"</code></p>
<p><strong>Type:</strong> character scalar (<code>"hvg_metric"</code> | <code>"intersection"</code> | <code>"union"</code> | <code>"all"</code>)</p>
<p>How to combine HVGs from single-samples:</p>
<ul>
<li>
<code>"hvg_metric"</code>: combine HVG metrics (gene variance or CV2) and then apply HVG selection similar to procedure in the <code>02_norm_clustering</code> stage of the single-sample pipeline. Note that in order to combine HVG metrics, all samples had to have <code>HVG_METRIC</code> set to either <code>"gene_var"</code> (<code><a href="https://rdrr.io/pkg/scran/man/combineVar.html" class="external-link">scran::combineVar()</a></code>) or <code>"gene_cv2"</code> (<code><a href="https://rdrr.io/pkg/scran/man/combineVar.html" class="external-link">scran::combineCV2()</a></code>).</li>
<li>Combine HVGs by set operations:
<ul>
<li><code>"intersection"</code></li>
<li><code>"union"</code></li>
</ul>
</li>
<li>
<code>"all"</code>: use all common genes from all samples as HVGs (without duplicates).</li>
</ul>
<hr>
<div class="sourceCode" id="cb2"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">HVG_SELECTION_INT</span><span class="kw">:</span><span class="at"> </span><span class="st">"top"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">HVG_SELECTION_VALUE_INT</span><span class="kw">:</span><span class="at"> </span><span class="dv">3000</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar, integer scalar</p>
<p>HVG selection strategy. Only relevant when <code>HVG_COMBINATION_INT</code> is <code>"hvg_metric"</code>, i.e. this selection will be applied to combined HVG metrics.</p>
<p>These parameters are similar to <code>HVG_SELECTION</code> and <code>HVG_SELECTION_VALUE</code> in the <code>02_norm_clustering</code> stage of the single-sample pipeline (see <code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code> for more details).</p>
</div>
<div class="section level3">
<h3 id="integration-methods">Integration methods<a class="anchor" aria-label="anchor" href="#integration-methods"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_METHODS</span><span class="kw">:</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> </span><span class="fu">uncorrected</span><span class="kw">:</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_selection_method</span><span class="kw">:</span><span class="at"> </span><span class="st">"forced"</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_forced_pcs</span><span class="kw">:</span><span class="at"> </span><span class="dv">15</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_perp</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_max_iter</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">rescaling</span><span class="kw">:</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_selection_method</span><span class="kw">:</span><span class="at"> </span><span class="st">"forced"</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_forced_pcs</span><span class="kw">:</span><span class="at"> </span><span class="dv">15</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_perp</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_max_iter</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">integration_params</span><span class="kw">:</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">log.base</span><span class="kw">:</span><span class="at"> </span><span class="dv">2</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">pseudo.count</span><span class="kw">:</span><span class="at"> </span><span class="dv">1</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">regression</span><span class="kw">:</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_selection_method</span><span class="kw">:</span><span class="at"> </span><span class="st">"corrected"</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_forced_pcs</span><span class="kw">:</span><span class="at"> </span><span class="dv">15</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_perp</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_max_iter</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">integration_params</span><span class="kw">:</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">d</span><span class="kw">:</span><span class="at"> </span><span class="dv">50</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">mnn</span><span class="kw">:</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_selection_method</span><span class="kw">:</span><span class="at"> </span><span class="st">"corrected"</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_forced_pcs</span><span class="kw">:</span><span class="at"> </span><span class="dv">15</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_perp</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_max_iter</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">integration_params</span><span class="kw">:</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">k</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">prop.k</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">cos.norm</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">ndist</span><span class="kw">:</span><span class="at"> </span><span class="dv">3</span></span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">d</span><span class="kw">:</span><span class="at"> </span><span class="dv">50</span></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">merge.order</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">auto.merge</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span></code></pre></div>
<p><strong>Type:</strong> list of named lists</p>
<p>Each named list is specifying parameters for one integration method. The ones specified in the <code>integration_params</code> list are used directly in an integration method, while the others are used prior/after integration. The <code>"uncorrected"</code> and at least one integration method must be set (by default, all methods are performed). Here are descriptions and possible parameters (some of them are used by multiple methods, and thus described only once) for each integration method:</p>
<ul>
<li>
<code>uncorrected</code>: correction for inter-batch sequencing depth with <code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchNorm.html" class="external-link">batchelor::multiBatchNorm()</a></code>. This method is required, because it is used for differential expression analysis (cluster markers and contrasts).
<ul>
<li>
<code>pca_selection_method: "forced"</code> (character scalar: <code>"forced"</code> | <code>"elbow"</code> | <code>"technical_noise"</code>): see the <code>PCA_SELECTION_METHOD</code> parameter in <code>vignette("norm_clustering")</code>.</li>
<li>
<code>pca_forced_pcs: 15</code> (integer scalar): force this number of selected PCs (if <code>pca_selection_method</code> is <code>"forced"</code>).</li>
<li>
<code>tsne_perp: 20</code> (numeric scalar): t-SNE perplexity.</li>
<li>
<code>tsne_max_iter: 1000</code> (positive integer scalar): a maximum number of t-SNE iterations.</li>
</ul>
</li>
<li>
<code>rescaling</code>: scale counts so that the average count within each batch is the same for each gene (<code><a href="https://rdrr.io/pkg/batchelor/man/rescaleBatches.html" class="external-link">batchelor::rescaleBatches()</a></code>).
<ul>
<li><code>pca_selection_method: "forced"</code></li>
<li><code>pca_forced_pcs: 15</code></li>
<li><code>tsne_perp: 20</code></li>
<li><code>tsne_max_iter: 1000</code></li>
<li>
<code>integration_params</code>: directly passed to <code><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">batchelor::RescaleParam()</a></code> used for <code><a href="https://rdrr.io/pkg/batchelor/man/rescaleBatches.html" class="external-link">batchelor::rescaleBatches()</a></code>. Default values from this function are used by default.
<ul>
<li>
<code>log.base: 2</code> (numeric scalar): a base of the log-transformation.</li>
<li>
<code>pseudo.count: 1</code> (numeric scalar): a pseudo-count used for the log-transformation.</li>
</ul>
</li>
</ul>
</li>
<li>
<code>regression</code>: fit a linear model to each gene regress out uninteresting factors of variation, returning a matrix of residuals (<code><a href="https://rdrr.io/pkg/batchelor/man/regressBatches.html" class="external-link">batchelor::regressBatches()</a></code>).
<ul>
<li>
<code>pca_selection_method: "corrected"</code>: in this integration method, <code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchPCA.html" class="external-link">batchelor::multiBatchPCA()</a></code> is used, and all PCs (50 by default) from this method will be used if <code>"corrected"</code> is specified. Otherwise any other PCA selection method can be used.</li>
<li><code>pca_forced_pcs: 15</code></li>
<li><code>tsne_perp: 20</code></li>
<li><code>tsne_max_iter: 1000</code></li>
<li>
<code>integration_params</code>: directly passed to <code><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">batchelor::RegressParam()</a></code> used for <code><a href="https://rdrr.io/pkg/batchelor/man/regressBatches.html" class="external-link">batchelor::regressBatches()</a></code>. Default values from this function are used by default.
<ul>
<li>
<code>d: 50</code> (integer scalar): a number of PCs to compute in <code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchPCA.html" class="external-link">batchelor::multiBatchPCA()</a></code>.</li>
</ul>
</li>
</ul>
</li>
<li>
<code>mnn</code>: fast mutual nearest neighbors (<code><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html" class="external-link">batchelor::fastMNN()</a></code>).
<ul>
<li><code>pca_selection_method: "corrected"</code></li>
<li><code>pca_forced_pcs: 15</code></li>
<li><code>tsne_perp: 20</code></li>
<li><code>tsne_max_iter: 1000</code></li>
<li>
<code>integration_params</code>: directly passed to <code><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">batchelor::FastMnnParam()</a></code> used for <code><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html" class="external-link">batchelor::fastMNN()</a></code>. Default values from this function are used by default.
<ul>
<li>
<code>k: 20</code> (integer scalar): a number of nearest neighbors to consider when identifying MNNs.</li>
<li>
<code>prop.k: null</code> (numeric scalar): a proportion of cells in each dataset to use for mutual nearest neighbor searching. If set, the number of nearest neighbors used for the MNN search in each batch is redefined as <code>max(k, prop.k*N)</code> where <code>N</code> is the number of cells in that batch.</li>
<li>
<code>cos.norm: True</code> (logical scalar): if <code>True</code>, cosine normalization is performed on the input data prior to PCA.</li>
<li>
<code>ndist: 3</code> (numeric scalar): a threshold beyond which neighbours are to be ignored when computing correction vectors. Each threshold is defined as a multiple of the number of median distances.</li>
<li><code>d: 50</code></li>
<li>
<code>merge.order: null</code> (integer vector, list of lists, or <code>null</code>): a merge order of single-samples. Alternatively, a list of lists representing a tree structure specifying a hierarchical merge order (not tested, see <code><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html" class="external-link">batchelor::fastMNN()</a></code> for more details). <code>auto.merge: True</code> (logical scalar): if <code>True</code>, automatically identify the “best” merge order.</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="integration-diagnostics">Integration diagnostics<a class="anchor" aria-label="anchor" href="#integration-diagnostics"></a>
</h3>
<div class="section level4">
<h4 id="fast-mnn-clustering">Fast MNN clustering<a class="anchor" aria-label="anchor" href="#fast-mnn-clustering"></a>
</h4>
<p>Used to obtain basic <a href="https://bioconductor.org/books/release/OSCA/integrating-datasets.html#correction-diagnostics" class="external-link">integration diagnostics</a>. See <code><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">scran::buildSNNGraph()</a></code> and <code><a href="https://rdrr.io/pkg/bluster/man/makeSNNGraph.html" class="external-link">bluster::makeSNNGraph()</a></code> for more details.</p>
<hr>
<p><code>INTEGRATION_SNN_K: 10</code></p>
<p><strong>Type:</strong> integer scalar</p>
<p>A number of nearest neighbors to consider during graph construction.</p>
<hr>
<p><code>INTEGRATION_SNN_TYPE: "rank"</code></p>
<p><strong>Type:</strong> character scalar (<code>"rank"</code> | <code>"number"</code> | <code>"jaccard"</code>)</p>
<p>A type of weighting scheme to use for shared neighbors.</p>
<hr>
<p><code>INTEGRATION_SNN_CLUSTERING_METHOD: "walktrap"</code></p>
<p><strong>Type:</strong> character scalar (<code>"walktrap"</code> | <code>"louvain"</code>)</p>
<p>A type of MNN clustering algorithm. See <code><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html" class="external-link">igraph::cluster_walktrap()</a></code> and <code><a href="https://rdrr.io/pkg/igraph/man/cluster_louvain.html" class="external-link">igraph::cluster_louvain()</a></code> for more details.</p>
</div>
</div>
<div class="section level3">
<h3 id="input-files">Input files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h3>
<p><code>SELECTED_MARKERS_INT_FILE: !code system.file("extdata", "selected_markers.csv", package = "scdrake", mustWork = TRUE)</code></p>
<p><strong>Type:</strong> character scalar or <code>null</code></p>
<p>Similar to the <code>SELECTED_MARKERS_FILE</code> parameter in <code>vignette("norm_clustering")</code>.</p>
<hr>
<p><code>INTEGRATION_REPORT_RMD_FILE: "Rmd/integration/01_integration.Rmd"</code></p>
<p><strong>Type:</strong> character scalar</p>
<p>A path to RMarkdown file used for HTML report of this pipeline stage.</p>
</div>
<div class="section level3">
<h3 id="output-files">Output files<a class="anchor" aria-label="anchor" href="#output-files"></a>
</h3>
<p><code>INTEGRATION_BASE_OUT_DIR: "01_integration"</code></p>
<p><strong>Type:</strong> character scalar</p>
<p>A path to base output directory for this stage. It will be created under <code>BASE_OUT_DIR</code> specified in <code>00_main.yaml</code> config.</p>
<hr>
<div class="sourceCode" id="cb4"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_SELECTED_MARKERS_OUT_DIR</span><span class="kw">:</span><span class="at"> </span><span class="st">"selected_markers"</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_REPORT_HTML_FILE</span><span class="kw">:</span><span class="at"> </span><span class="st">"01_integration.html"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>Names of files and directories created under <code>INTEGRATION_BASE_OUT_DIR</code>. Subdirectories are not allowed.</p>
</div>
<div class="section level3">
<h3 id="html-output-parameters">HTML output parameters<a class="anchor" aria-label="anchor" href="#html-output-parameters"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_KNITR_MESSAGE</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_KNITR_WARNING</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_KNITR_ECHO</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>These are passed to <code><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">knitr::opts_chunk()</a></code> and used for rendering of stage’s HTML report.</p>
</div>
</div>
<div class="section level2">
<h2 id="targets">Targets<a class="anchor" aria-label="anchor" href="#targets"></a>
</h2>
<p>Here you can find description of the most important targets for this stage. However, for a full overview, you have to inspect the <a href="https://github.com/bioinfocz/scdrake/blob/main/R/plans_integration.R" class="external-link">source code</a> of the <code><a href="../reference/get_subplan_integration.html">get_integration_subplan()</a></code> function.</p>
<div class="section level3">
<h3 id="singlecellexperiment-objects">
<code>SingleCellExperiment</code> objects<a class="anchor" aria-label="anchor" href="#singlecellexperiment-objects"></a>
</h3>
<p><code>sce_int_import</code>: list of imported single-sample SCE objects, as defined in the <code>INTEGRATION_SOURCES</code> parameter. Because <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cannot watch for changes in a cache, this target is always rerun. Also, some constraints are checked (common normalization method and HVG metrics).</p>
<hr>
<p><code>sce_int_raw_snn_clustering</code>: <code>sce_int_import</code> with computed fast SNN clustering for each single-sample, used for integration diagnostics.</p>
<hr>
<p><code>sce_int_processed</code>: <code>sce_int_import</code> subsetted to common colData, and features (rows) and their corresponding metadata (<code>hvg_ids</code>, <code>hvg_metric_fit</code>).</p>
<hr>
<p><code>sce_int_multibatchnorm</code>: <code>sce_int_processed</code> with SCE objects normalized for inter-batch sequencing depth (<code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchNorm.html" class="external-link">batchelor::multiBatchNorm()</a></code>).</p>
<hr>
<p><code>sce_int_df</code>: a tibble with integrated SCE objects. Each row is one method defined in the <code>INTEGRATION_METHODS</code> parameter and either with or without removed cell cycle-related genes from HVGs. See <code><a href="../reference/sce_int_df_fn.html">?sce_int_df_fn</a></code> for information about what is added or modified in <code>metadata()</code> of each integrated SCE object.</p>
<hr>
<p><code>sce_int_pca_df</code>: <code>sce_int_df</code> with computed PCA and selected number of PCs. Also includes PC selection statistics and plot.</p>
<hr>
<p><code>sce_int_clustering_df</code>: <code>sce_int_pca_df</code> with computed MNN clustering used for integration diagnostics.</p>
<hr>
<p><code>sce_int_dimred_df</code>: <code>sce_int_pca_df</code> with computed t-SNE and UMAP dimensionality reductions.</p>
</div>
<div class="section level3">
<h3 id="highly-variable-genes-hvgs-selection">Highly variable genes (HVGs) selection<a class="anchor" aria-label="anchor" href="#highly-variable-genes-hvgs-selection"></a>
</h3>
<p><code>hvg_int</code>, <code>hvg_int_with_cc</code>: lists of HVGs and selection parameters. The latter is not <code>NULL</code> when any of the single-samples defined in the <code>INTEGRATION_SOURCES</code> parameter has <code>hvg_rm_cc_genes</code> set to <code>True</code>.</p>
<hr>
<p><code>hvg_plots_int_df</code>: a tibble with HVG diagnostic plots.</p>
</div>
<div class="section level3">
<h3 id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a>
</h3>
<p><code>sce_int_dimred_plots_df</code>: a tibble with dimensionality reduction plots.</p>
<div class="section level4">
<h4 id="selected-markers">Selected markers<a class="anchor" aria-label="anchor" href="#selected-markers"></a>
</h4>
<p><code>selected_markers_plots_int_df</code>: a tibble with plots of selected markers, similar to <code>selected_markers_plots</code> in the <code>02_norm_clustering</code> stage of the single-sample pipeline.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jiri Novotny, Jan Kubovciak.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.9000.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '937cedcd467c204b19fc155c5f35019b',
    indexName: 'scdrake',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
