<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="02_norm_clustering stage of the single-sample pipeline.">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/5/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Normalization and clustering stage (`02_norm_clustering`) • scdrake</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Normalization and clustering stage (`02_norm_clustering`)">
<meta property="og:description" content="02_norm_clustering stage of the single-sample pipeline.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/5/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/5/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="../darkswitch.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scdrake</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scdrake.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/scdrake_integration.html">Integration pipeline guide</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/pipeline_overview.html">Pipeline overview</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/scdrake_spatial.html">Spatial extension</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/scdrake_docker.html">Using the Docker image</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_integration.html">02 Integration pipeline guide</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_advanced.html">Advanced topics</a></li>
    <li><a class="dropdown-item" href="../articles/interaction_analysis.html">Interaction analysis</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_extend.html">Extending the pipeline</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_development.html">Developing scdrake</a></li>
    <li><a class="dropdown-item" href="../articles/drake_basics.html">drake basics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>General information</h6></li>
    <li><a class="dropdown-item" href="../articles/pipeline_overview.html">Pipeline overview</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_cli.html">Command line interface</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_config.html">scdrake config internals</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_envvars.html">Environment variables</a></li>
    <li><a class="dropdown-item" href="../articles/scdrake_spatial.html">Spatial extension</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>General configs</h6></li>
    <li><a class="dropdown-item" href="../articles/config_pipeline.html">Pipeline config</a></li>
    <li><a class="dropdown-item" href="../articles/config_main.html">Main config</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Single-sample pipeline stages</h6></li>
    <li><a class="dropdown-item" href="../articles/stage_input_qc.html">Input and quality control stage (`01_input_qc`)</a></li>
    <li><a class="dropdown-item" href="../articles/stage_norm_clustering.html">Normalization and clustering stage (`02_norm_clustering`)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Integration pipeline stages</h6></li>
    <li><a class="dropdown-item" href="../articles/stage_integration.html">Integration stage (`01_integration`)</a></li>
    <li><a class="dropdown-item" href="../articles/stage_int_clustering.html">Post-integration clustering stage (`02_int_clustering`)</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Common stages</h6></li>
    <li><a class="dropdown-item" href="../articles/stage_cluster_markers.html">Cluster markers stage (`cluster_markers`)</a></li>
    <li><a class="dropdown-item" href="../articles/stage_contrasts.html">Contrasts stage (`contrasts`)</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bioinfocz/scdrake" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>


        <li>
          <a class="external-link nav-link" id="css-toggle-btn" aria-label="github">
            <span class="fas fa fas fa-adjust fa-lg"></span>
          </a>
        </li>



      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Normalization and clustering stage (`02_norm_clustering`)</h1>
            
            <h4 data-toc-skip class="date"><sup>Document generated:
2025-08-26 19:26:47 UTC+0000</sup></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bioinfocz/scdrake/blob/master/vignettes/stage_norm_clustering.Rmd" class="external-link"><code>vignettes/stage_norm_clustering.Rmd</code></a></small>
      <div class="d-none name"><code>stage_norm_clustering.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2 tabset">
<h2 class="tabset" id="section"></h2>



<ul class="nav nav-tabs" id="section" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#overview" id="overview-tab" type="button" role="tab" aria-controls="overview" aria-selected="true" class="active nav-link">Overview</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#config-parameters" id="config-parameters-tab" type="button" role="tab" aria-controls="config-parameters" aria-selected="false" class="nav-link">Config parameters</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#outputs" id="outputs-tab" type="button" role="tab" aria-controls="outputs" aria-selected="false" class="nav-link">Outputs</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="overview" role="tabpanel" aria-labelledby="overview-tab">

<p>This stage build upon the <code>01_input_qc</code> stage and performs
perhaps the most important steps in scRNA-seq and sequence-based SRT
data analysis: normalization, highly and spatially variable gene
selection, dimensionality reduction, clustering, deconvolution, cell
type annotation, and visualization.</p>
<p>⚙️ Config file:
<code>config/single_sample/02_norm_clustering.yaml</code></p>
<p>📋 HTML report target (in <code>config/pipeline.yaml</code>):
<code>DRAKE_TARGETS: ["report_norm_clustering", "report_norm_clustering_simple"]</code></p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/single_sample/pbmc1k/02_norm_clustering/02_norm_clustering.html" class="external-link">📜
Example report</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/single_sample_pbmc1k_full/02_norm_clustering.default.yaml" class="external-link">used
config</a>)</p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/single_sample/pbmc1k/02_norm_clustering/02_norm_clustering_simple.html" class="external-link">📜
Example simplified report</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/single_sample_pbmc1k_full/02_norm_clustering.default.yaml" class="external-link">used
config</a>)</p>
<p>🪜 Structure</p>
<ul>
<li>Calculation of cell cycle score and assignment of cells to phases
(G1, G2M, S) using the list of cell cycle genes in
<code><a href="https://satijalab.org/seurat/reference/cc.genes.updated.2019.html" class="external-link">Seurat::cc.genes.updated.2019</a></code> and
<code><a href="https://satijalab.org/seurat/reference/CellCycleScoring.html" class="external-link">Seurat::CellCycleScoring()</a></code> function</li>
<li>
<a href="https://bioconductor.org/books/3.15/OSCA.basic/normalization.html" class="external-link">Normalization</a>
of UMI counts by either:
<ul>
<li>
<a href="https://bioconductor.org/books/3.15/OSCA.basic/normalization.html#normalization-by-deconvolution" class="external-link">Normalization
by deconvolution</a> implemented in the <a href="https://bioconductor.org/packages/3.15/bioc/html/scran.html" class="external-link">scran</a>
package (<code><a href="https://rdrr.io/pkg/scuttle/man/computePooledFactors.html" class="external-link">scuttle::computePooledFactors()</a></code>,
<code><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">scuttle::logNormCounts()</a></code>)</li>
<li>Regularized negative binomial regression to normalize UMI count data
implemented in the <a href="https://github.com/ChristophH/sctransform" class="external-link">sctransform</a> package
and wrapped by the <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>
package. This method also allows to regress out confounding variables
such as cell cycle score or percentage of mitochondrial genes
expression. Note that for cell cycle score, a better method seems to be
removal of CC-related genes prior to detection of highly variable genes
(HVGs, see below). <code>sctransform</code> also returns a specified
number of HVGs.</li>
</ul>
</li>
<li>Selection of <a href="https://bioconductor.org/books/3.15/OSCA.basic/feature-selection.html" class="external-link">highly
variable genes</a> (HVGs):
<ul>
<li>HVGs are used downstream for dimensionality reduction and
clustering</li>
<li>Three metrics for HVGs are available:
<ul>
<li><a href="https://bioconductor.org/books/3.15/OSCA.basic/feature-selection.html#quantifying-per-gene-variation" class="external-link">Gene
variance</a></li>
<li>
<a href="http://bioconductor.org/books/3.15/OSCA.advanced/more-hvgs.html#using-the-coefficient-of-variation" class="external-link">Coefficient
of variation</a> (CV2)</li>
<li>HVGs obtained from the <code>sctransform</code> normalization</li>
</ul>
</li>
<li>Based on the selected one from the first two metrics above, HVGs are
selected either by:
<ul>
<li>Top N genes based on a metric (e.g. top 1000 genes with the highest
variance)</li>
<li>Significance (FDR) threshold</li>
<li>Threshold on a metric value</li>
</ul>
</li>
<li>Prior to HVG selection, cell cycle-related genes can be removed.
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> is using a method based on the percentage of
variance explained by the cell cycle phase in the expression profile for
each gene (<a href="https://bioconductor.org/books/3.15/OSCA.advanced/cell-cycle-assignment.html#removing-cell-cycle-related-genes" class="external-link">details</a>)</li>
</ul>
</li>
<li>Selection of <a href="https://lmweber.org/OSTA/pages/ind-feature-selection-testing.html" class="external-link">spatialy
variable genes</a> (SVGs):
<ul>
<li>SVGs are combined with HVGs so that both spatial patterns and
overall expression variability are leveraged in dimensionality reduction
and clustering</li>
</ul>
</li>
<li>Removal of <a href="https://bioconductor.org/books/3.15/OSCA.advanced/doublet-detection.html" class="external-link">cell
doublets</a> (<code><a href="https://plger.github.io/scDblFinder/reference/computeDoubletDensity.html" class="external-link">scDblFinder::computeDoubletDensity()</a></code>).</li>
<li>
<a href="https://bioconductor.org/books/3.15/OSCA.basic/dimensionality-reduction.html#principal-components-analysis" class="external-link">Calculation
of PCA</a> and <a href="https://bioconductor.org/books/3.15/OSCA.basic/dimensionality-reduction.html#choosing-the-number-of-pcs" class="external-link">selection
of number of principal components</a> (PCs) which will be used
downstream. There are three methods to select a proper number of PCs:
<ul>
<li>Using the <a href="http://bioconductor.org/books/3.15/OSCA.advanced/dimensionality-reduction-redux.html#more-choices-for-the-number-of-pcs" class="external-link">elbow
point</a>
</li>
<li>Using the <a href="http://bioconductor.org/books/3.15/OSCA.advanced/dimensionality-reduction-redux.html#using-the-technical-noise" class="external-link">technical
noise</a>
</li>
<li>Using a fixed number of PCs (<code>forced</code>)</li>
</ul>
</li>
<li>
<a href="https://bioconductor.org/books/3.15/OSCA.basic/clustering.html" class="external-link">Cell
clustering</a> using several algorithms:
<ul>
<li>
<a href="https://bioconductor.org/books/3.15/OSCA.basic/clustering.html#clustering-graph" class="external-link">Graph-based
clustering</a> (mutual nearest neighbors) using Leiden, Louvain or
Walktrap algorithms. To control for the number of resulting clusters
(granularity), multiple resolutions can be specified for the former two
algorithms.</li>
<li>
<a href="https://bioconductor.org/books/3.15/OSCA.basic/clustering.html#vector-quantization-with-k-means" class="external-link">K-means
clustering</a> using both defined numbers of K and best K selection</li>
<li>
<a href="https://bioconductor.org/packages/3.15/bioc/html/SC3.html" class="external-link">Single-Cell
Consensus Clustering</a> (SC3) with defined numbers of clusters</li>
</ul>
</li>
<li>Reference-based cell type annotation using <a href="https://github.com/SingleR-inc/SingleR" class="external-link">SingleR</a>.
More details are given in the accompanying <a href="http://bioconductor.org/books/3.15/SingleRBook/" class="external-link">book</a> and the
underlying method is described <a href="http://bioconductor.org/books/3.15/SingleRBook/introduction.html#method-description" class="external-link">here</a>.
<ul>
<li>Reference datasets are taken by default from the
<a href="https://github.com/LTLA/celldex" class="external-link">celldex</a> package.</li>
<li>Note that the reference datasets included in the
<a href="https://github.com/LTLA/celldex" class="external-link">celldex</a> package don’t contain every possible cell type
and the annotation may bring inaccurate labels for some datasets. Thus,
it is also possible to supply <a href="https://github.com/SingleR-inc/SingleR" class="external-link">SingleR</a> a custom reference
datasets in the form of <code>{SingleCellExperiment}</code> or
<a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a> object.</li>
<li>Plots of annotation diagnostics:
<ul>
<li>Score heatmaps show distribution of predicted cell types in computed
clusters (if any), along with per-cell annotation scores,</li>
<li>Marker heatmaps show genes that are markers for a given cell type in
both the reference and current datasets,, i.e. those markers have driven
the decision to label cells by the chosen cell type,</li>
<li>Delta scores show poor-quality or ambiguous assignments based on the
per-cell ‘delta’, i.e., the difference between, the score for the
assigned label and the median across all labels for each cell., See <a href="https://bioconductor.org/books/3.15/SingleRBook/annotation-diagnostics.html#based-on-the-deltas-across-cells" class="external-link">OSCA</a>
for more details</li>
</ul>
</li>
</ul>
</li>
<li>Manual cell type annotation based on signatures score of specified
cell type markers</li>
<li>Assignment of cells to predefined groupings based on existing data.
That means you can reuse an assignment of cells from e.g. clustering,
and rename or merge its levels (cluster numbers).
<ul>
<li>Example: you identify a biological meaning of clusters computed by
k-means with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>k</mi><mo>=</mo><mn>3</mn></mrow><annotation encoding="application/x-tex">k = 3</annotation></semantics></math>.
You can use this information, and rename and merge the clusters, such
that:
<ul>
<li>Cluster 1 -&gt; T cells</li>
<li>Cluster 2 -&gt; B cells</li>
<li>Cluster 3 -&gt; T cells</li>
</ul>
</li>
<li>Then you can use the new cell grouping for e.g. plotting of reduced
dimensions or calculation of cluster markers.</li>
</ul>
</li>
<li>
<a href="https://bioconductor.org/books/3.15/OSCA.basic/dimensionality-reduction.html" class="external-link">Dimensionality
reduction</a>:
<ul>
<li><a href="https://bioconductor.org/books/3.15/OSCA.basic/dimensionality-reduction.html#principal-components-analysis" class="external-link">PCA</a></li>
<li><a href="https://bioconductor.org/books/3.15/OSCA.basic/dimensionality-reduction.html#t-stochastic-neighbor-embedding" class="external-link">t-SNE</a></li>
<li><a href="https://bioconductor.org/books/3.15/OSCA.basic/dimensionality-reduction.html#uniform-manifold-approximation-and-projection" class="external-link">UMAP</a></li>
</ul>
</li>
<li>
<a href="https://lmweber.org/OSTA/pages/seq-deconvolution.html" class="external-link">Spot
deconvolution</a>:
<ul>
<li><a href="https://github.com/YMa-lab/CARD" class="external-link">CARD</a></li>
<li><a href="https://github.com/ggrajeda/spacexr" class="external-link">RCTD</a></li>
</ul>
</li>
<li>Expression plots of selected groups of genes (markers) in reduced
dimensions</li>
<li>A simplified report for non-technical audience which contains only
dimred plots, selected markers and cell annotation results</li>
</ul>
</div>
<div class="tab-pane" id="config-parameters" role="tabpanel" aria-labelledby="config-parameters-tab">

<p>Config for this stage is stored in the
<code>config/single_sample/02_norm_clustering.yaml</code> file.
Directory with this file is read from
<code>SCDRAKE_SINGLE_SAMPLE_CONFIG_DIR</code> environment variable upon
<a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> load or attach, and saved as
<code>scdrake_single_sample_config_dir</code> option. This option is
used as the default argument value in several <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>
functions.</p>
<hr>
<div class="section level4">
<h4 id="spatial-extension">Spatial extension<a class="anchor" aria-label="anchor" href="#spatial-extension"></a>
</h4>
<div class="sourceCode" id="cb1"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">SPATIAL</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>True</code>, pipeline enables spatial extension.</p>
<p>This option enables pseudo tissue visualization and other
visualization options for marker-based annotation.</p>
<hr>
</div>
<div class="section level4">
<h4 id="normalization-parameters">Normalization parameters<a class="anchor" aria-label="anchor" href="#normalization-parameters"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">NORMALIZATION_TYPE</span><span class="kw">:</span><span class="at"> </span><span class="st">"scran"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar (<code>"scran"</code> |
<code>"sctransform"</code> | <code>"none"</code>)</p>
<p>A normalization type:</p>
<ul>
<li>
<code>"scran"</code>: normalization by deconvolution.</li>
<li>
<code>"sctransform"</code>: regularized negative binomial
regression. The returned counts in log1p (natural log) scale are
converted to log2. Note: <code>SingleCellExperiment</code> normalized by
this method cannot be used for integration.</li>
<li>
<code>"none"</code>: skip normalization. This should be used only in
case you are importing a <code>SingleCellExperiment</code> object in the
<code>01_input_qc</code> stage and this object was normalized before
within this pipeline, i.e. you are importing a modified
<code>sce_norm</code> target or one of its downstream, dependent
<code>SingleCellExperiment</code> targets.
<ul>
<li>For example, you can take <code>sce_final_norm_clustering</code>,
subset cells to e.g. three clusters, save the result to Rds file, and
run the pipeline again while importing this file in the
<code>01_input_qc</code> stage. Note that in that case you should
disable empty droplets removal, and cell and gene filtering, as data are
already of high quality. Overall, this is an example of reclustering,
which is otherwise hard to achieve within the pipeline.</li>
</ul>
</li>
</ul>
<p>See <code><a href="../articles/pipeline_overview.html">vignette("pipeline_overview")</a></code> for more details about
normalization methods.</p>
<div class="section level5">
<h5 id="scran-normalization-parameters">
<code>scran</code> normalization parameters<a class="anchor" aria-label="anchor" href="#scran-normalization-parameters"></a>
</h5>
<p>These parameters are used when <code>NORMALIZATION_TYPE</code> is
<code>"scran"</code>.</p>
<hr>
<div class="sourceCode" id="cb3"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">SCRAN_USE_QUICKCLUSTER</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>Whether to cluster cells prior to normalization. See
<code><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">?scran::quickCluster</a></code> for more details.</p>
<hr>
<div class="sourceCode" id="cb4"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">SCRAN_QUICKCLUSTER_METHOD</span><span class="kw">:</span><span class="at"> </span><span class="st">"igraph"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar (<code>"igraph"</code> |
<code>"hclust"</code>)</p>
<p>A clustering method: <code>"igraph"</code> uses graph-based
clustering, while <code>"hclust"</code> uses hierarchical clustering.
See <code><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">?scran::quickCluster</a></code> for more details.</p>
</div>
<div class="section level5">
<h5 id="sctransform-normalization-parameters">
<code>sctransform</code> normalization parameters<a class="anchor" aria-label="anchor" href="#sctransform-normalization-parameters"></a>
</h5>
<p>These parameters are used when <code>NORMALIZATION_TYPE</code> is
<code>"sctransform"</code>.</p>
<hr>
<div class="sourceCode" id="cb5"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">SCT_VARS_TO_REGRESS</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span></code></pre></div>
<p><strong>Type:</strong> list of character scalar or
<code>null</code></p>
<p>Batch variables to regress out during the <code>SCTransform</code>
normalization. You can regress out variables such as
<code>["percent.mt", "nFeature_RNA", "nCount_RNA"]</code> or cell cycle
using <code>["S.Score", "G2M.Score"]</code> or
<code>["CC.Difference"]</code>.</p>
<p>However, for correction of cell cycle effect, a better method is the
removal of cell cycle-related genes during selection of highly variable
genes (HVGs). This can be set by <a href="#highly-variable-genes-hvgs-selection">HVG selection
parameters</a>, and the approach is explained in more details in
<code><a href="../articles/pipeline_overview.html">vignette("pipeline_overview")</a></code>.</p>
<hr>
<div class="sourceCode" id="cb6"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">SCT_N_HVG</span><span class="kw">:</span><span class="at"> </span><span class="dv">3000</span></span></code></pre></div>
<p><strong>Type:</strong> integer scalar</p>
<p>A number of top HVGs to return from the <code>SCTransform</code>
normalization.</p>
</div>
</div>
<div class="section level4">
<h4 id="highly-variable-genes-hvgs-selection">Highly variable genes (HVGs) selection<a class="anchor" aria-label="anchor" href="#highly-variable-genes-hvgs-selection"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">HVG_METRIC</span><span class="kw">:</span><span class="at"> </span><span class="st">"gene_var"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar (<code>"gene_var"</code> |
<code>"gene_cv2"</code> | <code>"sctransform"</code>)</p>
<p>A metric used to find HVGs. See <a href="https://bioconductor.org/books/3.15/OSCA.basic/feature-selection.html" class="external-link uri">https://bioconductor.org/books/3.15/OSCA.basic/feature-selection.html</a>
for more details.</p>
<ul>
<li>
<code>"gene_var"</code>: variance of the log-counts
(<code><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">scran::modelGeneVar()</a></code>).</li>
<li>
<code>"gene_cv2"</code>: coefficient of variation
(<code><a href="https://rdrr.io/pkg/scran/man/modelGeneCV2.html" class="external-link">scran::modelGeneCV2()</a></code>).</li>
<li>
<code>"sctransform"</code>: use HVGs selected by
<code>SCTransform</code>. Only relevant when the
<code>NORMALIZATION_TYPE</code> parameter is <code>"sctransform"</code>.
You can adjust number of these HVGs by the <code>SCT_N_HVG</code>
parameter.</li>
</ul>
<hr>
<div class="sourceCode" id="cb8"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">HVG_SELECTION</span><span class="kw">:</span><span class="at"> </span><span class="st">"top"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar (<code>"top"</code> |
<code>"significance"</code> | <code>"threshold"</code>)</p>
<p>A HVG selection strategy. Only relevant when the
<code>HVG_METRIC</code> parameter is <code>"gene_var"</code> or
<code>"gene_cv2"</code>.</p>
<ul>
<li>
<code>"top"</code>: take top X genes according to the metric (<a href="https://bioconductor.org/books/3.15/OSCA.basic/feature-selection.html#hvg-selection" class="external-link">details</a>).</li>
<li>
<code>"significance"</code>: use a FDR threshold (<a href="https://bioconductor.org/books/3.15/OSCA.advanced/more-hvgs.html#more-hvg-selection-strategies" class="external-link">details</a>).</li>
<li>
<code>"threshold"</code>: use a threshold on minimum value of the
metric (<a href="https://bioconductor.org/books/3.15/OSCA.advanced/more-hvgs.html#more-hvg-selection-strategies" class="external-link">details</a>).</li>
</ul>
<p>For <code>"top"</code> and <code>"threshold"</code>, <code>bio</code>
and <code>ratio</code> columns are used for <code>HVG_METRIC</code> of
<code>"gene_var"</code> and <code>"gene_cv2"</code>, respectively. These
columns are present in the DataFrame returned from the underlying method
(<code><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">scran::modelGeneVar()</a></code> or
<code><a href="https://rdrr.io/pkg/scran/man/modelGeneCV2.html" class="external-link">scran::modelGeneCV2()</a></code>).</p>
<hr>
<div class="sourceCode" id="cb9"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">HVG_SELECTION_VALUE</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span></code></pre></div>
<p><strong>Type:</strong> depends on value of the
<code>HVG_SELECTION</code> parameter:</p>
<ul>
<li>A positive integer scalar for
<code>HVG_SELECTION: "top"</code>.</li>
<li>A numeric scalar
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>&lt;</mo><mn>0</mn><mo>;</mo><mn>1</mn><mo>&gt;</mo></mrow><annotation encoding="application/x-tex">&lt;0; 1&gt;</annotation></semantics></math>
for <code>HVG_SELECTION: "significance"</code>.</li>
<li>A numeric scalar for <code>HVG_SELECTION: "threshold"</code>.</li>
</ul>
<p>A threshold value for HVG selection. The following parameter
combinations are recommended defaults:</p>
<ul>
<li>Top 1000 genes by a HVG metric.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">HVG_SELECTION</span><span class="kw">:</span><span class="at"> </span><span class="st">"top"</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">HVG_SELECTION_VALUE</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span></code></pre></div>
<ul>
<li>Genes with a HVG metric significance (FDR) lesser than
<code>0.05</code>.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">HVG_SELECTION</span><span class="kw">:</span><span class="at"> </span><span class="st">"significance"</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">HVG_SELECTION_VALUE</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.05</span></span></code></pre></div>
<ul>
<li>Genes with a HVG metric value greater than <code>0</code> or
<code>1</code>.</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">HVG_METRIC</span><span class="kw">:</span><span class="at"> </span><span class="st">"gene_var"</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="fu">HVG_SELECTION</span><span class="kw">:</span><span class="at"> </span><span class="st">"threshold"</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="fu">HVG_SELECTION_VALUE</span><span class="kw">:</span><span class="at"> </span><span class="dv">0</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="fu">HVG_METRIC</span><span class="kw">:</span><span class="at"> </span><span class="st">"gene_cv2"</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="fu">HVG_SELECTION</span><span class="kw">:</span><span class="at"> </span><span class="st">"threshold"</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="fu">HVG_SELECTION_VALUE</span><span class="kw">:</span><span class="at"> </span><span class="dv">1</span></span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="spatialy-variable-genes-svgs-selection">Spatialy variable genes (SVGs) selection<a class="anchor" aria-label="anchor" href="#spatialy-variable-genes-svgs-selection"></a>
</h4>
<div class="sourceCode" id="cb14"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">SVG</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p>When the SVG option is enabled, spatially variable genes are
automatically identified alongside HVGs. This is done using the
nnSVG::nnSVG() function with the “MoransI” selection method. Moran’s I
is a statistic that measures spatial autocorrelation—it detects whether
expression values for a gene are more similar in nearby spots than would
be expected by chance.</p>
<p>The resulting union of unique SVGs and HVGs is carried forward for
downstream analyses, ensuring that both spatial structure and overall
variability are preserved in dimensionality reduction and clustering
steps.</p>
<p>For methodological details, see: <a href="https://www.biorxiv.org/content/10.1101/2021.08.27.457741v1" class="external-link uri">https://www.biorxiv.org/content/10.1101/2021.08.27.457741v1</a>.</p>
<hr>
</div>
<div class="section level4">
<h4 id="removal-of-cell-cycle-related-genes-prior-to-hvg-selection">Removal of cell cycle-related genes prior to HVG selection<a class="anchor" aria-label="anchor" href="#removal-of-cell-cycle-related-genes-prior-to-hvg-selection"></a>
</h4>
<div class="sourceCode" id="cb15"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">HVG_RM_CC_GENES</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>Whether to apply a correction for cell cycle prior to HVG selection.
We implement a cell cycle-related genes removal strategy described <a href="https://bioconductor.org/books/3.15/OSCA.advanced/cell-cycle-assignment.html#removing-cell-cycle-related-genes" class="external-link">here</a>.
It is based on the percentage of variance explained by the cell cycle
phase in the expression profile for each gene.</p>
<p>Note that when <code>SCTransform</code> is used for normalization
including cell cycle variables to regress out, than the results of this
strategy remain unclear.</p>
<hr>
<div class="sourceCode" id="cb16"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="fu">HVG_CC_GENES_VAR_EXPL_THRESHOLD</span><span class="kw">:</span><span class="at"> </span><span class="dv">5</span></span></code></pre></div>
<p><strong>Type:</strong> positive numeric scalar</p>
<p>A threshold on gene cell cycle variance explained, i.e. genes with
var. expl. greater than this threshold will be marked as cell
cycle-related and removed <em>prior</em> to selection of HVGs.</p>
</div>
<div class="section level4">
<h4 id="doublet-detection-and-filtering">Doublet detection and filtering<a class="anchor" aria-label="anchor" href="#doublet-detection-and-filtering"></a>
</h4>
<p>Doublets are droplets in which multiple cells were captured.</p>
<hr>
<div class="sourceCode" id="cb17"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="fu">MAX_DOUBLET_SCORE</span><span class="kw">:</span><span class="at"> </span><span class="fl">3.5</span></span></code></pre></div>
<p><strong>Type:</strong> positive numeric scalar or
<code>null</code></p>
<p>A threshold for doublet score computed by
<code><a href="https://plger.github.io/scDblFinder/reference/computeDoubletDensity.html" class="external-link">scDblFinder::computeDoubletDensity()</a></code>. Cells with doublet
score greater than the threshold will be removed. If <code>null</code>,
doublets are not removed, but doublet score is still computed.</p>
</div>
<div class="section level4">
<h4 id="dimensionality-reduction">Dimensionality reduction<a class="anchor" aria-label="anchor" href="#dimensionality-reduction"></a>
</h4>
<div class="section level5">
<h5 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h5>
<div class="sourceCode" id="cb18"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">PCA_SELECTION_METHOD</span><span class="kw">:</span><span class="at"> </span><span class="st">"forced"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar (<code>"forced"</code> |
<code>"elbow"</code> | <code>"technical_noise"</code>)</p>
<p>A method to select first N principal components (PCs):</p>
<ul>
<li>
<code>"forced"</code>: use the number of PCs specified by the
<code>PCA_FORCED_PCS</code> parameter.</li>
<li>
<code>"elbow"</code>: use the <a href="https://bioconductor.org/books/3.15/OSCA.advanced/dimensionality-reduction-redux.html#using-the-elbow-point" class="external-link">elbow
point method</a>.</li>
<li>
<code>"technical_noise"</code>: use the <a href="https://bioconductor.org/books/3.15/OSCA.advanced/dimensionality-reduction-redux.html#using-the-technical-noise" class="external-link">technical
noise</a>.</li>
</ul>
<hr>
<div class="sourceCode" id="cb19"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">PCA_FORCED_PCS</span><span class="kw">:</span><span class="at"> </span><span class="dv">15</span></span></code></pre></div>
<p><strong>Type:</strong> positive integer scalar</p>
<p>If the <code>PCA_SELECTION_METHOD</code> parameter is
<code>"forced"</code>, use this number of first principal
components.</p>
</div>
<div class="section level5">
<h5 id="t-sne">t-SNE<a class="anchor" aria-label="anchor" href="#t-sne"></a>
</h5>
<p><a href="https://bioconductor.org/books/3.15/OSCA.basic/dimensionality-reduction.html#t-stochastic-neighbor-embedding" class="external-link">t-stochastic
neighbor embedding</a></p>
<p>See <code><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">?scater::runTSNE</a></code> for the underlying method.</p>
<hr>
<div class="sourceCode" id="cb20"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">TSNE_PERP</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span></code></pre></div>
<p><strong>Type:</strong> numeric scalar</p>
<p>t-SNE perplexity.</p>
<hr>
<div class="sourceCode" id="cb21"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">TSNE_MAX_ITER</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span></code></pre></div>
<p><strong>Type:</strong> positive integer scalar</p>
<p>A maximum number of t-SNE iterations.</p>
</div>
</div>
<div class="section level4">
<h4 id="clustering">Clustering<a class="anchor" aria-label="anchor" href="#clustering"></a>
</h4>
<div class="section level5">
<h5 id="graph-based-clustering">Graph-based clustering<a class="anchor" aria-label="anchor" href="#graph-based-clustering"></a>
</h5>
<p>Parameters for <a href="https://bioconductor.org/books/3.15/OSCA.basic/clustering.html#clustering-graph" class="external-link">graph-based
clustering</a>.</p>
<hr>
<div class="sourceCode" id="cb22"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="fu">CLUSTER_GRAPH_SNN_K</span><span class="kw">:</span><span class="at"> </span><span class="dv">10</span></span></code></pre></div>
<p><strong>Type:</strong> positive integer scalar</p>
<p>A number of shared nearest neighbors in graph used for community
detection algorithms (graph-based clustering). Passed as the
<code>k</code> parameter to <code><a href="https://rdrr.io/pkg/bluster/man/makeSNNGraph.html" class="external-link">bluster::makeSNNGraph()</a></code>. From
the function’s help page:</p>
<blockquote>
<p>The choice of k controls the connectivity of the graph and the
resolution of community detection algorithms. Smaller values of k will
generally yield smaller, finer clusters, while increasing k will
increase the connectivity of the graph and make it more difficult to
resolve different communities. The value of k can be roughly interpreted
as the anticipated size of the smallest subpopulation.</p>
</blockquote>
<hr>
<div class="sourceCode" id="cb23"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="fu">CLUSTER_GRAPH_SNN_TYPE</span><span class="kw">:</span><span class="at"> </span><span class="st">"rank"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar (<code>"rank"</code> |
<code>"number"</code> | <code>"jaccard"</code>)</p>
<p>A type of weighting scheme to use for shared neighbors. Passed as
<code>type</code> parameter to <code><a href="https://rdrr.io/pkg/bluster/man/makeSNNGraph.html" class="external-link">bluster::makeSNNGraph()</a></code> (see
the function’s help page for more information).</p>
<hr>
<div class="sourceCode" id="cb24"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="fu">CLUSTER_GRAPH_LEIDEN_ENABLED</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>True</code>, enable Leiden community detection
algorithm.</p>
<hr>
<div class="sourceCode" id="cb25"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="fu">CLUSTER_GRAPH_LEIDEN_RESOLUTIONS</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="fl">0.2</span><span class="kw">,</span><span class="at"> </span><span class="fl">0.4</span><span class="kw">,</span><span class="at"> </span><span class="fl">0.6</span><span class="kw">,</span><span class="at"> </span><span class="fl">0.8</span><span class="kw">,</span><span class="at"> </span><span class="fl">1.0</span><span class="kw">]</span></span></code></pre></div>
<p><strong>Type:</strong> numeric vector</p>
<p>Resolutions for Leiden clustering. Lower values result in more
coarse-grained clustering, while higher ones to more fine-grained
structures. Cell-cluster memberships will be named as
<code>cluster_graph_leiden_r{r}</code> (e.g. in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>
of a SCE object).</p>
<hr>
<div class="sourceCode" id="cb26"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="fu">CLUSTER_GRAPH_LOUVAIN_ENABLED</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>True</code>, enable Louvain community detection
algorithm.</p>
<hr>
<div class="sourceCode" id="cb27"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="fu">CLUSTER_GRAPH_LOUVAIN_RESOLUTIONS</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="fl">0.2</span><span class="kw">,</span><span class="at"> </span><span class="fl">0.4</span><span class="kw">,</span><span class="at"> </span><span class="fl">0.6</span><span class="kw">,</span><span class="at"> </span><span class="fl">0.8</span><span class="kw">,</span><span class="at"> </span><span class="fl">1.0</span><span class="kw">]</span></span></code></pre></div>
<p><strong>Type:</strong> numeric vector</p>
<p>Resolutions for Louvain clustering. Lower values result in more
coarse-grained clustering, while higher ones to more fine-grained
structures. Cell-cluster memberships will be named as
<code>cluster_graph_louvain_r{r}</code> (e.g. in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>
of a SCE object).</p>
<hr>
<div class="sourceCode" id="cb28"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="fu">CLUSTER_GRAPH_WALKTRAP_ENABLED</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>True</code>, enable Walktrap community detection
algorithm.</p>
</div>
<div class="section level5">
<h5 id="k-means-clustering">k-means clustering<a class="anchor" aria-label="anchor" href="#k-means-clustering"></a>
</h5>
<div class="sourceCode" id="cb29"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="fu">CLUSTER_KMEANS_K_ENABLED</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>True</code>, enable k-means clustering with selected numbers
of clusters (<code>k</code>s).</p>
<hr>
<div class="sourceCode" id="cb30"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="fu">CLUSTER_KMEANS_K</span><span class="kw">:</span><span class="at"> !code 3:6</span></span></code></pre></div>
<p><strong>Type:</strong> integer vector</p>
<p>A vector of number of clusters for which to compute k-means.</p>
<p>Note: <code>!code 3:6</code> will evaluate to
<code>c(3, 4, 5, 6)</code>.</p>
<hr>
<div class="sourceCode" id="cb31"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="fu">CLUSTER_KMEANS_KBEST_ENABLED</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>True</code>, enable k-means clustering with best
<code>k</code> based on gap statistics.</p>
</div>
<div class="section level5">
<h5 id="sc3-clustering">SC3 clustering<a class="anchor" aria-label="anchor" href="#sc3-clustering"></a>
</h5>
<div class="sourceCode" id="cb32"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="fu">CLUSTER_SC3_ENABLED</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>True</code>, enable SC3 with selected numbers of clusters
(<code>k</code>s).</p>
<hr>
<div class="sourceCode" id="cb33"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="fu">CLUSTER_SC3_K</span><span class="kw">:</span><span class="at"> !code 5:6</span></span></code></pre></div>
<p><strong>Type:</strong> integer vector</p>
<p>A vector of number of clusters for which to compute SC3
clustering.</p>
<p>Note: <code>!code 3:6</code> will evaluate to
<code>c(3, 4, 5, 6)</code>.</p>
<hr>
<div class="sourceCode" id="cb34"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="fu">CLUSTER_SC3_N_CORES</span><span class="kw">:</span><span class="at"> </span><span class="dv">1</span></span></code></pre></div>
<p><strong>Type:</strong> integer scalar</p>
<p>A number of CPU cores to use for SC3 computation. This may lead to
out-of-memory errors for a number of cores greater than 1.</p>
<p>Note that <code>drake</code> may have problems when the pipeline is
run in parallel mode and at the same time a target is using a parallel
code (<em>within target parallelism</em>). For the SC3 computation is
recommended to run the pipeline in sequential mode
(<code>DRAKE_PARALLELISM: "loop"</code>) and set
<code>DRAKE_TARGETS: ["cluster_sce_sc3"]</code>. Once the
<code>cluster_sce_sc3</code> target is finished, you can switch the
parallelism back to <code>clustermq</code> or <code>future</code>.</p>
<hr>
<div class="sourceCode" id="cb35"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="fu">CLUSTER_SC3_ENABLED</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>If <code>False</code>, skip SC3 clustering.</p>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="cell-type-annotation">Cell type annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation"></a>
</h4>
<div class="sourceCode" id="cb36"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="fu">CELL_ANNOTATION_SOURCES_DEFAULTS</span><span class="kw">:</span></span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a><span class="at">  </span><span class="fu">TRAIN_PARAMS</span><span class="kw">:</span></span>
<span id="cb36-3"><a href="#cb36-3" tabindex="-1"></a><span class="at">    </span><span class="fu">GENES</span><span class="kw">:</span><span class="at"> </span><span class="st">"de"</span></span>
<span id="cb36-4"><a href="#cb36-4" tabindex="-1"></a><span class="at">    </span><span class="fu">SD_THRESH</span><span class="kw">:</span><span class="at"> </span><span class="dv">1</span></span>
<span id="cb36-5"><a href="#cb36-5" tabindex="-1"></a><span class="at">    </span><span class="fu">DE_METHOD</span><span class="kw">:</span><span class="at"> </span><span class="st">"classic"</span></span>
<span id="cb36-6"><a href="#cb36-6" tabindex="-1"></a><span class="at">    </span><span class="fu">DE_N</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span>
<span id="cb36-7"><a href="#cb36-7" tabindex="-1"></a><span class="at">    </span><span class="fu">ASSAY_TYPE</span><span class="kw">:</span><span class="at"> </span><span class="st">"logcounts"</span></span>
<span id="cb36-8"><a href="#cb36-8" tabindex="-1"></a><span class="at">  </span><span class="fu">CLASSIFY_PARAMS</span><span class="kw">:</span></span>
<span id="cb36-9"><a href="#cb36-9" tabindex="-1"></a><span class="at">    </span><span class="fu">QUANTILE</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.8</span></span>
<span id="cb36-10"><a href="#cb36-10" tabindex="-1"></a><span class="at">    </span><span class="fu">TUNE_THRESH</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.05</span></span>
<span id="cb36-11"><a href="#cb36-11" tabindex="-1"></a><span class="at">    </span><span class="fu">ASSAY_TYPE</span><span class="kw">:</span><span class="at"> </span><span class="st">"logcounts"</span></span>
<span id="cb36-12"><a href="#cb36-12" tabindex="-1"></a><span class="at">  </span><span class="fu">PRUNE_SCORE_PARAMS</span><span class="kw">:</span></span>
<span id="cb36-13"><a href="#cb36-13" tabindex="-1"></a><span class="at">    </span><span class="fu">N_MADS</span><span class="kw">:</span><span class="at"> </span><span class="dv">3</span></span>
<span id="cb36-14"><a href="#cb36-14" tabindex="-1"></a><span class="at">    </span><span class="fu">MIN_DIFF_MED</span><span class="kw">:</span><span class="at"> </span><span class="fl">-.inf</span></span>
<span id="cb36-15"><a href="#cb36-15" tabindex="-1"></a><span class="at">    </span><span class="fu">MIN_DIFF_NEXT</span><span class="kw">:</span><span class="at"> </span><span class="dv">0</span></span>
<span id="cb36-16"><a href="#cb36-16" tabindex="-1"></a><span class="at">  </span><span class="fu">DIAGNOSTICS_PARAMS</span><span class="kw">:</span></span>
<span id="cb36-17"><a href="#cb36-17" tabindex="-1"></a><span class="at">    </span><span class="fu">HEATMAP_N_TOP_MARKERS</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span></code></pre></div>
<p><strong>Type:</strong> named list (dictionary) of named lists</p>
<p>Default parameters for computation and reporting of cell type
annotation. These can be overriden for each of the cell annotation
reference source in <code>CELL_ANNOTATION_SOURCES</code> (see
below).</p>
<ul>
<li>
<code>TRAIN_PARAMS</code>: used for training, passed to
<code><a href="https://rdrr.io/pkg/SingleR/man/trainSingleR.html" class="external-link">SingleR::trainSingleR()</a></code>.
<ul>
<li>
<code>GENES: "de"</code> (character scalar: <code>"de"</code> |
<code>"sd"</code> | <code>"all"</code>): feature selection mode, see
Details in <code>?SingleR::trainSingleR()</code>.</li>
<li>
<code>SD_THRESH: 1</code> (numeric scalar): the minimum threshold on
the standard deviation per gene. Only used when
<code>GENES: "sd"</code>.</li>
<li>
<code>DE_METHOD: "classic"</code> (character scalar:
<code>"classic"</code> | <code>"wilcox"</code> | <code>"t"</code>): how
DE genes should be detected between pairs of labels. Defaults to
<code>"classic"</code>, which sorts genes by the log-fold changes and
takes the top <code>DE_N</code>. Setting to <code>"wilcox"</code> or
<code>"t"</code> will use Wilcoxon ranked sum test or Welch t-test
between labels, respectively, and take the top <code>DE_N</code>
upregulated genes per comparison.</li>
<li>
<code>DE_N: null</code> (<code>null</code> or integer scalar): the
number of DE genes to use when <code>GENES: "de"</code>. If
<code>DE_METHOD: "classic"</code>, defaults to
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>500</mn><mo>*</mo><mo stretchy="false" form="prefix">(</mo><mn>2</mn><mi>/</mi><mn>3</mn><msup><mo stretchy="false" form="postfix">)</mo><mi>l</mi></msup><mi>o</mi><mi>g</mi><mn>2</mn><mo stretchy="false" form="prefix">(</mo><mi>N</mi><mo stretchy="false" form="postfix">)</mo></mrow><annotation encoding="application/x-tex">500 * (2/3) ^ log2(N)</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>N</mi><annotation encoding="application/x-tex">N</annotation></semantics></math>
is the number of unique labels. Otherwise, <code>null</code> defaults to
<code>10</code>.</li>
<li>
<code>ASSAY_TYPE: "logcounts"</code> (character scalar or integer):
the assay of the reference dataset containing the relevant expression
matrix.</li>
</ul>
</li>
<li>
<code>CLASSIFY_PARAMS</code>: used for classification, passed to
<code><a href="https://rdrr.io/pkg/SingleR/man/classifySingleR.html" class="external-link">SingleR::classifySingleR()</a></code>.
<ul>
<li>
<code>QUANTILE: 0.8</code> (numeric scalar): the quantile of the
correlation distribution to use to compute the score for each
label.</li>
<li>
<code>TUNE_THRESH: 0.05</code> (numeric scalar): the maximum
difference from the maximum correlation to use in fine-tuning.</li>
<li>
<code>ASSAY_TYPE: "logcounts"</code> (character scalar or integer):
the assay of the test dataset containing the relevant expression
matrix.</li>
</ul>
</li>
<li>
<code>PRUNE_SCORE_PARAMS</code>: used for label pruning after
classification, passed to <code><a href="https://rdrr.io/pkg/SingleR/man/pruneScores.html" class="external-link">SingleR::pruneScores()</a></code>.
<ul>
<li>
<code>N_MADS: 3</code> (numeric scalar): the number of MADs to use
for defining low outliers in the per-label distribution of delta values
(i.e., difference from median).</li>
<li>
<code>MIN_DIFF_MED: -.inf</code> (numeric scalar): the minimum
acceptable delta for each cell.</li>
<li>
<code>MIN_DIFF_NEXT: 0</code> (numeric scalar): the minimum
difference between the best score and the next best score in
fine-tuning.</li>
</ul>
</li>
<li>
<code>DIAGNOSTICS_PARAMS</code>: parameters for post-classification
diagnostics.
<ul>
<li>
<code>HEATMAP_N_TOP_MARKERS: 20</code> (integer scalar): how many
top markers to put into heatmaps after classification. Markers are
computed after classification using the predicted cell labels.</li>
</ul>
</li>
</ul>
<hr>
<div class="sourceCode" id="cb37"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="fu">CELL_ANNOTATION_SOURCES</span><span class="kw">:</span></span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> </span><span class="fu">human_primary_cell_atlas_main</span><span class="kw">:</span></span>
<span id="cb37-3"><a href="#cb37-3" tabindex="-1"></a><span class="at">      </span><span class="fu">reference_type</span><span class="kw">:</span><span class="at"> </span><span class="st">"celldex"</span></span>
<span id="cb37-4"><a href="#cb37-4" tabindex="-1"></a><span class="at">      </span><span class="fu">reference</span><span class="kw">:</span><span class="at"> </span><span class="st">"HumanPrimaryCellAtlasData"</span></span>
<span id="cb37-5"><a href="#cb37-5" tabindex="-1"></a><span class="fu">      description</span><span class="kw">: </span><span class="ch">&gt;</span></span>
<span id="cb37-6"><a href="#cb37-6" tabindex="-1"></a>        Microarray datasets derived from human primary cells (Mabbott et al. 2013).</span>
<span id="cb37-7"><a href="#cb37-7" tabindex="-1"></a>        Most of the labels refer to blood subpopulations but cell types from other tissues are also available.</span>
<span id="cb37-8"><a href="#cb37-8" tabindex="-1"></a><span class="at">      </span><span class="fu">label_column</span><span class="kw">:</span><span class="at"> </span><span class="st">"label.main"</span></span>
<span id="cb37-9"><a href="#cb37-9" tabindex="-1"></a><span class="at">      </span><span class="fu">label_subsets</span><span class="kw">:</span><span class="at"> </span><span class="kw">[]</span></span>
<span id="cb37-10"><a href="#cb37-10" tabindex="-1"></a><span class="at">    </span><span class="fu">monaco_immune_fine</span><span class="kw">:</span></span>
<span id="cb37-11"><a href="#cb37-11" tabindex="-1"></a><span class="at">      </span><span class="fu">reference_type</span><span class="kw">:</span><span class="at"> </span><span class="st">"celldex"</span></span>
<span id="cb37-12"><a href="#cb37-12" tabindex="-1"></a><span class="at">      </span><span class="fu">reference</span><span class="kw">:</span><span class="at"> </span><span class="st">"MonacoImmuneData"</span></span>
<span id="cb37-13"><a href="#cb37-13" tabindex="-1"></a><span class="at">      </span><span class="fu">description</span><span class="kw">:</span><span class="at"> </span><span class="st">"This is the human immune reference that best covers all of the bases for a typical PBMC sample."</span></span>
<span id="cb37-14"><a href="#cb37-14" tabindex="-1"></a><span class="at">      </span><span class="fu">label_column</span><span class="kw">:</span><span class="at"> </span><span class="st">"label.fine"</span></span>
<span id="cb37-15"><a href="#cb37-15" tabindex="-1"></a><span class="at">      </span><span class="fu">label_subsets</span><span class="kw">:</span><span class="at"> </span><span class="kw">[]</span></span></code></pre></div>
<p><strong>Type:</strong> list of named lists or <code>null</code></p>
<p>Specifies reference dataset(s) to be used for cell type annotation.
The method used in <a href="https://github.com/SingleR-inc/SingleR" class="external-link">SingleR</a> for training and
classification is described <a href="http://bioconductor.org/books/3.15/SingleRBook/introduction.html#method-description" class="external-link">here</a>.</p>
<p>The names of named lists are used to name the reference datasets,
e.g. as the prefixes in names of <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> columns
containing per-cell labels. More info about the structure of related
targets is given in the <a href="#cell-type-annotation-1">Targets</a>
section. To skip the cell annotation step, define
<code>CELL_ANNOTATION_SOURCES</code> as <code>null</code>.</p>
<p>Each reference dataset must have the following parameters:</p>
<ul>
<li>
<code>reference_type: "celldex"</code> (character scalar:
<code>"celldex"</code> | <code>"file"</code>): type of the reference
dataset.</li>
<li>
<code>reference: "HumanPrimaryCellAtlasData"</code> (character
scalar):
<ul>
<li>If <code>reference_type: "celldex"</code>: a name of function to
call in the <a href="https://github.com/LTLA/celldex" class="external-link">celldex</a> package. This function must return a
<code>SummarizedExperiment</code> object. See <a href="https://github.com/LTLA/celldex" class="external-link">celldex</a>’s
<a href="https://bioconductor.org/packages/3.15/data/experiment/vignettes/celldex/inst/doc/userguide.html" class="external-link">vignette</a>
for an overview of available datasets.</li>
<li>If <code>reference_type: "file"</code>: a path to Rds file with
saved <code>SingleCellExperiment</code> or
<code>SummarizedExperiment</code> object. The assay type specified
<code>TRAIN_PARAMS/assay_type</code> must be present in
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames()</a></code> of the object. The <code>label_column</code>
must be present in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> of the object.</li>
</ul>
</li>
<li>
<code>description: "..."</code> (character scalar): a description of
the reference dataset, will appear in some plots.</li>
<li>
<code>label_column: "label.main"</code> (character scalar):
specifies which column from <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> of the reference
dataset will be used for training. For datasets from the
<a href="https://github.com/LTLA/celldex" class="external-link">celldex</a> package, those are <code>label.main</code>,
<code>label.fine</code> (more granular labels), or
<code>label.ont</code> (cell ontology IDs).</li>
<li>
<code>label_subsets: []</code> (character vector): labels to subset
the reference dataset prior to training,
e.g. <code>["T cells", "B cells", "Progenitors"]</code>.</li>
</ul>
<p>In each entry (e.g. <code>human_primary_cell_atlas_main</code> above)
you can override default parameters in
<code>CELL_ANNOTATION_SOURCES_DEFAULTS</code> (by using lowercase
names), e.g.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="fu">CELL_ANNOTATION_SOURCES</span><span class="kw">:</span></span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> </span><span class="fu">human_primary_cell_atlas_main</span><span class="kw">:</span></span>
<span id="cb38-3"><a href="#cb38-3" tabindex="-1"></a><span class="at">      ...</span></span>
<span id="cb38-4"><a href="#cb38-4" tabindex="-1"></a><span class="at">      </span><span class="fu">train_params</span><span class="kw">:</span></span>
<span id="cb38-5"><a href="#cb38-5" tabindex="-1"></a><span class="at">        </span><span class="fu">de_method</span><span class="kw">:</span><span class="at"> </span><span class="st">"wilcox"</span></span>
<span id="cb38-6"><a href="#cb38-6" tabindex="-1"></a><span class="at">        </span><span class="fu">de_n</span><span class="kw">:</span><span class="at"> </span><span class="dv">30</span></span></code></pre></div>
<p>will override corresponding parameters in <code>TRAIN_PARAMS</code>
in <code>CELL_ANNOTATION_SOURCES_DEFAULTS</code>.</p>
<hr>
<hr>
</div>
<div class="section level4">
<h4 id="marker-based-cell-annotation-signatures">Marker-based cell annotation signatures<a class="anchor" aria-label="anchor" href="#marker-based-cell-annotation-signatures"></a>
</h4>
<div class="sourceCode" id="cb39"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="fu">MANUAL_ANNOTATION</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>Whether marker-based annotation is enabled. Marker-based annotation
is based on expression profiles, markers are taken from user-defined
file. Resulted annotation is stored in <code>annotation_metadata</code>
target. Marker-based cell annotation implemented from Giotto package
enrichment function, which is based on PAGE method from <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-6-144" class="external-link">Kim
SY et al</a>.</p>
<hr>
<div class="sourceCode" id="cb40"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="fu">ANNOTATION_MARKERS</span><span class="kw">:</span><span class="at"> </span><span class="ch">Null</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar or Null</p>
<p>Path to user-define file contacting markers for each expected cell
type. Each column name is one cell type, in rows are gene markers.
Example:</p>
<pre><code>Astrocytes,Endothelial,Neurons
GFAP,PECAM1,SLC6A1
SLC1A3,A2M,CHAT
GLUL,,GAD2</code></pre>
<hr>
<div class="sourceCode" id="cb42"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="fu">SCALE_ANNOTATION</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>Whether to scale gene profiles.</p>
<hr>
<div class="sourceCode" id="cb43"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a><span class="fu">OVERLAP</span><span class="kw">:</span><span class="at"> </span><span class="dv">5</span></span></code></pre></div>
<p><strong>Type:</strong> numeric scalar</p>
<p>Minimal overlap of genes in cell/cell in selected clusters, to assign
cell type.</p>
<hr>
<div class="sourceCode" id="cb44"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="fu">ANNOTATION_CLUSTERING</span><span class="kw">:</span><span class="at"> </span><span class="st">"cluster_graph_leiden_r0.2"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>On which clustering (column of cell metadata) to perform
annotation.</p>
<hr>
<div class="sourceCode" id="cb45"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="fu">SHOW_VALUE</span><span class="kw">:</span><span class="at"> </span><span class="st">"value"</span><span class="at"> </span><span class="er">(`"value"`</span><span class="at"> </span><span class="er">|</span><span class="at"> </span><span class="er">`"zscores"`</span><span class="at"> </span><span class="er">|</span><span class="at"> </span><span class="er">`"zscores_rescaled"`)</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>Which value to show in resulting heatmap.</p>
<hr>
<div class="sourceCode" id="cb46"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a><span class="fu">MAKE_CELL_PLOT</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>Only together with spatial option. Whether to show pseudo tissue cell
plots with enrichment of annotation.</p>
<hr>
<div class="sourceCode" id="cb47"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="fu">HEATMAP_DIMRED</span><span class="kw">:</span><span class="at"> </span><span class="st">"umap"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>Which dimension reduction use to show result of marker-based
annotation.</p>
<hr>
</div>
<div class="section level4">
<h4 id="spot-deconvolution">Spot deconvolution<a class="anchor" aria-label="anchor" href="#spot-deconvolution"></a>
</h4>
<div class="sourceCode" id="cb48"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="fu">SPOT_DECONVOLUTION</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>Wether spot deconvolution for SRT data is enabled. Spot deconvolution
is reference based, user must define reference scRNA-seq object and
column in colData that contain information about cell types for each
cell. Currently two main methods are available.</p>
<hr>
<div class="sourceCode" id="cb49"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="fu">DECONV_METHOD</span><span class="kw">:</span><span class="at"> </span><span class="st">"CARD"</span><span class="at"> </span><span class="er">(`"CARD"`</span><span class="at"> </span><span class="er">|</span><span class="at"> </span><span class="er">`"RCTD"`)</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>Which deconvolution method use. ***</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="fu">REFERENCE_SC_DECONV</span><span class="kw">:</span><span class="at"> </span><span class="ch">Null</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>Path to user-define reference <code>SingleCellExperiment</code>
object in rds format used for spot deconvolution, containing expected
cell types.</p>
<hr>
<div class="sourceCode" id="cb51"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="fu">REFERENCE_SC_LABEL</span><span class="kw">:</span><span class="at"> </span><span class="ch">Null</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>Column in colData in defined reference
<code>SingleCellExperiment</code> object, that contains cell types
labels/cell characteristics.</p>
<hr>
</div>
<div class="section level4">
<h4 id="cell-grouping-assignment">Cell grouping assignment<a class="anchor" aria-label="anchor" href="#cell-grouping-assignment"></a>
</h4>
<p>Using the two parameters below, it is possible to add custom cell
metadata or rearrange the current one into new groupings (saved as
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> columns of the <code>SingleCellExperiment</code>
object). These groupings can be then referenced in other parameters,
e.g. <code>NORM_CLUSTERING_REPORT_DIMRED_PLOTS_OTHER</code> or in the
<code>cluster_markers</code> and <code>contrasts</code> stages.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a><span class="fu">ADDITIONAL_CELL_DATA_FILE</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar or <code>null</code></p>
<p>A path to file with additional cell data which will be added to
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> of the final SCE object of this stage. The
additional data will overwrite existing columns
(e.g. <code>cluster_graph_louvain</code>), but are appended before the
<code>CELL_GROUPINGS</code> below is applied, so you can refer to
additional data columns inside <code>CELL_GROUPINGS</code>. In the
additional data, there can be only cell barcodes that are present after
cell filtering in the <code>01_input_qc</code> stage - the missing ones
will have <code>NA</code> set in the resulting data. Rownames are not
mandatory.</p>
<p>The supported file formats are CSV (comma separated,
<code>.csv</code> extension) and Rds (dataframe-like objects,
<code>.Rds</code> extension). The additional data must contain
<code>Barcode</code> column by which they will be joined with the
original cell data (<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>).</p>
<hr>
<div class="sourceCode" id="cb53"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a><span class="fu">CELL_GROUPINGS</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span></code></pre></div>
<p><strong>Type:</strong> list of named lists or <code>null</code></p>
<p>This parameter specifies new cell groupings based on existing columns
(primarily the clustering ones). Each key is a name of new cell grouping
(column in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> of the <code>sce_dimred</code> target)
and contains a named list in the form:</p>
<ul>
<li>
<code>source_column</code> (character scalar): a name of existing
column in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>.</li>
<li>
<code>description</code> (character scalar, optional): a description
of the cell grouping, will appear in plots. If not set, the name of the
assignment will be used.</li>
<li>
<code>assignments</code> (named list/vector of character scalars):
<ul>
<li>
<code>&lt;old level&gt;: "&lt;new level&gt;"</code>: assignment of
an old level to a new level. Different old levels with the same new
level will be merged. Order of assignments does not matter, levels are
just replaced (under the hood, <code><a href="https://dplyr.tidyverse.org/reference/recode.html" class="external-link">dplyr::recode()</a></code> is used).
Unspecified levels will be kept as they are. You cannot replace existing
groupings (columns).</li>
</ul>
</li>
</ul>
<p>For possible column names for <code>source_column</code>, see the
<code>sce_final_norm_clustering</code> target in the <a href="#targets">section</a> below.</p>
<p>Let’s look at example cell grouping:</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a><span class="fu">CELL_GROUPINGS</span><span class="kw">:</span></span>
<span id="cb54-2"><a href="#cb54-2" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> </span><span class="fu">cluster_graph_leiden_r0.4_annotated</span><span class="kw">:</span></span>
<span id="cb54-3"><a href="#cb54-3" tabindex="-1"></a><span class="at">      </span><span class="fu">source_column</span><span class="kw">:</span><span class="at"> </span><span class="st">"cluster_graph_leiden_r0.4"</span></span>
<span id="cb54-4"><a href="#cb54-4" tabindex="-1"></a><span class="at">      </span><span class="fu">description</span><span class="kw">:</span><span class="at"> </span><span class="st">"Graph-based clustering (Leiden alg.), annotated clusters"</span></span>
<span id="cb54-5"><a href="#cb54-5" tabindex="-1"></a><span class="at">      </span><span class="fu">assignments</span><span class="kw">:</span></span>
<span id="cb54-6"><a href="#cb54-6" tabindex="-1"></a><span class="at">        </span><span class="fu">1</span><span class="kw">:</span><span class="at"> </span><span class="st">"memory_CD4+"</span></span>
<span id="cb54-7"><a href="#cb54-7" tabindex="-1"></a><span class="at">        </span><span class="fu">6</span><span class="kw">:</span><span class="at"> </span><span class="st">"B"</span></span>
<span id="cb54-8"><a href="#cb54-8" tabindex="-1"></a><span class="at">        </span><span class="fu">7</span><span class="kw">:</span><span class="at"> </span><span class="st">"memory_CD4+"</span></span></code></pre></div>
<ul>
<li>
<code>cluster_graph_leiden_r0.4_annotated</code>: a name of new cell
grouping added to <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>.
<ul>
<li>
<code>source_column: "cluster_graph_leiden_r0.4"</code>: a name of
column in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> whose level will be used for the new
assignments.</li>
<li>
<code>description: "Graph-based clustering ..."</code>: a
description of the new assignment. If it was not set, the name of the
grouping (<code>cluster_graph_leiden_r0.4_annotated</code>) would be
used instead.</li>
<li>
<code>assignments</code>: assignments of the old levels to new ones.
That is, level (cluster) <code>6</code> will be renamed to
<code>"B"</code>, and levels <code>1</code> and <code>7</code> will be
merged to a new level <code>"memory_CD4+"</code>. All unspecified levels
will be kept as they are.</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="dimensionality-reduction-plots">Dimensionality reduction plots<a class="anchor" aria-label="anchor" href="#dimensionality-reduction-plots"></a>
</h4>
<div class="sourceCode" id="cb55"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb55-1"><a href="#cb55-1" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_REPORT_DIMRED_NAMES</span><span class="kw">:</span><span class="at"> </span><span class="kw">[</span><span class="st">"umap"</span><span class="kw">,</span><span class="at"> </span><span class="st">"pca"</span><span class="kw">,</span><span class="at"> </span><span class="st">"tsne"</span><span class="kw">]</span></span></code></pre></div>
<p><strong>Type:</strong> character vector (<code>"umap"</code> |
<code>"pca"</code> | <code>"tsne"</code>)</p>
<p>A vector of names of dimensionality reduction methods for which plots
in the report will be made.</p>
<hr>
<div class="sourceCode" id="cb56"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_REPORT_DIMRED_PLOTS_OTHER</span><span class="kw">:</span></span>
<span id="cb56-2"><a href="#cb56-2" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> </span><span class="st">"phase"</span><span class="kw">:</span><span class="at"> </span><span class="st">"Cell cycle phases"</span></span>
<span id="cb56-3"><a href="#cb56-3" tabindex="-1"></a><span class="at">    </span><span class="fu">"doublet_score"</span><span class="kw">:</span><span class="at"> </span><span class="st">"Doublet score"</span></span>
<span id="cb56-4"><a href="#cb56-4" tabindex="-1"></a><span class="at">    </span><span class="fu">"total"</span><span class="kw">:</span><span class="at"> </span><span class="st">"Total number of UMI"</span></span>
<span id="cb56-5"><a href="#cb56-5" tabindex="-1"></a><span class="at">    </span><span class="fu">"detected"</span><span class="kw">:</span><span class="at"> </span><span class="st">"Detected number of genes"</span></span></code></pre></div>
<p><strong>Type:</strong> list of named lists with character scalars</p>
<p>Names of other variables in <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> to plot dimreds
and color by, displayed in the report.</p>
<p>The format is <code>"variable_name": "description"</code>, where
description will appear in plot title.</p>
<p>Variables defined in the <code>CELL_GROUPINGS</code> parameter can
have <code>description</code> set to <code>null</code> - their
description from <code>CELL_GROUPINGS</code> will be used.</p>
</div>
<div class="section level4">
<h4 id="input-files">Input files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h4>
<div class="sourceCode" id="cb57"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb57-1"><a href="#cb57-1" tabindex="-1"></a><span class="fu">SELECTED_MARKERS_FILE</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar or <code>null</code></p>
<p>A path to CSV file with groups of markers to make expression dimred
plots for. Set to <code>null</code> to skip. The plots will be created
for each of <code>NORM_CLUSTERING_REPORT_DIMRED_NAMES</code>.</p>
<p>The format is following: each row defines gene symbols for a group.
First column is a name of group and second column are gene symbols
delimited by “:”. Example:</p>
<p><code>Naive_CD4+_T,IL7R:CCR7</code></p>
<p><strong>Do not use a header!</strong></p>
<p>An example CSV file for PBMC 1k dataset is automatically copied to a
newly initialized project (as
<code>system.file("extdata", "selected_markers.csv", package = "scdrake", mustWork = TRUE)</code>).
You can also view this file on GitHub <a href="https://github.com/bioinfocz/scdrake/blob/main/inst/extdata/selected_markers.csv" class="external-link">here</a>.</p>
<hr>
<div class="sourceCode" id="cb58"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_REPORT_RMD_FILE</span><span class="kw">:</span><span class="at"> </span><span class="st">"Rmd/single_sample/02_norm_clustering.Rmd"</span></span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a><span class="co">#NORM_CLUSTERING_REPORT_RMD_FILE: "Rmd/single_sample/02_norm_clustering_spatial.Rmd"</span></span>
<span id="cb58-3"><a href="#cb58-3" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_REPORT_SIMPLE_RMD_FILE</span><span class="kw">:</span><span class="at"> </span><span class="st">"Rmd/single_sample/02_norm_clustering_simple.Rmd"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>Paths to RMarkdown files used for HTML reports of this pipeline
stage.</p>
<p><code>NORM_CLUSTERING_REPORT_RMD_FILE</code> is technically more
detailed, while <code>NORM_CLUSTERING_REPORT_SIMPLE_RMD_FILE</code>
outputs only the dimensionality reduction plots along with cell
annotation.</p>
</div>
<div class="section level4">
<h4 id="output-files">Output files<a class="anchor" aria-label="anchor" href="#output-files"></a>
</h4>
<div class="sourceCode" id="cb59"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb59-1"><a href="#cb59-1" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_BASE_OUT_DIR</span><span class="kw">:</span><span class="at"> </span><span class="st">"02_norm_clustering"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>A path to base output directory for this stage. It will be created
under <code>BASE_OUT_DIR</code> specified in <code>00_main.yaml</code>
config.</p>
<hr>
<div class="sourceCode" id="cb60"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_SELECTED_MARKERS_OUT_DIR</span><span class="kw">:</span><span class="at"> </span><span class="st">"selected_markers"</span></span>
<span id="cb60-2"><a href="#cb60-2" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_DIMRED_PLOTS_OUT_DIR</span><span class="kw">:</span><span class="at"> </span><span class="st">"dimred_plots"</span></span>
<span id="cb60-3"><a href="#cb60-3" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_OTHER_PLOTS_OUT_DIR</span><span class="kw">:</span><span class="at"> </span><span class="st">"other_plots"</span></span>
<span id="cb60-4"><a href="#cb60-4" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_CELL_ANNOTATION_OUT_DIR</span><span class="kw">:</span><span class="at"> </span><span class="st">"cell_annotation"</span></span>
<span id="cb60-5"><a href="#cb60-5" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_REPORT_HTML_FILE</span><span class="kw">:</span><span class="at"> </span><span class="st">"02_norm_clustering.html"</span></span>
<span id="cb60-6"><a href="#cb60-6" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_REPORT_SIMPLE_HTML_FILE</span><span class="kw">:</span><span class="at"> </span><span class="st">"02_norm_clustering_simple.html"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>Names of files and directories created under
<code>NORM_CLUSTERING_BASE_OUT_DIR</code>. Subdirectories are not
allowed.</p>
</div>
<div class="section level4">
<h4 id="html-output-parameters">HTML output parameters<a class="anchor" aria-label="anchor" href="#html-output-parameters"></a>
</h4>
<div class="sourceCode" id="cb61"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb61-1"><a href="#cb61-1" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_KNITR_MESSAGE</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span>
<span id="cb61-2"><a href="#cb61-2" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_KNITR_WARNING</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span>
<span id="cb61-3"><a href="#cb61-3" tabindex="-1"></a><span class="fu">NORM_CLUSTERING_KNITR_ECHO</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>These are passed to <code><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">knitr::opts_chunk()</a></code> and used for
rendering of stage’s HTML report.</p>
</div>
</div>
<div class="tab-pane" id="outputs" role="tabpanel" aria-labelledby="outputs-tab">

<p>Here you can find description of the most important targets for this
stage. However, for a full overview, you have to inspect the <a href="https://github.com/bioinfocz/scdrake/blob/main/R/plans_single_sample.R" class="external-link">source
code</a> of the <code>get_input_norm_clustering_subplan()</code>
function.</p>
<p>HTML report target names: <code>report_norm_clustering</code>,
<code>report_norm_clustering_simple</code> (only dimred plots)</p>
<div class="section level4">
<h4 id="singlecellexperiment-or-spatialexperiment-objects">
<code>SingleCellExperiment</code> or <code>SpatialExperiment</code>
objects<a class="anchor" aria-label="anchor" href="#singlecellexperiment-or-spatialexperiment-objects"></a>
</h4>
<p><code>sce_cc</code>: a SCE object (the
<code>sce_final_input_qc</code> target from the <code>input_qc</code>
stage) with added cell cycle information to <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>:
<code>phase</code> (<code>"G1"</code>, <code>"G2m"</code> or
<code>"S"</code>), <code>s_score</code>, <code>g2m_score</code>,
<code>cc_difference</code> (<code>s_score</code> -
<code>g2m_score</code>).</p>
<hr>
<p><code>sce_norm</code>: <code>sce_cc</code> with normalized counts in
<code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">SingleCellExperiment::logcounts()</a></code> assay and
<code>normalization_type</code> item added to
<code>metadata()</code>.</p>
<hr>
<p><code>sce_norm_hvg</code>: <code>sce_norm</code> with added HVG
information:</p>
<ul>
<li>
<code>metadata()</code>: <code>hvg_metric</code>,
<code>hvg_selection</code>, <code>hvg_selection_value</code>,
<code>hvg_metric_fit</code>, <code>hvg_rm_cc_genes</code>,
<code>hvg_ids</code>
</li>
<li>
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>: <code>is_hvg</code> (logical)</li>
</ul>
<p>If the parameter <code>HVG_RM_CC_GENES</code> is <code>True</code>,
<code>sce_norm_hvg</code> will also contain a PCA dimensionality
reduction (named <code>pca_with_cc</code>) for whose computation all
HVGs were used.</p>
<hr>
<p><code>sce_rm_doublets</code>: <code>sce_norm_hvg</code> with removed
doublets and added information:</p>
<ul>
<li>
<code>metadata()</code>: <code>has_filtered_doublets</code>
(<code>FALSE</code> if <code>MAX_DOUBLET_SCORE</code> parameter is
<code>null</code>), <code>max_doublet_score</code>
</li>
<li>
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>: <code>doublet_score</code> (numeric),
<code>is_doublet</code> (logical)</li>
</ul>
<hr>
<p><code>sce_pca</code>: <code>sce_rm_doublets</code> with calculated
PCA (50 PCs). Reduced dimensions can be retrieved by
<code>SingleCellExperiment::reducedDim(sce_pca, "pca")</code></p>
<hr>
<p><code>sce_pca_selected_pcs</code>: <code>sce_pca</code> with selected
number of PCs. Reduced dimensions in <code>pca</code> slot are subsetted
to the selected number of PCs, and <code>pca_all</code> slot contains
the full matrix of 50 PCs. Information about selection of PCs is added
to <code>metadata()</code>: <code>pca_selection_method</code>,
<code>pca_selected_pcs</code>.</p>
<hr>
<p><code>sce_dimred</code>: <code>sce_pca_selected_pcs</code> with
calculated t-SNE and UMAP, both using the selected number of PCs for
calculation</p>
<hr>
<p><code>sce_final_norm_clustering</code>: a final SCE object of this
stage which will be used in the <code>cluster_markers</code> and
<code>contrasts</code> stages. This object is derived from
<code>sce_dimred</code>, but cell clusterings along with
<code>CELL_GROUPINGS</code> are added to <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>.</p>
<p>The names of columns containing cell clusterings are:</p>
<ul>
<li>Graph-based clustering: <code>cluster_graph_walktrap</code>,
<code>cluster_graph_louvain_r&lt;r&gt;</code>
(e.g. <code>cluster_graph_louvain_r0.4</code>),
<code>cluster_graph_leiden_r&lt;r&gt;</code>
(e.g. <code>cluster_graph_leiden_r0.4</code>)
<ul>
<li>
<code>&lt;r&gt;</code> is based on the
<code>CLUSTER_GRAPH_LEIDEN_RESOLUTIONS</code> and
<code>CLUSTER_GRAPH_LOUVAIN_RESOLUTIONS</code> parameters</li>
</ul>
</li>
<li>k-means clustering: <code>cluster_kmeans_kbest</code>,
<code>cluster_kmeans_k&lt;k&gt;</code>
(e.g. <code>cluster_kmeans_k5</code>)
<ul>
<li>
<code>&lt;k&gt;</code> is based on the <code>CLUSTER_KMEANS_K</code>
parameter</li>
</ul>
</li>
<li>SC3 clustering: <code>cluster_sc3_&lt;k&gt;</code>
(e.g. <code>cluster_sc3_k5</code>)
<ul>
<li>
<code>&lt;k&gt;</code> is based on the <code>CLUSTER_SC3_K</code>
parameter</li>
</ul>
</li>
</ul>
<p>The names of reduces dimensionality matrices (which can be retrieved
by <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">SingleCellExperiment::reducedDim()</a></code>):</p>
<ul>
<li>PCA: <code>"pca"</code>
</li>
<li>t-SNE: <code>"tsne"</code>
</li>
<li>UMAP: <code>"umap"</code>
</li>
</ul>
<p>Or you can view their names with
<code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">SingleCellExperiment::reducedDimNames()</a></code></p>
</div>
<div class="section level4">
<h4 id="selection-of-pcs">Selection of PCs<a class="anchor" aria-label="anchor" href="#selection-of-pcs"></a>
</h4>
<p><code>pca_elbow_pcs</code>, <code>pca_gene_var_pcs</code>: a number
of selected PCs for elbow point and technical variance methods,
respectively</p>
<p><code>pca_selected_pcs_plot</code>: a <a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a> object
showing cummulative variance explained and the selected number of first
PCs for each of the methods</p>
</div>
<div class="section level4">
<h4 id="cell-clustering">Cell clustering<a class="anchor" aria-label="anchor" href="#cell-clustering"></a>
</h4>
<p><code>clusters_all</code>: a named list of integer vectors which
wraps all clusterings below</p>
<div class="section level5">
<h5 id="graph-based-clustering-1">Graph-based clustering<a class="anchor" aria-label="anchor" href="#graph-based-clustering-1"></a>
</h5>
<p><code>graph_snn</code>: an <a href="https://r.igraph.org/" class="external-link">igraph</a> object of shared
nearest-neighbor graph, returned from
<code><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">scran::buildSNNGraph()</a></code></p>
<p><code>cluster_graph_leiden</code>,
<code>cluster_graph_walktrap</code>, <code>cluster_graph_louvain</code>:
lists with cell clusters returned from
<code><a href="https://r.igraph.org/reference/cluster_walktrap.html" class="external-link">igraph::cluster_walktrap()</a></code> and
<code><a href="https://r.igraph.org/reference/cluster_louvain.html" class="external-link">igraph::cluster_louvain()</a></code>, respectively. Both functions are
using <code>graph_snn</code> as the input</p>
<p><code>cluster_graph_louvain_df</code>,
<code>cluster_graph_walktrap_df</code>,
<code>cluster_graph_leiden_df</code>: dataframes with graph-based
clustering results</p>
<p><code>cluster_graph_leiden_clustree</code>,
<code>cluster_graph_louvain_clustree</code>: <a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a>
objects of <code><a href="https://lazappi.github.io/clustree/reference/clustree.html" class="external-link">clustree::clustree()</a></code> plots</p>
</div>
<div class="section level5">
<h5 id="k-means">k-means<a class="anchor" aria-label="anchor" href="#k-means"></a>
</h5>
<p><code>cluster_kmeans_kbest_gaps</code>: a <code>clusGap</code> object
used to calculate the best K. Returned from
<code><a href="https://rdrr.io/pkg/cluster/man/clusGap.html" class="external-link">cluster::clusGap()</a></code>, to which matrix of selected PCs is
passed</p>
<p><code>cluster_kmeans_kbest_k</code>: the best K for k-means, returned
from <code><a href="https://rdrr.io/pkg/cluster/man/clusGap.html" class="external-link">cluster::maxSE()</a></code></p>
<p><code>cluster_kmeans_kbest</code>, <code>cluster_kmeans_k</code>:
lists with cell clusters returned from <code><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">stats::kmeans()</a></code>. In
the former, <code>cluster_kmeans_kbest_k</code> is used as the number of
clusters. In the latter, Ks from the <code>CLUSTER_KMEANS_K</code>
parameter are used.</p>
<p><code>cluster_kmeans_k_df</code>,
<code>cluster_kmeans_kbest_df</code>: dataframes with k-means clustering
results</p>
<p><code>cluster_kmeans_k_clustree</code>: <a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a> object
of <code><a href="https://lazappi.github.io/clustree/reference/clustree.html" class="external-link">clustree::clustree()</a></code> plot. If
<code>CLUSTER_KMEANS_KBEST_ENABLED</code> is <code>True</code>, it will
also contain k-means result with the best <code>k</code>.</p>
</div>
<div class="section level5">
<h5 id="sc3">SC3<a class="anchor" aria-label="anchor" href="#sc3"></a>
</h5>
<p><code>sce_sc3</code>: a <code>SingleCellExperiment</code> object
returned from <code><a href="https://rdrr.io/pkg/SC3/man/sc3.html" class="external-link">SC3::sc3()</a></code></p>
<p><code>cluster_sc3</code>: a list with cell clusters</p>
<p><code>cluster_sc3_df</code>: a dataframe with SC3 clustering
results</p>
<p><code>cluster_sc3_stability_plots</code>: a list of
<a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a> objects, each returned from
<code>make_sc3_stability_plots()</code> (where
<code><a href="https://rdrr.io/pkg/SC3/man/sc3_plot_cluster_stability.html" class="external-link">SC3::sc3_plot_cluster_stability()</a></code> is used internally)</p>
<p><code>cluster_sc3_clustree</code>: <a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a> object of
<code><a href="https://lazappi.github.io/clustree/reference/clustree.html" class="external-link">clustree::clustree()</a></code> plot</p>
<hr>
</div>
</div>
<div class="section level4">
<h4 id="cell-type-annotation-1">Cell type annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation-1"></a>
</h4>
<p><code>cell_annotation_params</code>: a tibble with parameters parsed
from <code>CELL_ANNOTATION_SOURCES</code></p>
<p><code>cell_annotation</code>: <code>cell_annotation_params</code>
with added <code>cell_annotation</code> list column holding outputs
(<code>DataFrame</code>s) from <code><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR::SingleR()</a></code>. The
structure of those <code>DataFrame</code>s is described in
<code>?SingleR::classifySingleR()</code>.</p>
<hr>
<p><code>cell_annotation_labels</code>: a named list of character
vectors with per-cell label assignments. Later added to the
<code>cell_data</code> target and in turn, to <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> of
the <code>sce_final_norm_clustering</code> and
<code>sce_int_clustering_final</code> (in integration pipeline)
targets/objects.</p>
<p>By default, there are three vectors with labels for each reference
dataset named as:</p>
<ul>
<li>
<code>&lt;reference_name&gt;_labels_raw</code>: “raw” labels. Taken
from <code>first.labels</code> column of <code><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR::SingleR()</a></code>
output</li>
<li>
<code>&lt;reference_name&gt;_labels</code>: fine-tuned labels. Taken
from <code>labels</code> column of <code><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR::SingleR()</a></code>
output</li>
<li>
<code>&lt;reference_name&gt;_labels_pruned</code>: fine-tuned and
pruned labels (contains <code>NA</code>s for low quality labels). Taken
from <code>pruned.labels</code> column of
<code><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR::SingleR()</a></code> output.</li>
</ul>
<p>Note that you can use names of <code>cell_annotation_labels</code>’s
items in the <code>CELL_GROUPINGS</code> and
<code>NORM_CLUSTERING_REPORT_DIMRED_PLOTS_OTHER</code> parameters, and
also use them in <code>cluster_markers</code>
(<code>CLUSTER_MARKERS_SOURCES</code>) and <code>contrasts</code>
(<code>CONTRASTS_SOURCES</code>) stages</p>
<hr>
<p><code>cell_annotation_diagnostic_plots</code>:
<code>cell_annotation</code> with added list columns holding diagnostic
plots, i.e. for each reference dataset:</p>
<ul>
<li>
<code>score_heatmaps</code>: heatmaps of per-cell label scores
created for each clustering (as column annotation) (<a href="http://bioconductor.org/books/3.15/SingleRBook/annotation-diagnostics.html#based-on-the-scores-within-cells" class="external-link">details</a>).</li>
<li>
<code>marker_heatmaps</code>: <code>NULL</code> if
<code>cell_annotation$train_params$de</code> is not <code>"de"</code>,
otherwise heatmap for each label containing top upregulated markers from
pairwise t-tests (<a href="http://bioconductor.org/books/3.15/SingleRBook/annotation-diagnostics.html#based-on-marker-gene-expression" class="external-link">details</a>).
Number of top markers is specified in
<code>cell_annotation$diagnostics_params$heatmap_n_top_markers</code>.</li>
<li>
<code>delta_distribution_plot</code>: violin plots (in one figure /
object) of per-cell deltas for each label. Deltas are differences
between the score for the assigned label and the median across all
labels for each cell (<a href="http://bioconductor.org/books/3.15/SingleRBook/annotation-diagnostics.html#based-on-the-deltas-across-cells" class="external-link">details</a>).</li>
</ul>
<p><code>cell_annotation_diagnostic_plots_files</code>: paths to PDF
files of diagnostic plots in
<code>cell_annotation_diagnostic_plots</code>
(<code>score_heatmaps_out_file</code>,
<code>delta_distribution_plot_out_file</code>,
<code>marker_heatmaps_out_file</code>). That means when you make this
target, those files will be created.</p>
</div>
<div class="section level4">
<h4 id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a>
</h4>
<p><code>hvg_plot</code>: a <a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a> object with HVG
metric statistics (average expression vs. variance)</p>
<hr>
<p><code>pca_phase_plots</code>: a list of one or two
<code>{ggplot2</code>} objects. The first one is a plot of the first two
PCs colored by cell cycle phase, where PCA was computed using HVGs. If
the parameter <code>HVG_RM_CC_GENES</code> is <code>True</code>, the
second plot is the same as the first one, but with PCA computed on all
genes. In that case it also means that in the first plot, PCA was
computed using HVGs with removed cell cycle-related genes.</p>
<hr>
<p><code>pca_doublet_plot</code>, <code>pca_total_plot</code>: a
<code>{ggplot2</code>} plots of the first two PCs colored by doublet
score and total number of UMIs, respectively</p>
<hr>
<p><code>dimred_plots_clustering</code>: a tibble holding plots of
reduced dimensions colored by a clustering</p>
<p><code>dimred_plots_clustering_files_out</code>,
<code>dimred_plots_clustering_united_files</code>: make this target to
export the plots (the latter produces a single multipage PDF)</p>
<hr>
<p><code>dimred_plots_other_vars</code>: a tibble holding plots of
reduced dimensions colored by a specified variable, as defined by the
<code>NORM_CLUSTERING_REPORT_DIMRED_NAMES</code> and
<code>NORM_CLUSTERING_REPORT_DIMRED_PLOTS_OTHER</code> parameters.</p>
<p><code>dimred_plots_other_vars_files_out</code>: make this target to
export the plots</p>
<hr>
<p><code>plot_deconv_results</code>: a tibble holding plots of
deconvolution results, barplots as sumarizes statistics and spatial
pseudo-tissue plots ***</p>
<div class="section level5">
<h5 id="selected-markers">Selected markers<a class="anchor" aria-label="anchor" href="#selected-markers"></a>
</h5>
<p><code>selected_markers_plots</code>: a tibble holding selected marker
plots (<a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a> objects) for each dimensional reduction
specified in the <code>NORM_CLUSTERING_REPORT_DIMRED_NAMES</code>
parameter. Selected markers are read from CSV file defined in the
<code>SELECTED_MARKERS_FILE</code> parameter.</p>
<p><code>selected_markers_plots_files_out</code>: make this target to
export the plots</p>
<p>You can also use the underlying function
<code><a href="../reference/selected_markers_dimplot.html">selected_markers_dimplot()</a></code>:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/selected_markers_dimplot.html">selected_markers_dimplot</a></span><span class="op">(</span></span>
<span>  sce <span class="op">=</span> <span class="va">sce_final_norm_clustering</span>,</span>
<span>  dimred <span class="op">=</span> <span class="st">"umap"</span>,</span>
<span>  selected_markers_df <span class="op">=</span> <span class="va">selected_markers_df</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"logcounts"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code>selected_markers_df</code> tibble can be read from the CSV
file with</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_markers_df</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="st">"markers.csv"</span>, col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"group"</span>, <span class="st">"markers"</span><span class="op">)</span>, col_types <span class="op">=</span> <span class="st">"cc"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="other-targets">Other targets<a class="anchor" aria-label="anchor" href="#other-targets"></a>
</h4>
<p><code>config_norm_clustering</code>: a list holding parameters for
this stage</p>
<hr>
<p><code>cc_genes</code>: a dataframe of cell cycle genes, based on
<code><a href="https://satijalab.org/seurat/reference/cc.genes.updated.2019.html" class="external-link">Seurat::cc.genes.updated.2019</a></code> data. If the
<code>ORGANISM</code> parameter in the <code>00_main.yaml</code> config
is <code>"mouse"</code>, then gene symbols are first converted to
sentence-case (<code>"MKI67"</code> -&gt; <code>"Mki67"</code>).</p>
<hr>
<p><code>cell_data</code>: a DataFrame holding <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> of
<code>sce_rm_doublets</code>, clusterings from the
<code>clusters_all</code> target, cell annotation labels from the
<code>cell_annotation_labels</code> target, manual annotation labels
from the <code>manual_annotation_labels</code> target, deconvolution
labels from the <code>spot_deconvolution_labels</code> target, and the
new cell groupings defined in the <code>CELL_GROUPINGS</code>
parameter.</p>
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="copyright">
  <p></p>
<p>Developed by Jiri Novotny, Jan Kubovciak, Lucie Pfeiferova, Michal Kolar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

    </footer>
</div>





  </body>
</html>
