<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="01_integration stage of the integration pipeline.">
<!-- Inform modern browsers that this page supports both dark and light color schemes,
  and the page author prefers light. --><meta name="color-scheme" content="dark light">
<script>
  // If `prefers-color-scheme` is not supported, fall back to light mode.
  // i.e. In this case, inject the `light` CSS before the others, with
  // no media filter so that it will be downloaded with highest priority.
  if (window.matchMedia("(prefers-color-scheme: dark)").media === "not all") {
    document.documentElement.style.display = "none";
    document.head.insertAdjacentHTML(
      "beforeend",
      "<link id=\"css\" rel=\"stylesheet\" href=\"https://bootswatch.com/5/flatly/bootstrap.css\" onload=\"document.documentElement.style.display = ''\">"
    );
  }
</script><title>Integration stage (`01_integration`) ‚Ä¢ scdrake</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Integration stage (`01_integration`)">
<meta property="og:description" content="01_integration stage of the integration pipeline.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=281550496"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', '281550496');
</script><!-- Flatly Theme - Light  --><link id="css-light" rel="stylesheet" href="https://bootswatch.com/5/flatly/bootstrap.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<!-- Darkly Theme - Dark --><link id="css-dark" rel="stylesheet" href="https://bootswatch.com/5/darkly/bootstrap.css" media="(prefers-color-scheme: dark)">
<!-- preferably CSS --><link rel="stylesheet" href="../preferably.css">
<link id="css-code-light" rel="stylesheet" href="../code-color-scheme-light.css" media="(prefers-color-scheme: light), (prefers-color-scheme: no-preference)">
<link id="css-code-dark" rel="stylesheet" href="../code-color-scheme-dark.css" media="(prefers-color-scheme: dark)">
<script src="../darkswitch.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scdrake</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake_integration.html">Integration pipeline guide</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/pipeline_overview.html">Pipeline overview</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Guides</h6>
    <a class="dropdown-item" href="../articles/scdrake_docker.html">Using the Docker image</a>
    <a class="dropdown-item" href="../articles/scdrake_integration.html">02 Integration pipeline guide</a>
    <a class="dropdown-item" href="../articles/scdrake_advanced.html">Advanced topics</a>
    <a class="dropdown-item" href="../articles/scdrake_extend.html">Extending the pipeline</a>
    <a class="dropdown-item" href="../articles/drake_basics.html">drake basics</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General information</h6>
    <a class="dropdown-item" href="../articles/pipeline_overview.html">Pipeline overview</a>
    <a class="dropdown-item" href="../articles/scdrake_faq.html">FAQ &amp; Howtos</a>
    <a class="dropdown-item" href="../articles/scdrake_cli.html">Command line interface</a>
    <a class="dropdown-item" href="../articles/scdrake_config.html">scdrake config internals</a>
    <a class="dropdown-item" href="../articles/scdrake_envvars.html">Environment variables</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>General configs</h6>
    <a class="dropdown-item" href="../articles/config_pipeline.html">Pipeline config</a>
    <a class="dropdown-item" href="../articles/config_main.html">Main config</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Single-sample pipeline stages</h6>
    <a class="dropdown-item" href="../articles/stage_input_qc.html">Input and quality control stage (`01_input_qc`)</a>
    <a class="dropdown-item" href="../articles/stage_norm_clustering.html">Normalization and clustering stage (`02_norm_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Integration pipeline stages</h6>
    <a class="dropdown-item" href="../articles/stage_integration.html">Integration stage (`01_integration`)</a>
    <a class="dropdown-item" href="../articles/stage_int_clustering.html">Post-integration clustering stage (`02_int_clustering`)</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Common stages</h6>
    <a class="dropdown-item" href="../articles/stage_cluster_markers.html">Cluster markers stage (`cluster_markers`)</a>
    <a class="dropdown-item" href="../articles/stage_contrasts.html">Contrasts stage (`contrasts`)</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bioinfocz/scdrake" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>

        
        <li>
          <a class="external-link nav-link" id="css-toggle-btn" aria-label="github">
            <span class="fas fa fas fa-adjust fa-lg"></span>
          </a>
        </li>
        
        
        
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Integration stage (`01_integration`)</h1>
            
            <h4 data-toc-skip class="date"><sup>Document generated:
2023-12-19 08:41:40 UTC+0000</sup></h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bioinfocz/scdrake/blob/mastervignettes/stage_integration.Rmd" class="external-link"><code>vignettes/stage_integration.Rmd</code></a></small>
      <div class="d-none name"><code>stage_integration.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2 tabset">
<h2 class="tabset" id="section"></h2>



<ul class="nav nav-tabs" id="section" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#overview" id="overview-tab" type="button" role="tab" aria-controls="overview" aria-selected="true" class="active nav-link">Overview</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#config-parameters" id="config-parameters-tab" type="button" role="tab" aria-controls="config-parameters" aria-selected="false" class="nav-link">Config parameters</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#outputs" id="outputs-tab" type="button" role="tab" aria-controls="outputs" aria-selected="false" class="nav-link">Outputs</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="overview" role="tabpanel" aria-labelledby="overview-tab">

<p>This stage imports multiple <code>SingleCellExperiment</code> objects
and performs integration using one or several methods.</p>
<p>‚öôÔ∏è Config file:
<code>config/integration/01_integration.yaml</code></p>
<p>üìã HTML report target (in <code>config/pipeline.yaml</code>):
<code>DRAKE_TARGETS: ["report_integration"]</code></p>
<p><a href="https://onco.img.cas.cz/novotnyj/scdrake/integration/01_integration/01_integration.html" class="external-link">üìú
Example report</a> (<a href="https://github.com/bioinfocz/scdrake/blob/main/tests/testthat/run_pipeline_config_patches/integration/01_integration.default.yaml" class="external-link">used
config</a>)</p>
<p>ü™ú Structure</p>
<ul>
<li>Reading in data from:
<ul>
<li>
<code>drake</code> caches of multiple single-sample pipelines</li>
<li>Rds files with <code>SingleCellExperiment</code> objects</li>
</ul>
</li>
<li>Subsetting to common column and row data, and metadata</li>
<li>Normalization of each sample for inter-batch sequencing depth
(<code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchNorm.html" class="external-link">batchelor::multiBatchNorm()</a></code>, <a href="http://bioconductor.org/books/3.15/OSCA.multisample/integrating-datasets.html#slower-setup" class="external-link">theory</a>)</li>
<li>Combining of HVGs by either (<a href="http://bioconductor.org/books/3.15/OSCA.multisample/integrating-datasets.html#slower-setup" class="external-link">theory</a>):
<ul>
<li>A HVG metric (gene variance or CV2), then apply top HVGs selection
as in the single-sample pipeline. All samples must have the same HVG
metric used before.</li>
<li>Set operations (without duplicates): intersection, union, or take
them all</li>
</ul>
</li>
<li>Perform integration using one or more methods:
<ul>
<li>Rescaling (<code><a href="https://rdrr.io/pkg/batchelor/man/rescaleBatches.html" class="external-link">batchelor::rescaleBatches()</a></code>, <a href="http://bioconductor.org/books/3.15/OSCA.multisample/integrating-datasets.html#by-rescaling-the-counts" class="external-link">theory</a>):
scale counts so that the average count within each batch is the same for
each gene</li>
<li>Regression (<code><a href="https://rdrr.io/pkg/batchelor/man/regressBatches.html" class="external-link">batchelor::regressBatches()</a></code>, <a href="http://bioconductor.org/books/3.15/OSCA.multisample/integrating-datasets.html#by-fitting-a-linear-model" class="external-link">theory</a>):
fit a linear model to each gene regress out uninteresting factors of
variation, returning a matrix of residuals</li>
<li>Fast mutual nearest neighbors (<code><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html" class="external-link">batchelor::fastMNN()</a></code>, <a href="http://bioconductor.org/books/3.15/OSCA.multisample/integrating-datasets.html#mnn-correction" class="external-link">theory</a>)</li>
<li><a href="https://portals.broadinstitute.org/harmony/index.html" class="external-link">Harmony</a></li>
</ul>
</li>
</ul>
<p>The steps below are performed for each integration method, plus
separately for HVGs with removed cell cycle-related genes (if this
removal was requested).</p>
<ul>
<li>Calculation of PCA and selection of number of principal components
(PCs) which will be used downstream. This is similar to what is done in
the single-sample pipeline (<code>02_norm_clustering</code> stage), but,
in this case, the selection of PCs can be configured for each
integration method.</li>
<li>Dimensionality reduction, which is same as in the single-sample
pipeline, but as PCA, it can be configured for each integration
method</li>
<li>Expression plots of selected markers</li>
<li>Mutual nearest neighbors (MNN) clustering to assess integration
results</li>
<li>Integration diagnostics:
<ul>
<li>Assignment of cells of each sample to MNN clusters (tables,
plots)</li>
<li>
<a href="https://en.wikipedia.org/wiki/Rand_index" class="external-link">Rand indices</a>:
used to evaluate biological heterogeneity preservation by summarizing
the agreement between clusterings</li>
</ul>
</li>
</ul>
</div>
<div class="tab-pane" id="config-parameters" role="tabpanel" aria-labelledby="config-parameters-tab">

<p>Config for this stage is stored in the
<code>config/integration/01_integration.yaml</code> file. Directory with
this file is read from <code>SCDRAKE_INTEGRATION_CONFIG_DIR</code>
environment variable upon <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a> load or attach, and
saved as <code>scdrake_integration_config_dir</code> option. This option
is used as the default argument value in several <a href="https://github.com/bioinfocz/scdrake" class="external-link">scdrake</a>
functions.</p>
<hr>
<div class="section level4">
<h4 id="integration-sources">Integration sources<a class="anchor" aria-label="anchor" href="#integration-sources"></a>
</h4>
<div class="sourceCode" id="cb1"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_SOURCES</span><span class="kw">:</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> </span><span class="fu">pbmc1k</span><span class="kw">:</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">path</span><span class="kw">:</span><span class="at"> </span><span class="st">"path/to/.drake or path/to/sce.Rds"</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">path_type</span><span class="kw">:</span><span class="at"> </span><span class="st">"drake_cache"</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">description</span><span class="kw">:</span><span class="at"> </span><span class="st">"10x Genomics PBMC 1k dataset"</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">hvg_rm_cc_genes</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">hvg_cc_genes_var_expl_threshold</span><span class="kw">:</span><span class="at"> </span><span class="dv">5</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">pbmc3k</span><span class="kw">:</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">path</span><span class="kw">:</span><span class="at"> </span><span class="st">"path/to/.drake or path/to/sce.Rds"</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">path_type</span><span class="kw">:</span><span class="at"> </span><span class="st">"drake_cache"</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">description</span><span class="kw">:</span><span class="at"> </span><span class="st">"10x Genomics PBMC 3k dataset"</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">hvg_rm_cc_genes</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">hvg_cc_genes_var_expl_threshold</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span></code></pre></div>
<p><strong>Type:</strong> list of named lists</p>
<p>This parameter specifies datasets to be integrated. Those are read in
from single-sample pipeline results (the
<code>sce_final_norm_clustering</code> targets) taken from corresponding
<a href="https://docs.ropensci.org/drake" class="external-link">drake</a> caches or from SCE objects saved in Rds format. The
latter should correspond to <code>sce_final_norm_clustering</code>
targets, but one can modify their e.g.¬†<code>colData()</code> as
needed.</p>
<p>Each name of a nested list specifies the name of a single-sample,
e.g.¬†<code>pbmc1k</code>. Parameters for each single-sample are:</p>
<ul>
<li>
<code>path: "path/to/.drake or path/to/sce.Rds"</code> (character
scalar): a path to <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cache directory or SCE object
Rds file (depends on <code>path_type</code>).</li>
<li>
<code>path_type: "drake_cache"</code> (character scalar:
<code>"drake_cache"</code> | <code>"sce"</code>): a type of input data -
either a <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cache directory or SCE object Rds
file.</li>
<li>
<code>description: "10x Genomics PBMC 1k dataset"</code> (character
scalar): a description of the single-sample, will appear in
reports.</li>
<li>
<code>hvg_rm_cc_genes: True</code> (logical scalar): if
<code>True</code>, cell cycle-related genes will be removed prior to
selection of highly variable genes (HVGs). This is the same method of
cell cycle correction as in the <code>02_norm_clustering</code> stage of
the single-sample pipeline (see
<code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code> for more details).</li>
<li>
<code>hvg_cc_genes_var_expl_threshold: 5</code> (numeric scalar): a
threshold on cell cycle variance explained. Genes with var. expl.
greater than this threshold will be marked as CC-related.</li>
</ul>
<p>Currently, there are some limitations of which datasets can be
integrated:</p>
<ul>
<li>All single-samples must be normalized by <code>{scran}</code>
method.</li>
<li>If the HVG selection in integrated data is based on combined HVG
metrics (see the relevant parameters below), all single-samples had to
use the same HVG metric (<code>HVG_METRIC</code> of
<code>"gene_var"</code> or <code>"gene_cv2"</code>).</li>
</ul>
</div>
<div class="section level4">
<h4 id="selection-of-highly-variable-genes-hvgs">Selection of highly variable genes (HVGs)<a class="anchor" aria-label="anchor" href="#selection-of-highly-variable-genes-hvgs"></a>
</h4>
<p>This is done prior to integration, and is same for all methods
specified in the <code>INTEGRATION_METHODS</code> (see below).</p>
<p>If a single-sample has the <code>hvg_rm_cc_genes</code> set to
<code>True</code>, cell cycle-related genes will be removed prior to HVG
combination and selection.</p>
<hr>
<div class="sourceCode" id="cb2"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">HVG_COMBINATION_INT</span><span class="kw">:</span><span class="at"> </span><span class="st">"hvg_metric"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar (<code>"hvg_metric"</code> |
<code>"intersection"</code> | <code>"union"</code> |
<code>"all"</code>)</p>
<p>How to combine HVGs from single-samples:</p>
<ul>
<li>
<code>"hvg_metric"</code>: combine HVG metrics (gene variance or
CV2) and then apply HVG selection similar to procedure in the
<code>02_norm_clustering</code> stage of the single-sample pipeline.
Note that in order to combine HVG metrics, all samples had to have
<code>HVG_METRIC</code> set to either <code>"gene_var"</code>
(<code><a href="https://rdrr.io/pkg/scran/man/combineVar.html" class="external-link">scran::combineVar()</a></code>) or <code>"gene_cv2"</code>
(<code><a href="https://rdrr.io/pkg/scran/man/combineVar.html" class="external-link">scran::combineCV2()</a></code>).</li>
<li>Combine HVGs by set operations:
<ul>
<li><code>"intersection"</code></li>
<li><code>"union"</code></li>
</ul>
</li>
<li>
<code>"all"</code>: use all common genes from all samples as HVGs
(without duplicates).</li>
</ul>
<hr>
<div class="sourceCode" id="cb3"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">HVG_SELECTION_INT</span><span class="kw">:</span><span class="at"> </span><span class="st">"top"</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">HVG_SELECTION_VALUE_INT</span><span class="kw">:</span><span class="at"> </span><span class="dv">3000</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar, integer scalar</p>
<p>HVG selection strategy. Only relevant when
<code>HVG_COMBINATION_INT</code> is <code>"hvg_metric"</code>, i.e.¬†this
selection will be applied to combined HVG metrics.</p>
<p>These parameters are similar to <code>HVG_SELECTION</code> and
<code>HVG_SELECTION_VALUE</code> in the <code>02_norm_clustering</code>
stage of the single-sample pipeline (see
<code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code> for more details).</p>
</div>
<div class="section level4">
<h4 id="integration-methods">Integration methods<a class="anchor" aria-label="anchor" href="#integration-methods"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_METHODS</span><span class="kw">:</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="at">  </span><span class="kw">-</span><span class="at"> </span><span class="fu">uncorrected</span><span class="kw">:</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_selection_method</span><span class="kw">:</span><span class="at"> </span><span class="st">"forced"</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_forced_pcs</span><span class="kw">:</span><span class="at"> </span><span class="dv">15</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_perp</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_max_iter</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">rescaling</span><span class="kw">:</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_selection_method</span><span class="kw">:</span><span class="at"> </span><span class="st">"forced"</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_forced_pcs</span><span class="kw">:</span><span class="at"> </span><span class="dv">15</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_perp</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_max_iter</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">integration_params</span><span class="kw">:</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">log.base</span><span class="kw">:</span><span class="at"> </span><span class="dv">2</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">pseudo.count</span><span class="kw">:</span><span class="at"> </span><span class="dv">1</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">regression</span><span class="kw">:</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_selection_method</span><span class="kw">:</span><span class="at"> </span><span class="st">"corrected"</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_forced_pcs</span><span class="kw">:</span><span class="at"> </span><span class="dv">15</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_perp</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_max_iter</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">integration_params</span><span class="kw">:</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">d</span><span class="kw">:</span><span class="at"> </span><span class="dv">50</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">mnn</span><span class="kw">:</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_selection_method</span><span class="kw">:</span><span class="at"> </span><span class="st">"corrected"</span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_forced_pcs</span><span class="kw">:</span><span class="at"> </span><span class="dv">15</span></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_perp</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_max_iter</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">integration_params</span><span class="kw">:</span></span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">k</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">prop.k</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">cos.norm</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">ndist</span><span class="kw">:</span><span class="at"> </span><span class="dv">3</span></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">d</span><span class="kw">:</span><span class="at"> </span><span class="dv">50</span></span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">merge.order</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">auto.merge</span><span class="kw">:</span><span class="at"> </span><span class="ch">True</span></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a><span class="at">    </span><span class="fu">harmony</span><span class="kw">:</span></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_selection_method</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">pca_forced_pcs</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_perp</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">tsne_max_iter</span><span class="kw">:</span><span class="at"> </span><span class="dv">1000</span></span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a><span class="at">      </span><span class="fu">integration_params</span><span class="kw">:</span></span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">dims.use</span><span class="kw">:</span><span class="at"> </span><span class="dv">50</span></span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">theta</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">lambda</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">sigma</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.1</span></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">nclust</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">tau</span><span class="kw">:</span><span class="at"> </span><span class="dv">0</span></span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">block.size</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.05</span></span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">max.iter.harmony</span><span class="kw">:</span><span class="at"> </span><span class="dv">10</span></span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">max.iter.cluster</span><span class="kw">:</span><span class="at"> </span><span class="dv">20</span></span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">epsilon.cluster</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.00001</span></span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a><span class="at">        </span><span class="fu">epsilon.harmony</span><span class="kw">:</span><span class="at"> </span><span class="fl">0.0001</span></span></code></pre></div>
<p><strong>Type:</strong> list of named lists</p>
<p>Each named list is specifying parameters for one integration method.
The ones specified in the <code>integration_params</code> list are used
directly in an integration method, while the others are used prior/after
integration. The <code>"uncorrected"</code> and at least one integration
method must be set (by default, all methods are performed). Here are
descriptions and possible parameters (some of them are used by multiple
methods, and thus described only once) for each integration method:</p>
<ul>
<li>
<code>uncorrected</code>: correction for inter-batch sequencing
depth with <code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchNorm.html" class="external-link">batchelor::multiBatchNorm()</a></code>. This method is
required, because it is used for differential expression analysis
(cluster markers and contrasts).
<ul>
<li>
<code>pca_selection_method: "forced"</code> (character scalar:
<code>"forced"</code> | <code>"elbow"</code> |
<code>"technical_noise"</code>): see the
<code>PCA_SELECTION_METHOD</code> parameter in
<code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code>.</li>
<li>
<code>pca_forced_pcs: 15</code> (integer scalar): force this number
of selected PCs (if <code>pca_selection_method</code> is
<code>"forced"</code>).</li>
<li>
<code>tsne_perp: 20</code> (numeric scalar): t-SNE perplexity.</li>
<li>
<code>tsne_max_iter: 1000</code> (positive integer scalar): a
maximum number of t-SNE iterations.</li>
</ul>
</li>
<li>
<code>rescaling</code>: scale counts so that the average count
within each batch is the same for each gene
(<code><a href="https://rdrr.io/pkg/batchelor/man/rescaleBatches.html" class="external-link">batchelor::rescaleBatches()</a></code>).
<ul>
<li><code>pca_selection_method: "forced"</code></li>
<li><code>pca_forced_pcs: 15</code></li>
<li><code>tsne_perp: 20</code></li>
<li><code>tsne_max_iter: 1000</code></li>
<li>
<code>integration_params</code>: directly passed to
<code><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">batchelor::RescaleParam()</a></code> used for
<code><a href="https://rdrr.io/pkg/batchelor/man/rescaleBatches.html" class="external-link">batchelor::rescaleBatches()</a></code>. Default values from this
function are used by default.
<ul>
<li>
<code>log.base: 2</code> (numeric scalar): a base of the
log-transformation.</li>
<li>
<code>pseudo.count: 1</code> (numeric scalar): a pseudo-count used
for the log-transformation.</li>
</ul>
</li>
</ul>
</li>
<li>
<code>regression</code>: fit a linear model to each gene regress out
uninteresting factors of variation, returning a matrix of residuals
(<code><a href="https://rdrr.io/pkg/batchelor/man/regressBatches.html" class="external-link">batchelor::regressBatches()</a></code>).
<ul>
<li>
<code>pca_selection_method: "corrected"</code>: in this integration
method, <code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchPCA.html" class="external-link">batchelor::multiBatchPCA()</a></code> is used, and all PCs (50
by default) from this method will be used if <code>"corrected"</code> is
specified. Otherwise any other PCA selection method can be used.</li>
<li><code>pca_forced_pcs: 15</code></li>
<li><code>tsne_perp: 20</code></li>
<li><code>tsne_max_iter: 1000</code></li>
<li>
<code>integration_params</code>: directly passed to
<code><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">batchelor::RegressParam()</a></code> used for
<code><a href="https://rdrr.io/pkg/batchelor/man/regressBatches.html" class="external-link">batchelor::regressBatches()</a></code>. Default values from this
function are used by default.
<ul>
<li>
<code>d: 50</code> (integer scalar): a number of PCs to compute in
<code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchPCA.html" class="external-link">batchelor::multiBatchPCA()</a></code>.</li>
</ul>
</li>
</ul>
</li>
<li>
<code>mnn</code>: fast mutual nearest neighbors
(<code><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html" class="external-link">batchelor::fastMNN()</a></code>).
<ul>
<li><code>pca_selection_method: "corrected"</code></li>
<li><code>pca_forced_pcs: 15</code></li>
<li><code>tsne_perp: 20</code></li>
<li><code>tsne_max_iter: 1000</code></li>
<li>
<code>integration_params</code>: directly passed to
<code><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">batchelor::FastMnnParam()</a></code> used for
<code><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html" class="external-link">batchelor::fastMNN()</a></code>. Default values from this function are
used by default.
<ul>
<li>
<code>k: 20</code> (integer scalar): a number of nearest neighbors
to consider when identifying MNNs.</li>
<li>
<code>prop.k: null</code> (numeric scalar): a proportion of cells in
each dataset to use for mutual nearest neighbor searching. If set, the
number of nearest neighbors used for the MNN search in each batch is
redefined as <code>max(k, prop.k*N)</code> where <code>N</code> is the
number of cells in that batch.</li>
<li>
<code>cos.norm: True</code> (logical scalar): if <code>True</code>,
cosine normalization is performed on the input data prior to PCA.</li>
<li>
<code>ndist: 3</code> (numeric scalar): a threshold beyond which
neighbors are to be ignored when computing correction vectors. Each
threshold is defined as a multiple of the number of median
distances.</li>
<li><code>d: 50</code></li>
<li>
<code>merge.order: null</code> (integer vector, list of lists, or
<code>null</code>): a merge order of single-samples. Alternatively, a
list of lists representing a tree structure specifying a hierarchical
merge order (not tested, see <code><a href="https://rdrr.io/pkg/batchelor/man/fastMNN.html" class="external-link">batchelor::fastMNN()</a></code> for more
details).</li>
<li>
<code>auto.merge: True</code> (logical scalar): if
<code>True</code>, automatically identify the ‚Äúbest‚Äù merge order.</li>
</ul>
</li>
</ul>
</li>
<li>
<code>harmony</code>: projects cells into a shared embedding in
which cells group by cell type rather than dataset-specific conditions
(<code><a href="https://rdrr.io/pkg/harmony/man/RunHarmony.html" class="external-link">harmony::RunHarmony()</a></code>, <a href="https://www.nature.com/articles/s41592-019-0619-0" class="external-link">Nature
Methods</a>). Default values from this function are used by default.
Note that Harmony does not compute an integrated expression matrix, but
only reduced dimensions corrected for batch effect that are subsequently
used for UMAP and t-SNE, and clustering.
<ul>
<li>
<code>pca_selection_method: null; pca_forced_pcs: null</code>: not
used, see <code>dims.use</code> below</li>
<li><code>tsne_perp: 20</code></li>
<li><code>tsne_max_iter: 1000</code></li>
<li>
<code>integration_params</code>: directly passed to
<code><a href="https://rdrr.io/pkg/harmony/man/RunHarmony.html" class="external-link">harmony::RunHarmony()</a></code>. Default values from this function
are used by default, except <code>dims.use</code>.
<ul>
<li>
<code>dims.use: 50</code> (integer scalar): number of first PCs to
use for Harmony integration. Selection of PCs is not applied for Harmony
- PCA with <code>dims.use</code> PCs is computed prior to the
integration.</li>
<li>
<code>theta: null</code>: diversity clustering penalty parameter.
Specify for each variable in <code>group.by.vars</code>
(<code>"batch"</code>). Default <code>theta = 2</code>,
<code>theta = 0</code> does not encourage any diversity. Larger values
of theta result in more diverse clusters.</li>
<li>
<code>lambda: null</code>: didge regression penalty parameter.
Specify for each variable in <code>group.by.vars</code>
(<code>"batch"</code>). Default <code>lambda = 1</code>. Lambda must be
strictly positive. Smaller values result in more aggressive
correction.</li>
<li>
<code>sigma: 0.1</code>: width of soft kmeans clusters. Default
<code>sigma = 0.1</code>. Sigma scales the distance from a cell to
cluster centroids. Larger values of sigma result in cells assigned to
more clusters. Smaller values of sigma make soft k-means cluster
approach hard clustering.</li>
<li>
<code>nclust: null</code>: number of clusters in model.
<code>nclust = 1</code> equivalent to simple linear regression.</li>
<li>
<code>tau: 0</code>: protection against overclustering small
datasets with large ones. <code>tau</code> is the expected number of
cells per cluster.</li>
<li>
<code>block.size: 0.05</code>: what proportion of cells to update
during clustering. Between <code>0</code> to <code>1</code>, default
<code>0.05</code>. Larger values may be faster but less accurate.</li>
<li>
<code>max.iter.harmony: 10</code>: maximum number of rounds to run
Harmony. One round of Harmony involves one clustering and one correction
step.</li>
<li>
<code>max.iter.cluster: 20</code>: maximum number of rounds to run
clustering at each round of Harmony.</li>
<li>
<code>epsilon.cluster: 0.00001</code>: convergence tolerance for
clustering round of Harmony. Set to <code>-inf</code> to never stop
early.</li>
<li>
<code>epsilon.harmony: 0.0001</code>: convergence tolerance for
Harmony. Set to <code>-inf</code> to never stop early.</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="integration-diagnostics">Integration diagnostics<a class="anchor" aria-label="anchor" href="#integration-diagnostics"></a>
</h4>
<div class="section level5">
<h5 id="fast-mnn-clustering">Fast MNN clustering<a class="anchor" aria-label="anchor" href="#fast-mnn-clustering"></a>
</h5>
<p>Used to obtain basic <a href="http://bioconductor.org/books/3.15/OSCA.multisample/correction-diagnostics.html" class="external-link">integration
diagnostics</a>. See <code><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">scran::buildSNNGraph()</a></code> and
<code><a href="https://rdrr.io/pkg/bluster/man/makeSNNGraph.html" class="external-link">bluster::makeSNNGraph()</a></code> for more details.</p>
<hr>
<div class="sourceCode" id="cb5"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_SNN_K</span><span class="kw">:</span><span class="at"> </span><span class="dv">10</span></span></code></pre></div>
<p><strong>Type:</strong> integer scalar</p>
<p>A number of nearest neighbors to consider during graph
construction.</p>
<hr>
<div class="sourceCode" id="cb6"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_SNN_TYPE</span><span class="kw">:</span><span class="at"> </span><span class="st">"rank"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar (<code>"rank"</code> |
<code>"number"</code> | <code>"jaccard"</code>)</p>
<p>A type of weighting scheme to use for shared neighbors.</p>
<hr>
<div class="sourceCode" id="cb7"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_SNN_CLUSTERING_METHOD</span><span class="kw">:</span><span class="at"> </span><span class="st">"walktrap"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar (<code>"walktrap"</code> |
<code>"louvain"</code>)</p>
<p>A type of MNN clustering algorithm. See
<code><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html" class="external-link">igraph::cluster_walktrap()</a></code> and
<code><a href="https://rdrr.io/pkg/igraph/man/cluster_louvain.html" class="external-link">igraph::cluster_louvain()</a></code> for more details.</p>
</div>
</div>
<div class="section level4">
<h4 id="input-files">Input files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SELECTED_MARKERS_FILE</span><span class="kw">:</span><span class="at"> </span><span class="ch">null</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar or <code>null</code></p>
<p>Similar to the <code>SELECTED_MARKERS_FILE</code> parameter in
<code><a href="../articles/stage_norm_clustering.html">vignette("stage_norm_clustering")</a></code>.</p>
<hr>
<div class="sourceCode" id="cb9"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_REPORT_RMD_FILE</span><span class="kw">:</span><span class="at"> </span><span class="st">"Rmd/integration/01_integration.Rmd"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>A path to RMarkdown file used for HTML report of this pipeline
stage.</p>
</div>
<div class="section level4">
<h4 id="output-files">Output files<a class="anchor" aria-label="anchor" href="#output-files"></a>
</h4>
<div class="sourceCode" id="cb10"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_BASE_OUT_DIR</span><span class="kw">:</span><span class="at"> </span><span class="st">"01_integration"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>A path to base output directory for this stage. It will be created
under <code>BASE_OUT_DIR</code> specified in <code>00_main.yaml</code>
config.</p>
<hr>
<div class="sourceCode" id="cb11"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_SELECTED_MARKERS_OUT_DIR</span><span class="kw">:</span><span class="at"> </span><span class="st">"selected_markers"</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_REPORT_HTML_FILE</span><span class="kw">:</span><span class="at"> </span><span class="st">"01_integration.html"</span></span></code></pre></div>
<p><strong>Type:</strong> character scalar</p>
<p>Names of files and directories created under
<code>INTEGRATION_BASE_OUT_DIR</code>. Subdirectories are not
allowed.</p>
</div>
<div class="section level4">
<h4 id="html-output-parameters">HTML output parameters<a class="anchor" aria-label="anchor" href="#html-output-parameters"></a>
</h4>
<div class="sourceCode" id="cb12"><pre class="sourceCode yaml"><code class="sourceCode yaml"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_KNITR_MESSAGE</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_KNITR_WARNING</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">INTEGRATION_KNITR_ECHO</span><span class="kw">:</span><span class="at"> </span><span class="ch">False</span></span></code></pre></div>
<p><strong>Type:</strong> logical scalar</p>
<p>These are passed to <code><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">knitr::opts_chunk()</a></code> and used for
rendering of stage‚Äôs HTML report.</p>
</div>
</div>
<div class="tab-pane" id="outputs" role="tabpanel" aria-labelledby="outputs-tab">

<p>Here you can find description of the most important targets for this
stage. However, for a full overview, you have to inspect the <a href="https://github.com/bioinfocz/scdrake/blob/main/R/plans_integration.R" class="external-link">source
code</a> of the <code><a href="../reference/get_subplan_integration.html">get_integration_subplan()</a></code> function.</p>
<p>HTML report target name: <code>report_integration</code></p>
<div class="section level4">
<h4 id="singlecellexperiment-objects">
<code>SingleCellExperiment</code> objects<a class="anchor" aria-label="anchor" href="#singlecellexperiment-objects"></a>
</h4>
<p><code>sce_int_import</code>: list of imported single-sample SCE
objects, as defined in the <code>INTEGRATION_SOURCES</code> parameter.
Because <a href="https://docs.ropensci.org/drake" class="external-link">drake</a> cannot watch for changes in a cache, this
target is always rerun. Also, some constraints are checked (common
normalization method and HVG metrics).</p>
<hr>
<p><code>sce_int_raw_snn_clustering</code>: <code>sce_int_import</code>
with computed fast SNN clustering for each single-sample, used for
integration diagnostics</p>
<hr>
<p><code>sce_int_processed</code>: <code>sce_int_import</code> subsetted
to common colData, and features (rows) and their corresponding metadata
(<code>hvg_ids</code>, <code>hvg_metric_fit</code>)</p>
<hr>
<p><code>sce_int_multibatchnorm</code>: <code>sce_int_processed</code>
with SCE objects normalized for inter-batch sequencing depth
(<code><a href="https://rdrr.io/pkg/batchelor/man/multiBatchNorm.html" class="external-link">batchelor::multiBatchNorm()</a></code>)</p>
<hr>
<p><code>sce_int_df</code>: a tibble with integrated SCE objects. Each
row is one method defined in the <code>INTEGRATION_METHODS</code>
parameter and either with or without removed cell cycle-related genes
from HVGs. See <code><a href="../reference/sce_int_df_fn.html">?sce_int_df_fn</a></code> for information about what is
added or modified in <code>metadata()</code> of each integrated SCE
object. The integrated assay is named <code>"integrated"</code>, except
for Harmony integration, which only computes a batch-corrected reduced
dimensions (available in <code>reducedDims(sce, "harmony")</code>).</p>
<hr>
<p><code>sce_int_pca_df</code>: <code>sce_int_df</code> with computed
PCA and selected number of PCs. Also includes PC selection statistics
and plot</p>
<hr>
<p><code>sce_int_clustering_df</code>: <code>sce_int_pca_df</code> with
computed MNN clustering used for integration diagnostics</p>
<hr>
<p><code>sce_int_dimred_df</code>: <code>sce_int_pca_df</code> with
computed t-SNE and UMAP dimensionality reductions</p>
</div>
<div class="section level4">
<h4 id="highly-variable-genes-hvgs-selection">Highly variable genes (HVGs) selection<a class="anchor" aria-label="anchor" href="#highly-variable-genes-hvgs-selection"></a>
</h4>
<p><code>hvg_int</code>, <code>hvg_int_with_cc</code>: lists of HVGs and
selection parameters. The latter is not <code>NULL</code> when any of
the single-samples defined in the <code>INTEGRATION_SOURCES</code>
parameter has <code>hvg_rm_cc_genes</code> set to <code>True</code>.</p>
<hr>
<p><code>hvg_plots_int_df</code>: a tibble with HVG diagnostic plots</p>
</div>
<div class="section level4">
<h4 id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a>
</h4>
<p><code>sce_int_dimred_plots_df</code>: a tibble with dimensionality
reduction plots</p>
<div class="section level5">
<h5 id="selected-markers">Selected markers<a class="anchor" aria-label="anchor" href="#selected-markers"></a>
</h5>
<p><code>selected_markers_int_plots_df</code>: a tibble with plots of
selected markers, similar to <code>selected_markers_plots</code> in the
<code>02_norm_clustering</code> stage of the single-sample pipeline</p>
<p><code>selected_markers_plots_files_out</code>: make this target to
export the plots</p>
</div>
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="copyright">
  <p></p>
<p>Developed by Jiri Novotny, Jan Kubovciak, Michal Kolar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
  <p class="preferably">Using <a href="https://preferably.amirmasoudabdol.name/?source=footer" class="external-link">preferably</a> template.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
